{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='sec0'></a>\n",
    "# Feature Engineering1\n",
    "- Importing Data\n",
    "- <a href='#sec1'>Exemplary Text Analysis</a>\n",
    "  - <a href='#sec1_1'>Old way</a>\n",
    "  - <a href='#sec1_2'>Some useful regex and lists</a>\n",
    "  - <a href='#sec1_3'>Replace periods, commas, hyphens, brackets with a space, and then tokenize</a>\n",
    "\n",
    "- <a href='#sec2'>Surveying the whole text. Looking at each type of feature</a>\n",
    "  - <a href='#sec2_1'>Words with '-positive' and '-negative'</a>\n",
    "  - <a href='#sec2_2'>Detect genes with hyphens, including fusions</a>\n",
    "  - <a href='#sec2_3'>Mutation Types</a>\n",
    "  - <a href='#sec2_4'>Gene Like Names</a>\n",
    "\n",
    "- <a href='#sec3'>Functions for Processing Features</a>\n",
    "- <a href='#sec4'>Process Features</a>\n",
    "- <a href='#sec5'>Test with Random Forest</a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import re\n",
    "\n",
    "from nltk import PorterStemmer\n",
    "from nltk.tokenize import word_tokenize\n",
    "from nltk.corpus import stopwords\n",
    "\n",
    "sns.set_context(\"paper\")\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>Importing train_text</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class_train = pd.read_csv('train_variants')\n",
    "text_train = pd.read_csv(\"train_text\", sep=\"\\|\\|\", engine='python', header=None, skiprows=1, names=[\"ID\",\"Text\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>FAM58A</td>\n",
       "      <td>Truncating Mutations</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>CBL</td>\n",
       "      <td>W802*</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>CBL</td>\n",
       "      <td>Q249E</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>CBL</td>\n",
       "      <td>N454D</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>CBL</td>\n",
       "      <td>L399V</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID    Gene             Variation  Class\n",
       "0   0  FAM58A  Truncating Mutations      1\n",
       "1   1     CBL                 W802*      2\n",
       "2   2     CBL                 Q249E      2\n",
       "3   3     CBL                 N454D      3\n",
       "4   4     CBL                 L399V      4"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class_train.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Cyclin-dependent kinases (CDKs) regulate a var...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>Recent evidence has demonstrated that acquired...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>Oncogenic mutations in the monomeric Casitas B...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID                                               Text\n",
       "0   0  Cyclin-dependent kinases (CDKs) regulate a var...\n",
       "1   1   Abstract Background  Non-small cell lung canc...\n",
       "2   2   Abstract Background  Non-small cell lung canc...\n",
       "3   3  Recent evidence has demonstrated that acquired...\n",
       "4   4  Oncogenic mutations in the monomeric Casitas B..."
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='sec1'></a>\n",
    "# Exemplary Text Analysis  (<a href='#sec0'>Back To Top</a>)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "txt1 = text_train.iloc[903, 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ID                           903\n",
       "Gene                      PDGFRA\n",
       "Variation    KIF5B-PDGFRA Fusion\n",
       "Class                          7\n",
       "Name: 903, dtype: object"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class_train.iloc[903, :]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='sec1_1'></a>\n",
    "## Old way (<a href='#sec0'>Back To Top</a>)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "word_tokens = word_tokenize(txt1)\n",
    "word_tokens = np.array(word_tokens)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "initial leng 6402\n"
     ]
    }
   ],
   "source": [
    "print('initial leng %d' % len(word_tokens))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After removing stop words 4632\n"
     ]
    }
   ],
   "source": [
    "stop_words = set(stopwords.words('english'))\n",
    "txt1_words = [w for w in word_tokens if not w in stop_words]\n",
    "print('After removing stop words %d' % len(txt1_words))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tokens</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>We</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>identified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>two</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>patients</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>(</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       tokens\n",
       "0          We\n",
       "1  identified\n",
       "2         two\n",
       "3    patients\n",
       "4           ("
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = pd.DataFrame(txt1_words)\n",
    "df1.columns = ['tokens']\n",
    "df1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "gene_ish_pattern = r\"[A-Z]{2,7}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# get gene-ish words in a simple list\n",
    "gene_ish_words1 = [word for word in txt1_words if re.match(gene_ish_pattern, word)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "445"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gene_ish_words1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['BCR-ABL',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'STRN-PDGFRA',\n",
       " 'ETV6-PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'CEL',\n",
       " 'CEL',\n",
       " 'PDGFRA',\n",
       " 'PDGFRB',\n",
       " 'FGFR1',\n",
       " 'JAK2',\n",
       " 'PDGFRA',\n",
       " 'PDGFRB',\n",
       " 'PDGFRA',\n",
       " 'BCR',\n",
       " 'FIP1L1',\n",
       " 'KIF5B',\n",
       " 'CDK5RAP2',\n",
       " 'CEL',\n",
       " 'CEL',\n",
       " 'MPD',\n",
       " 'MPD',\n",
       " 'XY',\n",
       " 'XY',\n",
       " 'PCR',\n",
       " 'RT-PCR',\n",
       " 'PDGFRA',\n",
       " 'PCR',\n",
       " 'STRN-PDGFRA',\n",
       " 'DNA',\n",
       " 'STRN.Fusion.Ex6.1F',\n",
       " 'PDGFRA.Fusion.Ex12.2R',\n",
       " 'ETV6-PDGFRA',\n",
       " 'ETV6.Fusion.Ex6.1F',\n",
       " 'PDGFRA.Fusion.Ex12.2R',\n",
       " 'MRD',\n",
       " 'FISH',\n",
       " 'FISH',\n",
       " 'BAC',\n",
       " 'PDGFRA',\n",
       " 'ETV6',\n",
       " 'BAC',\n",
       " 'DNA',\n",
       " 'UK',\n",
       " 'BACs',\n",
       " 'RP11-24O10',\n",
       " 'PDGFRA',\n",
       " 'RP11-434C1',\n",
       " 'ETV6',\n",
       " 'PCR',\n",
       " 'PDGFRA',\n",
       " 'BCR-ABL',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'RT-PCR',\n",
       " 'FISH',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PCR',\n",
       " 'STRN-PDGFRA',\n",
       " 'PCR',\n",
       " 'PCR',\n",
       " 'PCR',\n",
       " 'PDAI12-R4',\n",
       " 'DNA',\n",
       " 'PCR',\n",
       " 'STRN-PDGFRA',\n",
       " 'STRN-PDGFRA',\n",
       " 'DNA',\n",
       " 'WW-like',\n",
       " 'WW',\n",
       " 'WD40',\n",
       " 'WW',\n",
       " 'STRN-PDGFRA',\n",
       " 'STRN',\n",
       " 'STRN',\n",
       " 'NM_003162',\n",
       " 'PDGFRA',\n",
       " 'DNA',\n",
       " 'DNA',\n",
       " 'PDGFRA-STRN',\n",
       " 'PCR',\n",
       " 'PDGFRA',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'AG',\n",
       " 'STRN',\n",
       " 'ETV6-PDGFRA',\n",
       " 'PCR',\n",
       " 'ETV6',\n",
       " 'NM_001987',\n",
       " 'PDGFRA',\n",
       " 'DNA',\n",
       " 'DNA',\n",
       " 'PDGFRA-ETV6',\n",
       " 'PCR',\n",
       " 'ETV6-PDGFRA',\n",
       " 'PCR',\n",
       " 'ETV6',\n",
       " 'PDGFRA',\n",
       " 'ETV6-PDGFRA',\n",
       " 'FISH',\n",
       " 'BACs',\n",
       " 'RP11-24O10',\n",
       " 'RP11-434C1',\n",
       " 'ETV6',\n",
       " 'PDGFRA',\n",
       " 'ETV6-PDGFRA',\n",
       " 'DNA',\n",
       " 'PCR',\n",
       " 'ETV6',\n",
       " 'PDGFRA',\n",
       " 'RT-PCR',\n",
       " 'ETV6-PDGFRA',\n",
       " 'DNA',\n",
       " 'FISH',\n",
       " 'ETS',\n",
       " 'MRD',\n",
       " 'DNA',\n",
       " 'DNA',\n",
       " 'CEL',\n",
       " 'STRN',\n",
       " 'ETV6',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'ETV6',\n",
       " 'CEL',\n",
       " 'STRN',\n",
       " 'WD',\n",
       " 'WD',\n",
       " 'DNA',\n",
       " 'STRN',\n",
       " 'PDGFRA',\n",
       " 'WD',\n",
       " 'WW-like',\n",
       " 'ETV6',\n",
       " 'CHIC2/BTL',\n",
       " 'AML',\n",
       " 'PGFRA',\n",
       " 'CHIC2/BTL',\n",
       " 'ETV6-PDGFRA',\n",
       " 'ETV6-CHIC2',\n",
       " 'PDGFRA',\n",
       " 'STRN',\n",
       " 'WW-like',\n",
       " 'ETV6-PDGFRA',\n",
       " 'WW-like',\n",
       " 'WW-like',\n",
       " 'ETV6-PDGFRA',\n",
       " 'WW-like',\n",
       " 'WW-like',\n",
       " 'WW-like',\n",
       " 'WW-like',\n",
       " 'IHES',\n",
       " 'FIP1-Like-1',\n",
       " 'FIP1L1',\n",
       " 'PDGFRA',\n",
       " 'IHES',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'FIP1L1-PDGFRA-positive',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'PCR',\n",
       " 'KIF5B',\n",
       " 'PDGFRA',\n",
       " 'PDGFRalpha',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'PCR',\n",
       " 'DNA',\n",
       " 'PDGFRA',\n",
       " 'IHES',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'KIF5B',\n",
       " 'HES',\n",
       " 'HES',\n",
       " 'IHES',\n",
       " 'CEL',\n",
       " 'CEL',\n",
       " 'FIP1-Like-1',\n",
       " 'FIP1L1',\n",
       " 'PDGFRA',\n",
       " 'PDGFRalpha',\n",
       " 'FIP1L1-PDGFRA-positive',\n",
       " 'CEL',\n",
       " 'PDGFRB',\n",
       " 'PDGFR',\n",
       " 'IHES',\n",
       " 'PDGFRB',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'FISH',\n",
       " 'FIP1L1',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'RT-PCR',\n",
       " 'PDGFRA',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'IHES',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'RT-PCR',\n",
       " 'BCR-ABL',\n",
       " 'FIP1L1-PDGFRA-positive',\n",
       " 'FIP1L1-PDGFRA-negative',\n",
       " 'PCR',\n",
       " 'RNA',\n",
       " 'RNeasy',\n",
       " 'RNA',\n",
       " 'BCR-ABL',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'RT-PCR',\n",
       " 'FIP1E6-F1',\n",
       " 'PDA-R1',\n",
       " 'FIP1E6-F2',\n",
       " 'PDA-R2',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'KIF5B-exon19F1',\n",
       " 'PDA-exon14R1',\n",
       " 'DNA',\n",
       " 'KIF5B-exon22-F1',\n",
       " 'PDAI12-R3',\n",
       " 'KIF5B-exon19F1',\n",
       " 'PDGFRA-exon14R1',\n",
       " 'KIF5B-exon19F2',\n",
       " 'PDGFRA-exon13/14R2',\n",
       " 'PCR',\n",
       " 'DNA',\n",
       " 'KIF5B-exon22-F1',\n",
       " 'PDAI12-R3',\n",
       " 'KIF5B',\n",
       " 'PDAI12-R4',\n",
       " 'PCR',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'RT-PCR',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDA-exon15R',\n",
       " 'PDGFRA',\n",
       " 'PDA-exon10F',\n",
       " 'PDA-exon13F',\n",
       " 'UK',\n",
       " 'PDA-exon10F',\n",
       " 'PDA-exon13F',\n",
       " 'PDA-exon15R',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'MG63',\n",
       " 'PCR',\n",
       " 'PCR',\n",
       " 'PCR',\n",
       " 'BCR-ABL',\n",
       " 'DNA',\n",
       " 'DNA',\n",
       " 'PCR',\n",
       " 'UK',\n",
       " 'PDGFRA',\n",
       " 'PDAI12-R3',\n",
       " 'DNA',\n",
       " 'UK',\n",
       " 'PCR',\n",
       " 'PDGFRA',\n",
       " 'NVAMP-1',\n",
       " 'TGCTCGTAGTAATCGTTCGCAC-3',\n",
       " 'PDAI12-R4',\n",
       " 'NVAMP-2',\n",
       " 'UK',\n",
       " 'TOPO',\n",
       " 'PCR4',\n",
       " 'UK',\n",
       " 'BAC',\n",
       " 'PDGFRA',\n",
       " 'KIF5B',\n",
       " 'DNA',\n",
       " 'HIV',\n",
       " 'TPHA',\n",
       " 'BCR-ABL',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'RT-PCR',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'BCR-PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'RT-PCR',\n",
       " 'PDGFRA',\n",
       " 'RT-PCR',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDA-exon10F',\n",
       " 'PDA-exon15R',\n",
       " 'PDGFRA',\n",
       " 'PDA-exon13F',\n",
       " 'PDA-exon15R',\n",
       " 'PDGFRA',\n",
       " 'MG-63',\n",
       " 'PDGFRA',\n",
       " 'FIP1-Like-1',\n",
       " 'FIP1L1',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'DNA',\n",
       " 'PDGFRA',\n",
       " 'MG-63',\n",
       " 'PDGFRA',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'DNA',\n",
       " 'PDGFRA',\n",
       " 'IHES',\n",
       " 'PDGFRA',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'RT-PCR',\n",
       " 'PDGFRA',\n",
       " 'FIP1L1-PDGFRA-negative',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PCR',\n",
       " 'DNA',\n",
       " 'BCR-ABL',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'NCP',\n",
       " 'PDAI12-R4',\n",
       " 'PDGFRA',\n",
       " 'PCR',\n",
       " 'PDGFRA',\n",
       " 'PCR',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PCR',\n",
       " 'DNA',\n",
       " 'PDAI12-R4',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'FIP1L1',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'RT-PCR',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'KIF5B',\n",
       " 'NM_004521',\n",
       " 'PDGFRA',\n",
       " 'DNA',\n",
       " 'DNA',\n",
       " 'KIF5B-exon22-F1',\n",
       " 'PDAI12-R3',\n",
       " 'PDGFRA-KIF5B',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'DNA',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'DNA',\n",
       " 'PB',\n",
       " 'BM',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'KIF5B-exon19F1',\n",
       " 'PDA-exon14R1',\n",
       " 'RT-PCR',\n",
       " 'DNA',\n",
       " 'KIF5B',\n",
       " 'PDGFRA',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'FISH',\n",
       " 'RP11-460H18',\n",
       " 'RP11-167O6',\n",
       " 'KIF5B',\n",
       " 'KIF5B',\n",
       " 'RP11-460H18',\n",
       " 'BAC',\n",
       " 'RP11-24O10',\n",
       " 'PDGFRA',\n",
       " 'KIF5B',\n",
       " 'PDGFRA',\n",
       " 'BACs',\n",
       " 'PDGFRA',\n",
       " 'RP11-571I18',\n",
       " 'RP11-24O10',\n",
       " 'KIF5B',\n",
       " 'PDGFRA',\n",
       " 'FISH',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'KIF5B',\n",
       " 'FISH',\n",
       " 'KIF5B',\n",
       " 'KIF5B',\n",
       " 'KIF5B',\n",
       " 'PDGFRA',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'RT-PCR',\n",
       " 'DNA',\n",
       " 'PCR',\n",
       " 'IHES/CEL,2',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'IHES',\n",
       " 'USA',\n",
       " 'IHES',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'IHES',\n",
       " 'RT-PCR',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'IHES',\n",
       " 'PCR',\n",
       " 'PDGFRA',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'RT-PCR',\n",
       " 'CHIC2',\n",
       " 'FIP1L1-PDGFRA-positive',\n",
       " 'PDGFRA',\n",
       " 'IHES',\n",
       " 'RT-PCR',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'PCR',\n",
       " 'PCR',\n",
       " 'FIP1L1-PDGFRA-positive',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'KIF5B',\n",
       " 'PDGFRA',\n",
       " 'BCR',\n",
       " 'FIP1L1',\n",
       " 'KIF5B',\n",
       " 'ATP',\n",
       " 'KIF5B-PDGFRalpha',\n",
       " 'KIF5B',\n",
       " 'PDGFRalpha',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'PDGFRalpha',\n",
       " 'KIF5B',\n",
       " 'KIF5B-PDGFRalpha',\n",
       " 'PDGFRA',\n",
       " 'IHES',\n",
       " 'PDGFRA',\n",
       " 'FIP1L1-PDGFRA']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_ish_words1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='sec1_2'></a>\n",
    "## Some useful regex and lists (<a href='#sec0'>Back To Top</a>)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>Keep on adding words that are too common in 'commoners' list!</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "commoners = ['RT', 'PCR', 'RT-PCR', 'DNA', 'cDNA', 'RNA', 'mRNA', 'protein', 'cell', 'cancer', 'CHIP', 'FISH', 'UK', 'USA']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>Keep words with '-positive' and '-negative'</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['FIP1L1-PDGFRA-positive',\n",
       " 'FIP1L1-PDGFRA-positive',\n",
       " 'FIP1L1-PDGFRA-positive',\n",
       " 'FIP1L1-PDGFRA-positive',\n",
       " 'FIP1L1-PDGFRA-positive']"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[word for word in txt1_words if '-positive' in word.lower()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['FIP1L1-PDGFRA-negative', 'FIP1L1-PDGFRA-negative']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[word for word in txt1_words if '-negative' in word.lower()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>Detect gene fusions</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['BCR-ABL',\n",
       " 'STRN-PDGFRA',\n",
       " 'STRN-PDGFRA',\n",
       " 'BCR-ABL',\n",
       " 'STRN-PDGFRA',\n",
       " 'STRN-PDGFRA',\n",
       " 'STRN-PDGFRA',\n",
       " 'STRN-PDGFRA',\n",
       " 'PDGFRA-STRN',\n",
       " 'PDGFRA-ETV6',\n",
       " 'BCR-ABL',\n",
       " 'BCR-ABL',\n",
       " 'BCR-ABL',\n",
       " 'BCR-ABL',\n",
       " 'BCR-PDGFRA',\n",
       " 'BCR-ABL',\n",
       " 'PDGFRA-KIF5B']"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fuse_pattern = r\"[A-Z]{2,7}-[A-Z]{2,7}\"\n",
    "[word for word in txt1_words \\\n",
    " if word not in commoners \\\n",
    " if re.search(fuse_pattern, word)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>Detect genes with hyphens, including fusions</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Remove nucleotide sequences\n",
    "nts = r\"[ACTGU]{4,}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['BCR-ABL',\n",
       " 'STRN-PDGFRA',\n",
       " 'ETV6-PDGFRA',\n",
       " 'STRN-PDGFRA',\n",
       " 'ETV6-PDGFRA',\n",
       " 'BCR-ABL',\n",
       " 'STRN-PDGFRA',\n",
       " 'PDAI12-R4',\n",
       " 'STRN-PDGFRA',\n",
       " 'STRN-PDGFRA',\n",
       " 'STRN-PDGFRA',\n",
       " 'PDGFRA-STRN',\n",
       " 'ETV6-PDGFRA',\n",
       " 'PDGFRA-ETV6',\n",
       " 'ETV6-PDGFRA',\n",
       " 'ETV6-PDGFRA',\n",
       " 'ETV6-PDGFRA',\n",
       " 'ETV6-PDGFRA',\n",
       " 'ETV6-PDGFRA',\n",
       " 'ETV6-CHIC2',\n",
       " 'ETV6-PDGFRA',\n",
       " 'ETV6-PDGFRA',\n",
       " 'BCR-ABL',\n",
       " 'BCR-ABL',\n",
       " 'PDA-R1',\n",
       " 'PDA-R2',\n",
       " 'PDAI12-R3',\n",
       " 'PDAI12-R3',\n",
       " 'PDAI12-R4',\n",
       " 'BCR-ABL',\n",
       " 'PDAI12-R3',\n",
       " 'NVAMP-1',\n",
       " 'PDAI12-R4',\n",
       " 'NVAMP-2',\n",
       " 'BCR-ABL',\n",
       " 'BCR-PDGFRA',\n",
       " 'MG-63',\n",
       " 'MG-63',\n",
       " 'BCR-ABL',\n",
       " 'PDAI12-R4',\n",
       " 'PDAI12-R4',\n",
       " 'PDAI12-R3']"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_pattern1 = r\"([A-Z]{2,7})([0-9]{1,4})?-([A-Z]{0,7})([0-9]{1,4})?$\"\n",
    "[word for word in txt1_words \\\n",
    " if word not in commoners \\\n",
    " if re.search(gene_pattern1, word) \\\n",
    " if not re.search(nts, word)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>Mutation Types (previously explored)</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mutation_patterns = ['truncation', 'deletion', 'promoter', \n",
    "                     'amplification', 'epigenetic', 'frame', \n",
    "                     'overexpression', 'duplication', 'insertion',\n",
    "                     'subtype', 'fusion', 'splice', 'wildtype']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['fusion',\n",
       " 'amplification',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'amplification',\n",
       " 'frame',\n",
       " 'amplification',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'splice',\n",
       " 'splice',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'amplification',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'truncation',\n",
       " 'truncation',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'truncation',\n",
       " 'deletion',\n",
       " 'fusion',\n",
       " 'overexpression',\n",
       " 'fusion',\n",
       " 'overexpression',\n",
       " 'amplification',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'overexpression',\n",
       " 'fusion',\n",
       " 'deletion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'overexpression',\n",
       " 'fusion',\n",
       " 'amplification',\n",
       " 'fusion',\n",
       " 'amplification',\n",
       " 'amplification',\n",
       " 'overexpression',\n",
       " 'amplification',\n",
       " 'overexpression',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'amplification',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'amplification',\n",
       " 'overexpression',\n",
       " 'fusion',\n",
       " 'amplification',\n",
       " 'amplification',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'amplification',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'deletion',\n",
       " 'fusion',\n",
       " 'deletion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'overexpression',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'fusion',\n",
       " 'overexpression']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[word.lower() for word in txt1_words if word.lower() in mutation_patterns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='sec1_3'></a>\n",
    "## Replace periods, commas, hyphens, brackets with a space, and then tokenize (<a href='#sec0'>Back To Top</a>)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "initial length 6402\n"
     ]
    }
   ],
   "source": [
    "tokens = word_tokenize(txt1)\n",
    "tokens = np.array(tokens)\n",
    "print('initial length %d' % len(tokens))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "txt1_white = txt1.encode().decode()  # copy a string?!\n",
    "txt1_white = txt1_white.replace('\"', ' ')\n",
    "txt1_white = txt1_white.replace('.', ' ')\n",
    "txt1_white = txt1_white.replace('\\'', ' ')\n",
    "txt1_white = txt1_white.replace('_', ' ')    # This should be done after _pos, etc checked\n",
    "#txt1_white = txt1_white.replace('-', ' ')    # This should be done after -pos, etc checked\n",
    "txt1_white = txt1_white.replace('=', ' ')\n",
    "txt1_white = txt1_white.replace('\\n', ' ')\n",
    "txt1_white = txt1_white.replace('\\\\n', ' ')\n",
    "txt1_white = txt1_white.replace('\\'', ' ')\n",
    "txt1_white = re.sub(' +',' ', txt1_white) \n",
    "txt1_white = txt1_white.replace('\\'', ' ')\n",
    "txt1_white = txt1_white.replace('(', ' ')\n",
    "txt1_white = txt1_white.replace(')', ' ')\n",
    "txt1_white = txt1_white.replace('[', ' ')\n",
    "txt1_white = txt1_white.replace(']', ' ')\n",
    "txt1_white = txt1_white.replace('{', ' ')\n",
    "txt1_white = txt1_white.replace('}', ' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "initial length 5891\n"
     ]
    }
   ],
   "source": [
    "tokens_white = word_tokenize(txt1_white)\n",
    "tokens_white = np.array(tokens_white)\n",
    "print('initial length %d' % len(tokens_white))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After removing stop-words & commoners 3987\n"
     ]
    }
   ],
   "source": [
    "tokens_white = [word for word in tokens_white if not word in stop_words]\n",
    "tokens_white = [word for word in tokens_white if not word in commoners]\n",
    "print('After removing stop-words & commoners %d' % len(tokens_white))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After removing DNA/RNA seqs 3967\n"
     ]
    }
   ],
   "source": [
    "# Remove nucleotide sequences\n",
    "nts = r\"[ACTGU]{4,}\"\n",
    "tokens_white = [word for word in tokens_white if not re.search(nts, word)]\n",
    "print('After removing DNA/RNA seqs %d' % len(tokens_white))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This time switch to re.search t pickup pattern not just from the beginning of the words"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# of gene-ish words: 364\n"
     ]
    }
   ],
   "source": [
    "gene_ish_words_white = [word for word in tokens_white if re.search(gene_ish_pattern, word)]\n",
    "print('# of gene-ish words: %d' % len(gene_ish_words_white))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['BCR-ABL',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'STRN-PDGFRA',\n",
       " 'ETV6-PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'CEL',\n",
       " 'CEL',\n",
       " 'PDGFRA',\n",
       " 'PDGFRB',\n",
       " 'FGFR1',\n",
       " 'JAK2',\n",
       " 'PDGFRA',\n",
       " 'PDGFRB',\n",
       " 'PDGFRA',\n",
       " 'BCR',\n",
       " 'FIP1L1',\n",
       " 'KIF5B',\n",
       " 'CDK5RAP2',\n",
       " 'CEL',\n",
       " 'CEL',\n",
       " 'MPD',\n",
       " 'MPD',\n",
       " 'XY',\n",
       " 'XY',\n",
       " 'PDGFRA',\n",
       " 'STRN-PDGFRA',\n",
       " 'STRN',\n",
       " 'PDGFRA',\n",
       " 'ETV6-PDGFRA',\n",
       " 'ETV6',\n",
       " 'PDGFRA',\n",
       " 'MRD',\n",
       " 'BAC',\n",
       " 'PDGFRA',\n",
       " 'ETV6',\n",
       " 'BAC',\n",
       " 'BACs',\n",
       " 'RP11-24O10',\n",
       " 'PDGFRA',\n",
       " 'RP11-434C1',\n",
       " 'ETV6',\n",
       " 'PDGFRA',\n",
       " 'BCR-ABL',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'STRN-PDGFRA',\n",
       " 'HaeIII',\n",
       " 'PDAI12-R4',\n",
       " 'HaeIII',\n",
       " 'STRN-PDGFRA',\n",
       " 'STRN-PDGFRA',\n",
       " 'WW-like',\n",
       " 'WW',\n",
       " 'WD40',\n",
       " 'WW',\n",
       " 'ΔWW',\n",
       " 'STRN-PDGFRA',\n",
       " 'STRN',\n",
       " 'STRN',\n",
       " 'NM',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA-STRN',\n",
       " 'PDGFRA',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'AG',\n",
       " 'STRN',\n",
       " 'ETV6-PDGFRA',\n",
       " 'ETV6',\n",
       " 'NM',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA-ETV6',\n",
       " 'ETV6-PDGFRA',\n",
       " 'ETV6',\n",
       " 'PDGFRA',\n",
       " 'ETV6-PDGFRA',\n",
       " 'BACs',\n",
       " 'RP11-24O10',\n",
       " 'RP11-434C1',\n",
       " 'ETV6',\n",
       " 'PDGFRA',\n",
       " 'ETV6-PDGFRA',\n",
       " 'ETV6',\n",
       " 'PDGFRA',\n",
       " 'ETV6-PDGFRA',\n",
       " 'ETS',\n",
       " 'MRD',\n",
       " 'CEL',\n",
       " 'STRN',\n",
       " 'ETV6',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'ETV6',\n",
       " 'CEL',\n",
       " 'STRN',\n",
       " 'WD',\n",
       " 'WD',\n",
       " 'STRN',\n",
       " 'PDGFRA',\n",
       " 'WD',\n",
       " 'WW-like',\n",
       " 'ETV6',\n",
       " 'CHIC2/BTL',\n",
       " 'AML',\n",
       " 'PGFRA',\n",
       " 'CHIC2/BTL',\n",
       " 'ETV6-PDGFRA',\n",
       " 'ETV6-CHIC2',\n",
       " 'PDGFRA',\n",
       " 'STRN',\n",
       " 'WW-like',\n",
       " 'ETV6-PDGFRA',\n",
       " 'WW-like',\n",
       " 'WW-like',\n",
       " 'ETV6-PDGFRA',\n",
       " 'WW-like',\n",
       " 'WW-like',\n",
       " 'WW-like',\n",
       " 'WW-like',\n",
       " 'IHES',\n",
       " 'FIP1-Like-1',\n",
       " 'FIP1L1',\n",
       " 'PDGFRA',\n",
       " 'IHES',\n",
       " 'PDGFRA',\n",
       " 'transcriptase-PCR',\n",
       " 'PDGFRA',\n",
       " 'FIP1L1-PDGFRA-positive',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'KIF5B',\n",
       " 'PDGFRA',\n",
       " 'PDGFRalpha',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'IHES',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'KIF5B',\n",
       " 'HES',\n",
       " 'HES',\n",
       " 'IHES',\n",
       " 'CEL',\n",
       " 'CEL',\n",
       " 'FIP1-Like-1',\n",
       " 'FIP1L1',\n",
       " 'PDGFRA',\n",
       " 'PDGFRalpha',\n",
       " 'FIP1L1-PDGFRA-positive',\n",
       " 'CEL',\n",
       " 'PDGFRB',\n",
       " 'PDGFR',\n",
       " 'IHES',\n",
       " 'PDGFRB',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'transcriptase-PCR',\n",
       " 'FIP1L1',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'IHES',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'BCR-ABL',\n",
       " 'FIP1L1-PDGFRA-positive',\n",
       " 'FIP1L1-PDGFRA-negative',\n",
       " 'transcriptase-PCR',\n",
       " 'RNeasy',\n",
       " 'BCR-ABL',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'FIP1E6-F1',\n",
       " 'PDA-R1',\n",
       " 'FIP1E6-F2',\n",
       " 'PDA-R2',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'KIF5B-exon19F1',\n",
       " 'PDA-exon14R1',\n",
       " 'KIF5B-exon22-F1',\n",
       " 'PDAI12-R3',\n",
       " 'KIF5B-exon19F1',\n",
       " 'PDGFRA-exon14R1',\n",
       " 'KIF5B-exon19F2',\n",
       " 'PDGFRA-exon13/14R2',\n",
       " 'KIF5B-exon22-F1',\n",
       " 'PDAI12-R3',\n",
       " 'KIF5B',\n",
       " 'PDAI12-R4',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDA-exon15R',\n",
       " 'PDGFRA',\n",
       " 'PDA-exon10F',\n",
       " 'PDA-exon13F',\n",
       " 'PDA-exon10F',\n",
       " 'PDA-exon13F',\n",
       " 'PDA-exon15R',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'MG63',\n",
       " 'BCR-ABL',\n",
       " 'HaeIII',\n",
       " 'PDGFRA',\n",
       " 'PDAI12-R3',\n",
       " 'PDGFRA',\n",
       " 'NVAMP-1',\n",
       " 'PDAI12-R4',\n",
       " 'NVAMP-2',\n",
       " 'TOPO',\n",
       " 'PCR4',\n",
       " 'BAC',\n",
       " 'PDGFRA',\n",
       " 'KIF5B',\n",
       " 'HIV',\n",
       " 'TPHA',\n",
       " 'BCR-ABL',\n",
       " 'FIP1L1-PDGFRA',\n",
       " '46XY',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'BCR-PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'transcriptase-PCR',\n",
       " 'PDGFRA',\n",
       " 'PDA-exon10F',\n",
       " 'PDA-exon15R',\n",
       " 'PDGFRA',\n",
       " 'X-PDGFRA',\n",
       " 'PDA-exon13F',\n",
       " 'PDA-exon15R',\n",
       " 'PDGFRA',\n",
       " 'X-PDGFRA',\n",
       " 'MG-63',\n",
       " 'PDGFRA',\n",
       " 'FIP1-Like-1',\n",
       " 'FIP1L1',\n",
       " '-PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'MG-63',\n",
       " 'PDGFRA',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'mRNA-derived',\n",
       " 'PDGFRA',\n",
       " 'IHES',\n",
       " 'PDGFRA',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'FIP1L1-PDGFRA-negative',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'BCR-ABL',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'NCP',\n",
       " 'HaeIII',\n",
       " 'PDAI12-R4',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'HaeIII',\n",
       " 'PDAI12-R4',\n",
       " 'HaeIII',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'FIP1L1',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'KIF5B',\n",
       " 'NM',\n",
       " 'PDGFRA',\n",
       " 'KIF5B-exon22-F1',\n",
       " 'PDAI12-R3',\n",
       " 'PDGFRA-KIF5B',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'PB',\n",
       " 'BM',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'KIF5B-exon19F1',\n",
       " 'PDA-exon14R1',\n",
       " 'KIF5B',\n",
       " 'PDGFRA',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'RP11-460H18',\n",
       " 'RP11-167O6',\n",
       " 'KIF5B',\n",
       " 'KIF5B',\n",
       " 'RP11-460H18',\n",
       " 'BAC',\n",
       " 'RP11-24O10',\n",
       " 'PDGFRA',\n",
       " 'KIF5B',\n",
       " 'PDGFRA',\n",
       " 'BACs',\n",
       " 'PDGFRA',\n",
       " 'RP11-571I18',\n",
       " 'RP11-24O10',\n",
       " 'KIF5B',\n",
       " 'PDGFRA',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'KIF5B',\n",
       " 'KIF5B',\n",
       " 'KIF5B',\n",
       " 'KIF5B',\n",
       " 'PDGFRA',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'IHES/CEL,2',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'IHES',\n",
       " 'IHES',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'IHES',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'IHES',\n",
       " 'PDGFRA',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'CHIC2',\n",
       " 'FIP1L1-PDGFRA-positive',\n",
       " 'PDGFRA',\n",
       " 'IHES',\n",
       " 'FIP1L1-PDGFRA',\n",
       " 'FIP1L1-PDGFRA-positive',\n",
       " 'PDGFRA',\n",
       " 'PDGFRA',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'KIF5B',\n",
       " 'PDGFRA',\n",
       " 'BCR',\n",
       " 'FIP1L1',\n",
       " 'KIF5B',\n",
       " 'ATP',\n",
       " 'KIF5B-PDGFRalpha',\n",
       " 'KIF5B',\n",
       " 'PDGFRalpha',\n",
       " 'KIF5B-PDGFRA',\n",
       " 'PDGFRalpha',\n",
       " 'KIF5B',\n",
       " 'KIF5B-PDGFRalpha',\n",
       " 'PDGFRA',\n",
       " 'IHES',\n",
       " 'PDGFRA',\n",
       " 'FIP1L1-PDGFRA']"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_ish_words_white"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "364\n"
     ]
    }
   ],
   "source": [
    "df1 = pd.DataFrame(gene_ish_words_white)\n",
    "df1.columns = ['tokens']\n",
    "print(len(df1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "table1 = df1.groupby('tokens').size().reset_index()\n",
    "table1.columns = ['tokens', 'appearances']  # Use rename!!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tokens</th>\n",
       "      <th>appearances</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>PDGFRA</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>FIP1L1-PDGFRA</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>KIF5B</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>KIF5B-PDGFRA</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>IHES</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ETV6</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ETV6-PDGFRA</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>WW-like</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CEL</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>STRN</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>FIP1L1</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>BCR-ABL</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>STRN-PDGFRA</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>HaeIII</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>PDAI12-R4</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           tokens  appearances\n",
       "58         PDGFRA           91\n",
       "24  FIP1L1-PDGFRA           24\n",
       "33          KIF5B           19\n",
       "34   KIF5B-PDGFRA           14\n",
       "30           IHES           13\n",
       "16           ETV6           10\n",
       "18    ETV6-PDGFRA           10\n",
       "80        WW-like            9\n",
       "12            CEL            9\n",
       "73           STRN            8\n",
       "23         FIP1L1            7\n",
       "8         BCR-ABL            7\n",
       "74    STRN-PDGFRA            6\n",
       "29         HaeIII            6\n",
       "56      PDAI12-R4            5"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table1.sort_values('appearances', ascending=False).head(15)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='sec2'></a>\n",
    "# Surveying the whole text. Looking at each type of feature <br>(<a href='#sec0'>Back To Top</a>)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 74.1 ms, sys: 160 ms, total: 234 ms\n",
      "Wall time: 235 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "textome = ''\n",
    "for i in range(len(text_train)):\n",
    "    text = text_train.loc[i, 'Text'] + ''\n",
    "    textome += text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3321"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(textome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "211268641"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(textome)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>Words with '-positive' and '-negative'</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['FIP1L1-PDGFRA-positive',\n",
       " 'FIP1L1-PDGFRA-positive',\n",
       " 'FIP1L1-PDGFRA-positive',\n",
       " 'FIP1L1-PDGFRA-positive',\n",
       " 'FIP1L1-PDGFRA-positive']"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[word for word in txt1_words if '-positive' in word.lower()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3min 5s, sys: 596 ms, total: 3min 5s\n",
      "Wall time: 3min 6s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "textome_tokens = word_tokenize(textome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "initial leng 37460802\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "initial leng 37460802\n"
     ]
    }
   ],
   "source": [
    "print('initial leng %d' % len(textome_tokens))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2.41 s, sys: 50.9 ms, total: 2.46 s\n",
      "Wall time: 2.46 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "stop_words = set(stopwords.words('english'))\n",
    "textome_tokens = [token for token in textome_tokens if not token in stop_words]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After removing stop words 27557071\n"
     ]
    }
   ],
   "source": [
    "print('After removing stop words %d' % len(textome_tokens))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove (E)GFP tags"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>Words with '-positive' and '-negative'</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3.95 s, sys: 1e+03 ns, total: 3.95 s\n",
      "Wall time: 3.95 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Remove (E)GFP taggs\n",
    "for i, word in enumerate(textome_tokens):\n",
    "    if 'EGFP/EYFP' in word:\n",
    "        textome_tokens[i] = word.replace('EGFP/EYFP' + '-', '')\n",
    "    if 'EGFP' in word:\n",
    "        textome_tokens[i] = word.replace('EGFP' + '-', '')\n",
    "    if 'GFP' in word:\n",
    "        textome_tokens[i] = word.replace('GFP' + '-', '')\n",
    "    if 'GST' in word:\n",
    "        textome_tokens[i] = word.replace('GST' + '-', '')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='sec2_1'></a>\n",
    "## Words with '-positive' and '-negative' (<a href='#sec0'>Back To Top</a>)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 4.62 s, sys: 3.01 ms, total: 4.62 s\n",
      "Wall time: 4.62 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "pos_neg = [token.lower() for token in textome_tokens \\\n",
    "           if (('-positive' in token.lower()) | ('-negative' in token.lower())) \\\n",
    "           if not (('false' in token.lower()) | ('true' in token.lower()))]\n",
    "\n",
    "for i, word in enumerate(pos_neg):\n",
    "    if 'egfp/eyfp' in word: \n",
    "        pos_neg[i] = word.replace('egfp/eyfp-', '')\n",
    "    elif 'egfp' in word:\n",
    "        pos_neg[i] = word.replace('egfp-', '')\n",
    "    elif 'gfp' in word:\n",
    "        pos_neg[i] = word.replace('gfp'+\"–\", '') #This hyphen is somehow different from the other one\n",
    "\n",
    "pos_neg = [token for token in pos_neg \\\n",
    "           if not ((token == '-positive') | (token == '-negative'))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "gfp_pattern = r\"e?gfp-\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "textome_tokens = re.sub(gfp_pattern, '', textome_tokens) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "word = 'gfp-positive'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'positive'"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "word.replace('gfp'+'-', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "word = 'gfp–vhl213-positive'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'gfpvhl213-positive'"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "word.replace('–', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['erα-positive',\n",
       " 'cbl-positive',\n",
       " 'mutation-positive',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'phospho-stat5-positive',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-positive',\n",
       " 'bcr-abl1-negative',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-negative',\n",
       " 'v617jak2-positive',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-positive',\n",
       " 'v-positive',\n",
       " 'v-positive',\n",
       " 'dominant-negative',\n",
       " 'v-positive',\n",
       " 'v-positive',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'mutation-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'mutation-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'cbl-positive',\n",
       " 'mutation-positive',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'mutation-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'mutation-negative',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-positive',\n",
       " 'bcr-abl1-negative',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-negative',\n",
       " 'v617jak2-positive',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-negative',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-positive',\n",
       " 'v617fjak2-positive',\n",
       " '11q-aupd-positive',\n",
       " 'aupd-positive',\n",
       " 'cd34-negative',\n",
       " 'cd34-negative',\n",
       " 'double-positive',\n",
       " 'dominant-negative',\n",
       " '11q-aupd-positive',\n",
       " 'aupd-positive',\n",
       " 'cd34-negative',\n",
       " 'cd34-negative',\n",
       " 'double-positive',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'cbl-positive',\n",
       " 'mutation-positive',\n",
       " 'cbl-positive',\n",
       " 'mutation-positive',\n",
       " 'cbl-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-negative',\n",
       " 'mutation-positive',\n",
       " 'p53-negative',\n",
       " 'p53-negative',\n",
       " 'telomerase-negative',\n",
       " 'telomerase-positive',\n",
       " 'dominant-negative',\n",
       " 'telomerase-negative',\n",
       " 'telomerase-positive',\n",
       " 'telomerase-negative',\n",
       " 'telomerase-positive',\n",
       " 'mcv-positive',\n",
       " 'mcv-negative',\n",
       " 'mcv-positive',\n",
       " 'mcv-negative',\n",
       " 'mcv-negative',\n",
       " 'mcv-negative',\n",
       " 'mcv-negative',\n",
       " 'virus-positive',\n",
       " 'mcv-positive',\n",
       " 'mcv-negative',\n",
       " 'mcv-positive',\n",
       " 'mcv-negative',\n",
       " 'c250t-positive',\n",
       " 'c250t-positive',\n",
       " 'c250t-positive',\n",
       " 'c250t-positive',\n",
       " 'p53-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-negative',\n",
       " 'mutation-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-negative',\n",
       " 'mutation-negative',\n",
       " 'triple-negative',\n",
       " 'lineage-positive',\n",
       " 'lineage-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-negative',\n",
       " 'mutation-negative',\n",
       " 'lysm-negative',\n",
       " 'lineage-positive',\n",
       " 'lineage-negative',\n",
       " 'dominant-negative',\n",
       " 'methylation-positive',\n",
       " 'methylation-positive',\n",
       " 'methylation-positive',\n",
       " 'methylation-negative',\n",
       " 'methylation-negative',\n",
       " 'hpv-positive',\n",
       " 'hpv-negative',\n",
       " 'dominant-negative',\n",
       " 'mutation-negative',\n",
       " 'mutation-negative',\n",
       " 'stat-negative',\n",
       " 'stat-mutation-negative',\n",
       " 'mutation-negative',\n",
       " 'mutation-negative',\n",
       " 'double-negative',\n",
       " 'mutation-negative',\n",
       " 'mutation-negative',\n",
       " 'mutation-negative',\n",
       " 'dominant-negative',\n",
       " 'hpv-positive',\n",
       " 'hpv-negative',\n",
       " 'dominant-negative',\n",
       " 'hpv-positive',\n",
       " 'hpv-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'fusion-positive',\n",
       " 'fusion-positive',\n",
       " 'fusion-positive',\n",
       " 'fusion-positive',\n",
       " 'dominant-negative',\n",
       " 'mutation-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-negative',\n",
       " 'mutation-positive',\n",
       " 'dominant-negative',\n",
       " 'γ-h2ax-positive',\n",
       " 'dominant-negative',\n",
       " 'fish-positive',\n",
       " 'ki67-positive',\n",
       " 'fish-positive',\n",
       " 'ki67-positive',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'fish-positive',\n",
       " 'ki67-positive',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'fish-positive',\n",
       " 'ki67-positive',\n",
       " 'oncogene-negative',\n",
       " 'oncogene-negative',\n",
       " 'oncogene-negative',\n",
       " 'oncogene-negative',\n",
       " 'oncogene-negative',\n",
       " 'oncogene-negative',\n",
       " 'oncogene-positive',\n",
       " 'oncogene-negative',\n",
       " 'oncogene-positive',\n",
       " 'oncogene-negative',\n",
       " 'oncogene-negative',\n",
       " 'oncogene-negative',\n",
       " 'oncogene-negative',\n",
       " 'oncogene-negative',\n",
       " 'oncogene-negative',\n",
       " 'oncogene-positive',\n",
       " 'oncogene-negative',\n",
       " 'oncogene-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'rit1-mutation-positive',\n",
       " 'rit1-mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'oncogene-negative',\n",
       " 'oncogene-negative',\n",
       " 'oncogene-negative',\n",
       " 'oncogene-negative',\n",
       " 'oncogene-negative',\n",
       " 'oncogene-negative',\n",
       " 'oncogene-positive',\n",
       " 'oncogene-negative',\n",
       " 'oncogene-positive',\n",
       " 'ihc-positive',\n",
       " 'ihc-positive',\n",
       " 'ihc-positive',\n",
       " 'e-positive',\n",
       " 'cdkn1b-negative',\n",
       " 'ihc-positive',\n",
       " 'ihc-positive',\n",
       " 'ihc-positive',\n",
       " 'e-positive',\n",
       " 'cdkn1b-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negatives',\n",
       " 'dominant-negative',\n",
       " 'tgf-β1-negative',\n",
       " 'tgf-β1-positive',\n",
       " 'cancer-negative',\n",
       " 'receptor-positive',\n",
       " 'er-positive',\n",
       " 'shh-positive',\n",
       " 'hydrochloride-positive',\n",
       " 'bethesda-positive',\n",
       " 'bethesda-negative',\n",
       " 'bethesda-positive',\n",
       " 'bethesda-positive',\n",
       " 'bethesda-positive',\n",
       " 'bethesda-negative',\n",
       " 'bethesda-positive',\n",
       " 'bethesda-positive',\n",
       " 'bethesda-negative',\n",
       " 'bethesda-positive',\n",
       " 'bethesda-negative',\n",
       " 'bethesda-positive',\n",
       " 'hmsh6-negative',\n",
       " 'bethesda-positive',\n",
       " 'cimp-positive',\n",
       " 'msi-positive',\n",
       " 'mutation-positive',\n",
       " 'mut-positive',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'msi-positive',\n",
       " 'pbrm1-positive',\n",
       " 'pbrm1-positive',\n",
       " 'pbrm1-positive',\n",
       " 'pbrm1-negative',\n",
       " 'pbrm1-positive',\n",
       " 'pbrm1-negative',\n",
       " 'pbrm1-negative',\n",
       " 'pbrm1-positive',\n",
       " 'pbrm1-positive',\n",
       " 'pbrm1-negative',\n",
       " 'pbrm1-positive',\n",
       " 'pbrm1-positive',\n",
       " 'pbrm1-positive',\n",
       " 'pbrm1-negative',\n",
       " 'pbrm1-positive',\n",
       " 'pbrm1-negative',\n",
       " 'pbrm1-negative',\n",
       " 'pbrm1-positive',\n",
       " 'pbrm1-positive',\n",
       " 'pbrm1-negative',\n",
       " 'mutant-positive',\n",
       " 'protein-positive',\n",
       " 'mutant-positive',\n",
       " 'protein-positive',\n",
       " 'v-positive',\n",
       " 'dominant-negative',\n",
       " 'v-positive',\n",
       " 'cd34-negative',\n",
       " 'protein-positive',\n",
       " 'v-positive',\n",
       " 'dominant-negative',\n",
       " 'v-positive',\n",
       " 'cd34-negative',\n",
       " 'protein-positive',\n",
       " 'v-positive',\n",
       " 'dominant-negative',\n",
       " 'v-positive',\n",
       " 'cd34-negative',\n",
       " 'mutant-positive',\n",
       " 'mutant-positive',\n",
       " 'mutant-positive',\n",
       " 'protein-positive',\n",
       " 'v-positive',\n",
       " 'dominant-negative',\n",
       " 'v-positive',\n",
       " 'cd34-negative',\n",
       " 'mutant-positive',\n",
       " 'protein-positive',\n",
       " 'v-positive',\n",
       " 'dominant-negative',\n",
       " 'v-positive',\n",
       " 'cd34-negative',\n",
       " 'er-positive',\n",
       " 'er-positive',\n",
       " 'er-positive',\n",
       " 'er-positive',\n",
       " 'er-negative',\n",
       " 'er-negative',\n",
       " 'er-positive',\n",
       " 'er-negative',\n",
       " 'er-positive',\n",
       " 'er-positive',\n",
       " 'er-positive',\n",
       " 'er-positive',\n",
       " 'er-positive',\n",
       " 'er-positive',\n",
       " 'er-positive',\n",
       " 'er-negative',\n",
       " 'mutant-positive',\n",
       " 'γ-h2ax-positive',\n",
       " 'cd3-positive',\n",
       " 'calcein-am-positive',\n",
       " 'calcein-am-positive',\n",
       " 'b-positive',\n",
       " 'calcein-am-positive',\n",
       " 'γ-h2ax-positive',\n",
       " 'γ-h2ax-positive',\n",
       " 'protein-positive',\n",
       " 'cd34-positive',\n",
       " 'cd34-positive',\n",
       " 'bcr/abl1-negative',\n",
       " 'cd38-positive',\n",
       " 'cd34-positive',\n",
       " 'cd34-positive',\n",
       " 'cre-positive',\n",
       " 'mx1-cre-negative',\n",
       " 'stain-positive',\n",
       " 'v-positive',\n",
       " 'dominant-negative',\n",
       " 'v-positive',\n",
       " 'cd34-negative',\n",
       " 'cd34-positive',\n",
       " 'cd34-positive',\n",
       " 'bcr/abl1-negative',\n",
       " 'cd38-positive',\n",
       " 'cd34-positive',\n",
       " 'cd34-positive',\n",
       " 'merlin-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutant-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'fish-positive',\n",
       " 'fish-negative',\n",
       " 'fish-positive',\n",
       " 'fish-negative',\n",
       " 'fish-negative',\n",
       " 'fish-positive',\n",
       " 'fish-positive',\n",
       " 'fish-negative',\n",
       " 'fish-positive',\n",
       " 'hybridization-positive',\n",
       " 'fish-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'driver-negative',\n",
       " 'her2-positive',\n",
       " 'her2-positive',\n",
       " 'her2-positive',\n",
       " 'her2-positive',\n",
       " 'her2-positive',\n",
       " 'her2-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'p-akt1-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutant-positive',\n",
       " 'akp-positive',\n",
       " 'akp-positive',\n",
       " 'akp-negative',\n",
       " '10-67-positive',\n",
       " 'tunel-positive',\n",
       " 'egfrviii-positive',\n",
       " 'dominant-negative',\n",
       " 'egfrviii-positive',\n",
       " 'egfr-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'bcr-abl-positive',\n",
       " 'fish-positive',\n",
       " 'fish-negative',\n",
       " 'fish-positive',\n",
       " 'fish-negative',\n",
       " 'fish-negative',\n",
       " 'fish-positive',\n",
       " 'fish-positive',\n",
       " 'fish-negative',\n",
       " 'fish-positive',\n",
       " 'hybridization-positive',\n",
       " 'fish-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutant-positive',\n",
       " 'pakt-negative',\n",
       " 'pakt-positive',\n",
       " 'pakt-positive',\n",
       " 'perk-negative',\n",
       " 'perk-positive',\n",
       " 'perk-positive',\n",
       " 'perk-negative',\n",
       " 'perk-positive',\n",
       " 'pakt-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutant-positive',\n",
       " 'mutation-positive',\n",
       " 'fish-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'fish-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'marker-positive',\n",
       " 'marker-negative',\n",
       " 'mutation-positive',\n",
       " 'fish-positive',\n",
       " 'fish-negative',\n",
       " 'fish-positive',\n",
       " 'fish-negative',\n",
       " 'fish-negative',\n",
       " 'fish-positive',\n",
       " 'fish-positive',\n",
       " 'fish-negative',\n",
       " 'fish-positive',\n",
       " 'hybridization-positive',\n",
       " 'fish-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutant-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'fish-positive',\n",
       " 'fish-negative',\n",
       " 'fish-positive',\n",
       " 'fish-negative',\n",
       " 'fish-negative',\n",
       " 'fish-positive',\n",
       " 'fish-positive',\n",
       " 'fish-negative',\n",
       " 'fish-positive',\n",
       " 'hybridization-positive',\n",
       " 'fish-positive',\n",
       " 'aberration-positive',\n",
       " 'bcr-abl-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'alk-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'aberration-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'py1173-egfr-positive',\n",
       " 'egfr-negative',\n",
       " 'p-egfr-positive',\n",
       " 'py1173-egfr-positive',\n",
       " 'her2-positive',\n",
       " 'her2-positive',\n",
       " 'her2-positive',\n",
       " 'her2-positive',\n",
       " 'her2-positive',\n",
       " 'her2-positive',\n",
       " 'mutation-positive',\n",
       " 'driver-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutant-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutant-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutant-positive',\n",
       " 'mutation-positive',\n",
       " 'fish-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'fish-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'marker-positive',\n",
       " 'marker-negative',\n",
       " 'mutation-positive',\n",
       " 'fish-positive',\n",
       " 'fish-negative',\n",
       " 'fish-positive',\n",
       " 'fish-negative',\n",
       " 'fish-negative',\n",
       " 'fish-positive',\n",
       " 'fish-positive',\n",
       " 'fish-negative',\n",
       " 'fish-positive',\n",
       " 'hybridization-positive',\n",
       " 'fish-positive',\n",
       " 'her2-positive',\n",
       " 'her2-negative',\n",
       " 'her3-positive',\n",
       " 'her3-positive',\n",
       " 'her3-negative',\n",
       " 'her2-positive/her3-',\n",
       " 'her2-positive/her3-positive',\n",
       " 'egfr-mutated/her2-positive',\n",
       " 'her2-positive/her3-positive',\n",
       " 'negative/her3-negative',\n",
       " 'her2-positive',\n",
       " 'her3-positive',\n",
       " 'her3-negative',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutant-positive',\n",
       " 'fish-positive',\n",
       " 'fish-negative',\n",
       " 'fish-positive',\n",
       " 'fish-negative',\n",
       " 'fish-negative',\n",
       " 'fish-positive',\n",
       " 'fish-positive',\n",
       " 'fish-negative',\n",
       " 'fish-positive',\n",
       " 'hybridization-positive',\n",
       " 'fish-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 's768i-positive',\n",
       " 'protein-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutant-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutant-positive',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'pc9-luciferase-negative',\n",
       " 'oncogene-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutant-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutant-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutant-positive',\n",
       " 'mutation-positive',\n",
       " 'bcr-abl-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'protein-positive',\n",
       " 'mutation-positive',\n",
       " 'driver-negative',\n",
       " 'mutation-positive',\n",
       " 'driver-negative',\n",
       " 'dominant-negative',\n",
       " 'hpv-positive',\n",
       " 'hpv-positive',\n",
       " 'hpv-positive',\n",
       " 'hpv-positive',\n",
       " 'hpv-positive',\n",
       " 'hpv-positive',\n",
       " 'hpv-positive',\n",
       " 'hpv-positive',\n",
       " 'hpv-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutant-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutant-positive',\n",
       " 'fish-positive',\n",
       " 'fish-negative',\n",
       " 'fish-positive',\n",
       " 'fish-negative',\n",
       " 'fish-negative',\n",
       " 'fish-positive',\n",
       " 'fish-positive',\n",
       " 'fish-negative',\n",
       " 'fish-positive',\n",
       " 'hybridization-positive',\n",
       " 'fish-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutant-positive',\n",
       " 'her2-positive',\n",
       " 'her2-negative',\n",
       " 'her3-positive',\n",
       " 'her3-positive',\n",
       " 'her3-negative',\n",
       " 'her2-positive/her3-',\n",
       " 'her2-positive/her3-positive',\n",
       " 'egfr-mutated/her2-positive',\n",
       " 'her2-positive/her3-positive',\n",
       " 'negative/her3-negative',\n",
       " 'her2-positive',\n",
       " 'her3-positive',\n",
       " 'her3-negative',\n",
       " 't790m-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'egfrviii-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutant-positive',\n",
       " 'dominant-negative',\n",
       " 'dominant-negative',\n",
       " 'cd31-positive',\n",
       " 't790m-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'dominant-negative',\n",
       " 'mutation-positive',\n",
       " 'driver-negative',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'pan-negative',\n",
       " 'pan-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutant-positive',\n",
       " 'dominant-negative',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'protein-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'mutation-positive',\n",
       " 'driver-negative',\n",
       " 'fish-positive',\n",
       " 'fish-positive',\n",
       " 'her2-positive',\n",
       " 'her2-positive',\n",
       " 'triple-negative',\n",
       " ...]"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pos_neg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_pos_neg = pd.DataFrame(pos_neg, columns=['pos_neg'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pos_neg\n",
       "dominant-negative    1853\n",
       "mutation-positive     588\n",
       "alk-positive          566\n",
       "her2-positive         277\n",
       "mutation-negative     265\n",
       "pan-negative          255\n",
       "fusion-positive       253\n",
       "dsred-positive        216\n",
       "er-positive           191\n",
       "triple-negative       191\n",
       "her2-negative         126\n",
       "fish-positive         101\n",
       "ros1-positive          98\n",
       "double-positive        97\n",
       "cd117-positive         76\n",
       "dtype: int64"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pos_neg.groupby('pos_neg').size().sort_values(ascending=False).head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([], dtype=object)"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pos_neg[df_pos_neg['pos_neg'].str.contains('gfp')]['pos_neg'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='sec2_2'></a>\n",
    "## Detect genes with hyphens, including fusions (<a href='#sec0'>Back To Top</a>)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "commoners = ['RT', 'PCR', 'RT-PCR', 'DNA', 'cDNA', 'RNA', 'mRNA', 'siRNA', 'shRNA', 'protein', \n",
    "             'cell', 'cancer', 'CHIP', 'FISH', 'SDS-PAGE', 'UK', 'USA', 'GST', 'GFP', 'SDS', 'PAGE',\n",
    "             'qPCR', 'PBS', 'TBS', 'DTT', 'BSA', 'HSA', 'HCl', 'NCBI', 'PBST', 'ANOVA', 'RIKEN',\n",
    "             'COHORT', 'OUTCOME', 'AIRWAY', 'EMSA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "word = 'gfp-positive'dd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'positive'"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "word.replace('gfp'+'-', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[word for word in pos_neg if 'gfp' in word]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 21 s, sys: 8.96 ms, total: 21 s\n",
      "Wall time: 21 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Remove nucleotide sequences\n",
    "nts = r\"[ACTGU]{4,}\"\n",
    "\n",
    "# fusion-like pattern\n",
    "fusion_like_pattern = r\"([A-Z]{2,7})([0-9]{1,4})?-([A-Z]{0,7})([0-9]{1,4})?$\"\n",
    "\n",
    "fusion_like = [word.lower() for word in textome_tokens \\\n",
    "               if word not in commoners \\\n",
    "               if re.search(fusion_like_pattern, word) \\\n",
    "               if not re.search(nts, word)]\n",
    "\n",
    "for i, word in enumerate(fusion_like):\n",
    "    if 'gst-' in word: \n",
    "        fusion_like[i] = word.replace('gst-', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['pjg4-5',\n",
       " 'sc-2033',\n",
       " 'cdk10-',\n",
       " 'cbl-3',\n",
       " 'rpmi-1640',\n",
       " 'exosap-it',\n",
       " 'np-40',\n",
       " 'pca-96',\n",
       " 'flt-1',\n",
       " 'cbl-3',\n",
       " 'rpmi-1640',\n",
       " 'exosap-it',\n",
       " 'np-40',\n",
       " 'pca-96',\n",
       " 'flt-1',\n",
       " 'bcr-abl',\n",
       " 'bcr-abl',\n",
       " 'pcmv-ha',\n",
       " 'bcr-abl',\n",
       " 'il-3',\n",
       " 'il-3',\n",
       " 'ns-1',\n",
       " 'nup98-hoxd13',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-s',\n",
       " 'mm-pbsa',\n",
       " 'cbl-s',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'ring-tkbd',\n",
       " 'cbl-e2',\n",
       " 'ring-tkbd',\n",
       " 'cbl-e2',\n",
       " 'cbl-e3',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-s',\n",
       " 'mm-pbsa',\n",
       " 'cbl-s',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'ring-tkbd',\n",
       " 'cbl-e2',\n",
       " 'ring-tkbd',\n",
       " 'cbl-e2',\n",
       " 'cbl-e3',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-s',\n",
       " 'mm-pbsa',\n",
       " 'cbl-s',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'ring-tkbd',\n",
       " 'cbl-e2',\n",
       " 'ring-tkbd',\n",
       " 'cbl-e2',\n",
       " 'cbl-e3',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'flt3-wt',\n",
       " 'flt3-wt',\n",
       " 'cbl-wt',\n",
       " 'rpmi-1640',\n",
       " 'rpmi-1640',\n",
       " 'rpmi-1640',\n",
       " 'il-3',\n",
       " 'flt3-wt',\n",
       " 'cbl-wt',\n",
       " 'cbl-',\n",
       " 'cbl-wt',\n",
       " 'flt3-wt',\n",
       " 'mscv-ires-eyfp',\n",
       " 'mscvires-egfp',\n",
       " 'v5-flt3-wt',\n",
       " 'wst-1',\n",
       " 'tcs-sp2',\n",
       " 'molm-13',\n",
       " 'molm-14',\n",
       " 'cbl-wt',\n",
       " 'cbl-wt',\n",
       " 'cbl-wt',\n",
       " 'molm-13',\n",
       " 'flt3-wt',\n",
       " 'flt3-wt',\n",
       " 'flt3-wt',\n",
       " 'flt3-wt',\n",
       " 'cbl-wt',\n",
       " 'cbl-wt',\n",
       " 'cbl-wt',\n",
       " 'flt3-wt',\n",
       " 'flt3-wt',\n",
       " 'cbl-wt',\n",
       " 'cbl-wt',\n",
       " 'cbl-wt',\n",
       " 'molm-13',\n",
       " 'flt3-wt',\n",
       " 'flt3-wt',\n",
       " 'flt3-wt',\n",
       " 'cbl-wt',\n",
       " 'flt3-wt',\n",
       " 'mscv-yfp',\n",
       " 'flt3-',\n",
       " 'cbl-wt',\n",
       " 'flt3-wt',\n",
       " 'flt3-wt',\n",
       " 'il-3',\n",
       " 'il-3',\n",
       " 'flt3-wt',\n",
       " 'flt3-wt',\n",
       " 'flt3-wt',\n",
       " 'flt3-wt',\n",
       " 'flt3-wt',\n",
       " 'cbl-wt',\n",
       " 'flt3-wt',\n",
       " 'flt3-wt',\n",
       " 'il-3',\n",
       " 'flt3-wt',\n",
       " 'flt3-wt',\n",
       " 'flt3-',\n",
       " 'flt3-wt-',\n",
       " 'flt3-wt/cbl-wt',\n",
       " 'flt3-wt',\n",
       " 'cbl-wt',\n",
       " 'cbl-wt',\n",
       " 'cbl-wt',\n",
       " 'flt3-wt',\n",
       " 'flt3-wt/cbl-wt',\n",
       " 'flt3-wt-',\n",
       " 'flt3-wt/cbl-wt',\n",
       " 'flt3-wt/cbl-wt',\n",
       " 'flt3-',\n",
       " 'il-3',\n",
       " 'flt3-wt',\n",
       " 'cbl-wt',\n",
       " 'cbl-wt',\n",
       " 'flt3-wt',\n",
       " 'flt3-wt',\n",
       " 'cbl-wt',\n",
       " 'cbl-',\n",
       " 'molm-13',\n",
       " 'molm-14',\n",
       " 'flt3-wt',\n",
       " 'expressingflt3-wt',\n",
       " 'flt3-kd',\n",
       " 'flt3-wt',\n",
       " 'flt3-wt',\n",
       " 'cbl-wt',\n",
       " 'flt3-wt',\n",
       " 'cbl-wt',\n",
       " 'il-3',\n",
       " 'cbl-',\n",
       " 'flt3-wt',\n",
       " 'il-3',\n",
       " 'il-3',\n",
       " 'flt3-wt',\n",
       " 'cbl-wt',\n",
       " 'flt3-wt',\n",
       " 'flt3-itd',\n",
       " 'flt3-tkd',\n",
       " 'flt3-jm-pm',\n",
       " 'snp-a',\n",
       " 'npm-1',\n",
       " 'snp-a',\n",
       " 'snp-a',\n",
       " 'snp-a',\n",
       " 'snp-a',\n",
       " 'snp-a',\n",
       " 'snp-a',\n",
       " 'snp-a',\n",
       " 'snp-a',\n",
       " 'cml-bc',\n",
       " 'snp-a',\n",
       " 'snp-a',\n",
       " 'snp-a',\n",
       " 'cml-bc',\n",
       " 'cml-bc',\n",
       " 'snp-a',\n",
       " 'csf-1',\n",
       " 'ra/rars/rcmd/rcmd-rs/5q/mds-u',\n",
       " 'rcmd-rs',\n",
       " 'jak-stat',\n",
       " 'flt3-itd',\n",
       " 'bcr-abl1',\n",
       " 'arms-pcr',\n",
       " 'uke-1',\n",
       " 'set-2',\n",
       " 'pcmv6-ac-gfp',\n",
       " 'pcmv6-ac-rfp',\n",
       " 'pcmv6-ac',\n",
       " 'set-2',\n",
       " 'uke-1',\n",
       " 'pcmv6-ac-gfp',\n",
       " 'set-2',\n",
       " 'uke-1',\n",
       " 'set-2',\n",
       " 'uke-1',\n",
       " 'gm-csf',\n",
       " 'nf-1',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csr',\n",
       " 'gm-csfr',\n",
       " 'il-3',\n",
       " 'il-5',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'gm-csf·gm-csfr',\n",
       " 'α·gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'jak-stat',\n",
       " 'gm-csf',\n",
       " 'bcr-abl',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'il-3',\n",
       " 'gm-csf',\n",
       " 'il-5',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'pmscv-ha-cbl',\n",
       " 'pvsv-g',\n",
       " 'pmscv-gfp',\n",
       " 'vsv-g',\n",
       " 'pmscv-ires-ha-cbl',\n",
       " 'vsv-g',\n",
       " 'pmscv-ha-cbl',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'vsv-g',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'ha-cbl',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'anti-il-3/il-5/gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'il-3/il-5/gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'mg-132',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'cbl-wt',\n",
       " 'mg-132',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'cbl-',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'cbl-wt',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'pdgf-r',\n",
       " 'flt-3',\n",
       " 'il-3',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'cbl-',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'wt-cbl',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'cbl-',\n",
       " 'gm-csfr',\n",
       " 'il-5',\n",
       " 'il-3/il-5/gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'il-5',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'pi3-k',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'gm-csfr',\n",
       " 'egf-r',\n",
       " 'epo-r',\n",
       " 'ba/f3-epo-r',\n",
       " 'cbl-y371',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-s',\n",
       " 'mm-pbsa',\n",
       " 'cbl-s',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'ring-tkbd',\n",
       " 'cbl-e2',\n",
       " 'ring-tkbd',\n",
       " 'cbl-e2',\n",
       " 'cbl-e3',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'gm-csf',\n",
       " 'nf-1',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csr',\n",
       " 'gm-csfr',\n",
       " 'il-3',\n",
       " 'il-5',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'gm-csf·gm-csfr',\n",
       " 'α·gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'jak-stat',\n",
       " 'gm-csf',\n",
       " 'bcr-abl',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'il-3',\n",
       " 'gm-csf',\n",
       " 'il-5',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'pmscv-ha-cbl',\n",
       " 'pvsv-g',\n",
       " 'pmscv-gfp',\n",
       " 'vsv-g',\n",
       " 'pmscv-ires-ha-cbl',\n",
       " 'vsv-g',\n",
       " 'pmscv-ha-cbl',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'vsv-g',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'ha-cbl',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'anti-il-3/il-5/gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'il-3/il-5/gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'mg-132',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'cbl-wt',\n",
       " 'mg-132',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'cbl-',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'cbl-wt',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'pdgf-r',\n",
       " 'flt-3',\n",
       " 'il-3',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'cbl-',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'wt-cbl',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'tf-1',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'cbl-',\n",
       " 'gm-csfr',\n",
       " 'il-5',\n",
       " 'il-3/il-5/gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'il-5',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'pi3-k',\n",
       " 'gm-csf',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'tf-1',\n",
       " 'gm-csfr',\n",
       " 'tf-1',\n",
       " 'gm-csfr',\n",
       " 'egf-r',\n",
       " 'epo-r',\n",
       " 'ba/f3-epo-r',\n",
       " 'cbl-y371',\n",
       " 'tf-1',\n",
       " 'gm-csf',\n",
       " 'gm-csfr',\n",
       " 'gm-csfr',\n",
       " 'gm-csf',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-s',\n",
       " 'mm-pbsa',\n",
       " 'cbl-s',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'ring-tkbd',\n",
       " 'cbl-e2',\n",
       " 'ring-tkbd',\n",
       " 'cbl-e2',\n",
       " 'cbl-e3',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ap-2',\n",
       " 'egfr-ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'pcdna3-dest',\n",
       " 'pcdna3-egfr',\n",
       " 'cos-7',\n",
       " 'pcdna3-dest-cbl',\n",
       " 'anti-ha-hrp',\n",
       " 'cos-7',\n",
       " 'sulfo-nhs-ss',\n",
       " 'cos-7',\n",
       " 'cos-7',\n",
       " 'cos-7',\n",
       " 'cos-7',\n",
       " 'cos-7',\n",
       " 'cos-7',\n",
       " 'cos-7',\n",
       " 'cos-7',\n",
       " 'cos-7',\n",
       " 'cos-7',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'cos-7',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'cos-7',\n",
       " 'ras-mapk',\n",
       " 'raf-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'egfr-ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-s',\n",
       " 'mm-pbsa',\n",
       " 'cbl-s',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-s',\n",
       " 'cbl-s',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'ring-tkbd',\n",
       " 'cbl-e2',\n",
       " 'ring-tkbd',\n",
       " 'cbl-e2',\n",
       " 'cbl-e3',\n",
       " 'cbl-e2',\n",
       " 'cbl-e2',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ap-2',\n",
       " 'egfr-ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'pcdna3-dest',\n",
       " 'pcdna3-egfr',\n",
       " 'cos-7',\n",
       " 'pcdna3-dest-cbl',\n",
       " 'anti-ha-hrp',\n",
       " 'cos-7',\n",
       " 'sulfo-nhs-ss',\n",
       " 'cos-7',\n",
       " 'cos-7',\n",
       " 'cos-7',\n",
       " 'cos-7',\n",
       " 'cos-7',\n",
       " 'cos-7',\n",
       " 'cos-7',\n",
       " 'cos-7',\n",
       " 'cos-7',\n",
       " 'cos-7',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'cos-7',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'cos-7',\n",
       " 'ras-mapk',\n",
       " 'raf-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'egfr-ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'ras-mapk',\n",
       " 'shp-2',\n",
       " 'il-3',\n",
       " 'cos-7',\n",
       " 'il-3',\n",
       " 'cos-7',\n",
       " 'cos-7',\n",
       " 'il-3',\n",
       " 'tbs-t',\n",
       " 'tbs-t',\n",
       " 'il-3',\n",
       " 'cos-7',\n",
       " 'cos-7',\n",
       " 'il-3',\n",
       " 'cos-7',\n",
       " 'il-3',\n",
       " 'il-3',\n",
       " 'il-3',\n",
       " 'cos-7',\n",
       " 'pi3-k',\n",
       " 'pi3-',\n",
       " 'bcr-abl',\n",
       " 'bcr-abl',\n",
       " 'pcmv-ha',\n",
       " 'bcr-abl',\n",
       " 'il-3',\n",
       " 'il-3',\n",
       " 'ns-1',\n",
       " 'nup98-hoxd13',\n",
       " 'ns-1',\n",
       " 'il-3',\n",
       " 'il-3',\n",
       " 'pp-1',\n",
       " 'pp-2',\n",
       " 'cos-7',\n",
       " 'il-3',\n",
       " 'mil-3',\n",
       " 'mil-6',\n",
       " 'cos-7',\n",
       " 'il-3',\n",
       " 'pp-1',\n",
       " 'pp-2',\n",
       " 'cos-7',\n",
       " 'il-3',\n",
       " 'il-3',\n",
       " 'il-3',\n",
       " 'il-3',\n",
       " ...]"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fusion_like"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_fusion_like = pd.DataFrame(fusion_like, columns=['fusion_like'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "fusion_like\n",
       "nih-pa           3399\n",
       "il-3             3297\n",
       "bcr-abl          2091\n",
       "flt3-itd         2031\n",
       "eml4-alk         1557\n",
       "cos-7            1086\n",
       "src-3             994\n",
       "shp-2             970\n",
       "mcf-7             911\n",
       "npm-alk           773\n",
       "dr-gfp            745\n",
       "ar-v7             728\n",
       "wm278-nic-gfp     688\n",
       "msi-h             674\n",
       "wtb-raf           672\n",
       "dtype: int64"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_fusion_like.groupby('fusion_like').size().sort_values(ascending=False).head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='sec2_3'></a>\n",
    "## Mutation Types (<a href='#sec0'>Back To Top</a>)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mutation_patterns = ['truncation', 'deletion', 'promoter', \n",
    "                     'amplification', 'epigenetic', 'frame', \n",
    "                     'overexpression', 'duplication', 'insertion',\n",
    "                     'subtype', 'fusion', 'splice', 'wildtype']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 5.59 s, sys: 3.99 ms, total: 5.59 s\n",
      "Wall time: 5.6 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "mutation_types = [word.lower() for word in textome_tokens if word.lower() in mutation_patterns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_mutation_types = pd.DataFrame(mutation_types, columns=['mutation_types'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "mutation_types\n",
       "fusion            17133\n",
       "deletion          15717\n",
       "amplification     13849\n",
       "promoter          10147\n",
       "overexpression     7572\n",
       "splice             5974\n",
       "insertion          4925\n",
       "subtype            2440\n",
       "truncation         2122\n",
       "duplication        1925\n",
       "epigenetic         1889\n",
       "frame              1611\n",
       "wildtype           1498\n",
       "dtype: int64"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mutation_types.groupby('mutation_types').size().sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='sec2_4'></a>\n",
    "## Gene Like Names (<a href='#sec0'>Back To Top</a>)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 10.5 s, sys: 2.65 s, total: 13.1 s\n",
      "Wall time: 13.1 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "txt1_white = textome.encode().decode()  # copy a string?!\n",
    "txt1_white = txt1_white.replace('\"', ' ')\n",
    "txt1_white = txt1_white.replace('.', ' ')\n",
    "txt1_white = txt1_white.replace('/', ' ')\n",
    "txt1_white = txt1_white.replace('\\'', ' ')\n",
    "txt1_white = txt1_white.replace('_', ' ')    # This should be done after _pos, etc checked\n",
    "txt1_white = txt1_white.replace('-', ' ')    # This should be done after -pos, etc checked\n",
    "txt1_white = txt1_white.replace('=', ' ')\n",
    "txt1_white = txt1_white.replace('\\n', ' ')\n",
    "txt1_white = txt1_white.replace('\\\\n', ' ')\n",
    "txt1_white = txt1_white.replace('\\'', ' ')\n",
    "txt1_white = re.sub(' +',' ', txt1_white) \n",
    "txt1_white = txt1_white.replace('\\'', ' ')\n",
    "txt1_white = txt1_white.replace('(', ' ')\n",
    "txt1_white = txt1_white.replace(')', ' ')\n",
    "txt1_white = txt1_white.replace('[', ' ')\n",
    "txt1_white = txt1_white.replace(']', ' ')\n",
    "txt1_white = txt1_white.replace('{', ' ')\n",
    "txt1_white = txt1_white.replace('}', ' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1min 48s, sys: 500 ms, total: 1min 49s\n",
      "Wall time: 1min 49s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "tokens_white = word_tokenize(txt1_white)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 24.6 s, sys: 43.9 ms, total: 24.7 s\n",
      "Wall time: 24.7 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Remove nucleotide sequences\n",
    "nts = r\"[ACTGU]{4,}\"\n",
    "tokens_white = [word for word in tokens_white \\\n",
    "                if not word in stop_words \\\n",
    "                if not word in commoners \\\n",
    "                if not re.search(nts, word)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# of gene-ish words: 364\n",
      "CPU times: user 29.6 s, sys: 0 ns, total: 29.6 s\n",
      "Wall time: 29.7 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "gene_ish_pattern1 = r\"[A-Z]{2,7}[0-9]{0,4}$\"\n",
    "gene_ish_pattern2 = r\"p[0-9]{1,3}$\"\n",
    "gene_ish_words = [word for word in tokens_white if re.match(gene_ish_pattern1, word)]\n",
    "gene_ish_words += [word for word in tokens_white if re.match(gene_ish_pattern2, word)]\n",
    "print('# of gene-ish words: %d' % len(gene_ish_words_white))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'KEGG',\n",
       " 'ASS',\n",
       " 'FRASER',\n",
       " 'PHD',\n",
       " 'FY23',\n",
       " 'OHA',\n",
       " 'WHSC2',\n",
       " 'DFS',\n",
       " 'KIAA1509',\n",
       " 'FFAR2',\n",
       " 'SOS1',\n",
       " 'NSCPO',\n",
       " 'FYVE',\n",
       " 'RNF4',\n",
       " 'UBI',\n",
       " 'PTCH1',\n",
       " 'RASU2',\n",
       " 'PVALB',\n",
       " 'NB',\n",
       " 'RIKEN',\n",
       " 'HCC2429',\n",
       " 'AXIN1',\n",
       " 'BT4',\n",
       " 'MSN2',\n",
       " 'EPHA2',\n",
       " 'HM48',\n",
       " 'ARV7',\n",
       " 'NOX4',\n",
       " 'QTR2',\n",
       " 'BME',\n",
       " 'FINGER',\n",
       " 'NCOA6',\n",
       " 'PGCTYS',\n",
       " 'LAML',\n",
       " 'AB1442',\n",
       " 'FLVRDAS',\n",
       " 'SLIDE',\n",
       " 'CXCL3',\n",
       " 'CYVD',\n",
       " 'ROBETTA',\n",
       " 'CSE1',\n",
       " 'WHATIF',\n",
       " 'NACC2',\n",
       " 'META',\n",
       " 'NXPH1',\n",
       " 'CAPZB',\n",
       " 'OD2',\n",
       " 'CAND',\n",
       " 'ZNF639',\n",
       " 'DAB2',\n",
       " 'IRC',\n",
       " 'GAG',\n",
       " 'BZS',\n",
       " 'AVF',\n",
       " 'LYS505',\n",
       " 'TYYL',\n",
       " 'EN26',\n",
       " 'ZWILCH',\n",
       " 'PGBM14',\n",
       " 'IIA110',\n",
       " 'TALL',\n",
       " 'LSL',\n",
       " 'SKBR',\n",
       " 'VHL30',\n",
       " 'HOXB8',\n",
       " 'AZD2644',\n",
       " 'GOPC',\n",
       " 'RAS10',\n",
       " 'ASPL',\n",
       " 'ATHB6',\n",
       " 'PRCC',\n",
       " 'CUCRC040',\n",
       " 'GHE0536',\n",
       " 'EA4340',\n",
       " 'MARCKS',\n",
       " 'FOXA',\n",
       " 'PD0119',\n",
       " 'RCPHHER',\n",
       " 'LSD1',\n",
       " 'EPOCH',\n",
       " 'GMS',\n",
       " 'SCCOHT1',\n",
       " 'TSS1',\n",
       " 'ALH3000',\n",
       " 'CAC',\n",
       " 'RAL',\n",
       " 'CLD',\n",
       " 'TRAF7',\n",
       " 'ML3000',\n",
       " 'UMUC17',\n",
       " 'WO',\n",
       " 'IDH1',\n",
       " 'ABI7000',\n",
       " 'RENZO',\n",
       " 'GMCSFR',\n",
       " 'RSB2',\n",
       " 'COLO205',\n",
       " 'BRCAX',\n",
       " 'LZTR1',\n",
       " 'VAMP',\n",
       " 'PR110',\n",
       " 'DOGS',\n",
       " 'CEN8',\n",
       " 'DOMAIN',\n",
       " 'TFBS',\n",
       " 'ZAP70',\n",
       " 'DMRT1',\n",
       " 'MF1',\n",
       " 'PM1',\n",
       " 'CD4',\n",
       " 'UHRF1',\n",
       " 'LUCA',\n",
       " 'UQCRH',\n",
       " 'SET3',\n",
       " 'PKP2',\n",
       " 'PCM25',\n",
       " 'NNFF103',\n",
       " 'TMEM49',\n",
       " 'PHRC',\n",
       " 'ULBP',\n",
       " 'DIE',\n",
       " 'RTR',\n",
       " 'OCT4',\n",
       " 'GIST780',\n",
       " 'CAF11',\n",
       " 'BIK',\n",
       " 'PRL',\n",
       " 'TBET',\n",
       " 'SETBP1',\n",
       " 'RIZ1',\n",
       " 'BRF2',\n",
       " 'SRP',\n",
       " 'GATOR1',\n",
       " 'RADS0',\n",
       " 'IQGAP',\n",
       " 'JUNQ',\n",
       " 'AXIN2',\n",
       " 'GOLPH3',\n",
       " 'SAS14',\n",
       " 'BCC9',\n",
       " 'GRK',\n",
       " 'HCA',\n",
       " 'BHQ1',\n",
       " 'CCNB2',\n",
       " 'MESO',\n",
       " 'LL6',\n",
       " 'SFA7',\n",
       " 'STK12',\n",
       " 'XL139',\n",
       " 'MCF10',\n",
       " 'DDB',\n",
       " 'ACER3',\n",
       " 'RPS6',\n",
       " 'DFP',\n",
       " 'JBW902',\n",
       " 'BLP1',\n",
       " 'REMLEN',\n",
       " 'SV',\n",
       " 'CECE',\n",
       " 'YPB2',\n",
       " 'DS86',\n",
       " 'CFA6',\n",
       " 'SREBP1',\n",
       " 'NHUC',\n",
       " 'MAB636',\n",
       " 'SCH6636',\n",
       " 'NB0073',\n",
       " 'EVQWKV',\n",
       " 'HCC4006',\n",
       " 'QHQH',\n",
       " 'SFRP5',\n",
       " 'TRP2',\n",
       " 'RBF29',\n",
       " 'CNIO',\n",
       " 'CC1',\n",
       " 'XLIII',\n",
       " 'MYCN4',\n",
       " 'BCPAP',\n",
       " 'HGDC',\n",
       " 'ING2',\n",
       " 'DR105',\n",
       " 'PMF',\n",
       " 'PCAF8',\n",
       " 'GOLD',\n",
       " 'QYL',\n",
       " 'RAD7',\n",
       " 'ETV5',\n",
       " 'MAPK9',\n",
       " 'AQ5',\n",
       " 'UVS',\n",
       " 'POSF2',\n",
       " 'TE',\n",
       " 'SPAFAS',\n",
       " 'SAPS',\n",
       " 'COHORT',\n",
       " 'MACROD1',\n",
       " 'MATCH',\n",
       " 'UKE',\n",
       " 'OUTCOME',\n",
       " 'SLFN5',\n",
       " 'GPR18',\n",
       " 'RAUTR',\n",
       " 'ROSJ',\n",
       " 'NRAS3',\n",
       " 'MOI',\n",
       " 'BYERS',\n",
       " 'LFC',\n",
       " 'GCF2',\n",
       " 'ALPHA',\n",
       " 'ERRB2',\n",
       " 'MDC1',\n",
       " 'SSBP2',\n",
       " 'CEP128',\n",
       " 'IBGC4',\n",
       " 'p50',\n",
       " 'RS3',\n",
       " 'GRR',\n",
       " 'VAIK',\n",
       " 'ITP25',\n",
       " 'DEAD',\n",
       " 'FXYD6',\n",
       " 'TEX14',\n",
       " 'ATG12',\n",
       " 'IDL',\n",
       " 'SHOWN',\n",
       " 'BFTC909',\n",
       " 'ZSTK474',\n",
       " 'DFCI076',\n",
       " 'p43',\n",
       " 'KIAA1109',\n",
       " 'CSCS',\n",
       " 'MX3000',\n",
       " 'FANCD10',\n",
       " 'VEGFF',\n",
       " 'RIP2',\n",
       " 'AAUH',\n",
       " 'HR36',\n",
       " 'CHD6',\n",
       " 'WILEY',\n",
       " 'MDM4',\n",
       " 'GIMEMA',\n",
       " 'CD41',\n",
       " 'RING',\n",
       " 'SEF99',\n",
       " 'SMYD3',\n",
       " 'MPV',\n",
       " 'MMP',\n",
       " 'NVPBVU972',\n",
       " 'IKK2',\n",
       " 'p385',\n",
       " 'TAAD',\n",
       " 'ARG13',\n",
       " 'NTB',\n",
       " 'JCRB0077',\n",
       " 'CTSB',\n",
       " 'CCNE',\n",
       " 'RXY',\n",
       " 'FJ',\n",
       " 'MDA231',\n",
       " 'EHRC',\n",
       " 'TEAD',\n",
       " 'SMARCC2',\n",
       " 'MMP11',\n",
       " 'ESD',\n",
       " 'CAF37',\n",
       " 'CAIRO2',\n",
       " 'DOJINDO',\n",
       " 'SAHF',\n",
       " 'ECL',\n",
       " 'IGL',\n",
       " 'OB1',\n",
       " 'DBLCL',\n",
       " 'ROSA',\n",
       " 'SAE',\n",
       " 'TRPI',\n",
       " 'LC1',\n",
       " 'VPC',\n",
       " 'STIS71',\n",
       " 'VHL19',\n",
       " 'RPA40',\n",
       " 'GFAP',\n",
       " 'NC5',\n",
       " 'PU',\n",
       " 'MAT3',\n",
       " 'EHE1',\n",
       " 'SEQUEST',\n",
       " 'SXM30',\n",
       " 'LAU149',\n",
       " 'FAX',\n",
       " 'GECO',\n",
       " 'PAG608',\n",
       " 'RINGH2',\n",
       " 'SBMA',\n",
       " 'MK2006',\n",
       " 'CEBPD',\n",
       " 'PITPNA',\n",
       " 'PB2',\n",
       " 'EMSY',\n",
       " 'SCREEN',\n",
       " 'DIC',\n",
       " 'NIHR',\n",
       " 'PGAAHY',\n",
       " 'TCF2',\n",
       " 'DUSP18',\n",
       " 'MEDI',\n",
       " 'KATP',\n",
       " 'YGRFTI',\n",
       " 'EDAID',\n",
       " 'CRC',\n",
       " 'UX13',\n",
       " 'SOD3',\n",
       " 'ETHODS',\n",
       " 'FITC',\n",
       " 'ABH3',\n",
       " 'BX51',\n",
       " 'JSR',\n",
       " 'HUT102',\n",
       " 'ISL1',\n",
       " 'ERBB2',\n",
       " 'p2',\n",
       " 'EDA',\n",
       " 'PI',\n",
       " 'RLUC',\n",
       " 'CRAF31',\n",
       " 'RASRAF',\n",
       " 'KIT67',\n",
       " 'GDP',\n",
       " 'RRAS2',\n",
       " 'PLDLS',\n",
       " 'GTP42',\n",
       " 'ONLUS',\n",
       " 'SHC',\n",
       " 'LIT',\n",
       " 'SPFH2',\n",
       " 'MSK111',\n",
       " 'CEN4',\n",
       " 'BRD',\n",
       " 'DDR16',\n",
       " 'TRKC',\n",
       " 'KSR2',\n",
       " 'GSG2',\n",
       " 'SMARCE1',\n",
       " 'ERTAM',\n",
       " 'MT3',\n",
       " 'CFA4',\n",
       " 'NT2',\n",
       " 'TRID',\n",
       " 'BID',\n",
       " 'TRIAP1',\n",
       " 'NFM',\n",
       " 'GDC',\n",
       " 'LTGC',\n",
       " 'GCT10',\n",
       " 'CALCR',\n",
       " 'SUN',\n",
       " 'SSCV',\n",
       " 'WAT',\n",
       " 'VHL213',\n",
       " 'ULTRA',\n",
       " 'KYSE270',\n",
       " 'MVSR',\n",
       " 'UA',\n",
       " 'BGJ',\n",
       " 'LN319',\n",
       " 'ADRBK1',\n",
       " 'SOC2',\n",
       " 'MODY3',\n",
       " 'ASCII',\n",
       " 'FLASH',\n",
       " 'ALL2',\n",
       " 'LB960',\n",
       " 'MGDC',\n",
       " 'NOL3',\n",
       " 'GI',\n",
       " 'EPOR',\n",
       " 'IVB',\n",
       " 'WGD29',\n",
       " 'SCC294',\n",
       " 'CASPASE',\n",
       " 'CHH',\n",
       " 'SKMG3',\n",
       " 'TP21',\n",
       " 'PAK6',\n",
       " 'SA530',\n",
       " 'BC112',\n",
       " 'UBL1',\n",
       " 'VEGF164',\n",
       " 'KEEG',\n",
       " 'MUSCLE',\n",
       " 'EZ1',\n",
       " 'VEC',\n",
       " 'TER',\n",
       " 'CCNG1',\n",
       " 'SHIP2',\n",
       " 'KRAS',\n",
       " 'ZF7',\n",
       " 'ERKMAP',\n",
       " 'RARB',\n",
       " 'REDD1',\n",
       " 'MVD',\n",
       " 'MP41',\n",
       " 'MKN',\n",
       " 'ABCFR',\n",
       " 'LB1',\n",
       " 'AZAP',\n",
       " 'IRTA21',\n",
       " 'PIK3',\n",
       " 'LGR',\n",
       " 'SVM',\n",
       " 'AREG55',\n",
       " 'TGI',\n",
       " 'LYDLLL',\n",
       " 'RALGDS',\n",
       " 'GENOME',\n",
       " 'TA2',\n",
       " 'PCK412',\n",
       " 'ISG',\n",
       " 'AJFL',\n",
       " 'KAF1400',\n",
       " 'MCPH4',\n",
       " 'DRG',\n",
       " 'MAPK8',\n",
       " 'ABED',\n",
       " 'GCLC',\n",
       " 'CSR',\n",
       " 'CD13',\n",
       " 'LVEF',\n",
       " 'CMML70',\n",
       " 'TOP0',\n",
       " 'PLAP',\n",
       " 'RPGN',\n",
       " 'ENU',\n",
       " 'ENO2',\n",
       " 'PTEN314',\n",
       " 'SPRY1',\n",
       " 'PGM',\n",
       " 'CFX',\n",
       " 'HIC1',\n",
       " 'POLE1',\n",
       " 'BEARING',\n",
       " 'PJS049',\n",
       " 'QMAA',\n",
       " 'GSK126',\n",
       " 'LEO1',\n",
       " 'RBP1',\n",
       " 'VEGFC',\n",
       " 'TAD2',\n",
       " 'MICZ',\n",
       " 'SCRN2',\n",
       " 'CD71',\n",
       " 'PMS134',\n",
       " 'TFL',\n",
       " 'JAK2',\n",
       " 'BBMM',\n",
       " 'PORT',\n",
       " 'CCNO',\n",
       " 'MST3',\n",
       " 'IDLSQV',\n",
       " 'NZM6',\n",
       " 'CRM326',\n",
       " 'MCG224',\n",
       " 'PCT',\n",
       " 'FGR',\n",
       " 'STAT2',\n",
       " 'WPRE',\n",
       " 'ANA',\n",
       " 'IWP',\n",
       " 'HD2',\n",
       " 'JGI',\n",
       " 'BIMS',\n",
       " 'GCN2',\n",
       " 'MGT',\n",
       " 'EOSHA',\n",
       " 'NELF',\n",
       " 'GIST74',\n",
       " 'SW756',\n",
       " 'VAP',\n",
       " 'ME1',\n",
       " 'ELVES',\n",
       " 'p130',\n",
       " 'TMLQHQH',\n",
       " 'PBX',\n",
       " 'SMID',\n",
       " 'PGL1',\n",
       " 'AGY70',\n",
       " 'ITGA2',\n",
       " 'CPO',\n",
       " 'p40',\n",
       " 'LD50',\n",
       " 'FDX1',\n",
       " 'SMCY',\n",
       " 'WDR48',\n",
       " 'SHGP',\n",
       " 'PE56',\n",
       " 'AB23',\n",
       " 'RHOF',\n",
       " 'MMG1',\n",
       " 'AMLRTR1',\n",
       " 'ZSG',\n",
       " 'FEN1',\n",
       " 'ALKBH8',\n",
       " 'BRSK1',\n",
       " 'HSP70',\n",
       " 'AZD2281',\n",
       " 'RBBP5',\n",
       " 'LPTSY',\n",
       " 'PCDNA',\n",
       " 'PBP2',\n",
       " 'GUAVA',\n",
       " 'MSKCC',\n",
       " 'TM8',\n",
       " 'HC',\n",
       " 'QUAD',\n",
       " 'IML',\n",
       " 'MECA',\n",
       " 'p537',\n",
       " 'IFNG',\n",
       " 'LSM1',\n",
       " 'CRTAC1',\n",
       " 'PVF2',\n",
       " 'KOPTK',\n",
       " 'PPH2',\n",
       " 'IGFJR',\n",
       " 'LDS5',\n",
       " 'EPD23',\n",
       " 'GENOMED',\n",
       " 'SUDHL7',\n",
       " 'HGS',\n",
       " 'MQM',\n",
       " 'SKIL',\n",
       " 'SOS',\n",
       " 'PI335',\n",
       " 'CSA',\n",
       " 'AMAG',\n",
       " 'JBP',\n",
       " 'MDD',\n",
       " 'SA494',\n",
       " 'ZNF493',\n",
       " 'AG2986',\n",
       " 'MC44',\n",
       " 'USC',\n",
       " 'NHE',\n",
       " 'CHK',\n",
       " 'PMA',\n",
       " 'HR1',\n",
       " 'PRIM2',\n",
       " 'NIBP',\n",
       " 'NAAIRS',\n",
       " 'THIS',\n",
       " 'DGDP',\n",
       " 'BXPC',\n",
       " 'SJDOS006',\n",
       " 'WILMS',\n",
       " 'GROUCHO',\n",
       " 'CPA',\n",
       " 'ICAM',\n",
       " 'CS21',\n",
       " 'EPD1',\n",
       " 'MPEC',\n",
       " 'BFTC905',\n",
       " 'FBXW7',\n",
       " 'YXXM',\n",
       " 'PTTG1',\n",
       " 'GHR',\n",
       " 'VQSR',\n",
       " 'IRCC',\n",
       " 'ABN11',\n",
       " 'IP',\n",
       " 'IC050',\n",
       " 'ARUP',\n",
       " 'RQ428',\n",
       " 'SRCC',\n",
       " 'CLUSTAL',\n",
       " 'ZINC',\n",
       " 'VDRC',\n",
       " 'AWS',\n",
       " 'VD862',\n",
       " 'CGT',\n",
       " 'ITC',\n",
       " 'TTIGFP',\n",
       " 'LUV',\n",
       " 'TPO',\n",
       " 'GLIPR1',\n",
       " 'EUFA30',\n",
       " 'LUCC3',\n",
       " 'EBER1',\n",
       " 'PEGPH20',\n",
       " 'NA934',\n",
       " 'ECTMWT',\n",
       " 'KYSE150',\n",
       " 'AJ',\n",
       " 'TYK2',\n",
       " 'RDA',\n",
       " 'SANT1',\n",
       " 'CTR1',\n",
       " 'CIT',\n",
       " 'USA10',\n",
       " 'WHIM8',\n",
       " 'MGRNA',\n",
       " 'CITED4',\n",
       " 'REVIEWS',\n",
       " 'MYC128',\n",
       " 'RESPER',\n",
       " 'NUP205',\n",
       " 'VSL',\n",
       " 'CFD80',\n",
       " 'SHEP1',\n",
       " 'UVR',\n",
       " 'XLII',\n",
       " 'BRC3',\n",
       " 'FWD1',\n",
       " 'ARHGAP',\n",
       " 'NP7',\n",
       " 'MSH1',\n",
       " 'LPHN3',\n",
       " 'MNES',\n",
       " 'WNK2',\n",
       " 'NZM9',\n",
       " 'MY5',\n",
       " 'SGX523',\n",
       " 'FIGURE4',\n",
       " 'LK79',\n",
       " 'AST487',\n",
       " 'BLI',\n",
       " 'INBS',\n",
       " 'PKRG1',\n",
       " 'CGH9577',\n",
       " 'SLIT',\n",
       " 'EFV',\n",
       " 'EXLM1',\n",
       " 'VSV',\n",
       " 'FG',\n",
       " 'ACH',\n",
       " 'HAL',\n",
       " 'LCT',\n",
       " 'CCCH',\n",
       " 'APE1',\n",
       " 'CA1',\n",
       " 'IPMN',\n",
       " 'EX96',\n",
       " 'CDH11',\n",
       " 'SW527',\n",
       " 'BGTP',\n",
       " 'KRIT1',\n",
       " 'ATF4',\n",
       " 'DR50',\n",
       " 'TRAILR3',\n",
       " 'SAVVLL',\n",
       " 'CD5',\n",
       " 'MYOD1',\n",
       " 'COS7',\n",
       " 'RAB11',\n",
       " 'NFT',\n",
       " 'CRL1651',\n",
       " 'UNC',\n",
       " 'ENTRY',\n",
       " 'CEBPB',\n",
       " 'APL2000',\n",
       " 'HCC1195',\n",
       " 'CFS1',\n",
       " 'PGP',\n",
       " 'JM',\n",
       " 'SETD22',\n",
       " 'FFS',\n",
       " 'GAFG',\n",
       " 'SNEV',\n",
       " 'KIAA1239',\n",
       " 'GRAALL',\n",
       " 'NCOA1',\n",
       " 'PT09',\n",
       " 'EL',\n",
       " 'WM3622',\n",
       " 'MVR',\n",
       " 'HGOC',\n",
       " 'WNK',\n",
       " 'BIX',\n",
       " 'PRKC',\n",
       " 'MOLT4',\n",
       " 'GXDD',\n",
       " 'HCT',\n",
       " 'RPTOR',\n",
       " 'BHRF1',\n",
       " 'PRODRG',\n",
       " 'ANK1',\n",
       " 'LIZ',\n",
       " 'HIAP',\n",
       " 'DMRXA',\n",
       " 'RGB',\n",
       " 'KPT',\n",
       " 'KITWK557',\n",
       " 'AF6',\n",
       " 'FASTP',\n",
       " 'VTA',\n",
       " 'IDO',\n",
       " 'SSNG',\n",
       " 'KIR',\n",
       " 'NH3',\n",
       " 'RPN',\n",
       " 'PDL36',\n",
       " 'ZC25',\n",
       " 'RTC14',\n",
       " 'CBF5',\n",
       " 'SPINT1',\n",
       " 'BRAF',\n",
       " 'MCH5',\n",
       " 'EOL3',\n",
       " 'SEVERAL',\n",
       " 'NMC1',\n",
       " 'NCI',\n",
       " 'BLMH',\n",
       " 'LOH4',\n",
       " 'CEEA',\n",
       " 'MFE319',\n",
       " 'KLHL7',\n",
       " 'ACA',\n",
       " 'NK4',\n",
       " 'HI1',\n",
       " 'RN',\n",
       " 'TIMPS',\n",
       " 'VSMC',\n",
       " 'LM200',\n",
       " 'RNF149',\n",
       " 'DEPDC4',\n",
       " 'WTKIT',\n",
       " 'RESULTS',\n",
       " 'ISOGEN',\n",
       " 'ARHGEF12',\n",
       " 'SMCD',\n",
       " 'MS9',\n",
       " 'FBXW10',\n",
       " 'ST6',\n",
       " 'ABN5',\n",
       " 'PDGRFR',\n",
       " 'QU',\n",
       " 'MLL2',\n",
       " 'YH',\n",
       " 'MYERS',\n",
       " 'PDL1',\n",
       " 'TRIS',\n",
       " 'EHS',\n",
       " 'VMCUB1',\n",
       " 'GAR1',\n",
       " 'LGG',\n",
       " 'GBC',\n",
       " 'HDC90',\n",
       " 'DSS1',\n",
       " 'FCSCOR',\n",
       " 'SIOP',\n",
       " 'GRB',\n",
       " 'p17',\n",
       " 'OMEGA',\n",
       " 'CUBII',\n",
       " 'SEF',\n",
       " 'COLD',\n",
       " 'TRPS1',\n",
       " 'DPP6',\n",
       " 'MISC',\n",
       " 'MAPKERK1',\n",
       " 'FDC',\n",
       " 'BCL',\n",
       " 'NFKB',\n",
       " 'CCDC27',\n",
       " 'KIAA0228',\n",
       " 'RGL2',\n",
       " 'COSM1146',\n",
       " 'p85',\n",
       " 'JAK235',\n",
       " 'LZTFL1',\n",
       " 'EHZ2',\n",
       " 'MD10',\n",
       " 'RLU',\n",
       " 'p800',\n",
       " 'BTZ',\n",
       " 'IMGENEX',\n",
       " 'CHO',\n",
       " 'TM182',\n",
       " 'CEPR',\n",
       " 'ST486',\n",
       " 'CAPB',\n",
       " 'XM',\n",
       " 'HIPK',\n",
       " 'PCDH19',\n",
       " 'MIP',\n",
       " 'IGT',\n",
       " 'COCOC',\n",
       " 'CFG',\n",
       " 'DC',\n",
       " 'NSCL',\n",
       " 'PE3700',\n",
       " 'MAA',\n",
       " 'GJA',\n",
       " 'AOCS',\n",
       " 'IG32',\n",
       " 'TRF',\n",
       " 'FF99',\n",
       " 'IRIF',\n",
       " 'RSC18',\n",
       " 'PW',\n",
       " 'MMU12',\n",
       " 'SETD222',\n",
       " 'SPT16',\n",
       " 'EP300',\n",
       " 'FST',\n",
       " 'MK2',\n",
       " 'MY10',\n",
       " 'THRAP3',\n",
       " 'TUBGCP6',\n",
       " 'DTX1',\n",
       " 'IVIG',\n",
       " 'SMCT',\n",
       " 'GT6',\n",
       " 'THO1',\n",
       " 'CRC525',\n",
       " 'NCOA7',\n",
       " 'CRG',\n",
       " 'CAPAN',\n",
       " 'GF68',\n",
       " 'EMSA',\n",
       " 'HK',\n",
       " 'LS1034',\n",
       " 'NTG',\n",
       " 'YXXQ',\n",
       " 'BO',\n",
       " 'FRT',\n",
       " 'THYROID',\n",
       " 'MCLR',\n",
       " 'CME',\n",
       " 'XA90',\n",
       " 'SPIRE2',\n",
       " 'TACH',\n",
       " 'GES',\n",
       " 'FC1',\n",
       " 'HT',\n",
       " 'RW2982',\n",
       " 'ICF',\n",
       " 'NP250',\n",
       " 'JR2355',\n",
       " 'BGJ398',\n",
       " 'AGC3',\n",
       " 'CGNP',\n",
       " 'DGOS',\n",
       " 'SGK3',\n",
       " 'TIAL1',\n",
       " 'SMS2',\n",
       " 'NJ',\n",
       " 'CHIC',\n",
       " 'COOT',\n",
       " 'APBS',\n",
       " 'SM2702',\n",
       " 'RKK',\n",
       " 'SYPRO',\n",
       " 'HSU4376',\n",
       " 'WORKUP',\n",
       " 'p3',\n",
       " 'POLE2',\n",
       " 'LUC1',\n",
       " 'WHO',\n",
       " 'CREB53',\n",
       " 'DEN',\n",
       " 'SMRT',\n",
       " 'PAPS',\n",
       " 'SFA19',\n",
       " 'SQY',\n",
       " 'DEP',\n",
       " 'IEDNE',\n",
       " 'LKBI',\n",
       " 'PCI',\n",
       " 'RUS08',\n",
       " 'IGFBP7',\n",
       " 'NAG',\n",
       " 'SHANK3',\n",
       " 'CH12',\n",
       " 'INICD2',\n",
       " 'MGC8602',\n",
       " 'RM2235',\n",
       " 'PHLDB1',\n",
       " 'THADA',\n",
       " 'CAPONY',\n",
       " 'WA1',\n",
       " 'BMG',\n",
       " 'PJS055',\n",
       " 'GTP30',\n",
       " 'BTG',\n",
       " 'LANCL2',\n",
       " 'CRC012',\n",
       " 'FLAME',\n",
       " 'HHF2',\n",
       " 'HNSC',\n",
       " 'AIRWAY',\n",
       " 'ETS2',\n",
       " 'MOV34',\n",
       " 'IRB100',\n",
       " 'RADW',\n",
       " 'SJSA',\n",
       " 'CPEB1',\n",
       " 'FKHRL',\n",
       " 'PRPF8',\n",
       " 'BIDMC2009',\n",
       " 'LFA',\n",
       " 'MLST8',\n",
       " 'UBN',\n",
       " 'GHITM',\n",
       " 'PSR',\n",
       " 'GSNAP',\n",
       " 'IVS13',\n",
       " 'PVF3',\n",
       " 'HCCS',\n",
       " 'ST480',\n",
       " 'RB2',\n",
       " 'DLL494',\n",
       " 'DM6000',\n",
       " 'IRTA17',\n",
       " 'FYB',\n",
       " 'DNR',\n",
       " 'p217',\n",
       " 'IQ',\n",
       " 'SQLE',\n",
       " 'TIL1362',\n",
       " 'NEO4',\n",
       " 'BRC8',\n",
       " 'ROI',\n",
       " 'NP40',\n",
       " 'MELT87',\n",
       " 'HBL100',\n",
       " 'LOT',\n",
       " 'UOK121',\n",
       " 'LIP1',\n",
       " 'CYTEL',\n",
       " 'SC255',\n",
       " 'SIC',\n",
       " 'SJLGG039',\n",
       " 'NAP57',\n",
       " 'FARAGE',\n",
       " 'ARR2',\n",
       " 'RL',\n",
       " 'DPH5',\n",
       " 'PTPN3',\n",
       " 'HELF',\n",
       " 'SEL10',\n",
       " 'FTI',\n",
       " 'DEV',\n",
       " 'FBXO1150',\n",
       " 'BRAF75',\n",
       " 'NZM33',\n",
       " 'LY',\n",
       " 'IP4',\n",
       " 'RAK',\n",
       " 'GEMININ',\n",
       " 'EFNA1',\n",
       " 'PR16',\n",
       " 'PV',\n",
       " 'AKTA',\n",
       " 'SF5',\n",
       " 'SPSS16',\n",
       " 'LINKAGE',\n",
       " 'MBO2',\n",
       " 'UMR5641',\n",
       " 'JAK249',\n",
       " 'HUC',\n",
       " 'MSE2',\n",
       " 'UC39',\n",
       " 'ALL51',\n",
       " 'PAM50',\n",
       " 'MLLT1',\n",
       " 'NO',\n",
       " 'NUMA1',\n",
       " 'CLL2',\n",
       " 'GR150',\n",
       " 'QUASAR2',\n",
       " 'RAB18',\n",
       " 'CD81',\n",
       " 'DR',\n",
       " 'KLHL9',\n",
       " 'MM1',\n",
       " 'JEG3',\n",
       " 'CPSF',\n",
       " 'SCNT',\n",
       " 'ADB',\n",
       " 'MRC',\n",
       " 'IGV15',\n",
       " 'NORE1',\n",
       " 'RBM5',\n",
       " 'TENN',\n",
       " 'BF264',\n",
       " 'PFB',\n",
       " 'IFAR7',\n",
       " 'RMSD',\n",
       " 'RB10',\n",
       " 'ECD1',\n",
       " 'RPL39',\n",
       " 'MCG758',\n",
       " 'REMARK',\n",
       " 'MP470',\n",
       " 'TKI13',\n",
       " 'MZFLIII',\n",
       " 'NZM24',\n",
       " 'PROSITE',\n",
       " 'LOX',\n",
       " 'STO',\n",
       " 'KAT18',\n",
       " ...}"
      ]
     },
     "execution_count": 191,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(gene_ish_words)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='sec3'></a>\n",
    "# Functions for Processing Features (<a href='#sec0'>Back To Top</a>)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "commoners = ['RT', 'PCR', 'RT-PCR', 'DNA', 'cDNA', 'RNA', 'mRNA', 'siRNA', 'shRNA', 'protein', \n",
    "             'cell', 'cancer', 'CHIP', 'FISH', 'SDS-PAGE', 'UK', 'USA', 'GST', 'GFP', 'SDS', 'PAGE',\n",
    "             'qPCR', 'PBS', 'TBS', 'DTT', 'BSA', 'HSA', 'HCl', 'NCBI', 'PBST', 'ANOVA', 'RIKEN',\n",
    "             'COHORT', 'OUTCOME', 'AIRWAY', 'EMSA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_positive_and_negative_words(tokens):\n",
    "    pos_neg = [token.lower() for token in tokens \\\n",
    "           if (('-positive' in token.lower()) | ('-negative' in token.lower())) \\\n",
    "           if not (('false' in token.lower()) | ('true' in token.lower()))]\n",
    "    \n",
    "    # Remove GFP tag\n",
    "    for i, word in enumerate(pos_neg):\n",
    "        if 'egfp/eyfp' in word: \n",
    "            pos_neg[i] = word.replace('egfp/eyfp-', '')\n",
    "        elif 'egfp' in word:\n",
    "            pos_neg[i] = word.replace('egfp-', '')\n",
    "        elif 'gfp' in word:\n",
    "            pos_neg[i] = word.replace('gfp'+\"–\", '') #This hyphen is somehow different from the other one\n",
    "\n",
    "    pos_neg = [token for token in pos_neg \\\n",
    "               if not ((token == '-positive') | (token == '-negative'))]\n",
    "    \n",
    "    return pos_neg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_fusion_like_words(tokens, commoners):\n",
    "    # Remove nucleotide sequences\n",
    "    nts = r\"[ACTGU]{4,}\"\n",
    "    \n",
    "    # fusion-like pattern\n",
    "    fusion_like_pattern = r\"([A-Z]{2,7})([0-9]{1,4})?-([A-Z]{0,7})([0-9]{1,4})?$\"\n",
    "\n",
    "    fusion_like = [word.lower() for word in tokens \\\n",
    "                   if word not in commoners \\\n",
    "                   if re.search(fusion_like_pattern, word) \\\n",
    "                   if not re.search(nts, word)]\n",
    "    \n",
    "    # Remove GST-tagg\n",
    "    for i, word in enumerate(fusion_like):\n",
    "        if 'gst-' in word: \n",
    "            fusion_like[i] = word.replace('gst-', '')\n",
    "    \n",
    "    return fusion_like"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_mutation_type_words(tokens):\n",
    "    mutation_patterns = ['truncation', 'deletion', 'promoter', \n",
    "                     'amplification', 'epigenetic', 'frame', \n",
    "                     'overexpression', 'duplication', 'insertion',\n",
    "                     'subtype', 'fusion', 'splice', 'wildtype']\n",
    "    \n",
    "    mutation_type_words = [word.lower() for word in tokens if word.lower() in mutation_patterns]\n",
    "    \n",
    "    return mutation_type_words"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_gene_like_words(text, commoners):\n",
    "    text_white = text.encode().decode()  # copy a string?!\n",
    "    text_white = text_white.replace('\"', ' ')\n",
    "    text_white = text_white.replace('.', ' ')\n",
    "    text_white = text_white.replace('/', ' ')\n",
    "    text_white = text_white.replace('\\'', ' ')\n",
    "    text_white = text_white.replace('_', ' ')    # This should be done after _pos, etc checked\n",
    "    text_white = text_white.replace('-', ' ')    # This should be done after -pos, etc checked\n",
    "    text_white = text_white.replace('=', ' ')\n",
    "    text_white = text_white.replace('\\n', ' ')\n",
    "    text_white = text_white.replace('\\\\n', ' ')\n",
    "    text_white = text_white.replace('\\'', ' ')\n",
    "    text_white = re.sub(' +',' ', text_white) \n",
    "    text_white = text_white.replace('\\'', ' ')\n",
    "    text_white = text_white.replace('(', ' ')\n",
    "    text_white = text_white.replace(')', ' ')\n",
    "    text_white = text_white.replace('[', ' ')\n",
    "    text_white = text_white.replace(']', ' ')\n",
    "    text_white = text_white.replace('{', ' ')\n",
    "    text_white = text_white.replace('}', ' ')\n",
    "\n",
    "    tokens_white = word_tokenize(text_white)\n",
    "    \n",
    "    # Remove nucleotide sequences\n",
    "    nts = r\"[ACTGU]{4,}\"\n",
    "    \n",
    "    tokens_white = [word for word in tokens_white \\\n",
    "                    if not word in stop_words \\\n",
    "                    if not word in commoners \\\n",
    "                    if not re.search(nts, word)]\n",
    "    \n",
    "    gene_ish_pattern1 = r\"[A-Z]{2,7}[0-9]{0,4}$\"\n",
    "    gene_ish_pattern2 = r\"p[0-9]{1,3}$\"\n",
    "    gene_ish_words = [word for word in tokens_white if re.match(gene_ish_pattern1, word)]\n",
    "    gene_ish_words += [word for word in tokens_white if re.match(gene_ish_pattern2, word)]\n",
    "    \n",
    "    return gene_ish_words"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def count_appearances(list_of_words):\n",
    "    unique_words = set(list_of_words)\n",
    "    appearances = {}\n",
    "    for unique_word in unique_words:\n",
    "        count = len([word for word in list_of_words if word == unique_word])\n",
    "        appearances[unique_word] = count\n",
    "    \n",
    "    return appearances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "text1 = text_train.loc[2105, 'Text']\n",
    "tokens1 = word_tokenize(text1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'mhc-positive': 2}"
      ]
     },
     "execution_count": 244,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_appearances(get_positive_and_negative_words(tokens1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'crl-1772': 1,\n",
       " 'ct-tc': 1,\n",
       " 'hg-u133': 1,\n",
       " 'mck-cat': 2,\n",
       " 'mf-20': 1,\n",
       " 'mlp-cat': 4,\n",
       " 'p3300mck-cat': 1,\n",
       " 'pax-foxo1': 3,\n",
       " 'pax3-foxo1': 1,\n",
       " 'pax7-foxo1': 1,\n",
       " 'pmlp-cat': 3,\n",
       " 'sms-ctr': 1}"
      ]
     },
     "execution_count": 245,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_appearances(get_fusion_like_words(tokens1, commoners))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'amplification': 3,\n",
       " 'duplication': 1,\n",
       " 'fusion': 8,\n",
       " 'promoter': 9,\n",
       " 'splice': 1,\n",
       " 'subtype': 1,\n",
       " 'wildtype': 1}"
      ]
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_appearances(get_mutation_type_words(tokens1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ABC': 1,\n",
       " 'AD': 2,\n",
       " 'AKT': 3,\n",
       " 'AL': 1,\n",
       " 'ALK': 1,\n",
       " 'ALN': 1,\n",
       " 'ALT': 1,\n",
       " 'AND': 1,\n",
       " 'AP4': 1,\n",
       " 'ARMS': 8,\n",
       " 'BAM': 1,\n",
       " 'BD': 2,\n",
       " 'BJ': 6,\n",
       " 'BWA': 3,\n",
       " 'CA': 11,\n",
       " 'CANNTG': 3,\n",
       " 'CASAVA': 1,\n",
       " 'CAT': 21,\n",
       " 'CBFJ': 1,\n",
       " 'CC': 2,\n",
       " 'CC1': 1,\n",
       " 'CGG': 1,\n",
       " 'CLN': 1,\n",
       " 'CMV': 3,\n",
       " 'CMYC': 1,\n",
       " 'CRL': 1,\n",
       " 'CT': 1,\n",
       " 'CTG': 1,\n",
       " 'CTR': 1,\n",
       " 'CZ': 1,\n",
       " 'DAPI': 1,\n",
       " 'DI': 1,\n",
       " 'DMEM': 6,\n",
       " 'DP': 1,\n",
       " 'DSHB': 1,\n",
       " 'EDTA': 2,\n",
       " 'EMD': 1,\n",
       " 'EO': 3,\n",
       " 'ERMS': 37,\n",
       " 'EXOME': 1,\n",
       " 'FASTQ': 1,\n",
       " 'FASTX': 2,\n",
       " 'FBS': 3,\n",
       " 'FFPE': 4,\n",
       " 'FGFR4': 2,\n",
       " 'FI': 1,\n",
       " 'FIG': 4,\n",
       " 'FOXO1': 5,\n",
       " 'GAC': 4,\n",
       " 'GAPDH': 1,\n",
       " 'GATK': 8,\n",
       " 'GSEA': 5,\n",
       " 'GTF': 1,\n",
       " 'HA': 2,\n",
       " 'HG': 1,\n",
       " 'HLH': 4,\n",
       " 'IA': 1,\n",
       " 'IHC': 2,\n",
       " 'II': 1,\n",
       " 'IIII': 1,\n",
       " 'IL': 1,\n",
       " 'INDEL': 1,\n",
       " 'JAF': 1,\n",
       " 'JB': 1,\n",
       " 'LOH': 2,\n",
       " 'MA': 3,\n",
       " 'MAB3878': 1,\n",
       " 'MACS': 2,\n",
       " 'MALDI': 3,\n",
       " 'MAPQ': 1,\n",
       " 'MCK': 9,\n",
       " 'METHODS': 2,\n",
       " 'MF': 1,\n",
       " 'MF20': 1,\n",
       " 'MHC': 3,\n",
       " 'MI': 1,\n",
       " 'MLP': 8,\n",
       " 'MLPCAT': 1,\n",
       " 'MO': 3,\n",
       " 'MRF4': 2,\n",
       " 'MS': 4,\n",
       " 'MSKCC': 4,\n",
       " 'MYC': 23,\n",
       " 'MYCN': 2,\n",
       " 'MYF5': 3,\n",
       " 'MYF6': 1,\n",
       " 'MYOD': 1,\n",
       " 'MYOD1': 110,\n",
       " 'MYOG': 3,\n",
       " 'ND': 6,\n",
       " 'NJ': 1,\n",
       " 'NRAF': 4,\n",
       " 'OM': 1,\n",
       " 'PA': 1,\n",
       " 'PAX': 3,\n",
       " 'PAX3': 1,\n",
       " 'PAX7': 1,\n",
       " 'PER': 1,\n",
       " 'PI3': 1,\n",
       " 'PICARD': 3,\n",
       " 'PM': 1,\n",
       " 'PMC': 1,\n",
       " 'PTEN': 2,\n",
       " 'PVDF': 1,\n",
       " 'RAS': 2,\n",
       " 'RD': 1,\n",
       " 'REF': 1,\n",
       " 'RESULTS': 1,\n",
       " 'RIN': 1,\n",
       " 'RMA': 1,\n",
       " 'RMS': 13,\n",
       " 'RMS559': 1,\n",
       " 'RS': 1,\n",
       " 'SAM': 3,\n",
       " 'SAMSE': 1,\n",
       " 'SBS': 2,\n",
       " 'SE': 2,\n",
       " 'SK': 1,\n",
       " 'SMS': 1,\n",
       " 'SNP': 2,\n",
       " 'SYBR': 1,\n",
       " 'TBST': 1,\n",
       " 'TC': 1,\n",
       " 'TMM': 1,\n",
       " 'TOF': 3,\n",
       " 'TUMOR': 1,\n",
       " 'UCSC': 1,\n",
       " 'USF': 1,\n",
       " 'UT': 1,\n",
       " 'VA': 2,\n",
       " 'VCF': 1,\n",
       " 'WT': 4,\n",
       " 'WTS': 2,\n",
       " 'XT': 2}"
      ]
     },
     "execution_count": 247,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_appearances(get_gene_like_words(text1, commoners))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='sec4'></a>\n",
    "# Process Features (<a href='#sec0'>Back To Top</a>)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 7min 12s, sys: 8.84 ms, total: 7min 12s\n",
      "Wall time: 7min 12s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "pos_neg_list = []\n",
    "fusion_like_list = []\n",
    "mutation_type_list = []\n",
    "gene_like_list = []\n",
    "\n",
    "for i in range(len(text_train)):\n",
    "    text = text_train.loc[i, 'Text']\n",
    "    tokens = word_tokenize(text)\n",
    "    \n",
    "    pos_neg_list.append(count_appearances(get_positive_and_negative_words(tokens)))\n",
    "    fusion_like_list.append(count_appearances(get_fusion_like_words(tokens, commoners)))\n",
    "    mutation_type_list.append(count_appearances(get_mutation_type_words(tokens)))\n",
    "    gene_like_list.append(count_appearances(get_gene_like_words(text, commoners)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pos_neg_table = pd.DataFrame(pos_neg_list)\n",
    "gene_like_table = pd.DataFrame(gene_like_list)\n",
    "fusion_like_table = pd.DataFrame(fusion_like_list)\n",
    "mutation_type_table = pd.DataFrame(mutation_type_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pos_neg_table = pos_neg_table.fillna(value=0)\n",
    "gene_like_table = gene_like_table.fillna(value=0)\n",
    "fusion_like_table = fusion_like_table.fillna(value=0)\n",
    "mutation_type_table = mutation_type_table.fillna(value=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>'dominant-negative</th>\n",
       "      <th>'mutation-negative</th>\n",
       "      <th>-stain-negative</th>\n",
       "      <th>.non-negative</th>\n",
       "      <th>/aml1-eto-positive</th>\n",
       "      <th>/dsred-positive</th>\n",
       "      <th>/kit-positive</th>\n",
       "      <th>10-67-positive</th>\n",
       "      <th>1100delc-positive</th>\n",
       "      <th>11q-aupd-positive</th>\n",
       "      <th>...</th>\n",
       "      <th>zsgreen-positive</th>\n",
       "      <th>α-actinin-positive</th>\n",
       "      <th>α-negative</th>\n",
       "      <th>α-positive</th>\n",
       "      <th>β-galactosidase-positive</th>\n",
       "      <th>β1-tubulin-positive</th>\n",
       "      <th>β2m-negative</th>\n",
       "      <th>β2m-positive</th>\n",
       "      <th>γ-h2ax-positive</th>\n",
       "      <th>γh2ax-positive</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 765 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   'dominant-negative  'mutation-negative  -stain-negative  .non-negative  \\\n",
       "0                 0.0                 0.0              0.0            0.0   \n",
       "1                 0.0                 0.0              0.0            0.0   \n",
       "2                 0.0                 0.0              0.0            0.0   \n",
       "3                 0.0                 0.0              0.0            0.0   \n",
       "4                 0.0                 0.0              0.0            0.0   \n",
       "5                 0.0                 0.0              0.0            0.0   \n",
       "6                 0.0                 0.0              0.0            0.0   \n",
       "7                 0.0                 0.0              0.0            0.0   \n",
       "8                 0.0                 0.0              0.0            0.0   \n",
       "9                 0.0                 0.0              0.0            0.0   \n",
       "\n",
       "   /aml1-eto-positive  /dsred-positive  /kit-positive  10-67-positive  \\\n",
       "0                 0.0              0.0            0.0             0.0   \n",
       "1                 0.0              0.0            0.0             0.0   \n",
       "2                 0.0              0.0            0.0             0.0   \n",
       "3                 0.0              0.0            0.0             0.0   \n",
       "4                 0.0              0.0            0.0             0.0   \n",
       "5                 0.0              0.0            0.0             0.0   \n",
       "6                 0.0              0.0            0.0             0.0   \n",
       "7                 0.0              0.0            0.0             0.0   \n",
       "8                 0.0              0.0            0.0             0.0   \n",
       "9                 0.0              0.0            0.0             0.0   \n",
       "\n",
       "   1100delc-positive  11q-aupd-positive       ...        zsgreen-positive  \\\n",
       "0                0.0                0.0       ...                     0.0   \n",
       "1                0.0                0.0       ...                     0.0   \n",
       "2                0.0                0.0       ...                     0.0   \n",
       "3                0.0                0.0       ...                     0.0   \n",
       "4                0.0                0.0       ...                     0.0   \n",
       "5                0.0                0.0       ...                     0.0   \n",
       "6                0.0                0.0       ...                     0.0   \n",
       "7                0.0                0.0       ...                     0.0   \n",
       "8                0.0                0.0       ...                     0.0   \n",
       "9                0.0                0.0       ...                     0.0   \n",
       "\n",
       "   α-actinin-positive  α-negative  α-positive  β-galactosidase-positive  \\\n",
       "0                 0.0         0.0         0.0                       0.0   \n",
       "1                 0.0         0.0         0.0                       0.0   \n",
       "2                 0.0         0.0         0.0                       0.0   \n",
       "3                 0.0         0.0         0.0                       0.0   \n",
       "4                 0.0         0.0         0.0                       0.0   \n",
       "5                 0.0         0.0         0.0                       0.0   \n",
       "6                 0.0         0.0         0.0                       0.0   \n",
       "7                 0.0         0.0         0.0                       0.0   \n",
       "8                 0.0         0.0         0.0                       0.0   \n",
       "9                 0.0         0.0         0.0                       0.0   \n",
       "\n",
       "   β1-tubulin-positive  β2m-negative  β2m-positive  γ-h2ax-positive  \\\n",
       "0                  0.0           0.0           0.0              0.0   \n",
       "1                  0.0           0.0           0.0              0.0   \n",
       "2                  0.0           0.0           0.0              0.0   \n",
       "3                  0.0           0.0           0.0              0.0   \n",
       "4                  0.0           0.0           0.0              0.0   \n",
       "5                  0.0           0.0           0.0              0.0   \n",
       "6                  0.0           0.0           0.0              0.0   \n",
       "7                  0.0           0.0           0.0              0.0   \n",
       "8                  0.0           0.0           0.0              0.0   \n",
       "9                  0.0           0.0           0.0              0.0   \n",
       "\n",
       "   γh2ax-positive  \n",
       "0             0.0  \n",
       "1             0.0  \n",
       "2             0.0  \n",
       "3             0.0  \n",
       "4             0.0  \n",
       "5             0.0  \n",
       "6             0.0  \n",
       "7             0.0  \n",
       "8             0.0  \n",
       "9             0.0  \n",
       "\n",
       "[10 rows x 765 columns]"
      ]
     },
     "execution_count": 251,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pos_neg_table.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA</th>\n",
       "      <th>AA0869</th>\n",
       "      <th>AA1077</th>\n",
       "      <th>AA1370</th>\n",
       "      <th>AA2</th>\n",
       "      <th>AA26</th>\n",
       "      <th>AA2711</th>\n",
       "      <th>AA2800</th>\n",
       "      <th>AA4345</th>\n",
       "      <th>AA491</th>\n",
       "      <th>...</th>\n",
       "      <th>p815</th>\n",
       "      <th>p83</th>\n",
       "      <th>p84</th>\n",
       "      <th>p85</th>\n",
       "      <th>p86</th>\n",
       "      <th>p89</th>\n",
       "      <th>p9</th>\n",
       "      <th>p90</th>\n",
       "      <th>p95</th>\n",
       "      <th>p97</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 18295 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    AA  AA0869  AA1077  AA1370  AA2  AA26  AA2711  AA2800  AA4345  AA491 ...   \\\n",
       "0  0.0     0.0     0.0     0.0  0.0   0.0     0.0     0.0     0.0    0.0 ...    \n",
       "1  0.0     0.0     0.0     0.0  0.0   0.0     0.0     0.0     0.0    0.0 ...    \n",
       "2  0.0     0.0     0.0     0.0  0.0   0.0     0.0     0.0     0.0    0.0 ...    \n",
       "3  0.0     0.0     0.0     0.0  0.0   0.0     0.0     0.0     0.0    0.0 ...    \n",
       "4  0.0     0.0     0.0     0.0  0.0   0.0     0.0     0.0     0.0    0.0 ...    \n",
       "5  0.0     0.0     0.0     0.0  0.0   0.0     0.0     0.0     0.0    0.0 ...    \n",
       "6  0.0     0.0     0.0     0.0  0.0   0.0     0.0     0.0     0.0    0.0 ...    \n",
       "7  0.0     0.0     0.0     0.0  0.0   0.0     0.0     0.0     0.0    0.0 ...    \n",
       "8  0.0     0.0     0.0     0.0  0.0   0.0     0.0     0.0     0.0    0.0 ...    \n",
       "9  0.0     0.0     0.0     0.0  0.0   0.0     0.0     0.0     0.0    0.0 ...    \n",
       "\n",
       "   p815  p83  p84  p85  p86  p89   p9  p90  p95  p97  \n",
       "0   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "1   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "2   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "4   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "5   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "6   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "7   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "8   0.0  0.0  0.0  3.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "9   0.0  0.0  0.0  3.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "\n",
       "[10 rows x 18295 columns]"
      ]
     },
     "execution_count": 252,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_like_table.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>'251i-labeledfo-</th>\n",
       "      <th>'whole-exome</th>\n",
       "      <th>**pax5-jak2</th>\n",
       "      <th>+ca-alk5</th>\n",
       "      <th>+il-3</th>\n",
       "      <th>-aif4-</th>\n",
       "      <th>-akt-mtor</th>\n",
       "      <th>-akt-s6</th>\n",
       "      <th>-bcor-rara</th>\n",
       "      <th>...</th>\n",
       "      <th>δbrip1-pcdna-dest47</th>\n",
       "      <th>δc·abt-737</th>\n",
       "      <th>δegf-myc</th>\n",
       "      <th>δg466vb-raf</th>\n",
       "      <th>δgh2oi-n</th>\n",
       "      <th>δgh2ou-i</th>\n",
       "      <th>δhm-akt1</th>\n",
       "      <th>δpr-lbd</th>\n",
       "      <th>δzf-ha</th>\n",
       "      <th>∆wd40-jak2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 7789 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        '251i-labeledfo-  'whole-exome  **pax5-jak2  +ca-alk5  +il-3  -aif4-  \\\n",
       "0  0.0               0.0           0.0          0.0       0.0    0.0     0.0   \n",
       "1  0.0               0.0           0.0          0.0       0.0    0.0     0.0   \n",
       "2  0.0               0.0           0.0          0.0       0.0    0.0     0.0   \n",
       "3  0.0               0.0           0.0          0.0       0.0    0.0     0.0   \n",
       "4  0.0               0.0           0.0          0.0       0.0    0.0     0.0   \n",
       "5  0.0               0.0           0.0          0.0       0.0    0.0     0.0   \n",
       "6  0.0               0.0           0.0          0.0       0.0    0.0     0.0   \n",
       "7  0.0               0.0           0.0          0.0       0.0    0.0     0.0   \n",
       "8  0.0               0.0           0.0          0.0       0.0    0.0     0.0   \n",
       "9  0.0               0.0           0.0          0.0       0.0    0.0     0.0   \n",
       "\n",
       "   -akt-mtor  -akt-s6  -bcor-rara     ...      δbrip1-pcdna-dest47  \\\n",
       "0        0.0      0.0         0.0     ...                      0.0   \n",
       "1        0.0      0.0         0.0     ...                      0.0   \n",
       "2        0.0      0.0         0.0     ...                      0.0   \n",
       "3        0.0      0.0         0.0     ...                      0.0   \n",
       "4        0.0      0.0         0.0     ...                      0.0   \n",
       "5        0.0      0.0         0.0     ...                      0.0   \n",
       "6        0.0      0.0         0.0     ...                      0.0   \n",
       "7        0.0      0.0         0.0     ...                      0.0   \n",
       "8        0.0      0.0         0.0     ...                      0.0   \n",
       "9        0.0      0.0         0.0     ...                      0.0   \n",
       "\n",
       "   δc·abt-737  δegf-myc  δg466vb-raf  δgh2oi-n  δgh2ou-i  δhm-akt1  δpr-lbd  \\\n",
       "0         0.0       0.0          0.0       0.0       0.0       0.0      0.0   \n",
       "1         0.0       0.0          0.0       0.0       0.0       0.0      0.0   \n",
       "2         0.0       0.0          0.0       0.0       0.0       0.0      0.0   \n",
       "3         0.0       0.0          0.0       0.0       0.0       0.0      0.0   \n",
       "4         0.0       0.0          0.0       0.0       0.0       0.0      0.0   \n",
       "5         0.0       0.0          0.0       0.0       0.0       0.0      0.0   \n",
       "6         0.0       0.0          0.0       0.0       0.0       0.0      0.0   \n",
       "7         0.0       0.0          0.0       0.0       0.0       0.0      0.0   \n",
       "8         0.0       0.0          0.0       0.0       0.0       0.0      0.0   \n",
       "9         0.0       0.0          0.0       0.0       0.0       0.0      0.0   \n",
       "\n",
       "   δzf-ha  ∆wd40-jak2  \n",
       "0     0.0         0.0  \n",
       "1     0.0         0.0  \n",
       "2     0.0         0.0  \n",
       "3     0.0         0.0  \n",
       "4     0.0         0.0  \n",
       "5     0.0         0.0  \n",
       "6     0.0         0.0  \n",
       "7     0.0         0.0  \n",
       "8     0.0         0.0  \n",
       "9     0.0         0.0  \n",
       "\n",
       "[10 rows x 7789 columns]"
      ]
     },
     "execution_count": 253,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fusion_like_table.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>amplification</th>\n",
       "      <th>deletion</th>\n",
       "      <th>duplication</th>\n",
       "      <th>epigenetic</th>\n",
       "      <th>frame</th>\n",
       "      <th>fusion</th>\n",
       "      <th>insertion</th>\n",
       "      <th>overexpression</th>\n",
       "      <th>promoter</th>\n",
       "      <th>splice</th>\n",
       "      <th>subtype</th>\n",
       "      <th>truncation</th>\n",
       "      <th>wildtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   amplification  deletion  duplication  epigenetic  frame  fusion  insertion  \\\n",
       "0            0.0      12.0          0.0         0.0    0.0     5.0        1.0   \n",
       "1            0.0       4.0          0.0         0.0    0.0     0.0        0.0   \n",
       "2            0.0       4.0          0.0         0.0    0.0     0.0        0.0   \n",
       "3            2.0       5.0          1.0         0.0    0.0     0.0        0.0   \n",
       "4            0.0       0.0          0.0         0.0    0.0     0.0        0.0   \n",
       "5            0.0       0.0          0.0         0.0    0.0     0.0        0.0   \n",
       "6            0.0       0.0          0.0         0.0    0.0     0.0        0.0   \n",
       "7            2.0      28.0          1.0         0.0    7.0     3.0        2.0   \n",
       "8            0.0       0.0          0.0         0.0    0.0     0.0        0.0   \n",
       "9            0.0       0.0          0.0         0.0    0.0     0.0        0.0   \n",
       "\n",
       "   overexpression  promoter  splice  subtype  truncation  wildtype  \n",
       "0             2.0       0.0     5.0      0.0         1.0       0.0  \n",
       "1             0.0       0.0     1.0      0.0         0.0       0.0  \n",
       "2             0.0       0.0     1.0      0.0         0.0       0.0  \n",
       "3             4.0       0.0     0.0      0.0         0.0       0.0  \n",
       "4             0.0       0.0     0.0      0.0         0.0       0.0  \n",
       "5             0.0       0.0     0.0      0.0         0.0       0.0  \n",
       "6             0.0       0.0     0.0      0.0         0.0       0.0  \n",
       "7             4.0       0.0    13.0      1.0         0.0       0.0  \n",
       "8             0.0       0.0     0.0      0.0         0.0       0.0  \n",
       "9             0.0       0.0     0.0      0.0         0.0       0.0  "
      ]
     },
     "execution_count": 254,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutation_type_table.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3321, 765)\n",
      "(3321, 18295)\n",
      "(3321, 7789)\n",
      "(3321, 13)\n"
     ]
    }
   ],
   "source": [
    "print(pos_neg_table.shape)\n",
    "print(gene_like_table.shape)\n",
    "print(fusion_like_table.shape)\n",
    "print(mutation_type_table.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 478,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from sklearn.preprocessing import OneHotEncoder, LabelEncoder\n",
    "from sklearn.preprocessing import scale, normalize, robust_scale\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import accuracy_score, recall_score, precision_score, f1_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Get Gene feature from 'train_variants' data\n",
    "X_gene = np.array(class_train.Gene)\n",
    "X_gene_int = LabelEncoder().fit_transform(X_gene.ravel()).reshape(-1, 1)\n",
    "X_gene_bin = OneHotEncoder().fit_transform(X_gene_int).toarray()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "full_gene_table = pd.DataFrame(X_gene_bin)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3321, 264)"
      ]
     },
     "execution_count": 260,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_gene_table.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>Combined All!</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "features = pd.concat([pos_neg_table, \n",
    "                      gene_like_table,\n",
    "                      fusion_like_table,\n",
    "                      mutation_type_table,\n",
    "                      full_gene_table],\n",
    "                      axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3321, 27126)"
      ]
     },
     "execution_count": 263,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3321,)"
      ]
     },
     "execution_count": 264,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class_train.Class.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='sec5'></a>\n",
    "# Test with Random Forest (<a href='#sec0'>Back To Top</a>)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 577,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X = np.array(features).astype(float)\n",
    "y = np.array(class_train.Class).astype(int).ravel()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>un-scaled</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 578,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.15, random_state=55)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 579,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 15.2 s, sys: 93.3 ms, total: 15.3 s\n",
      "Wall time: 4.08 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "rfc1 = RandomForestClassifier(n_estimators=50, max_depth=100, n_jobs=4)\n",
    "rfc1.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 580,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "y_pred = rfc1.predict(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 581,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.685370741483\n"
     ]
    }
   ],
   "source": [
    "print(accuracy_score(y_test, y_pred))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "<b>scaled</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 582,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X_scaled = scale(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 583,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.15, random_state=55)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 584,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 13.5 s, sys: 66.8 ms, total: 13.6 s\n",
      "Wall time: 3.65 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "rfc2 = RandomForestClassifier(n_estimators=50, max_depth=100, n_jobs=4)\n",
    "rfc2.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 585,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "y_pred = rfc2.predict(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 586,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.661322645291\n"
     ]
    }
   ],
   "source": [
    "print(accuracy_score(y_test, y_pred))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "<b>normalized</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 587,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X_normed = normalize(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 588,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X_train, X_test, y_train, y_test = train_test_split(X_normed, y, test_size=0.15, random_state=55)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 589,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 14.5 s, sys: 76.2 ms, total: 14.6 s\n",
      "Wall time: 3.95 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "rfc3 = RandomForestClassifier(n_estimators=50, max_depth=100, n_jobs=4)\n",
    "rfc3.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 590,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "y_pred = rfc3.predict(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 591,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.673346693387\n"
     ]
    }
   ],
   "source": [
    "print(accuracy_score(y_test, y_pred))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>Select Features from the Best RFC above</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 592,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "importances = rfc1.feature_importances_\n",
    "indices = np.argsort(importances)[::-1]\n",
    "features1000 = indices[:1000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 593,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X100 = X[:, features1000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 594,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X_train, X_test, y_train, y_test = train_test_split(X100, y, test_size=0.15, random_state=55)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 595,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 742 ms, sys: 5.03 ms, total: 747 ms\n",
      "Wall time: 231 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "rfc4 = RandomForestClassifier(n_estimators=50, max_depth=100, n_jobs=4)\n",
    "rfc4.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 596,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "y_pred = rfc4.predict(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 597,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.663326653307\n"
     ]
    }
   ],
   "source": [
    "print(accuracy_score(y_test, y_pred))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "<b>RobustScale</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 598,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X_robust = robust_scale(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 599,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X_train, X_test, y_train, y_test = train_test_split(X_robust, y, test_size=0.15, random_state=55)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 600,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 13.2 s, sys: 58.8 ms, total: 13.3 s\n",
      "Wall time: 3.55 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "rfc5 = RandomForestClassifier(n_estimators=50, max_depth=100, n_jobs=4)\n",
    "rfc5.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 601,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "y_pred = rfc5.predict(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 602,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.657314629259\n"
     ]
    }
   ],
   "source": [
    "print(accuracy_score(y_test, y_pred))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GridSearchCV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 603,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from sklearn.model_selection import GridSearchCV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 628,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "parameters = {'max_depth':[40, 60, 80, 100], 'criterion':['gini', 'entropy'], \n",
    "              'class_weight':[None, 'balanced', 'balanced_subsample']}\n",
    "rfc = RandomForestClassifier(n_jobs=1, n_estimators=60)\n",
    "clf = GridSearchCV(rfc, parameters, n_jobs=4, verbose=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 629,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.15, random_state=55)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 630,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting 3 folds for each of 24 candidates, totalling 72 fits\n",
      "[CV] class_weight=None, criterion=gini, max_depth=40 .................\n",
      "[CV] class_weight=None, criterion=gini, max_depth=40 .................\n",
      "[CV] class_weight=None, criterion=gini, max_depth=40 .................\n",
      "[CV] class_weight=None, criterion=gini, max_depth=60 .................\n",
      "[CV] .. class_weight=None, criterion=gini, max_depth=40, total=   8.5s\n",
      "[CV] class_weight=None, criterion=gini, max_depth=60 .................\n",
      "[CV] .. class_weight=None, criterion=gini, max_depth=40, total=   8.1s\n",
      "[CV] class_weight=None, criterion=gini, max_depth=60 .................\n",
      "[CV] .. class_weight=None, criterion=gini, max_depth=40, total=   8.7s\n",
      "[CV] class_weight=None, criterion=gini, max_depth=80 .................\n",
      "[CV] .. class_weight=None, criterion=gini, max_depth=60, total=  10.2s\n",
      "[CV] class_weight=None, criterion=gini, max_depth=80 .................\n",
      "[CV] .. class_weight=None, criterion=gini, max_depth=60, total=   9.8s\n",
      "[CV] class_weight=None, criterion=gini, max_depth=80 .................\n",
      "[CV] .. class_weight=None, criterion=gini, max_depth=60, total=  10.1s\n",
      "[CV] class_weight=None, criterion=gini, max_depth=100 ................\n",
      "[CV] .. class_weight=None, criterion=gini, max_depth=80, total=  11.0s\n",
      "[CV] class_weight=None, criterion=gini, max_depth=100 ................\n",
      "[CV] .. class_weight=None, criterion=gini, max_depth=80, total=  10.5s\n",
      "[CV] class_weight=None, criterion=gini, max_depth=100 ................\n",
      "[CV] .. class_weight=None, criterion=gini, max_depth=80, total=  10.7s\n",
      "[CV] class_weight=None, criterion=entropy, max_depth=40 ..............\n",
      "[CV] . class_weight=None, criterion=gini, max_depth=100, total=  10.9s\n",
      "[CV] class_weight=None, criterion=entropy, max_depth=40 ..............\n",
      "[CV] . class_weight=None, criterion=gini, max_depth=100, total=  10.9s\n",
      "[CV] class_weight=None, criterion=entropy, max_depth=40 ..............\n",
      "[CV] . class_weight=None, criterion=gini, max_depth=100, total=  11.0s\n",
      "[CV] class_weight=None, criterion=entropy, max_depth=60 ..............\n",
      "[CV]  class_weight=None, criterion=entropy, max_depth=40, total=   9.3s\n",
      "[CV] class_weight=None, criterion=entropy, max_depth=60 ..............\n",
      "[CV]  class_weight=None, criterion=entropy, max_depth=40, total=   9.3s\n",
      "[CV] class_weight=None, criterion=entropy, max_depth=60 ..............\n",
      "[CV]  class_weight=None, criterion=entropy, max_depth=40, total=   9.4s\n",
      "[CV] class_weight=None, criterion=entropy, max_depth=80 ..............\n",
      "[CV]  class_weight=None, criterion=entropy, max_depth=60, total=  10.3s\n",
      "[CV] class_weight=None, criterion=entropy, max_depth=80 ..............\n",
      "[CV]  class_weight=None, criterion=entropy, max_depth=60, total=  10.3s\n",
      "[CV] class_weight=None, criterion=entropy, max_depth=80 ..............\n",
      "[CV]  class_weight=None, criterion=entropy, max_depth=60, total=  10.5s\n",
      "[CV] class_weight=None, criterion=entropy, max_depth=100 .............\n",
      "[CV]  class_weight=None, criterion=entropy, max_depth=80, total=  10.9s\n",
      "[CV] class_weight=None, criterion=entropy, max_depth=100 .............\n",
      "[CV]  class_weight=None, criterion=entropy, max_depth=80, total=  10.6s\n",
      "[CV] class_weight=None, criterion=entropy, max_depth=100 .............\n",
      "[CV]  class_weight=None, criterion=entropy, max_depth=80, total=  10.6s\n",
      "[CV] class_weight=balanced, criterion=gini, max_depth=40 .............\n",
      "[CV]  class_weight=None, criterion=entropy, max_depth=100, total=  10.6s\n",
      "[CV] class_weight=balanced, criterion=gini, max_depth=40 .............\n",
      "[CV]  class_weight=None, criterion=entropy, max_depth=100, total=  10.7s\n",
      "[CV] class_weight=balanced, criterion=gini, max_depth=40 .............\n",
      "[CV]  class_weight=None, criterion=entropy, max_depth=100, total=  10.8s\n",
      "[CV] class_weight=balanced, criterion=gini, max_depth=60 .............\n",
      "[CV]  class_weight=balanced, criterion=gini, max_depth=40, total=   8.7s\n",
      "[CV] class_weight=balanced, criterion=gini, max_depth=60 .............\n",
      "[CV]  class_weight=balanced, criterion=gini, max_depth=40, total=   8.8s\n",
      "[CV] class_weight=balanced, criterion=gini, max_depth=60 .............\n",
      "[CV]  class_weight=balanced, criterion=gini, max_depth=40, total=   9.4s\n",
      "[CV] class_weight=balanced, criterion=gini, max_depth=80 .............\n",
      "[CV]  class_weight=balanced, criterion=gini, max_depth=60, total=  10.7s\n",
      "[CV] class_weight=balanced, criterion=gini, max_depth=80 .............\n",
      "[CV]  class_weight=balanced, criterion=gini, max_depth=60, total=  10.6s\n",
      "[CV] class_weight=balanced, criterion=gini, max_depth=80 .............\n",
      "[CV]  class_weight=balanced, criterion=gini, max_depth=60, total=  10.6s\n",
      "[CV] class_weight=balanced, criterion=gini, max_depth=100 ............\n",
      "[CV]  class_weight=balanced, criterion=gini, max_depth=80, total=  11.6s\n",
      "[CV] class_weight=balanced, criterion=gini, max_depth=100 ............\n",
      "[CV]  class_weight=balanced, criterion=gini, max_depth=80, total=  11.5s\n",
      "[CV] class_weight=balanced, criterion=gini, max_depth=100 ............\n",
      "[CV]  class_weight=balanced, criterion=gini, max_depth=80, total=  11.6s\n",
      "[CV] class_weight=balanced, criterion=entropy, max_depth=40 ..........\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=4)]: Done  33 tasks      | elapsed:  1.5min\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[CV]  class_weight=balanced, criterion=gini, max_depth=100, total=  11.9s\n",
      "[CV] class_weight=balanced, criterion=entropy, max_depth=40 ..........\n",
      "[CV]  class_weight=balanced, criterion=gini, max_depth=100, total=  11.8s\n",
      "[CV] class_weight=balanced, criterion=entropy, max_depth=40 ..........\n",
      "[CV]  class_weight=balanced, criterion=gini, max_depth=100, total=  11.8s\n",
      "[CV]  class_weight=balanced, criterion=entropy, max_depth=40, total=   9.3s\n",
      "[CV] class_weight=balanced, criterion=entropy, max_depth=60 ..........\n",
      "[CV] class_weight=balanced, criterion=entropy, max_depth=60 ..........\n",
      "[CV]  class_weight=balanced, criterion=entropy, max_depth=40, total=   9.4s\n",
      "[CV] class_weight=balanced, criterion=entropy, max_depth=60 ..........\n",
      "[CV]  class_weight=balanced, criterion=entropy, max_depth=40, total=   9.9s\n",
      "[CV] class_weight=balanced, criterion=entropy, max_depth=80 ..........\n",
      "[CV]  class_weight=balanced, criterion=entropy, max_depth=60, total=  10.6s\n",
      "[CV] class_weight=balanced, criterion=entropy, max_depth=80 ..........\n",
      "[CV]  class_weight=balanced, criterion=entropy, max_depth=60, total=  10.8s\n",
      "[CV] class_weight=balanced, criterion=entropy, max_depth=80 ..........\n",
      "[CV]  class_weight=balanced, criterion=entropy, max_depth=60, total=  10.6s\n",
      "[CV] class_weight=balanced, criterion=entropy, max_depth=100 .........\n",
      "[CV]  class_weight=balanced, criterion=entropy, max_depth=80, total=  11.1s\n",
      "[CV] class_weight=balanced, criterion=entropy, max_depth=100 .........\n",
      "[CV]  class_weight=balanced, criterion=entropy, max_depth=80, total=  11.0s\n",
      "[CV] class_weight=balanced, criterion=entropy, max_depth=100 .........\n",
      "[CV]  class_weight=balanced, criterion=entropy, max_depth=80, total=  10.9s\n",
      "[CV] class_weight=balanced_subsample, criterion=gini, max_depth=40 ...\n",
      "[CV]  class_weight=balanced, criterion=entropy, max_depth=100, total=  11.2s\n",
      "[CV] class_weight=balanced_subsample, criterion=gini, max_depth=40 ...\n",
      "[CV]  class_weight=balanced, criterion=entropy, max_depth=100, total=  11.4s\n",
      "[CV] class_weight=balanced_subsample, criterion=gini, max_depth=40 ...\n",
      "[CV]  class_weight=balanced_subsample, criterion=gini, max_depth=40, total=   9.2s\n",
      "[CV] class_weight=balanced_subsample, criterion=gini, max_depth=60 ...\n",
      "[CV]  class_weight=balanced_subsample, criterion=gini, max_depth=40, total=   9.0s\n",
      "[CV] class_weight=balanced_subsample, criterion=gini, max_depth=60 ...\n",
      "[CV]  class_weight=balanced, criterion=entropy, max_depth=100, total=  13.0s\n",
      "[CV] class_weight=balanced_subsample, criterion=gini, max_depth=60 ...\n",
      "[CV]  class_weight=balanced_subsample, criterion=gini, max_depth=40, total=  10.2s\n",
      "[CV] class_weight=balanced_subsample, criterion=gini, max_depth=80 ...\n",
      "[CV]  class_weight=balanced_subsample, criterion=gini, max_depth=60, total=  12.2s\n",
      "[CV] class_weight=balanced_subsample, criterion=gini, max_depth=80 ...\n",
      "[CV]  class_weight=balanced_subsample, criterion=gini, max_depth=60, total=  11.9s\n",
      "[CV] class_weight=balanced_subsample, criterion=gini, max_depth=80 ...\n",
      "[CV]  class_weight=balanced_subsample, criterion=gini, max_depth=60, total=  11.5s\n",
      "[CV] class_weight=balanced_subsample, criterion=gini, max_depth=100 ..\n",
      "[CV]  class_weight=balanced_subsample, criterion=gini, max_depth=80, total=  11.8s\n",
      "[CV] class_weight=balanced_subsample, criterion=gini, max_depth=100 ..\n",
      "[CV]  class_weight=balanced_subsample, criterion=gini, max_depth=80, total=  11.7s\n",
      "[CV] class_weight=balanced_subsample, criterion=gini, max_depth=100 ..\n",
      "[CV]  class_weight=balanced_subsample, criterion=gini, max_depth=80, total=  11.8s\n",
      "[CV] class_weight=balanced_subsample, criterion=entropy, max_depth=40 \n",
      "[CV]  class_weight=balanced_subsample, criterion=gini, max_depth=100, total=  12.1s\n",
      "[CV] class_weight=balanced_subsample, criterion=entropy, max_depth=40 \n",
      "[CV]  class_weight=balanced_subsample, criterion=gini, max_depth=100, total=  12.2s\n",
      "[CV] class_weight=balanced_subsample, criterion=entropy, max_depth=40 \n",
      "[CV]  class_weight=balanced_subsample, criterion=entropy, max_depth=40, total=   9.5s\n",
      "[CV] class_weight=balanced_subsample, criterion=entropy, max_depth=60 \n",
      "[CV]  class_weight=balanced_subsample, criterion=gini, max_depth=100, total=  12.4s\n",
      "[CV] class_weight=balanced_subsample, criterion=entropy, max_depth=60 \n",
      "[CV]  class_weight=balanced_subsample, criterion=entropy, max_depth=40, total=   9.6s\n",
      "[CV] class_weight=balanced_subsample, criterion=entropy, max_depth=60 \n",
      "[CV]  class_weight=balanced_subsample, criterion=entropy, max_depth=40, total=   9.7s\n",
      "[CV] class_weight=balanced_subsample, criterion=entropy, max_depth=80 \n",
      "[CV]  class_weight=balanced_subsample, criterion=entropy, max_depth=60, total=  11.4s\n",
      "[CV] class_weight=balanced_subsample, criterion=entropy, max_depth=80 \n",
      "[CV]  class_weight=balanced_subsample, criterion=entropy, max_depth=60, total=  11.1s\n",
      "[CV] class_weight=balanced_subsample, criterion=entropy, max_depth=80 \n",
      "[CV]  class_weight=balanced_subsample, criterion=entropy, max_depth=60, total=  11.0s\n",
      "[CV] class_weight=balanced_subsample, criterion=entropy, max_depth=100 \n",
      "[CV]  class_weight=balanced_subsample, criterion=entropy, max_depth=80, total=  11.5s\n",
      "[CV] class_weight=balanced_subsample, criterion=entropy, max_depth=100 \n",
      "[CV]  class_weight=balanced_subsample, criterion=entropy, max_depth=80, total=  11.1s\n",
      "[CV] class_weight=balanced_subsample, criterion=entropy, max_depth=100 \n",
      "[CV]  class_weight=balanced_subsample, criterion=entropy, max_depth=80, total=  11.1s\n",
      "[CV]  class_weight=balanced_subsample, criterion=entropy, max_depth=100, total=  11.3s\n",
      "[CV]  class_weight=balanced_subsample, criterion=entropy, max_depth=100, total=  10.7s\n",
      "[CV]  class_weight=balanced_subsample, criterion=entropy, max_depth=100, total=  10.1s\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=4)]: Done  72 out of  72 | elapsed:  3.4min finished\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1min 14s, sys: 780 ms, total: 1min 15s\n",
      "Wall time: 3min 38s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "GridSearchCV(cv=None, error_score='raise',\n",
       "       estimator=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',\n",
       "            max_depth=None, max_features='auto', max_leaf_nodes=None,\n",
       "            min_impurity_split=1e-07, min_samples_leaf=1,\n",
       "            min_samples_split=2, min_weight_fraction_leaf=0.0,\n",
       "            n_estimators=60, n_jobs=1, oob_score=False, random_state=None,\n",
       "            verbose=0, warm_start=False),\n",
       "       fit_params={}, iid=True, n_jobs=4,\n",
       "       param_grid={'max_depth': [40, 60, 80, 100], 'criterion': ['gini', 'entropy'], 'class_weight': [None, 'balanced', 'balanced_subsample']},\n",
       "       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,\n",
       "       scoring=None, verbose=2)"
      ]
     },
     "execution_count": 630,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "clf.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 631,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "best_clf = clf.best_estimator_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 632,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',\n",
       "            max_depth=80, max_features='auto', max_leaf_nodes=None,\n",
       "            min_impurity_split=1e-07, min_samples_leaf=1,\n",
       "            min_samples_split=2, min_weight_fraction_leaf=0.0,\n",
       "            n_estimators=60, n_jobs=1, oob_score=False, random_state=None,\n",
       "            verbose=0, warm_start=False)"
      ]
     },
     "execution_count": 632,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_clf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 633,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "y_pred = best_clf.predict(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 634,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.649298597194\n"
     ]
    }
   ],
   "source": [
    "print(accuracy_score(y_test, y_pred))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
