{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='sec0'></a>\n",
    "# Text Analysis1\n",
    "- Importing Data\n",
    "- <a href='#sec1'>Exemplary Text Analysis for Row3</a>\n",
    "- <a href='#sec2'>Write function to get gene-ish words list and mutation type table</a>\n",
    "- <a href='#sec3'>Compiling the entire text-ome - testing</a>\n",
    "- <a href='#sec4'>Compiling the entire text-ome - full mutation table</a>\n",
    "- <a href='#sec5'>Compiling the entire text-ome - full gene-like words table</a>\n",
    "- <a href='#sec6'>Compiling the entire gene-ome - full gene table (not genome)</a>\n",
    "- <a href='#sec7'>Convert Mutation_Types in Class file</a>\n",
    "- <a href='#sec8'>Combined All!</a>\n",
    "- <a href='#sec9'>Test with Random Forest</a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import re\n",
    "\n",
    "from nltk import PorterStemmer\n",
    "from nltk.tokenize import word_tokenize\n",
    "from nltk.corpus import stopwords\n",
    "\n",
    "sns.set_context(\"paper\")\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>Importing train_text</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class_train = pd.read_csv('train_variants')\n",
    "text_train = pd.read_csv(\"train_text\", sep=\"\\|\\|\", engine='python', header=None, skiprows=1, names=[\"ID\",\"Text\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>FAM58A</td>\n",
       "      <td>Truncating Mutations</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>CBL</td>\n",
       "      <td>W802*</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>CBL</td>\n",
       "      <td>Q249E</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>CBL</td>\n",
       "      <td>N454D</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>CBL</td>\n",
       "      <td>L399V</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID    Gene             Variation  Class\n",
       "0   0  FAM58A  Truncating Mutations      1\n",
       "1   1     CBL                 W802*      2\n",
       "2   2     CBL                 Q249E      2\n",
       "3   3     CBL                 N454D      3\n",
       "4   4     CBL                 L399V      4"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class_train.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Cyclin-dependent kinases (CDKs) regulate a var...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>Recent evidence has demonstrated that acquired...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>Oncogenic mutations in the monomeric Casitas B...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID                                               Text\n",
       "0   0  Cyclin-dependent kinases (CDKs) regulate a var...\n",
       "1   1   Abstract Background  Non-small cell lung canc...\n",
       "2   2   Abstract Background  Non-small cell lung canc...\n",
       "3   3  Recent evidence has demonstrated that acquired...\n",
       "4   4  Oncogenic mutations in the monomeric Casitas B..."
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text_train.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='sec1'></a>\n",
    "# Exemplary Text Analysis for Row3 (<a href='#sec0'>Back To Top</a>)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "txt1 = text_train.iloc[3, 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ID               3\n",
       "Gene           CBL\n",
       "Variation    N454D\n",
       "Class            3\n",
       "Name: 3, dtype: object"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class_train.iloc[3, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Recent evidence has demonstrated that acquired uniparental disomy (aUPD) is a novel mechanism by which pathogenetic mutations in cancer may be reduced to homozygosity. To help identify novel mutations in myeloproliferative neoplasms (MPNs), we performed a genome-wide single nucleotide polymorphism (SNP) screen to identify aUPD in 58 patients with atypical chronic myeloid leukemia (aCML; n = 30), JAK2 mutation–negative myelofibrosis (MF; n = 18), or JAK2 mutation–negative polycythemia vera (PV; n = 10). Stretches of homozygous, copy neutral SNP calls greater than 20Mb were seen in 10 (33%) aCML and 1 (6%) MF, but were absent in PV. In total, 7 different chromosomes were involved with 7q and 11q each affected in 10% of aCML cases. CBL mutations were identified in all 3 cases with 11q aUPD and analysis of 574 additional MPNs revealed a total of 27 CBL variants in 26 patients with aCML, myelofibrosis or chronic myelomonocytic leukemia. Most variants were missense substitutions in the RING or linker domains that abrogated CBL ubiquitin ligase activity and conferred a proliferative advantage to 32D cells overexpressing FLT3. We conclude that acquired, transforming CBL mutations are a novel and widespread pathogenetic abnormality in morphologically related, clinically aggressive MPNs.  Introduction  Myeloproliferative neoplasms (MPNs) are clonal hematopoietic stem cell disorders characterized by overproliferation of one or more myeloid cell lineages in the bone marrow and increased numbers of mature and immature myeloid cells in the peripheral blood. Excess proliferation is frequently associated with splenomegaly and cardiovascular complications as well as increased risk of transformation to acute leukemia. MPNs are categorized into subtypes based on specific morphologic, hematologic, and laboratory parameters, the best characterized being the 4 so-called classic MPNs: polycythemia vera (PV), essential thrombocythemia (ET), primary myelofibrosis (MF), and chronic myeloid leukemia (CML).1 In addition, several atypical MPNs are recognized, some of which show both dysplastic and proliferative features, such as atypical, BCR-ABL negative CML (aCML).2  MPNs are associated with acquired, activating mutations or gene fusions of tyrosine kinases, abnormalities that are believed to be critical drivers of excess proliferation as a result of deregulated or constitutive signaling.3 The 2 most prominent examples are BCR-ABL in CML4 and the V617F JAK2 mutation in PV, ET, and MF,5⇓⇓–8 but more than 40 variant tyrosine kinase fusions have been identified in MPNs as well as other mutations in JAK2 and FLT3.2 Activating mutations have been described in components that signal upstream (eg, MPL) or downstream (eg, NRAS) of tyrosine kinases9,10; however, the molecular pathogenesis of the majority of atypical MPNs and approximately 50% of ET and MF cases remains obscure.  V617F JAK2 was initially identified by several different routes, one of which was based on the observation that many PV patients show evidence of acquired uniparental disomy (aUPD) at chromosome 9p.11,12 Regions of aUPD exhibit loss of heterozygosity (LOH) compared with constitutional DNA without change of copy number and arise by mitotic recombination followed by selection for one of the products. After the initial observations in PV, it has emerged that aUPD is common in both hematologic and epithelial malignancies, and is associated with known oncogenic changes in a variety of genes within the affected regions.13 In this study we set out to determine whether aUPD characterizes MPNs of unknown molecular etiology and, if so, whether it could be used as a tool to help identify novel driver mutations.  Methods  Patients  Peripheral blood or bone marrow samples were received from patients diagnosed with an MPN or other hematologic malignancy according to standard morphologic, hematologic, and laboratory criteria. Clinical data were available from a subset of these cases. The study was approved by the Internal Review Boards from participating institutions and informed consent was obtained in accordance with the Declaration of Helsinki.  SNP array analysis  DNA labeling and hybridization to Affymetrix 50k XbaI chips was performed at the Deutsches Ressourcenzentrum für Genomforschung (RZPD, Berlin, Germany). Raw data were imported into the Affymetrix GeneChip Operating Software, analyzed using Affymetrix (High Wycombe, United Kingdom) GeneChip Genotyping Analysis Software (GTYPE 4.1) and copy number analysis tool (CNAT). Data were exported to custom-designed spreadsheets that display loss of heterozygosity and copy number changes in ideogram format. Data were also analyzed using Affymetrix Genotyping Console (version 2.1). Overall, a median of 98.2% (range, 91.5%-99.6%) of SNPs gave readable calls.  Mutation analysis  Detection of mutations by high resolution melting (HRM) analysis was performed as described14 using a Rotor-Gene 6000 (Corbett Life Sciences, St Neots, United Kingdom). Direct sequencing of polymerase chain reaction (PCR) products was performed by standard techniques on an ABI 3130 Genetic Analyzer (Applied Biosystems, Warrington, United Kingdom) and analyzed using Mutation Surveyor (SoftGenetics, State College, PA). CBL mutations are numbered relative to the ATG (A = 1) in the Ensembl cDNA sequence ENSG00000110395. A dropping factor (relative intensity drop of the wild-type allele peak relative to that seen in a concurrently run normal sample) of 60% or more was considered indicative of a biallelic mutation whereas a dropping factor of less than 60% was considered as monoallelic. Primer sequences for CBL exon 8 were CBL_i7f (5′-tgtggtttcactttaaaccctgga-3′) and CBL_i8r (5′-gccaggccaccccttgtatc-3′) and for CBL exon 9 were CBL_i8f (5′-ggcctggcttttgggttagg-3′) and CBL_i9r (5′-cacaatggattttgccagtctcc-3′). CBL reverse-transcriptase PCR (RT-PCR) was performed on random hexamer–reverse transcribed RNA using primers CBLe7F: 5′-gaatcctgatctgactggcttatg-3′ and CBLe10R 5′-gctgcggcagaaggtcaagt-3′. Other primer sequences are available on request.  Microsatellite analysis  DNA samples were amplified with a series of fluorescently labeled primer pairs flanking highly polymorphic microsatellite markers on chromosome 11q as described.10 aUPD was scored if 4 or more consecutive markers encompassing CBL were homozygous without change in allelic copy number (P < .05 of this occurring in the absence of aUPD based on published rates of heterozygosity). Copy number at 11q was estimated by multiplex ligation probe amplification (MLPA) using the Marfan probeset 1 (P065; MRC Holland, Amsterdam, The Netherlands), a kit that contains four 11q control probes including one at 11q23.3.  Constructs  pMY-wtCBL and pMY-CBLR420Q express wild-type and R420Q CBL, respectively, along with green fluorescent protein (GFP).15 Other CBL mutant constructs were derived from pMY-wtCBL using the QuikChange Site-Directed Mutagenesis Kit (Stratagene, La Jolla, CA). pAL/FLT3 and pCMV–HA ubiquitin have also been described previously.15  Cell lines  The IL-3–dependent murine myeloid cell 32Dcl3 (32D) was purchased from the Deutsche Sammlung von Mikroorganismen und Zellkulturen (Braunschweig, Germany) and was maintained in 90% RPMI 1640 medium with 10% fetal bovine serum (FBS) and 10% WEHI-3B conditioned medium (WEHI). All cells were maintained in a humidified incubator at 37°C and 5% carbon dioxide. 32D cells were stably transduced by electroporation with wt-FLT3 in a pAL vector as previously described16 and maintained in 15 ng/mL blastocidin (Invitrogen, Paisley United Kingdom). The Platinum-E (Plat-E) retrovirus packaging cell line17 was transfected using the calcium phosphate method with 5 μg CBL construct DNA: 106 32D/wtFLT3 cells were cocultivated with the Plat-E cells for 48 hours in a volume of 2 mL (50% RPMI, 50% DMEM, 10% FBS, 5% WEHI). GFP positive cells were flow sorted to highest possible purity using a BD FACSAria cell sorter (Becton Dickinson, Stanford, CA). If insufficient cells were obtained for experiments after flow sorting, the cells were maintained in culture and resorted to greater than 95% purity.  Growth factor withdrawal assays  Cell growth was compared with wt-FLT3, vector-only transfected cells and the parental 32D line in stably transfected cells. Cells (105/mL) were grown over 3 days in triplicate for each CBL mutant and a wt-CBL control in 96-well plates. Cultures were analyzed daily for proliferation using the Aqueous One Solution Cell Proliferation Assay kit (MTS assay; Promega, Southampton, United Kingdom). All experiments were performed at least 3 times independently and results were analyzed using GraphPad Prism 4 software (GraphPad Software, San Diego, CA).  Ubiquitin ligase activity  The ubiquitin ligase activity of CBL mutants in comparison to wild-type CBL was determined by transiently transfecting 107 32D/wtFLT3 cells expressing wild-type or CBL mutants by electroporation with 20 μg pCMV-HA ubiquitin. Cells were maintained in culture with WEHI for 12 hours after transfection before being washed, serum deprived overnight, exposed to FLT3 ligand for 10 minutes, and lysed in buffer as described.18 Lysates were incubated with a FLT3-specific polyclonal antibody overnight at 4°C. Protein G agarose beads (100 μL) were added and incubated at 4°C for 4 hours. The beads were washed 4 times in ice-cold radioimmunoprecipitation assay (RIPA) lysis buffer. The supernatant was mixed with sodium dodecyl sulphate (SDS) sample loading buffer before immunoblot analyses with antibodies against HA-ubiquitin (sc-805; Santa Cruz Biotechnology, Santa Cruz, CA), CBL (2111C3a; Abcam, Cambridge, United Kingdom), and FLT3 (sc-479; Cell Signaling Technology, Danvers, MA).  Results  SNP array analysis  We performed genome-wide SNP analysis on leukocyte DNA extracted from 58 MPD patients: aCML (n = 30), V617F JAK2 negative MF (n = 18), and JAK2 mutation negative PV/idiopathic erythrocytosis (n = 10). As comparators, we also analyzed cases of CML in transformation to blast crisis (n = 30, of which 20 were myeloid and 10 lymphoid) and chronic lymphocytic leukemia (CLL; n = 20). We conservatively defined likely regions of aUPD as at least 20 Mb of absent or significantly reduced heterozygous SNP calls (ie, < 5% SNPs called as heterozygous) running to a telomere without change in copy number (Figure 1A). Such large regions are uncommon in healthy individuals.19 Strikingly, we found that aUPD was common in aCML with 10 (33%) of cases affected. In contrast, aUPD was only seen in a single case of MF and was not seen in other patient subgroups. In total, 7 different chromosomes were affected, with abnormalities at 7q and 11q each seen recurrently in 10% of aCML cases. The affected regions ranged in size from 29 Mb to a complete copy of chromosome 13 (Table 1).  Figure 1 Download figureOpen in new tabDownload powerpoint Figure 1 Single nucleotide polymorphism array results and initial mutation analysis. (A) Array analysis of 2 control patients showing copy number (CN) and loss of heterozygosity (LOH) outputs. The x-axis for the CN plots show (copy number)−2 as determined by CNAT, with values approaching or exceeding −1 indicating deletions and greater than +1 indicating amplifications. The x-axis for LOH shows log10 P values (ie, 20 indicates 10−20). The 2 panels on the left are from a polycythemia vera (PV) case with a homozygous V617F JAK2 mutation. There is no gross copy number change (individual datapoints spread around the zero line), but a large block of homozygous SNP calls at 9p indicative of aUPD. The 2 panels on the right are from a chronic myeloid leukemia (CML) blast crisis case and show LOH accompanied by a 20q deletion. (B) For case UPN 5, a large block of 11q aUPD is apparent on analysis of granulocytes but almost completely absent from mononuclear cells (MNCs) extracted from the same sample. (C) Chromosome 1p aUPD in case UPN 27 is associated with a biallelic G>T MPL mutation that is predicted to result in a W515L substitution. (D) Whole chromosome 13 isodisomy in case UPN 28 associated with a homozygous FLT3 ITD. Two controls are shown: an acute myeloid leukemia (AML) patient previously known to have a heterozygous FLT3 ITD and a healthy individual. View inlineView popup Table 1 Chromosomal regions and cases affected by aUPD Association between aUPD and known acquired oncogenic mutations  Although we did not have constitutional DNA to formally confirm that the regions of homozygosity were acquired, in 1 individual we found a marked reduction in the proportion of 11q aUPD positive cells on comparison of mononuclear cells and granulocytes, consistent with an acquired abnormality (Figure 1B). Some of the affected regions harbor genes known to be relevant to hematologic malignancies and therefore we assessed their mutational status. Sequencing of the MPL gene in the case UPN 27 with 1p aUPD revealed a homozygous G>T change (Figure 1C) that results in the previously reported W515L substitution.9 Case UPN 28 with whole chromosome 13 UPD had a homozygous FLT3 internal tandem duplication (ITD; Figure 1D). Candidate genes for the abnormalities at 17q and 21q include NF1 and RUNX1, respectively; however, these were not investigated.  11q aUPD is associated with CBL mutations  Next, we focused our analysis on the 2 recurrent regions of aUPD at 7q and 11q. Because of the strong association between MPNs and aberrant tyrosine kinase signaling, we analyzed genes encoding tyrosine kinases and associated signaling components. The minimally affected region at chromosome 7q contains 3 tyrosine kinases: MET, EPHA1, and EPHB6. We fully sequenced the coding exons of these genes in the 3 cases with 7q aUPD and 2 normal controls but found no abnormalities. Similarly, we found no sequence changes in exons 11 through 17 of BRAF, a gene encoding a downstream component of RAS signaling that is mutated in malignant melanoma and other cancers.20 There are no tyrosine kinase genes on chromosome 11q, and although there are several genes encoding signal transduction components, CBL stood out as a known oncogene that negatively regulates tyrosine kinase signaling.21 Sequencing of CBL in the three 11q aUPD cases revealed that UPN 2 and UPN 5 each had a T>C change at nucleotide 1277 in exon 8 that predicts a L380P substitution. Case UPN 3 had a C>G missense change in exon 9 at nucleotide 1387 that predicts a P417A change (Figure 2). Consistent with the array results, the mutations in cases UPN 3 and UPN 5 were predominant with the residual wild-type alleles only weakly visible. For case UPN 2 both the mutation and level of aUPD was less prominent, presumably due to a greater background of normal cells.  Figure 2 Download figureOpen in new tabDownload powerpoint Figure 2 CBL mutations in the 3 cases with 11q aUPD. (A) Genotyping Console output indicating homozygous SNP calls in green, copy number state as estimated by a Hidden Markov Model in blue, and smoothed copy number log(2) ratio values in red. (B) Ideograms showing lack of large copy number changes and blocks of homozygosity at 11q. Abbreviations are as in Figure 1. (C) Sequence changes indicated by ↓ in each case compared with normal controls. UPN 2 and UPN 5 each have the c.1139T>C: L380P mutation whereas UPN 3 has c.1249C>G: P417A. Prevalence and nature of CBL mutations  To determine the prevalence of CBL mutations, we sequenced exons 8 and 9 (that encode amino acids 366 through 409 and 410 through 477, respectively) in an additional 574 MPN cases. A total of 27 sequence variants were identified in 26 patients, of whom 3 had MF, 10 had chronic myelomonocytic leukemia (CMML), 12 had aCML and 1 had hypereosinophilic syndrome/chronic eosinophilic leukemia (HES/CEL; Tables 2 and 3). To ensure that we were not missing mutations elsewhere in the gene we analyzed all coding exons in 70 CBL mutation negative aCML or unclassified, CML-like MPN cases by high resolution melt (HRM) analysis but failed to detect any further sequence changes apart from known polymorphisms. Similarly, no mutations in CBLB exons 9 or 10 or CBLC exons 7 and 8 (the exons that correspond to CBL exons 8 and 9) were identified. The activity of BCR-ABL is known to increase during progression of CML from chronic phase to blast crisis22 and we speculated that abrogation of CBL activity might be at least partially responsible for this increase. HRM analysis of CBL exons 8 and 9 in blast crisis cases (n = 31), however, failed to reveal any mutations.  View inlineView popup Table 2 Cases analyzed for CBL mutations View inlineView popup Table 3 List of CBL sequence variants Of the 27 CBL variants, 21 (78%) were missense substitutions affecting 12 amino acids (15 different substitutions), 5 (19%) were candidate splicing abnormalities, and 1 (3%) produced a stop codon (Table 3). Most of the changes were novel, although 4 cases had R420Q as previously identified in a single case of AML.15 Apart from N454D, all residues affected by missense mutations were completely conserved in CBL orthologues as well as the 2 other human CBL family members (Figure 3). The mutation to a stop codon was found in a patient (UPN 9) who also had the Y371H variant; PCR across the affected regions and sequencing indicated that the 2 changes were on different alleles.  Figure 3 Download figureOpen in new tabDownload powerpoint Figure 3 Location of missense mutations. (Top panel) Schematic representation of the CBL protein showing the tyrosine kinase binding (TKB), linker, RING, proline-rich (P-rich), and ubiquitin-associated/leucine zipper (UBA/LZ) domains. (Bottom panel) ClustalW alignment of the entire linker and RING domains plus part of the distal sequence. The RING domain is shaded in gray with the defining C and H residues shaded pale blue. Residues affected by missense mutations found in this study are indicated by ▼. CBL mutations are acquired and rarely seen in conjunction with other known mutations  For 2 cases (UPN 5 and UPN 14) we were able to compare the mutational status of leukocytes and T cells or buccal epithelia cells. In both cases the L380P mutation was only present in leukocytes and was therefore acquired (Figure 4A). A third case (UPN 15) originally presented with V617F JAK2 negative ET in 1989 but progressed to MF in 2004. Analysis of DNA extracted from sequential bone marrow slides indicated that the R420Q mutation was absent at the ET phase but acquired on transformation (Figure 4B). The 3 CBL-positive MF cases tested negative for MPL 515 mutations and all CBL mutation positive cases were negative for cytogenetic indicators of tyrosine kinase fusions, FLT3 ITD, activating NRAS mutations and V617F JAK2, with the exception of 1 case (UPN 14) with MF who tested positive for both L380P CBL and V617F JAK2. Analysis of retrospective bone marrow slides for this case indicated the presence of V617F JAK2 at low level initially but absence of L380P CBL. The CBL mutation was detectable 7 years later and completely displaced V617F after a further 3 years, corresponding with a significant rise is peripheral leukocyte counts (Figure 4C). A further case (UPN 21) presented with systemic mastocytosis before evolving to CBL mutation-positive CMML but retrospective samples were unavailable.  Figure 4 Download figureOpen in new tabDownload powerpoint Figure 4 Acquisition of CBL mutations. (A) Comparison of granulocytes (gran) with buccal epithelial cells (bucc) or T cells (T) demonstrates that the L380P mutations in cases UPN 5 and 14 are acquired. (B) Sequential leukocyte and platelet counts for case, who presented in 1989 with essential thrombocythemia (ET) but progressed to myelofibrosis (MF) in 2004. Homozygous CBL R420Q, indicated by *, was detected on transformation but was undetectable in previous specimens. (C) Clinical course of case UPN 14, who presented with MF in 1998 but experienced elevated leukocyte counts in 2005, concomitant with the appearance of the CBL L380P mutation. The JAK2 V617F mutation, which had been present at low level since diagnosis, was still detectable when CBL L380P first appeared but was undetectable in the 2008 sample, when the CBL mutated clone predominated (* on sequence trace). (D) RT-PCR analysis. All cases and controls show the expected product from CBL exons 7-10. Cases UPN 18 and UPN 20 show smaller bands that result from complete deletion of exon 8. Splicing abnormalities  RNA was available for 3 of the 5 cases with potential splicing mutations. Patients UPN 18 and UPN 20 both exhibited double bands by RT-PCR, sequencing of which showed one to be the normal product and the smaller band to result from the complete deletion of exon 8 in each case (Figure 4D), similar to that described in the cell line MOLM-13.23 Patient UPN 19 showed a single band by RT-PCR that was entirely normal on sequence analysis indicating that the exon 8+4 C>T change does not alter splicing. It seems likely, therefore, that the variant identified in case UPN 6 is also of no consequence. The pathogenicity of the intronic variant seen in case UPN 13 remains unknown.  The association between CBL mutations and UPD  To examine the association between CBL mutations and aUPD in more detail we performed microsatellite analysis on all mutated cases using 9 polymorphic markers spanning chromosome 11q13-qter. Of the 26 cases, 11 displayed patterns that indicated significant tracts of homozygosity, including all 3 with 11q aUPD detected by SNP arrays. Case UPN 9 with 2 mutations was heterozygous for all markers. Of the 11 cases with biallelic CBL mutations as judged by sequence analysis, 5 showed complete homozygosity at all microsatellite loci tested and 6 had at least 4 consecutive homozygous calls encompassing the location of the CBL gene at 11q23.3 (Figure 5). We performed multiplex ligation probe amplification (MLPA) analysis to determine whether the observed homozygosity was a consequence of a deletion but no copy number changes were detected (data not shown) and therefore it is likely that homozygosity arose by aUPD. The number of mutated alleles estimated by sequencing was concordant with the microsatellite analysis in 20 cases (Figure 5). Of those that were discordant, UPN 6 and UPN 25 had significant tracts of 11q homozygosity but had a monoallelic CBL mutation whereas UPN 10 and UPN 16 did not show 11q homozygosity but had clear biallelic CBL mutations. The reason for these discrepant cases is unclear, but it is possible that the aUPD seen in UPN 6 and UPN 25 targeted another gene at 11q before acquisition of the CBL mutation. Homozygosity for the CBL mutations in the absence of 11q aUPD in UPN 10 and UPN 16 might be explained by a subcytogenetic deletion of 1 allele.  Figure 5 Download figureOpen in new tabDownload powerpoint Figure 5 Microsatellite analysis of CBL mutation positive cases. The number of CBL mutated alleles (1 monoallelic; 2 biallelic) was estimated by sequence analysis as described in “Experimental procedures.” Microsatellites were scored as homozygous (Embedded Image) or heterozygous (□) for each case and their positions on chromosome 11q are indicated. Correlation between CBL mutations and clinical features  We compared clinical and hematologic features in patients with or without CBL mutations, restricting the analysis to cases with aCML, MF and CMML because these were the subgroups in which CBL mutations were most commonly found. Patients with CBL mutations had a shorter overall survival and progression-free survival compared with mutation negative cases (overall survival: 33 months vs 39 months; progression-free survival: 22 months vs 32 months) but the differences were not significant (Figure 6). Similarly there was no significant difference in gender distribution, age and standard hematologic parameters between mutation positive and mutation negative cases and no differences emerged when individual disease entities were considered.  Figure 6 Download figureOpen in new tabDownload powerpoint Figure 6 Clinical significance of CBL mutations. Overall survival (A) and progression-free survival (B) for atypical chronic myeloid leukemia (aCML), chronic myelomonocytic leukemia (CMML), and MF cases with (n = 19) or without (n = 87) CBL mutations. The differences between mutation positive and negative cases was not significant as determined by Kaplan-Meier analysis. Transforming activities of CBL mutants  We determined whether selected CBL variants had oncogenic activity by testing whether they could transform the interleukin-3 (IL-3)–dependent cell line 32D to growth factor independence in conjunction with overexpression of wild-type FLT3, as has been described previously for R420Q.15 We found that S376F, H398Y, P417A, and R420Q alone each conferred a degree of growth factor independence to 32D cells, but this effect was considerably enhanced in 32Ds that overexpressed wild-type FLT3. In contrast, wild-type CBL conferred no growth advantage either with or without FLT3 overexpression. The N454D change was not transforming, indicating that it is probably a rare polymorphism or a pathogenetically unimportant passenger mutation (Figure 7). Reproducible differences in survival between the different mutants were seen, as illustrated in Figure 8A.  Figure 7 Download figureOpen in new tabDownload powerpoint Figure 7 Transforming activity of CBL mutants. 32D cells or 32Ds overexpressing wild-type FLT3 were transfected with wild-type or mutant CBL constructs and assayed for IL-3 independent growth over 3 days. Results shown are the mean of at least 3 independent experiments, each of which were performed in triplicate. The y-axis shows proliferation in arbitrary units as determined by MTS assay (Promega) and the x-axis time in days. Figure 8 Download figureOpen in new tabDownload powerpoint Figure 8 Survival and abrogation of E3 ubiquitin ligase activity. (A) Relative survival/proliferation of 32D cells expressing wild-type or mutant CBL constructs 48 hours after growth factor withdrawal as determined by MTS assay. (B) 32D cells expressing wild-type FLT3 and CBL constructs as indicated were transfected with HA-tagged ubiquitin and stimulated with FLT3 ligand. After lysis, FLT3 was immunoprecipitated, blotted, and probed with anti-HA. Total blots for FLT3 AND CBL are also shown. Abrogation of E3 ubiquitin ligase activity  It has been shown that FLT3 physically associates with CBL and that, upon ligand stimulation, FLT3 is rapidly ubiquitinylated by wild-type but not R420Q CBL.15 We found that transforming activity correlated directly with the ubiquitin ligase activity of the CBL variants, as assayed by their ability to transfer HA-tagged ubiquitin to FLT3 after stimulation with FLT3 ligand. The S376F, H398Y, P417A and R420Q mutants showed loss of ubiquitin ligase activity, whereas N454D was comparable to wild-type CBL (Figure 8B).  Discussion  The initial aim of our study was to identify large regions of aUPD and therefore we undertook a relatively low resolution genome scan using 50k SNP arrays. Because we did not have constitutional DNA available from most of our cases, we were not able to unambiguously determine whether the extended blocks of homozygosity that we observed were a consequence of aUPD, constitutional UPD or autozygosity due to consaguinity, however we used a very conservative definition of candidate aUPD regions (at least 20 Mb homozygous SNPs calls running to a telomere) and therefore expected that most or all would be acquired. We found that aUPD was relatively common in aCML, but was uncommon in other MPNs that were negative for known mutations, plus also CML blast crisis and CLL. Six distinct regions of aUPD were identified in aCML, indicating substantial genetic heterogeneity in the genesis of this disorder. However, we identified 2 recurrent regions of aUPD at 7q and 11q that were each seen in 10% of cases indicating the presence of common molecular abnormalities.  Our mutation screening focused primarily on tyrosine kinases and associated signal transduction components because of the known association between deregulated tyrosine kinase signaling and MPNs.3 Although we failed thus far to identify any abnormality on chromosome 7q, we found missense CBL mutations in all three 11q aUPD cases. Screening of further MPNs revealed a total of 27 CBL variants in 26 patients who had been diagnosed with aCML (n = 12; 8%), MF (n = 3, 6%), CMML (n = 10, 13%), or HES/CEL (n = 1, 1%); that is, morphologically and clinically related diseases that generally exhibit a poor prognosis. We were unable to discern, however, any specific clinical or prognostic features specifically associated with CBL mutated cases.  Casitas B-lineage lymphoma (CBL) is a well characterized protein that plays both positive and negative regulatory roles in tyrosine kinase signaling. In its positive role, CBL binds to activated receptor tyrosine kinases via its N-terminal tyrosine kinase binding (TKB) domain and serves as an adaptor by recruiting downstream signal transduction components such as SHP2 and P13K. However the RING domain of CBL has E3 ligase activity and ubiquitinylates activated receptor tyrosine kinases on lysine residues, a signal that triggers internalization of the receptor/ligand complex and subsequent recycling or degradation.21,24,25 CBL was originally identified after the characterization of v-Cbl, the transforming component of the Cas NS-1 retrovirus, and thus its association with neoplasia is well established26; however, it is only very recently that CBL mutations were first identified in a human malignancy, specifically occasional cases of AML.15,23,27 We have found that CBL mutations are much more common in MPNs, and previous reports of 11q aUPD in MDS28 suggest that CBL mutations will also prove to be common in this disease. Indeed, 7 of 12 MDS cases with aUPD at 11q were recently shown to harbor CBL mutations.29 There are 2 other human CBL family members, CBLB and CBLC, but we did not identify aUPD in the regions containing these genes and did not detect any mutations of the linker/RING domain of CBLB.  We identified a total of 15 different CBL missense mutations affecting 12 residues (Table 2; Figure 3). Mutations of some of these residues (eg, Y371, C381, H398 and W408) have been analyzed previously in a study that concluded that loss of E3 ubiquitin ligase activity and concomitant impairment of EGFR internalization by RING finger mutations was insufficient for oncogenicity, as assayed in NIH3T3 cells. However mutations of key residues within the α-helix of the linker region abolished ubiquitin ligase activity and were transforming.30 Although the precise amino acid substitutions we identified were different from those studied, we found that mutations in both the RING finger and linker regions conferred autonomous growth to 32D/wt-FLT3 cells, consistent with a causal relationship to the MPN phenotype. The fact that one mutation (N454D) was not transforming emphasizes the importance of functional analysis to distinguish driver mutations from irrelevant passengers or infrequent polymorphisms.31 Notably, this mutation was only weakly conserved between species, in contrast to all other missense mutations that affected highly conserved residues and are therefore likely to be functionally significant.  Enhanced autonomous growth in the presence of CBL mutations and overexpression of wild-type FLT3 may simply reflect the paucity of receptor tyrosine kinase gene expression in 32D cells or, alternatively, may indicate that the mutants are only weakly transforming and require the cooperation of other events to give rise to clinically manifest disease. Consistent with the latter hypothesis, we identified 2 cases in which CBL mutations were acquired during progression of a preexisting MPN (Figure 4). In both cases the identity of the cooperating change is unknown: although one presented with a (relatively low level) V617F JAK2 mutation, this disappeared on progression to CBL L380P positive disease, indicating that the 2 mutations must have arisen independently in different clones. This is reminiscent of the observation that the leukemic blasts of approximately half of V617F positive MPNs that evolve to AML are negative for the JAK2 mutation.32,33 It remains to be established, however, if CBL mutations are usually secondary events or whether the 2 cases we identified are exceptional.  Further evidence for the requirement for additional cooperating genetic events comes from published mouse studies in which neither complete CBL knockout nor heterozygous knockin of the RING finger inactivating mutant C379A (mouse C379 is equivalent to human C381) resulted in evidence of an MPN or other malignancy.34,35 However, closer analysis revealed subtle hemopoietic perturbations that clearly relate to the pathogenesis of MPNs. Specifically, hemopoietic stem cells (HSCs) from these animals were increased in number, hyperresponsive to thrombopoietin and more potent than wild-type HSCs in repopulating the hemopoietic system. These effects were associated with increased STAT5 accumulation and phosphorylation.34,35 Hyperresponsiveness to growth factors and stem cell involvement are 2 of the cardinal features of MPNs. Furthermore, the association between STAT5 activation and MPNs is well documented.36,37 While it is possible that overt hematologic disease might have been induced by a homozygous knockin mutant, these findings suggest that CBL mutations may not be solely responsible for a full MPN phenotype.  The nature of any cooperating abnormalities remains largely obscure. Our results show that overexpression of a tyrosine kinase can cooperate with CBL mutants to promote transformation, however we failed to identify any kinase that was consistently overexpressed in mutation positive cases by tyrosine kinase expression profiling (data not shown). One CBL mutation positive case, however, also tested positive for 7q UPD by SNP array, suggesting that these 2 abnormalities may complement each other. A CBL mutation resulting in skipping of exon 8 has also been identified in a murine model of NUP98-HOXD13 on progression from MDS to AML,38 further corroborating the theory that acute leukemia arises from complementary mutations, one of which inhibits differentiation, and a second (in this case CBL) that enhances proliferation or inhibits apoptosis.  The fact that many CBL mutations are associated with aUPD strongly suggests that clones with homozygous mutations have a selective advantage over those that are heterozygous. This contrasts with the prevailing view that transforming CBL RING finger and linker region mutants act in a dominant negative fashion. Although it is conceivable that another, unknown locus on 11q is the target of aUPD, the finding of 1 case with a Y371H mutation and a stop codon on opposite alleles strongly suggests that homozygous CBL mutations do indeed confer a direct selective advantage. Because we did not screen the entire gene for mutations, it is possible that other cases with heterozygous mutations may also be functionally homozygous due to premature stop codons on the other allele. Alternatively, heterozygosity in these cases may only be apparent due to variable proportions of background normal cells that are not part of the malignant clone. Although the nontransforming N454D mutant was heterozygous, 2 cases with heterozygous R420Q were observed and thus absence of homozygosity cannot be used to infer the presence of an irrelevant, nontransforming sequence variant.  In summary, we have found that oncogenic CBL mutations are acquired in a subset of hematologically related, poor prognosis MPNs. Our findings further strengthen the notion that MPNs are primarily “tyrosine kinopathies,” that is, diseases caused by aberrant activation of proliferation and survival pathways as a consequence of mutations that either directly or indirectly promote excess tyrosine kinase signaling. In view of the fact that tyrosine kinases are such good drug targets, the possibility that CBL-mutated cases might be amenable to therapeutic inhibition is clearly attractive.'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "txt1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "word_tokens = word_tokenize(txt1)\n",
    "word_tokens = np.array(word_tokens)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "initial leng 6396\n"
     ]
    }
   ],
   "source": [
    "print('initial leng %d' % len(word_tokens))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<i>The below stemming operation was tried but did not work well for Gene names, so not implemented</i><br>\n",
    "stemmer = PorterStemmer()<br>\n",
    "for i in range(len(word_tokens)):<br>\n",
    "    word_tokens[i] = stemmer.stem(word_tokens[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After removing stop words 4627\n"
     ]
    }
   ],
   "source": [
    "stop_words = set(stopwords.words('english'))\n",
    "txt1_words = [w for w in word_tokens if not w in stop_words]\n",
    "print('After removing stop words %d' % len(txt1_words))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tokens</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Recent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>evidence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>demonstrated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>acquired</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>uniparental</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         tokens\n",
       "0        Recent\n",
       "1      evidence\n",
       "2  demonstrated\n",
       "3      acquired\n",
       "4   uniparental"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = pd.DataFrame(txt1_words)\n",
    "df1.columns = ['tokens']\n",
    "df1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "gene_ish_pattern = r\"[A-Z]{2,7}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# get gene-ish words in a simple list\n",
    "gene_ish_words1 = [word for word in txt1_words if re.match(gene_ish_pattern, word)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "401"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gene_ish_words1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "401\n"
     ]
    }
   ],
   "source": [
    "# Do the same with pd.DF\n",
    "gene_ish_words = df1[df1['tokens'].str.match(gene_ish_pattern)]\n",
    "print(len(gene_ish_words))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "gene_table = gene_ish_words.groupby('tokens').size().reset_index()\n",
    "gene_table.columns = ['tokens', 'appearances']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tokens</th>\n",
       "      <th>appearances</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CBL</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>UPN</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>MPNs</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>FLT3</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>JAK2</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>MF</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>SNP</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>RING</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>DNA</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>PV</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>MPN</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>CML</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>ET</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>CMML</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>RT-PCR</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    tokens  appearances\n",
       "11     CBL           99\n",
       "95     UPN           38\n",
       "66    MPNs           21\n",
       "39    FLT3           20\n",
       "52    JAK2           15\n",
       "58      MF           15\n",
       "89     SNP           11\n",
       "80    RING            9\n",
       "31     DNA            8\n",
       "77      PV            7\n",
       "65     MPN            7\n",
       "24     CML            6\n",
       "36      ET            6\n",
       "27    CMML            5\n",
       "84  RT-PCR            4"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_table.sort_values('appearances', ascending=False).head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mutation_patterns = ['Truncation', 'Deletion', 'Promoter','Amplification', 'Epigenetic', 'Frame', 'Overexpression',\n",
    "                     'Duplication', 'Insertion','Subtype', 'Fusion', 'Splice', 'Wildtype']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mutation_table = pd.DataFrame(index=[mutation_patterns])\n",
    "mutation_table['appearances'] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for pattern in mutation_patterns:\n",
    "    appearance = len(df1[df1['tokens'].str.contains(pattern, case=False)])\n",
    "    mutation_table.loc[pattern, 'appearances'] = appearance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>appearances</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Truncation</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Deletion</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Promoter</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Amplification</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Epigenetic</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Frame</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Overexpression</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Duplication</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Insertion</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Subtype</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Fusion</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Splice</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Wildtype</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                appearances\n",
       "Truncation                0\n",
       "Deletion                  6\n",
       "Promoter                  0\n",
       "Amplification             3\n",
       "Epigenetic                0\n",
       "Frame                     0\n",
       "Overexpression            4\n",
       "Duplication               1\n",
       "Insertion                 0\n",
       "Subtype                   1\n",
       "Fusion                    3\n",
       "Splice                    0\n",
       "Wildtype                  0"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutation_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "<a id='sec2'></a>\n",
    "# Write function to get gene-ish words list and mutation type table (<a href='#sec0'>Back To Top</a>)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def process_text1(text, print_on=False):\n",
    "    '''\n",
    "    Process the original text. Tokenize into words first, and then remove stop words and numbers\n",
    "    \n",
    "    INPUT:\n",
    "    ======\n",
    "    text : str\n",
    "        A string containing a writing to be analyzed\n",
    "    \n",
    "    OUTPUT:\n",
    "    =======\n",
    "    words : list\n",
    "        A list of tokenized words\n",
    "        \n",
    "    '''\n",
    "    # Tokenize the text\n",
    "    word_tokens = word_tokenize(text)\n",
    "    \n",
    "    # Remove some unwanted words (hyphen excluded), and numbers\n",
    "    remove_list = ['.', ',', '(', ')', '[', ']', '=', '+', '>', '<', ':', ';', '%']\n",
    "    word_tokens = [word for word in word_tokens if word not in remove_list]\n",
    "    word_tokens = [word for word in word_tokens if (word.isnumeric() == False)]\n",
    "    \n",
    "    # Remove Stop words\n",
    "    stop_words = set(stopwords.words('english'))\n",
    "    words = [w for w in word_tokens if not w in stop_words]\n",
    "    \n",
    "    # print if print_on=True\n",
    "    if print_on:\n",
    "        print('Length Before removing stop words %d' % len(word_tokens))\n",
    "        print('Length After removing stop words %d' % len(words))\n",
    "    \n",
    "    return words"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Recent',\n",
       " 'evidence',\n",
       " 'demonstrated',\n",
       " 'acquired',\n",
       " 'uniparental',\n",
       " 'disomy',\n",
       " 'aUPD',\n",
       " 'novel',\n",
       " 'mechanism',\n",
       " 'pathogenetic',\n",
       " 'mutations',\n",
       " 'cancer',\n",
       " 'may',\n",
       " 'reduced',\n",
       " 'homozygosity',\n",
       " 'To',\n",
       " 'help',\n",
       " 'identify',\n",
       " 'novel',\n",
       " 'mutations',\n",
       " 'myeloproliferative',\n",
       " 'neoplasms',\n",
       " 'MPNs',\n",
       " 'performed',\n",
       " 'genome-wide',\n",
       " 'single',\n",
       " 'nucleotide',\n",
       " 'polymorphism',\n",
       " 'SNP',\n",
       " 'screen',\n",
       " 'identify',\n",
       " 'aUPD',\n",
       " 'patients',\n",
       " 'atypical',\n",
       " 'chronic',\n",
       " 'myeloid',\n",
       " 'leukemia',\n",
       " 'aCML',\n",
       " 'n',\n",
       " 'JAK2',\n",
       " 'mutation–negative',\n",
       " 'myelofibrosis',\n",
       " 'MF',\n",
       " 'n',\n",
       " 'JAK2',\n",
       " 'mutation–negative',\n",
       " 'polycythemia',\n",
       " 'vera',\n",
       " 'PV',\n",
       " 'n',\n",
       " 'Stretches',\n",
       " 'homozygous',\n",
       " 'copy',\n",
       " 'neutral',\n",
       " 'SNP',\n",
       " 'calls',\n",
       " 'greater',\n",
       " '20Mb',\n",
       " 'seen',\n",
       " 'aCML',\n",
       " 'MF',\n",
       " 'absent',\n",
       " 'PV',\n",
       " 'In',\n",
       " 'total',\n",
       " 'different',\n",
       " 'chromosomes',\n",
       " 'involved',\n",
       " '7q',\n",
       " '11q',\n",
       " 'affected',\n",
       " 'aCML',\n",
       " 'cases',\n",
       " 'CBL',\n",
       " 'mutations',\n",
       " 'identified',\n",
       " 'cases',\n",
       " '11q',\n",
       " 'aUPD',\n",
       " 'analysis',\n",
       " 'additional',\n",
       " 'MPNs',\n",
       " 'revealed',\n",
       " 'total',\n",
       " 'CBL',\n",
       " 'variants',\n",
       " 'patients',\n",
       " 'aCML',\n",
       " 'myelofibrosis',\n",
       " 'chronic',\n",
       " 'myelomonocytic',\n",
       " 'leukemia',\n",
       " 'Most',\n",
       " 'variants',\n",
       " 'missense',\n",
       " 'substitutions',\n",
       " 'RING',\n",
       " 'linker',\n",
       " 'domains',\n",
       " 'abrogated',\n",
       " 'CBL',\n",
       " 'ubiquitin',\n",
       " 'ligase',\n",
       " 'activity',\n",
       " 'conferred',\n",
       " 'proliferative',\n",
       " 'advantage',\n",
       " '32D',\n",
       " 'cells',\n",
       " 'overexpressing',\n",
       " 'FLT3',\n",
       " 'We',\n",
       " 'conclude',\n",
       " 'acquired',\n",
       " 'transforming',\n",
       " 'CBL',\n",
       " 'mutations',\n",
       " 'novel',\n",
       " 'widespread',\n",
       " 'pathogenetic',\n",
       " 'abnormality',\n",
       " 'morphologically',\n",
       " 'related',\n",
       " 'clinically',\n",
       " 'aggressive',\n",
       " 'MPNs',\n",
       " 'Introduction',\n",
       " 'Myeloproliferative',\n",
       " 'neoplasms',\n",
       " 'MPNs',\n",
       " 'clonal',\n",
       " 'hematopoietic',\n",
       " 'stem',\n",
       " 'cell',\n",
       " 'disorders',\n",
       " 'characterized',\n",
       " 'overproliferation',\n",
       " 'one',\n",
       " 'myeloid',\n",
       " 'cell',\n",
       " 'lineages',\n",
       " 'bone',\n",
       " 'marrow',\n",
       " 'increased',\n",
       " 'numbers',\n",
       " 'mature',\n",
       " 'immature',\n",
       " 'myeloid',\n",
       " 'cells',\n",
       " 'peripheral',\n",
       " 'blood',\n",
       " 'Excess',\n",
       " 'proliferation',\n",
       " 'frequently',\n",
       " 'associated',\n",
       " 'splenomegaly',\n",
       " 'cardiovascular',\n",
       " 'complications',\n",
       " 'well',\n",
       " 'increased',\n",
       " 'risk',\n",
       " 'transformation',\n",
       " 'acute',\n",
       " 'leukemia',\n",
       " 'MPNs',\n",
       " 'categorized',\n",
       " 'subtypes',\n",
       " 'based',\n",
       " 'specific',\n",
       " 'morphologic',\n",
       " 'hematologic',\n",
       " 'laboratory',\n",
       " 'parameters',\n",
       " 'best',\n",
       " 'characterized',\n",
       " 'so-called',\n",
       " 'classic',\n",
       " 'MPNs',\n",
       " 'polycythemia',\n",
       " 'vera',\n",
       " 'PV',\n",
       " 'essential',\n",
       " 'thrombocythemia',\n",
       " 'ET',\n",
       " 'primary',\n",
       " 'myelofibrosis',\n",
       " 'MF',\n",
       " 'chronic',\n",
       " 'myeloid',\n",
       " 'leukemia',\n",
       " 'CML',\n",
       " '.1',\n",
       " 'In',\n",
       " 'addition',\n",
       " 'several',\n",
       " 'atypical',\n",
       " 'MPNs',\n",
       " 'recognized',\n",
       " 'show',\n",
       " 'dysplastic',\n",
       " 'proliferative',\n",
       " 'features',\n",
       " 'atypical',\n",
       " 'BCR-ABL',\n",
       " 'negative',\n",
       " 'CML',\n",
       " 'aCML',\n",
       " '.2',\n",
       " 'MPNs',\n",
       " 'associated',\n",
       " 'acquired',\n",
       " 'activating',\n",
       " 'mutations',\n",
       " 'gene',\n",
       " 'fusions',\n",
       " 'tyrosine',\n",
       " 'kinases',\n",
       " 'abnormalities',\n",
       " 'believed',\n",
       " 'critical',\n",
       " 'drivers',\n",
       " 'excess',\n",
       " 'proliferation',\n",
       " 'result',\n",
       " 'deregulated',\n",
       " 'constitutive',\n",
       " 'signaling.3',\n",
       " 'The',\n",
       " 'prominent',\n",
       " 'examples',\n",
       " 'BCR-ABL',\n",
       " 'CML4',\n",
       " 'V617F',\n",
       " 'JAK2',\n",
       " 'mutation',\n",
       " 'PV',\n",
       " 'ET',\n",
       " 'MF,5⇓⇓–8',\n",
       " 'variant',\n",
       " 'tyrosine',\n",
       " 'kinase',\n",
       " 'fusions',\n",
       " 'identified',\n",
       " 'MPNs',\n",
       " 'well',\n",
       " 'mutations',\n",
       " 'JAK2',\n",
       " 'FLT3.2',\n",
       " 'Activating',\n",
       " 'mutations',\n",
       " 'described',\n",
       " 'components',\n",
       " 'signal',\n",
       " 'upstream',\n",
       " 'eg',\n",
       " 'MPL',\n",
       " 'downstream',\n",
       " 'eg',\n",
       " 'NRAS',\n",
       " 'tyrosine',\n",
       " 'kinases9,10',\n",
       " 'however',\n",
       " 'molecular',\n",
       " 'pathogenesis',\n",
       " 'majority',\n",
       " 'atypical',\n",
       " 'MPNs',\n",
       " 'approximately',\n",
       " 'ET',\n",
       " 'MF',\n",
       " 'cases',\n",
       " 'remains',\n",
       " 'obscure',\n",
       " 'V617F',\n",
       " 'JAK2',\n",
       " 'initially',\n",
       " 'identified',\n",
       " 'several',\n",
       " 'different',\n",
       " 'routes',\n",
       " 'one',\n",
       " 'based',\n",
       " 'observation',\n",
       " 'many',\n",
       " 'PV',\n",
       " 'patients',\n",
       " 'show',\n",
       " 'evidence',\n",
       " 'acquired',\n",
       " 'uniparental',\n",
       " 'disomy',\n",
       " 'aUPD',\n",
       " 'chromosome',\n",
       " '9p.11,12',\n",
       " 'Regions',\n",
       " 'aUPD',\n",
       " 'exhibit',\n",
       " 'loss',\n",
       " 'heterozygosity',\n",
       " 'LOH',\n",
       " 'compared',\n",
       " 'constitutional',\n",
       " 'DNA',\n",
       " 'without',\n",
       " 'change',\n",
       " 'copy',\n",
       " 'number',\n",
       " 'arise',\n",
       " 'mitotic',\n",
       " 'recombination',\n",
       " 'followed',\n",
       " 'selection',\n",
       " 'one',\n",
       " 'products',\n",
       " 'After',\n",
       " 'initial',\n",
       " 'observations',\n",
       " 'PV',\n",
       " 'emerged',\n",
       " 'aUPD',\n",
       " 'common',\n",
       " 'hematologic',\n",
       " 'epithelial',\n",
       " 'malignancies',\n",
       " 'associated',\n",
       " 'known',\n",
       " 'oncogenic',\n",
       " 'changes',\n",
       " 'variety',\n",
       " 'genes',\n",
       " 'within',\n",
       " 'affected',\n",
       " 'regions.13',\n",
       " 'In',\n",
       " 'study',\n",
       " 'set',\n",
       " 'determine',\n",
       " 'whether',\n",
       " 'aUPD',\n",
       " 'characterizes',\n",
       " 'MPNs',\n",
       " 'unknown',\n",
       " 'molecular',\n",
       " 'etiology',\n",
       " 'whether',\n",
       " 'could',\n",
       " 'used',\n",
       " 'tool',\n",
       " 'help',\n",
       " 'identify',\n",
       " 'novel',\n",
       " 'driver',\n",
       " 'mutations',\n",
       " 'Methods',\n",
       " 'Patients',\n",
       " 'Peripheral',\n",
       " 'blood',\n",
       " 'bone',\n",
       " 'marrow',\n",
       " 'samples',\n",
       " 'received',\n",
       " 'patients',\n",
       " 'diagnosed',\n",
       " 'MPN',\n",
       " 'hematologic',\n",
       " 'malignancy',\n",
       " 'according',\n",
       " 'standard',\n",
       " 'morphologic',\n",
       " 'hematologic',\n",
       " 'laboratory',\n",
       " 'criteria',\n",
       " 'Clinical',\n",
       " 'data',\n",
       " 'available',\n",
       " 'subset',\n",
       " 'cases',\n",
       " 'The',\n",
       " 'study',\n",
       " 'approved',\n",
       " 'Internal',\n",
       " 'Review',\n",
       " 'Boards',\n",
       " 'participating',\n",
       " 'institutions',\n",
       " 'informed',\n",
       " 'consent',\n",
       " 'obtained',\n",
       " 'accordance',\n",
       " 'Declaration',\n",
       " 'Helsinki',\n",
       " 'SNP',\n",
       " 'array',\n",
       " 'analysis',\n",
       " 'DNA',\n",
       " 'labeling',\n",
       " 'hybridization',\n",
       " 'Affymetrix',\n",
       " '50k',\n",
       " 'XbaI',\n",
       " 'chips',\n",
       " 'performed',\n",
       " 'Deutsches',\n",
       " 'Ressourcenzentrum',\n",
       " 'für',\n",
       " 'Genomforschung',\n",
       " 'RZPD',\n",
       " 'Berlin',\n",
       " 'Germany',\n",
       " 'Raw',\n",
       " 'data',\n",
       " 'imported',\n",
       " 'Affymetrix',\n",
       " 'GeneChip',\n",
       " 'Operating',\n",
       " 'Software',\n",
       " 'analyzed',\n",
       " 'using',\n",
       " 'Affymetrix',\n",
       " 'High',\n",
       " 'Wycombe',\n",
       " 'United',\n",
       " 'Kingdom',\n",
       " 'GeneChip',\n",
       " 'Genotyping',\n",
       " 'Analysis',\n",
       " 'Software',\n",
       " 'GTYPE',\n",
       " '4.1',\n",
       " 'copy',\n",
       " 'number',\n",
       " 'analysis',\n",
       " 'tool',\n",
       " 'CNAT',\n",
       " 'Data',\n",
       " 'exported',\n",
       " 'custom-designed',\n",
       " 'spreadsheets',\n",
       " 'display',\n",
       " 'loss',\n",
       " 'heterozygosity',\n",
       " 'copy',\n",
       " 'number',\n",
       " 'changes',\n",
       " 'ideogram',\n",
       " 'format',\n",
       " 'Data',\n",
       " 'also',\n",
       " 'analyzed',\n",
       " 'using',\n",
       " 'Affymetrix',\n",
       " 'Genotyping',\n",
       " 'Console',\n",
       " 'version',\n",
       " '2.1',\n",
       " 'Overall',\n",
       " 'median',\n",
       " '98.2',\n",
       " 'range',\n",
       " '91.5',\n",
       " '-99.6',\n",
       " 'SNPs',\n",
       " 'gave',\n",
       " 'readable',\n",
       " 'calls',\n",
       " 'Mutation',\n",
       " 'analysis',\n",
       " 'Detection',\n",
       " 'mutations',\n",
       " 'high',\n",
       " 'resolution',\n",
       " 'melting',\n",
       " 'HRM',\n",
       " 'analysis',\n",
       " 'performed',\n",
       " 'described14',\n",
       " 'using',\n",
       " 'Rotor-Gene',\n",
       " 'Corbett',\n",
       " 'Life',\n",
       " 'Sciences',\n",
       " 'St',\n",
       " 'Neots',\n",
       " 'United',\n",
       " 'Kingdom',\n",
       " 'Direct',\n",
       " 'sequencing',\n",
       " 'polymerase',\n",
       " 'chain',\n",
       " 'reaction',\n",
       " 'PCR',\n",
       " 'products',\n",
       " 'performed',\n",
       " 'standard',\n",
       " 'techniques',\n",
       " 'ABI',\n",
       " 'Genetic',\n",
       " 'Analyzer',\n",
       " 'Applied',\n",
       " 'Biosystems',\n",
       " 'Warrington',\n",
       " 'United',\n",
       " 'Kingdom',\n",
       " 'analyzed',\n",
       " 'using',\n",
       " 'Mutation',\n",
       " 'Surveyor',\n",
       " 'SoftGenetics',\n",
       " 'State',\n",
       " 'College',\n",
       " 'PA',\n",
       " 'CBL',\n",
       " 'mutations',\n",
       " 'numbered',\n",
       " 'relative',\n",
       " 'ATG',\n",
       " 'A',\n",
       " 'Ensembl',\n",
       " 'cDNA',\n",
       " 'sequence',\n",
       " 'ENSG00000110395',\n",
       " 'A',\n",
       " 'dropping',\n",
       " 'factor',\n",
       " 'relative',\n",
       " 'intensity',\n",
       " 'drop',\n",
       " 'wild-type',\n",
       " 'allele',\n",
       " 'peak',\n",
       " 'relative',\n",
       " 'seen',\n",
       " 'concurrently',\n",
       " 'run',\n",
       " 'normal',\n",
       " 'sample',\n",
       " 'considered',\n",
       " 'indicative',\n",
       " 'biallelic',\n",
       " 'mutation',\n",
       " 'whereas',\n",
       " 'dropping',\n",
       " 'factor',\n",
       " 'less',\n",
       " 'considered',\n",
       " 'monoallelic',\n",
       " 'Primer',\n",
       " 'sequences',\n",
       " 'CBL',\n",
       " 'exon',\n",
       " 'CBL_i7f',\n",
       " '5′-tgtggtttcactttaaaccctgga-3′',\n",
       " 'CBL_i8r',\n",
       " '5′-gccaggccaccccttgtatc-3′',\n",
       " 'CBL',\n",
       " 'exon',\n",
       " 'CBL_i8f',\n",
       " '5′-ggcctggcttttgggttagg-3′',\n",
       " 'CBL_i9r',\n",
       " '5′-cacaatggattttgccagtctcc-3′',\n",
       " 'CBL',\n",
       " 'reverse-transcriptase',\n",
       " 'PCR',\n",
       " 'RT-PCR',\n",
       " 'performed',\n",
       " 'random',\n",
       " 'hexamer–reverse',\n",
       " 'transcribed',\n",
       " 'RNA',\n",
       " 'using',\n",
       " 'primers',\n",
       " 'CBLe7F',\n",
       " '5′-gaatcctgatctgactggcttatg-3′',\n",
       " 'CBLe10R',\n",
       " '5′-gctgcggcagaaggtcaagt-3′',\n",
       " 'Other',\n",
       " 'primer',\n",
       " 'sequences',\n",
       " 'available',\n",
       " 'request',\n",
       " 'Microsatellite',\n",
       " 'analysis',\n",
       " 'DNA',\n",
       " 'samples',\n",
       " 'amplified',\n",
       " 'series',\n",
       " 'fluorescently',\n",
       " 'labeled',\n",
       " 'primer',\n",
       " 'pairs',\n",
       " 'flanking',\n",
       " 'highly',\n",
       " 'polymorphic',\n",
       " 'microsatellite',\n",
       " 'markers',\n",
       " 'chromosome',\n",
       " '11q',\n",
       " 'described.10',\n",
       " 'aUPD',\n",
       " 'scored',\n",
       " 'consecutive',\n",
       " 'markers',\n",
       " 'encompassing',\n",
       " 'CBL',\n",
       " 'homozygous',\n",
       " 'without',\n",
       " 'change',\n",
       " 'allelic',\n",
       " 'copy',\n",
       " 'number',\n",
       " 'P',\n",
       " '.05',\n",
       " 'occurring',\n",
       " 'absence',\n",
       " 'aUPD',\n",
       " 'based',\n",
       " 'published',\n",
       " 'rates',\n",
       " 'heterozygosity',\n",
       " 'Copy',\n",
       " 'number',\n",
       " '11q',\n",
       " 'estimated',\n",
       " 'multiplex',\n",
       " 'ligation',\n",
       " 'probe',\n",
       " 'amplification',\n",
       " 'MLPA',\n",
       " 'using',\n",
       " 'Marfan',\n",
       " 'probeset',\n",
       " 'P065',\n",
       " 'MRC',\n",
       " 'Holland',\n",
       " 'Amsterdam',\n",
       " 'The',\n",
       " 'Netherlands',\n",
       " 'kit',\n",
       " 'contains',\n",
       " 'four',\n",
       " '11q',\n",
       " 'control',\n",
       " 'probes',\n",
       " 'including',\n",
       " 'one',\n",
       " '11q23.3',\n",
       " 'Constructs',\n",
       " 'pMY-wtCBL',\n",
       " 'pMY-CBLR420Q',\n",
       " 'express',\n",
       " 'wild-type',\n",
       " 'R420Q',\n",
       " 'CBL',\n",
       " 'respectively',\n",
       " 'along',\n",
       " 'green',\n",
       " 'fluorescent',\n",
       " 'protein',\n",
       " 'GFP',\n",
       " '.15',\n",
       " 'Other',\n",
       " 'CBL',\n",
       " 'mutant',\n",
       " 'constructs',\n",
       " 'derived',\n",
       " 'pMY-wtCBL',\n",
       " 'using',\n",
       " 'QuikChange',\n",
       " 'Site-Directed',\n",
       " 'Mutagenesis',\n",
       " 'Kit',\n",
       " 'Stratagene',\n",
       " 'La',\n",
       " 'Jolla',\n",
       " 'CA',\n",
       " 'pAL/FLT3',\n",
       " 'pCMV–HA',\n",
       " 'ubiquitin',\n",
       " 'also',\n",
       " 'described',\n",
       " 'previously.15',\n",
       " 'Cell',\n",
       " 'lines',\n",
       " 'The',\n",
       " 'IL-3–dependent',\n",
       " 'murine',\n",
       " 'myeloid',\n",
       " 'cell',\n",
       " '32Dcl3',\n",
       " '32D',\n",
       " 'purchased',\n",
       " 'Deutsche',\n",
       " 'Sammlung',\n",
       " 'von',\n",
       " 'Mikroorganismen',\n",
       " 'und',\n",
       " 'Zellkulturen',\n",
       " 'Braunschweig',\n",
       " 'Germany',\n",
       " 'maintained',\n",
       " 'RPMI',\n",
       " 'medium',\n",
       " 'fetal',\n",
       " 'bovine',\n",
       " 'serum',\n",
       " 'FBS',\n",
       " 'WEHI-3B',\n",
       " 'conditioned',\n",
       " 'medium',\n",
       " 'WEHI',\n",
       " 'All',\n",
       " 'cells',\n",
       " 'maintained',\n",
       " 'humidified',\n",
       " 'incubator',\n",
       " '37°C',\n",
       " 'carbon',\n",
       " 'dioxide',\n",
       " '32D',\n",
       " 'cells',\n",
       " 'stably',\n",
       " 'transduced',\n",
       " 'electroporation',\n",
       " 'wt-FLT3',\n",
       " 'pAL',\n",
       " 'vector',\n",
       " 'previously',\n",
       " 'described16',\n",
       " 'maintained',\n",
       " 'ng/mL',\n",
       " 'blastocidin',\n",
       " 'Invitrogen',\n",
       " 'Paisley',\n",
       " 'United',\n",
       " 'Kingdom',\n",
       " 'The',\n",
       " 'Platinum-E',\n",
       " 'Plat-E',\n",
       " 'retrovirus',\n",
       " 'packaging',\n",
       " 'cell',\n",
       " 'line17',\n",
       " 'transfected',\n",
       " 'using',\n",
       " 'calcium',\n",
       " 'phosphate',\n",
       " 'method',\n",
       " 'μg',\n",
       " 'CBL',\n",
       " 'construct',\n",
       " 'DNA',\n",
       " '32D/wtFLT3',\n",
       " 'cells',\n",
       " 'cocultivated',\n",
       " 'Plat-E',\n",
       " 'cells',\n",
       " 'hours',\n",
       " 'volume',\n",
       " 'mL',\n",
       " 'RPMI',\n",
       " 'DMEM',\n",
       " 'FBS',\n",
       " 'WEHI',\n",
       " 'GFP',\n",
       " 'positive',\n",
       " 'cells',\n",
       " 'flow',\n",
       " 'sorted',\n",
       " 'highest',\n",
       " 'possible',\n",
       " 'purity',\n",
       " 'using',\n",
       " 'BD',\n",
       " 'FACSAria',\n",
       " 'cell',\n",
       " 'sorter',\n",
       " 'Becton',\n",
       " 'Dickinson',\n",
       " 'Stanford',\n",
       " 'CA',\n",
       " 'If',\n",
       " 'insufficient',\n",
       " 'cells',\n",
       " 'obtained',\n",
       " 'experiments',\n",
       " 'flow',\n",
       " 'sorting',\n",
       " 'cells',\n",
       " 'maintained',\n",
       " 'culture',\n",
       " 'resorted',\n",
       " 'greater',\n",
       " 'purity',\n",
       " 'Growth',\n",
       " 'factor',\n",
       " 'withdrawal',\n",
       " 'assays',\n",
       " 'Cell',\n",
       " 'growth',\n",
       " 'compared',\n",
       " 'wt-FLT3',\n",
       " 'vector-only',\n",
       " 'transfected',\n",
       " 'cells',\n",
       " 'parental',\n",
       " '32D',\n",
       " 'line',\n",
       " 'stably',\n",
       " 'transfected',\n",
       " 'cells',\n",
       " 'Cells',\n",
       " '105/mL',\n",
       " 'grown',\n",
       " 'days',\n",
       " 'triplicate',\n",
       " 'CBL',\n",
       " 'mutant',\n",
       " 'wt-CBL',\n",
       " 'control',\n",
       " '96-well',\n",
       " 'plates',\n",
       " 'Cultures',\n",
       " 'analyzed',\n",
       " 'daily',\n",
       " 'proliferation',\n",
       " 'using',\n",
       " 'Aqueous',\n",
       " 'One',\n",
       " 'Solution',\n",
       " 'Cell',\n",
       " 'Proliferation',\n",
       " 'Assay',\n",
       " 'kit',\n",
       " 'MTS',\n",
       " 'assay',\n",
       " 'Promega',\n",
       " 'Southampton',\n",
       " 'United',\n",
       " 'Kingdom',\n",
       " 'All',\n",
       " 'experiments',\n",
       " 'performed',\n",
       " 'least',\n",
       " 'times',\n",
       " 'independently',\n",
       " 'results',\n",
       " 'analyzed',\n",
       " 'using',\n",
       " 'GraphPad',\n",
       " 'Prism',\n",
       " 'software',\n",
       " 'GraphPad',\n",
       " 'Software',\n",
       " 'San',\n",
       " 'Diego',\n",
       " 'CA',\n",
       " 'Ubiquitin',\n",
       " 'ligase',\n",
       " 'activity',\n",
       " 'The',\n",
       " 'ubiquitin',\n",
       " 'ligase',\n",
       " 'activity',\n",
       " 'CBL',\n",
       " 'mutants',\n",
       " 'comparison',\n",
       " 'wild-type',\n",
       " 'CBL',\n",
       " 'determined',\n",
       " 'transiently',\n",
       " 'transfecting',\n",
       " '32D/wtFLT3',\n",
       " 'cells',\n",
       " 'expressing',\n",
       " 'wild-type',\n",
       " 'CBL',\n",
       " 'mutants',\n",
       " 'electroporation',\n",
       " 'μg',\n",
       " 'pCMV-HA',\n",
       " 'ubiquitin',\n",
       " 'Cells',\n",
       " 'maintained',\n",
       " 'culture',\n",
       " 'WEHI',\n",
       " 'hours',\n",
       " 'transfection',\n",
       " 'washed',\n",
       " 'serum',\n",
       " 'deprived',\n",
       " 'overnight',\n",
       " 'exposed',\n",
       " 'FLT3',\n",
       " 'ligand',\n",
       " 'minutes',\n",
       " 'lysed',\n",
       " 'buffer',\n",
       " 'described.18',\n",
       " 'Lysates',\n",
       " 'incubated',\n",
       " 'FLT3-specific',\n",
       " 'polyclonal',\n",
       " 'antibody',\n",
       " 'overnight',\n",
       " '4°C',\n",
       " 'Protein',\n",
       " 'G',\n",
       " 'agarose',\n",
       " 'beads',\n",
       " 'μL',\n",
       " 'added',\n",
       " 'incubated',\n",
       " '4°C',\n",
       " 'hours',\n",
       " 'The',\n",
       " 'beads',\n",
       " 'washed',\n",
       " 'times',\n",
       " 'ice-cold',\n",
       " 'radioimmunoprecipitation',\n",
       " 'assay',\n",
       " 'RIPA',\n",
       " 'lysis',\n",
       " 'buffer',\n",
       " 'The',\n",
       " 'supernatant',\n",
       " 'mixed',\n",
       " 'sodium',\n",
       " 'dodecyl',\n",
       " 'sulphate',\n",
       " 'SDS',\n",
       " 'sample',\n",
       " 'loading',\n",
       " 'buffer',\n",
       " 'immunoblot',\n",
       " 'analyses',\n",
       " 'antibodies',\n",
       " 'HA-ubiquitin',\n",
       " 'sc-805',\n",
       " 'Santa',\n",
       " 'Cruz',\n",
       " 'Biotechnology',\n",
       " 'Santa',\n",
       " 'Cruz',\n",
       " 'CA',\n",
       " 'CBL',\n",
       " '2111C3a',\n",
       " 'Abcam',\n",
       " 'Cambridge',\n",
       " 'United',\n",
       " 'Kingdom',\n",
       " 'FLT3',\n",
       " 'sc-479',\n",
       " 'Cell',\n",
       " 'Signaling',\n",
       " 'Technology',\n",
       " 'Danvers',\n",
       " 'MA',\n",
       " 'Results',\n",
       " 'SNP',\n",
       " 'array',\n",
       " 'analysis',\n",
       " 'We',\n",
       " 'performed',\n",
       " 'genome-wide',\n",
       " 'SNP',\n",
       " 'analysis',\n",
       " 'leukocyte',\n",
       " 'DNA',\n",
       " 'extracted',\n",
       " 'MPD',\n",
       " 'patients',\n",
       " 'aCML',\n",
       " 'n',\n",
       " 'V617F',\n",
       " 'JAK2',\n",
       " 'negative',\n",
       " 'MF',\n",
       " 'n',\n",
       " 'JAK2',\n",
       " 'mutation',\n",
       " 'negative',\n",
       " 'PV/idiopathic',\n",
       " 'erythrocytosis',\n",
       " 'n',\n",
       " 'As',\n",
       " 'comparators',\n",
       " 'also',\n",
       " 'analyzed',\n",
       " 'cases',\n",
       " 'CML',\n",
       " 'transformation',\n",
       " 'blast',\n",
       " 'crisis',\n",
       " 'n',\n",
       " 'myeloid',\n",
       " 'lymphoid',\n",
       " 'chronic',\n",
       " 'lymphocytic',\n",
       " 'leukemia',\n",
       " ...]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check if it works\n",
    "txt2 = process_text1(txt1)\n",
    "txt2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_gene_like_words(tokenized_text, gene_list=None):\n",
    "    '''\n",
    "    Get Gene-name like words from the a list of tokenized words\n",
    "    \n",
    "    INPUT:\n",
    "    ======\n",
    "    tokenized_text : list\n",
    "        A list of tokenized words\n",
    "    \n",
    "    OUTPUT:\n",
    "    =======\n",
    "    gene_like_words : list\n",
    "        A list of gene name like words in the tokenized list\n",
    "    '''\n",
    "    gene_ish_pattern = r\"[A-Z]{2,7}\"\n",
    "    gene_like_words = [word for word in tokenized_text if re.match(gene_ish_pattern, word)]\n",
    "    \n",
    "    if gene_list is not None:\n",
    "        genes = gene_list\n",
    "        for gene in genes:\n",
    "            for i in range(len(gene_like_words)):\n",
    "                if gene in gene_like_words[i]:\n",
    "                    gene_like_words[i] = gene\n",
    "    \n",
    "    return gene_like_words"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['MPNs',\n",
       " 'SNP',\n",
       " 'JAK2',\n",
       " 'MF',\n",
       " 'JAK2',\n",
       " 'PV',\n",
       " 'SNP',\n",
       " 'MF',\n",
       " 'PV',\n",
       " 'CBL',\n",
       " 'MPNs',\n",
       " 'CBL',\n",
       " 'RING',\n",
       " 'CBL',\n",
       " 'FLT3',\n",
       " 'CBL',\n",
       " 'MPNs',\n",
       " 'MPNs',\n",
       " 'MPNs',\n",
       " 'MPNs',\n",
       " 'PV',\n",
       " 'ET',\n",
       " 'MF',\n",
       " 'CML',\n",
       " 'MPNs',\n",
       " 'BCR-ABL',\n",
       " 'CML',\n",
       " 'MPNs',\n",
       " 'BCR-ABL',\n",
       " 'CML4',\n",
       " 'JAK2',\n",
       " 'PV',\n",
       " 'ET',\n",
       " 'MF,5⇓⇓–8',\n",
       " 'MPNs',\n",
       " 'JAK2',\n",
       " 'FLT3.2',\n",
       " 'MPL',\n",
       " 'NRAS',\n",
       " 'MPNs',\n",
       " 'ET',\n",
       " 'MF',\n",
       " 'JAK2',\n",
       " 'PV',\n",
       " 'LOH',\n",
       " 'DNA',\n",
       " 'PV',\n",
       " 'MPNs',\n",
       " 'MPN',\n",
       " 'SNP',\n",
       " 'DNA',\n",
       " 'RZPD',\n",
       " 'GTYPE',\n",
       " 'CNAT',\n",
       " 'SNPs',\n",
       " 'HRM',\n",
       " 'PCR',\n",
       " 'ABI',\n",
       " 'PA',\n",
       " 'CBL',\n",
       " 'ATG',\n",
       " 'ENSG00000110395',\n",
       " 'CBL',\n",
       " 'CBL_i7f',\n",
       " 'CBL_i8r',\n",
       " 'CBL',\n",
       " 'CBL_i8f',\n",
       " 'CBL_i9r',\n",
       " 'CBL',\n",
       " 'PCR',\n",
       " 'RT-PCR',\n",
       " 'RNA',\n",
       " 'CBLe7F',\n",
       " 'CBLe10R',\n",
       " 'DNA',\n",
       " 'CBL',\n",
       " 'MLPA',\n",
       " 'MRC',\n",
       " 'CBL',\n",
       " 'GFP',\n",
       " 'CBL',\n",
       " 'CA',\n",
       " 'IL-3–dependent',\n",
       " 'RPMI',\n",
       " 'FBS',\n",
       " 'WEHI-3B',\n",
       " 'WEHI',\n",
       " 'CBL',\n",
       " 'DNA',\n",
       " 'RPMI',\n",
       " 'DMEM',\n",
       " 'FBS',\n",
       " 'WEHI',\n",
       " 'GFP',\n",
       " 'BD',\n",
       " 'FACSAria',\n",
       " 'CA',\n",
       " 'CBL',\n",
       " 'MTS',\n",
       " 'CA',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'WEHI',\n",
       " 'FLT3',\n",
       " 'FLT3-specific',\n",
       " 'RIPA',\n",
       " 'SDS',\n",
       " 'HA-ubiquitin',\n",
       " 'CA',\n",
       " 'CBL',\n",
       " 'FLT3',\n",
       " 'MA',\n",
       " 'SNP',\n",
       " 'SNP',\n",
       " 'DNA',\n",
       " 'MPD',\n",
       " 'JAK2',\n",
       " 'MF',\n",
       " 'JAK2',\n",
       " 'PV/idiopathic',\n",
       " 'CML',\n",
       " 'CLL',\n",
       " 'SNP',\n",
       " 'SNPs',\n",
       " 'MF',\n",
       " 'CN',\n",
       " 'LOH',\n",
       " 'CN',\n",
       " 'CNAT',\n",
       " 'LOH',\n",
       " 'PV',\n",
       " 'JAK2',\n",
       " 'SNP',\n",
       " 'CML',\n",
       " 'LOH',\n",
       " 'UPN',\n",
       " 'MNCs',\n",
       " 'UPN',\n",
       " 'MPL',\n",
       " 'UPN',\n",
       " 'FLT3',\n",
       " 'ITD',\n",
       " 'AML',\n",
       " 'FLT3',\n",
       " 'ITD',\n",
       " 'DNA',\n",
       " 'MPL',\n",
       " 'UPN',\n",
       " 'UPN',\n",
       " 'UPD',\n",
       " 'FLT3',\n",
       " 'ITD',\n",
       " 'NF1',\n",
       " 'RUNX1',\n",
       " 'CBL',\n",
       " 'MPNs',\n",
       " 'MET',\n",
       " 'EPHA1',\n",
       " 'EPHB6',\n",
       " 'BRAF',\n",
       " 'RAS',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'UPN',\n",
       " 'UPN',\n",
       " 'UPN',\n",
       " 'UPN',\n",
       " 'UPN',\n",
       " 'UPN',\n",
       " 'CBL',\n",
       " 'SNP',\n",
       " 'UPN',\n",
       " 'UPN',\n",
       " 'UPN',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'MPN',\n",
       " 'MF',\n",
       " 'CMML',\n",
       " 'HES/CEL',\n",
       " 'CBL',\n",
       " 'CML-like',\n",
       " 'MPN',\n",
       " 'HRM',\n",
       " 'CBLB',\n",
       " 'CBLC',\n",
       " 'CBL',\n",
       " 'BCR-ABL',\n",
       " 'CML',\n",
       " 'CBL',\n",
       " 'HRM',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'AML.15',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'UPN',\n",
       " 'PCR',\n",
       " 'CBL',\n",
       " 'TKB',\n",
       " 'RING',\n",
       " 'UBA/LZ',\n",
       " 'RING',\n",
       " 'RING',\n",
       " 'CBL',\n",
       " 'UPN',\n",
       " 'UPN',\n",
       " 'UPN',\n",
       " 'JAK2',\n",
       " 'ET',\n",
       " 'MF',\n",
       " 'DNA',\n",
       " 'ET',\n",
       " 'CBL-positive',\n",
       " 'MF',\n",
       " 'MPL',\n",
       " 'CBL',\n",
       " 'FLT3',\n",
       " 'ITD',\n",
       " 'NRAS',\n",
       " 'JAK2',\n",
       " 'UPN',\n",
       " 'MF',\n",
       " 'CBL',\n",
       " 'JAK2',\n",
       " 'JAK2',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'UPN',\n",
       " 'CBL',\n",
       " 'CMML',\n",
       " 'CBL',\n",
       " 'UPN',\n",
       " 'ET',\n",
       " 'MF',\n",
       " 'CBL',\n",
       " 'UPN',\n",
       " 'MF',\n",
       " 'CBL',\n",
       " 'JAK2',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'RT-PCR',\n",
       " 'CBL',\n",
       " 'UPN',\n",
       " 'UPN',\n",
       " 'RNA',\n",
       " 'UPN',\n",
       " 'UPN',\n",
       " 'RT-PCR',\n",
       " 'MOLM-13.23',\n",
       " 'UPN',\n",
       " 'RT-PCR',\n",
       " 'UPN',\n",
       " 'UPN',\n",
       " 'CBL',\n",
       " 'UPD',\n",
       " 'CBL',\n",
       " 'SNP',\n",
       " 'UPN',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'MLPA',\n",
       " 'UPN',\n",
       " 'UPN',\n",
       " 'CBL',\n",
       " 'UPN',\n",
       " 'UPN',\n",
       " 'CBL',\n",
       " 'UPN',\n",
       " 'UPN',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'UPN',\n",
       " 'UPN',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'MF',\n",
       " 'CMML',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'CMML',\n",
       " 'MF',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'IL-3',\n",
       " 'FLT3',\n",
       " 'FLT3',\n",
       " 'CBL',\n",
       " 'FLT3',\n",
       " 'CBL',\n",
       " 'FLT3',\n",
       " 'CBL',\n",
       " 'IL-3',\n",
       " 'MTS',\n",
       " 'CBL',\n",
       " 'MTS',\n",
       " 'FLT3',\n",
       " 'CBL',\n",
       " 'HA-tagged',\n",
       " 'FLT3',\n",
       " 'FLT3',\n",
       " 'FLT3',\n",
       " 'AND',\n",
       " 'CBL',\n",
       " 'FLT3',\n",
       " 'CBL',\n",
       " 'FLT3',\n",
       " 'CBL.15',\n",
       " 'CBL',\n",
       " 'HA-tagged',\n",
       " 'FLT3',\n",
       " 'FLT3',\n",
       " 'CBL',\n",
       " 'SNP',\n",
       " 'DNA',\n",
       " 'UPD',\n",
       " 'SNPs',\n",
       " 'MPNs',\n",
       " 'CML',\n",
       " 'CLL',\n",
       " 'MPNs.3',\n",
       " 'CBL',\n",
       " 'MPNs',\n",
       " 'CBL',\n",
       " 'MF',\n",
       " 'CMML',\n",
       " 'HES/CEL',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'TKB',\n",
       " 'SHP2',\n",
       " 'RING',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'NS-1',\n",
       " 'CBL',\n",
       " 'AML.15,23,27',\n",
       " 'CBL',\n",
       " 'MPNs',\n",
       " 'MDS28',\n",
       " 'CBL',\n",
       " 'MDS',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'CBLB',\n",
       " 'CBLC',\n",
       " 'CBLB',\n",
       " 'CBL',\n",
       " 'EGFR',\n",
       " 'RING',\n",
       " 'NIH3T3',\n",
       " 'RING',\n",
       " 'MPN',\n",
       " 'CBL',\n",
       " 'FLT3',\n",
       " 'CBL',\n",
       " 'MPN',\n",
       " 'JAK2',\n",
       " 'CBL',\n",
       " 'MPNs',\n",
       " 'AML',\n",
       " 'JAK2',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'RING',\n",
       " 'MPN',\n",
       " 'MPNs',\n",
       " 'HSCs',\n",
       " 'HSCs',\n",
       " 'STAT5',\n",
       " 'MPNs',\n",
       " 'STAT5',\n",
       " 'MPNs',\n",
       " 'CBL',\n",
       " 'MPN',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'UPD',\n",
       " 'SNP',\n",
       " 'CBL',\n",
       " 'NUP98-HOXD13',\n",
       " 'MDS',\n",
       " 'AML,38',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'RING',\n",
       " 'CBL',\n",
       " 'CBL',\n",
       " 'MPNs',\n",
       " 'MPNs',\n",
       " 'CBL-mutated']"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glike_words = get_gene_like_words(txt2)\n",
    "glike_words"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def create_mutation_words_table(tokenized_text, normed=False):\n",
    "    '''\n",
    "    Create table for words to describe the mutation types from a list of\n",
    "    tokenized words\n",
    "    \n",
    "    INPUT:\n",
    "    ======\n",
    "    text : list\n",
    "        a list of tokenized words\n",
    "    \n",
    "    OUTPUT:\n",
    "    =======\n",
    "    mutation table : a list of sets\n",
    "    '''\n",
    "    # List of words for mutation types\n",
    "    mutation_patterns = ['truncation', 'deletion', 'promoter','amplification', 'epigenetic', 'frame', 'overexpression',\n",
    "                     'duplication', 'insertion','subtype', 'fusion', 'splice', 'wildtype']\n",
    "    \n",
    "    appearances = []\n",
    "    for pattern in mutation_patterns:\n",
    "        appearance = len([word for word in tokenized_text if pattern in word.lower()])\n",
    "        appearances.append(appearance)\n",
    "    \n",
    "    if normed == 'mutation_types':\n",
    "        appearances = np.array(appearances)\n",
    "        if np.sum(appearances) != 0:\n",
    "            appearances = appearances / np.sum(appearances)\n",
    "        table = dict(zip(mutation_patterns, appearances))\n",
    "    elif normed == 'total_text':\n",
    "        appearances = np.array(appearances)\n",
    "        appearances = appearances / len(tokenized_text)\n",
    "        table = dict(zip(mutation_patterns, appearances))\n",
    "    else:\n",
    "        table = dict(zip(mutation_patterns, appearances))\n",
    "        table['Total'] = np.sum(appearances)\n",
    "    \n",
    "    return table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'amplification': 0.16666666666666666,\n",
       " 'deletion': 0.33333333333333331,\n",
       " 'duplication': 0.055555555555555552,\n",
       " 'epigenetic': 0.0,\n",
       " 'frame': 0.0,\n",
       " 'fusion': 0.16666666666666666,\n",
       " 'insertion': 0.0,\n",
       " 'overexpression': 0.22222222222222221,\n",
       " 'promoter': 0.0,\n",
       " 'splice': 0.0,\n",
       " 'subtype': 0.055555555555555552,\n",
       " 'truncation': 0.0,\n",
       " 'wildtype': 0.0}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "create_mutation_words_table(txt2, normed='mutation_types')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='sec3'></a>\n",
    "# Compiling the entire text-ome - testing (<a href='#sec0'>Back To Top</a>)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "txt3 = text_train.iloc[150, 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ID                150\n",
       "Gene             EGFR\n",
       "Variation    EGFRvIII\n",
       "Class               7\n",
       "Name: 150, dtype: object"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class_train.iloc[150, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Alterations ofthe EGFR gene occur frequently in human gliomas where the most common Is an jjj.f@.ntn@edeletion of exons 2â€”7from the extracel lular domain, resulting in a truncated mutant receptor (AEGFR or de 2-7 EGFR). We previously demonstrated that introduction of @.EGFRinto human US7MG giloblastoma cells (US7MG.@EGFR) conferred remark ably enhanced tumorigenlcity in vivo. Here, we show by cell-mixing ex periments that the enhanced tumorigenicity conferred by 4@EGFRis attributable to a growth advantage Intrinsic to cells expressing the mutant receptor. We analyzed the labeling Index of the proliferation markers 1(1-67 and bromodeoxyuridine and found that tumors derived from US7MGAEGFR cells had significantly higher labeling indexes than those of tumors derived from US7MG cells that were either naive, expressed kinase-deflcient mutants of @EGFR,or overexpressed exogenous wild type EGFR. We also utilized terminal deoxynudeotidyl transferase-medi ated nick end-labeling assays and showed that the apoptotic index of U87MG.@.EGFRtumors was more than 4-fold lower than that of parental US7MG tumors. This decrease in cell death was inversely correlated with the exprossion level of Bd-XL, a negative regulator of apoptosis, which was more than 3-fold higher in U87MG.4@EGFR-derived tumors than in those derived from parental cells. Similar observations were obtained in vitro in serum-free conditions. These results suggest that I@EGFR exerts its pronounced enhancement of glioblastoma tamorigenidty by stimulating proilferadon and inhibiting apoptosis and that the effects are directly attributable to Its constitutively active signal. INTRODUCTION Gliomas are the most common primary tumor of the CNS.3 The most malignant type of glioma, glioblastoma multiforme, has a sur vival rate of less than 5% over 5 years (1). Malignant gliomas originate and progress through clonal evolution, a multistep process by which genetic alterations that are advantageous to growth accu mulate, often in a defined relative order (2, 3). These genetic events include mutation of the p53 gene (4), p16/CDKN2 gene inactivation (5),amplificationandrearrangementoftheEGFRgene(6,7),and monosomy of chromosome 10 (2). Several lines of evidence have implicated abnormalities in EGFR signaling in tumorigenesis. First, the overexpression of EGFR leads to ligand-dependent transformation of immortalized rodent fibroblasts (8). Second, the v-erbB oncogene, derived from the avian erythro blastosis virus, encodes a truncated and mutated form of EGFR that has a high level of tyrosine kinase activity in the absence ofligand (9). Received 5/13/96; accepted 8/30/96. The costs of publication of this article were defrayed in part by the payment of page charges. This article must therefore be hereby marked advertisement in accordance with 18 U.S.C. Section 1734 solely to indicate this fact. I This work was supported in part by the Yasuda Medical Research Foundation, Osaka (to M. N.), NIH Training Grant 5-T32-NS 07342 (to F. C.), and California Division of the American Cancer Society Fellowship 1â€”62-95(to 0. B.). 2 To whom requests for reprints should be addressed, at Ludwig Institute for Cancer Research, 9500 Gilman Drive, La Jolla, CA 92093-0660. 3 The abbreviations used are: CNS, central nervous system; EGF, epidermal growth factor, EGFR, EGF receptor, Akp, alkaline phosphatase; Lux, luciferase; BrdUrd, bro modeoxyuridine; TIJNEL, terminal deoxynucleotidyl transferase-mediated nick end la beling; LI, labeling index; Al, apoptotic index. Finally, amplification and rearrangement of the EGFR gene is fre quently observed in a variety of human malignant neoplasms, includ ing carcinomas of the breast (10), lung (11), kidney (12), ovary (12), cervix (12), and squamous cells (13). In gliomas, EGFR gene ampli fication and overexpression are restricted to high-grade tumors, espe cially those of grade IV (glioblastoma), where they occur at a fre quency of 40â€”50%(14â€”16).Nearly one half of the tumors that show amplification also show rearrangement of the gene, resulting in tu mors expressing endogenous wild-type EGFR as well as the episomal mutant forms (7, 14). The most common of the abnormalities in the EGFR in gliomas are genomic rearrangements leading to specific elimination of a DNA fragment containing exons 2â€”7of the gene, which causes an in-frame deletion of 801 bp of the coding sequence in the extracellular domain of the receptor (6, 7, 17). Overexpression of this mutant to levels comparable to those seen in primary brain tumors in the human glioblastoma cell line U87MG leads to the presence of mutant receptors of 140â€”155-kilodaltons at the cell sur face in which tyrosine residues are constitutively phosphorylated independent of EGF stimulation, whereas the endogenous receptor of 170 kilodaltons is phosphorylated only in response to ligand (18). Although the expression of the mutant EGFR does not dramatically affect the in vitro growth rate of cells cultured with serum, their tumorigenic capacity is remarkably enhanced when implanted s.c. or mtracerebrally into immune-deficient mice (18). This is consistent with the observation that its presence is prognostic of a shorter interval to clinical relapse and poorer survival (15). Genetic alterations can bestow growth advantages on tumor cells by a variety of biological mechanisms. Accumulation of genetic alter ations may result in a substantial growth advantage for single cells, leading to subsequent clonal expansion (19). Some genetic events facilitate the interaction of cells with their microenvironment, thereby exerting a field effect, such as are seen in angiogenesis and paracrine growth factor stimulation, favoring the expansion of the tumor as a whole (16, 20). Tumor growth itself is determined by the balance between cell proliferation and cell death, the latter comprising the distinct processes of necrosis and apoptosis (21). In contrast to ne crosis, which usually occurs as the outcome of catastrophic injury to the cell (22), apoptosis is considered as a physiological regulator of cell number in normal and neoplastic tissues and can be regulated by specific genes including oncogenes (21, 22). In certain human solid malignancies, including gliomas, apoptosis has been shown to occur spontaneously, suggesting that it may have a profound effect on tumor formation and progression (23). All of these mechanisms could po tentially play a role in the enhanced tumorigenicity conferred by L@EGFRin vivo. Here, we demonstrate that a small number of U87MG.L@EGFR cells outgrow a large excess of the parental U87MG cells in an in vivo glioma formation model in mouse brain. This dramatic clonal expan sion results in tumors composed almost entirely of @EGFR-express ing cells and suggests that the growth advantage bestowed by this mutant protein is intrinsic and exerted at the single-cell level. This growth advantage was due not only to an increased proliferation rate, 5079 A Common Mutant Epidermal Growth Factor Receptor Confers Enhanced Tumorigemcity on Human Glioblastoma Cells by Increasing Proliferation and Reducing Apoptosis\\' Motoo Nagane, Frank Coufal, Hong Lin, Oliver BÃ¶gler, Webster K. Cavenee, and H-J. Su Huane Ludwig Institute for Cancer Research (M. N., F. C., H. L, 0. B., W. K. C., H-i. S. H], Department of Medicine (W. K. C.. H-I. S. H.], and Center for Molecular Genetics IW.K.C.],Universityof Californiaat SanDiego,LaJolla,CA 92093-0660 Downloaded from cancerres.aacrjournals.org on May 29, 2017. © 1996 American Association for Cancer Research.  ENHANCED TUMORIGENICITYCONFERRED BY A MUTANT EGFR but also to a reduced rate of apoptosis which was inversely correlated with expression levels of Bcl-XL, a negative regulator of apoptosis (24). Furthermore, a kinase-deficient form of L@EGFRwas unable to confer similar growth advantages. Overexpression of wild-type EGFR was similarly ineffective. These phenomena also became evident in vitro when cells were deprived of serum. These data suggest that in glioblastoma cells, L@EGFRexerts its biological effect by virtue of constitutive activation of its tyrosine kinase, which emanates signals leading to pronounce tumorigenic activity. MATERIALS AND METHODS Cells and Culture. The human glioma cell line U87MG was purchased from American Tissue Culture Collection (25). The generation of the U87MG.I@EGFR cells expressing the exogenous truncated EGFR which lacks a part of the extracellular domain of the receptor has been described previously (18). U87MG.I@EGFRcells were labeled by expression of the human placental Akp gene by infecting cells with a retrovirus carrying hygromycin resistance and Akpgenes (26). Akpexpression by U87MG.t@EGFR-Akpcells was studied prior to their injection into nude mouse brain, confirming that more than 95% of the cells expressed Akp in vitro (data not shown). U87MG cells were coinfected with a retrovirus carrying Lux and neomycin resistance genes along with a retrovirus vector harboring LacZ and hygromycin resistance genes. These U87MG.Lux-LacZ cells were used as controls for U87MG.@EGFR-Akp cells because they harbored a similar set of retroviral constructs and exhibited similar drug resistance, being resistant to both neomycin and hygromycin. The kinase-deficient human EGFR mutant DK, with a lysine 721 to methi onine substitution at the AlP-binding site, was obtained by site-directed mutagenesis. The retroviral vector pLRNL was used for gene transfer of these L@EGFR mutants into U87MG cells as described previously (18). A retroviral vector encoding wild-type EGFR, as well as the hygromycin resistance gene, was used to infect U87MG cells to obtain wild-type EGFR-overexpressing cells (U87M0.wtEGFR). The steady-state expression levels of these various EGFR proteins in U87MG cells were determined by fluorescence-activated cell sorter analysis, confirming equivalent amounts of cell surface protein for all of the constructs used. U87MG.@EGFR-Akp and U87MG.Lux-LacZ were maintained in medium containing 400 @.tg/ml0418 (Life Technologies, Inc., Grand Island, NY) and 200 ,.@g/mlhygromycin (Boehringer Mannheim, Indianapolis, IN). U87MG.I@EGFRand U87MG.DK were maintained in the presence of 400 @g/mlG418. U87MG.wtEGFR was maintained with 200 @g/mlhygromy cm. All cells were cultivated in DMEM supplemented with 10% heat inactivated cosmic bovine serum (Hyclone, Logan, UT), 2 mMglutamine, 100 units/mI penicillin, and 100 p@g/mlstreptomycin in a humidified atmosphere of 5% CO2 at 37Â°C. Stereotactic Inoculation of Tumor Cells into Nude Mouse Brain. For intracerebral stereotactic inoculation, cells were resuspended in 5 pJ of PBS and inoculated into the right corpus striatum of the brain of 4â€”5-week-old female nude mice of BALB/c background (18). Brains were removed at various time points and fixed. Tumor sizes were determined microscopically. In experiments where cell types were mixed prior to injection, 5 X 10@ U87MG.Lux-LacZ cells were injected with an admixture of either 10 or 50 U87MG.i@EGFR-Akpcells. As controls, 5 X l0@cells of U87MG.Lux-LacZ alone or 400 cells of U87MG.AEGFR-Akp cells alone were injected. In experiments where proliferation and apoptosis were determined, between 800 and 5 X l0@cells of a given cell type were injected to produce tumors of a similar size at a similar time from cells with very different growth rates. Akp Staining in Mixed Tumor Sections. Brainswere fixed by perfusion with 4% paraformaldehyde followed by further fixation in 4% paraformalde hyde after removal and then immersed in 30% sucrose in PBS at 4Â°C.Tissues were embedded in OCT (Miles, Elkhart, IN), frozen in liquid nitrogen, and stored at â€”80Â°C.Thin cryostat sections (5â€”7 @m)were immersed in 4% paraformaldehyde, washed in PBS containing 2 msi MgCI2, and heat treated in PBS at 65Â°C.Then sections were rinsed in X-P detection buffer [100 mM Tris-HCI (pH 9.5), 100mMNaC1,and 50 mMMgCI2Jfor 10mm and incubated in a X-P reaction mix [0.1 mg/ml 5-bromo-4--chloro-3-indolyl-phosphate,1 mg/mi nitroblue tetrazolium (Sigma, St. Louis, MO), and 0.5 mM levamisole in a X-P detection buffer) for 4 to 8 h at room temperature. The reaction was terminated by washing in PBS containing 20 mMEDTA, and the sections were mounted after counterstaining with 1 @xg/mlHoechst 33258 (Sigma). Akp positive cells and nuclei in the tumor were identified using light and fluores cence microscopy, respectively, and the proportion of Akp-positive cells was calculated. Ki-67 Immunostaining. Brains were removed, embedded in OCT corn pound, frozen in liquid nitrogen, and stored at â€”80Â°C.Thin (5 @.tm)cryostat sections were fixed in ice-cold acetone for 20 mm followed by air drying for 30 mm. After equilibration, endogenous peroxidase activity was quenched with a 2% hydrogen peroxide solution (in PBS) for 30 mm and rinsed in PBS for 15 mm. Sections were then blocked in 10% normal goat serum (in PBS) for 10 mm. Prediluted rabbit anti-human Ki-67 polyclonal antibody (DAKO, Carpinteria, CA) was then administered, and incubation was conducted for 12h at 4Â°C.After rinsing in PBS, a goat anti-rabbit immunoglobulin secondary antibody (Vector, Burlingame, CA) was applied at 1:200 dilution for 30 mm at roomtemperature.Sectionswererinsedin PBS and then incubatedwith the avidin-biotin complex (Vector) for 30 mm. A final PBS rinse then preceded application of the substrate mix consisting of 0.5 mg/mI diaminobenzidene (Sigma), 0.7 mg/ml sodium azide (Sigma), and 6.8 rng/ml imidazole (Sigma) in PBS. Sections were exposed to the substrate for 3 to 5 mm in the presence of hydrogen peroxide, rinsed in PBS, counterstained in hematoxylin, and then mounted. The percentage of reactive tumor nuclei to the Ki-67 antibody was estimated following examination of 10 high-power fields. An average of 1000 cells was examined for scoring in areas of tumor judged to be most represent ative. BrdUrd Incorporation in DNA Synthesis. DNA synthesis rates in vivo were evaluated with BrdUrd incorporation. BrdUrd (30 pg/kg; Amersham, Arlington Heights,IL) wasadministeredi.p. to the nudemice 30 mm before sacrifice at approximately 3â€”4weeks after inoculation. Removed brains con taming tumors were fixed in chilled 70% ethanol for 12 h, embedded in paraffin, and cut into sections 4-sm thick. The sections were deparaffinized, rehydrated, and incubated for 20 mm in 2 N HC1 to denature DNA. After washing in PBS, the sections were immunostained for BrdUrd by reaction with a prediluted mouse monoclonal antibody to BrdUrd (Amersham) for 1 h at room temperature and then a secondary rhodamine-conjugated antimouse immunoglobulin antibody (Jackson ImmunoResearch, West Grove, PA) at a 1:100 dilution for 30 mm. Nuclei of the cells were stained with 2 @.&Wml Hoechst 33258 for 10 mm. To calculate the BrdUrd LI, more than 2000 cells were examined in each tissue section. For in vitro assays, cells were plated at 60â€”70%confluency on 12-mm coverslips in fully supplemented medium for 24 h. Then the cells were subjected to serum starvation for 32 h followed by labeling with BrdUrd under serum-free conditions for 24 h. The cells were fixed in 95% ethanol and 5% glacial acetic acid and subsequently immuno stained for DNA synthesis as described above. To calculate BrdUrd LI, 500 cells were examined in each coverslip using fluorescence microscopy. In Situ (TUNEL) Labeling of Apoptotic DNA Fragmentation. In paraf fin-embedded, 10% formalin-fixed tissue sections of xenografted tumors in the nude mouse brain removed 3â€”5weeks after inoculation, apoptotic cells were visualized using TUNEL of apoptotic DNA strand breaks (27). Tissue sections at 4-tim thickness were deparaffinized, rehydrated, and treated with 20 @g/ml proteinase K (Boehringer Mannheim). The slides were then incubated with TUNEL reaction mixture containing terminal deoxynucleotidyl transferase and its substrate fluorescein-conjugated dUTP using an In Situ Cell Death Detec tion kit-fluorescein (Boehringer Mannheim) at 37Â°Cfor 1 h in the dark. Positive control sections were pretreated with 2 units/ml DNase I (Boehringer Mannheim) in 20 mM Tris (pH 7.0) and 5 mM MgCl2 for 10 mm at room temperature, and negative controls were incubated without terminal de oxynucleotidyl transferase. Following incubation, the sections were washed in PBS and incubated with 1 @.tg/mlHoechst 33258 in PBS for 10 mm at room temperature and washed in PBS. Apoptotic (FITC-labeled nuclei which were green) cells in each tissue section were visualized with fluorescence micros copy and counted by examination of 1000â€”8000cells. The Al was calculated as a ratio of apoptotic cell number:total tumor cell (Hoechst-stained nuclei which were blue) number within the areas. For in vitro assays, cells plated on coverslipswere starvedfor 75 h andsubsequentlyfixed in 4% paraformalde hyde for 30 mm. Following permeabilization in Triton X-lOO/sodiumcitrate, apoptotic cells were labeled by TUNEL reaction as described above. To calculate Al, more than 1000 cells were examined in each coverslip. 5080 Downloaded from cancerres.aacrjournals.org on May 29, 2017. © 1996 American Association for Cancer Research.  ENHANCED TUMORIGENICITYCONFERRED BY A MUTANT EGFR Northern Blotting. Total RNAs from frozen tumor tissues ground in liquid nitrogen and cultured cells under serum starvation for 48 h were isolated using Tnzol (Life Technologies, Inc.). Fifteen @.tgof total RNA were electrophoresed on a 1%agarose formaldehyde gel, transferred onto a nylon filter (Hybond N@; Amersham), and visualized using methylene blue staining (28). The filter was then hybridized with 32P-labeledhuman bcl-XL cDNA (gift from S. J. Kors meyer, Washington University, St. Louis, MO) at a specific activity of 106 cpm/ml in a hybridization buffer containing 50% formamide, 1X Denhardt\\'s solution, 0.5% SDS, 5X SSC, 10% dextran sulfate, and 100 @tg/mlherring sperm DNA at 42Â°Cfor 15 h. The filter was washed in 0.5X SSC and 0.1% SDS at 50Â°Cfor 15 mm and exposed to KOdakBiomax film at â€”80Â°Cwith an intensifying screen. Quantitation was carried out using laser scanning densi tometry of the autoradiogram. Western Blotting. To obtain whole tumor lysates, frozen tumor samples were ground in liquid nitrogen and lysed in sample buffer containing 0.1 M Tris-HCI (pH 6.8), 2% SDS, 3.2 M urea, 4% 2-mercaptoethanol, 0.25% bromphenol blue, 1 mi@tphenylmethylsulfonyl fluoride, 2 @g/m1aprotinin, 5 @i.g/mileupeptin, 5 @tg/mlpepstatin A, 5 pg/mi antipain, 2 mt@isodium o-vanadate, and 10 mMsodium PP. Cultured cells were lysed in RIPA buffer containing 50 mM Tris-HC1 (pH 7.4), 150 mM NaCl, 2 mr@iEDTA, 1% NP4O, 0.5% sodium deoxycholate, 0.1% SDS, and inhibitors of proteases and phos phatases. The lysates were boiled and centrifuged to remove cell debris prior to separation of proteins on 12.5% SDS-polyacrylamide gels and transfer to mtrocellulosemembranes(Bio-Rad,Hercules,CA).Membraneswereblocked for 1h at room temperature in 5% nonfat miIkITBST[0.1% Tween 20, 125m@i NaCl, 25 mM Tris (pH 7.5), and 0.1% sodium azide] and probed with anti Bcl-X rabbit polyclonal antibody (Transduction, Lexington, KY), anti-Bcl-2 hamster monoclonal antibody (PharMingen, San Diego, CA), or anti-Bax rabbit polyclonal antibody (Santa Cruz Biotechnology, Santa Cruz, CA) for 2 h at room temperature. Blots were then incubated with horseradish peroxidase linked secondary antibodies followed by enhanced chemiluminescence detec tion, according to the manufacturer\\'s instructions (Amersham). Quantitation was carried out using laser scanning densitometry of the autoradiogram. RESULTS Clonal Expansion of U87MG.@EGFR Cells. The enhancement of tumorigenesis conferred by expression of L@EGFR (18) could be mediated at the single-cell level or could be due to a positive field effect on surrounding tumor cells or both. To distinguish among these possibilities, we mixed U87MG.L@EGFR cells identifiable by their stable expression of the Akp gene (U87MG.i@EGFR-Akp) with 5 x iO@U87MGcells double-infectedwith luciferaseandLacZ gene-carrying retroviruses (U87MG.Lux-LacZ) at 1: 10,000 and 1:50,000 ratios and injected them into nude mouse brains. Within 4â€”5weeks after injection, U87MG.@EGFR-Akp cells, identified by Akp-positive staining, had outgrown the U87MG.Lux-LacZ cells dramatically, with the percentage of U87MG.@EGFR-Akp U87MG.L@SEGFR (% oftotal) Tumor 0% cells reaching 95% (20: 1 ratio) and 83% (5: 1 ratio) in tumors where the original ratio was 1: 10,000 and 1:50,000, respectively (Fig. 1, b and c). Control tumors derived from S X iO@ U87MG.Lux-LacZ cells formed brain tumors that were completely negative for Akp staining and grew much slower than tumors that also contained U87MG.@EGFR-Akp cells (Fig. la). Tumors formed from 400 U87MG.z@EGFR-Akp cells developed rapidly and showed 100% positive staining for Akp. The growth rate of these tumors was similar to that of the mixed tumors initiated with 1:50,000 U87MG.Lux-LacZ cells (Fig. 14 Similar results were obtained when U87MG cells were alternately labeled by Akp instead of U87MG.@EGFR cells. Since U87MG.L@EGFR cells were originally mixed with parental U87MG cells at extremely low ratios, it is unlikely that U87MG.L@EGFR cells promoted the growth of U87MG cells surrounding them, since this would have resulted in tumors with cell-type ratios closer to the original mixtures. Thus, these experiments suggest that @EGFRoverex pressed in U87MG cells exerts its biological functions not by a positive field effect on all surrounding cells, but rather by bestow ing a cell intrinsic growth advantage. Increased Proliferation in U87MG.@EGFR Cells in Vivo and in Vitro under Serum Starvation. We next tested whether this intrinsic growth advantage conferred by L@EGFRresulted from an elevated proliferation rate, reduced apoptosis rate, or both. To evaluate the proliferation rate of tumor cells, Ki-67 immunostaining was applied to frozen sections of xenografted brain tumors. Ki-67 is a proliferation associated nuclear antigen present throughout the cell cycle, and the Ki-67 LI provides a reliable means for evaluating the growth fraction of a tumor cell population (29). The Ki-67 LI of U87MG, U87MGAEGFR, U87MG.DK, and U87MG.wtEGFR brain tumors was compared in two different settings: tumors that had developed after a short incubation period (12 days after implantation) and tumors that had grown to a large size in longer incubation times (more than 22 days). The results showed that U87MG.@EGFR tumors had a significantly higher Ki-67 LI than parental U87MG tumors (Ki-67 LI of 58% and â€”20%, respectively) regardless of the incubation time (P < 0.001, Student\\'s t test). Similarly, Ki-67 LIs of tumor cells overexpressing the kinase-deficient form of L@EGFRor wild-type EGFR were also lower than that of U87MG.L@EGFR tumors in both short (up to 16%) (P < 0.001) and long incubation periods (up to 29%, P < 0.01; Fig. 2A). We further investigated BrdUrd LI in these tumors as a second and independent parameter of proliferation rate, it being a direct measure ment of the percentage of cells undergoing DNA synthesis at a given a b C d Inoculum 0% 0.002% 0.01% 100% 83% 95% 100% Fig. I. Selectionfor U87MG.I@EGFRcellsin xenograftedbraintumors.U87MG.L@EGFRcellsstablyexpressingtheplacentalAkp geneproduct(Akp positive,darkbluecytoplasmic staining) were mixed with U87MG. Lux-LacZ cells (Akp-negative, clear) at ratios of 1:50,000 (b) or 1:10,000 (c), inoculated into the nude mouse brain stereotactically, and brains were resected for analysis when tumors were grown. As controls, either U87MG, Lux-LacZ cells (a), or U87MG.AEGFR-Akp cells (d) were inoculated. Representative photographs of Akp-stained cells with Hoechst 33258 nuclear staining are shown at X 100.The percentages of U87MG.L@EGFRcells in the original inocula and the harvested tumors are indicated above and below the photographs, respectively. The data represent the means of four animals in each group. Similar results were obtained in experiments where U87MG cells were alternatelylabeled by Akp. 5081 Downloaded from cancerres.aacrjournals.org on May 29, 2017. © 1996 American Association for Cancer Research.  ENHANCED TIJMORIGENICITYCONFERRED BY A MUTANT EGFR A B C 80 0 â€¢0 C Dl C @ :@ 0 -J F,.. @ 20 0) C 0 0 a. 0 U C 0 0 C.) 0 time (30). The moderate-sized tumors which had grown for 3 to 4 weeks following implantation of U87MGAEGFR cells had a BrdUrd LI of 28.6%, which was significantly higher than that of U87MG derived tumors (16.0%, P < 0.01) and of any of the other types of tumors (P < 0.01; Fig. 2B). The BrdUrd LI of U87MG.L@EGFR tumors indicates a strikingly high proliferation rate when compared to that of glioblastomas in clinical specimens which average 7.3% (31). These differences in proliferation rate could be also observed in vitro when cells were cultured in serum-free medium. As shown in Fig. 2C, U87MG.@EGFR cells retained a significantly higher BrdUrd LI (36.4%) than parental U87MG cells (14.0%, P < 0.001) even after a 56-h serum starvation. The kinase-deficient mutant of I@.EGFRmostly lost this enhanced proliferative activity (P < 0.001), although a small increase in the BrdUrd LI was observed in U87MG.DK cells com pared to that in U87MG cells. Overexpression of wild-type EGFR also showed a significantly lower effect than L@EGFR(P < 0.001) under starvation conditions. Acute stimulation of overexpressed wild-type EGFR by addition of EGF to the serum-free medium, however, effectively increased the proliferation rate of U87MG.WtEGFR cells (Fig. 2C). These BrdUrd LI data were consistent with the observed Ki-67 LI in vivo, strongly suggesting a positive effect of @EGFRon the proliferative rate in these glioma cells. Furthermore, this property appears to be mediated through its constitutively active kinase activ ity, since the kinase-deficient mutant L@EGFRprotein did not confer similar enhanced proliferation. Reduced Apoptosis Rate in U87MG.@EGFR Cells in Vivo and in Vitro under Serum Starvation. We next sought to determine the effect of L@EGFRon the rate of cell death. Apoptosis was assessed @0 0) C 0 1@0 a. 5-0 U C In In C.) I.0 using the TUNEL assay which allows a clear and reliable detection of intemucleosomal DNA fragmentation characteristic of apoptotic cells (27). The TUNEL assay was applied to sections of 10% formalin fixed, paraffin-embedded brain tumors of medium size harvested at 3â€”4-weeks following inoculation. All cells in control sections pre treated with DNase I revealed positive nuclear staining, whereas control cells incubated in the absence of terminal deoxynucleotidyl transferase were negative (data not shown). The Al indicating the proportion of positive cells in the TUNEL assay was the lowest in U87MG.L@IEGFRtumors (0.22%). This value was more than 4-fold lower than that seen for U87MG-derived tumors (0.94%), tumors that overexpressed the wild-type EGFR (0.90%, P < 0.01), or kinase deficient DK (0.83%, P < 0.001; Fig. 3A). These results indicated that apoptotic cell loss was inhibited in tumor cells expressing L@EGFRbut not in the cells that overexpressed wild-type EGFR and that this inhibition was relieved when kinase activity was lost in L@EGFR.In tissue culture, U87MG cells also underwent apoptosis when cells were completely deprived of serum for 75 h or more. Similar to the apoptosis rates observed in vivo, U87MG.L@EGFR cells had a signif icantly lower Al than U87MG cells (P < 0.001), showing increased resistance to serum starvation. Neither overexpression of the kinase deficient mutant DK (P < 0.001) nor wild-type EGFR (P < 0.01) conferred resistance to apoptosis as seen with U87MG.AEGFR cells. In addition, stimulation of cells overexpressing wild-type EGFR with EGF had no significant effect on the apoptotic cell death rate (Fig. 3B). Taken together with data shown above, it appears that both an increased rate of cell proliferation and a decreased rate of cell death contribute to the drastically enhanced tumorigenicity of (6)@3) (3) (3) 0 U. 0wd N. Fig. 2. Increasedproliferationratein U87MG.@EGFRtumorsandcells.A, 10-67LI of the xenograftedtumorsresectedafter 12days(U) or morethan22 days(0) following inoculation. Either 5 x l0@or 1 X l0@cells were stereotactically inoculated into the brains of nude mice, and brains were resected 12 days or more than 22 days after implantation, respectively. For U87MG.L@EGFRcells, either 1 X l0@or 800 cells were injected in each group, respectively. Frozen sections were stained with anti-Ki-67 antibody, counterstained with hematoxylin, and the proportion of 10-67-positive cells was counted. B, BrdUrd LI of xenografted tumors. Cells were stereotactically inoculated into the nude mouse brain using appropriate numbers of cells. Three to 4 weeks after implantation, the mice were given an i.p. injection of BrdUrd 30 mm before brain resection and 70% ethanol fixation. Paraffin-embedded, moderate-size brain tumor sections were stained with anti-BrdUrd antibody and BrdUrd LIs were assessed. C, BrdUrd LI of cells cultured under serum-free condition. Cells grown on coverslips were completely deprived of serum for 32 h, followed by BrdUrd labeling for an additional 24 h under serum starvation, and fixed. U87MG.wtEGFR cells were also stimulated with 100 ng/ml EGF at 24 h after starvation (U87MG.wtEGFR + EGF). In each coverslip, 500 cells were counted. Experiments were repeatedfourtimes.Dataare means.TheLI of eachcelltypewascomparedwiththatof correspondentU87MG.AEGFR.Thenumbersof animals(AandB)andcoverslips(C)are shown in parentheses. Bars, SE. 5082 @ -l@ -@IJ- @ (@ @ LI. 0 U. N. (@ @ w w + @ â€˜1 @ C, m U. @.1 @ N. W N. C, N. Downloaded from cancerres.aacrjournals.org on May 29, 2017. © 1996 American Association for Cancer Research.  ENHANCEDTUMORIGENICITYCONFERREDBY A MUTANTEGFR B 20 A 1.5 (6) @1I@@ (8) I (7) (5) (4) (5) â€˜5â€• J@L â€˜5â€• Fig. 3. Reduced apoptotic rate of U87MG.U@EGFRtumors and cells. A, cells were stereotactically inoculated into the nude mouse brain. Three to 4 weeks after implantation when tumors reached a moderate size, the brains were resected and fixed in 10% formalin. Paraffin sections were stained with TUNEL as described. The A! or percentage of TUNEL-positive cells was assessed in each tumor and is shown as means. B, A! of cells under serum-free conditions. Cells grown on coverslips were completely deprived of serum for 75 h and fixed in 4% paraformaldehyde. U87MG.WIEGFR cells were also stimulated with 100 ng/ml EGF (U87MG.wtEGFR + EGF) at both 24 h and 48 h after starvation. On each coverslip, more than 1000cells were counted. Experiments were repeated three times. Values represent means. The A! of each cell type was compared with that of the correspondent U87MG.L@EGFR. The numbers of animals (A) and coverslips (B) are shown in parentheses, Bars, SE. 5083 15@ >1In @0 C .2 10@ 0 a. 0 a. 4 5: 1<In 1@ @0 C U 0 a. @ 0 a. 4 U I@ U U @ C, @ U- U- @ I.,. C@ ,,; (@ w @.â€˜w @ D <@ @ C, F,.. @ F,.. D U87MG.@EGFR cells. Furthermore, kinase-inactivating mutations abrogate these properties of @EGFR. Upregulation of Bcl-XL Expression in U87MG4@EGFR Cells. To further characterize the underlying mechanism by which apoptosis is inhibited in U87MG.L@EGFR cells, we investigated the expression levels of the Bcl-2 gene family of apoptosis regulators in these tumors and cells under starvation (32). Bcl-2, an apoptosis inhibitor, and Bax, an apoptosis promoter, were detectable in tumors and cells. However, the expression levels in each type of tumor or cell were similar (Fig. 4,C and D). In contrast, mRNA and protein expression of Bcl-XL, which functionally resembles the Bcl-2 protein as a potent inhib itor of apoptosis, was significantly higher (more than 3-fold) in U87MG.L@EGFR tumors when compared with U87MG-derived tumors (Fig. 4, A and B). Bcl-XL expression level was also signif icantly higher in U87MG.AEGFR cells cultured in the absence of serum than in parental cells (about 2-fold at the protein level). Although this difference is smaller than that seen in vivo, compar ison of the Bcl-XL mRNA and protein levels in U87MG.AEGFR and U87MG.DK clearly showed that it is due to the presence of a functional @EGFR.Tumors expressed only small amounts of Bcl xs, theproductof the short-formsplicevariantof Bcl-XLwhich functions as an antagonist of Bcl-XL and Bcl-2 products, detected by reverse transcription-PCR (data not shown). This up-regulation of Bcl-XL expression was markedly reduced in the tumors of U87MG.wtEGFR and kinase-deficient DK tumors (Fig. 4, A and B). A clear inverse correlation between the expression level of Bcl-XL and A! was observed in U87MG, U87MG.@EGFR, (4) (5) @1@ (7) J U U @1 i@L @ C, LI. @ U. LI. 0 @ F,.. 0 ,i@ 0 W w @-â€˜ w + @ â€˜1 @ C, - U. @1 @ F,.. w N. D C, N. U87MG.DK, and U87MG.wtEGFR (Figs. 3 and 4). These data suggested that Bcl-XL expression could be regulated through con stitutively active L@EGFRat the transcriptional level and that this up-regulation of Bcl-XL expression might play an important role in the inhibition of spontaneous apoptosis in U87MG.U@EGFR tumors, resulting in accelerating tumor expansion. DISCUSSION We previously showed that L@EGFRmarkedly enhanced the tumor igenicity of glioblastoma cells (18). However, an increased growth rate of cells was not evident when cultured in serum containing medium in vitro, leaving unresolved the question of whether the effect of A@EGFRwas primarily cell intrinsic, or due to a field effect, or a combination of the two. Here, we addressed this issue directly by mixing U87MG.I@.EGFRcells with large excesses of parental U87MG glioblastoma cells and allowing them to form tumors in the brains of nude mice. We hypothesized that if the tumorigenic effect of L@EGFR was due to a field effect, the ratio of the two cell populations should not change in the course of tumor growth. Alternately, if U@EGFR bestowed a cell intrinsic advantage, then cells bearing the receptor would increase in proportion. We observed an extreme form of the latter behavior: a starting ratio of U87MG.L@EGFR to U87MG cells of 1:50,000 at injection became 5:1 in the fully grown tumor. Direct analysis ofproliferation and cell death rates in tumors showed that this cell intrinsic growth advantage was due to both increased cell prolif eration and reduced apoptosis of cells bearing the AEGFR. In vitro, Downloaded from cancerres.aacrjournals.org on May 29, 2017. © 1996 American Association for Cancer Research.  ENHANCED TUMORIGENICITYCONFERRED BY A MUTANT EGFR - 2.7 kb 1.0 1.5 0.8 0.9 0.5 . 28 S B @ (Bcl-XL) = U -29kD @ @k 1.0 2.2 1.1 1.0 1.0 @ â€” â€” â€” â€” 1.0 1.2 1.1 1.0 0.9 @ = â€” â€” â€” 1.0 0.8 0.9 0.8 1.0 Fig. 4. Up-regulationof Bcl-XL expressionin U87MGAEGFR tumorsand cells. Stereotacticallyinoculatedbrain tumorswere resected3 or more weeksafter implantation, immediately frozen in liquid nitrogen, and stored at â€”80Â°C.For in vitro assays,cells cultured under complete serum starvation for 48 h were harvested. Additionally, U87MG.WtEGFR cells were also stimulated with 100 ng/ml EGF at 24 h after starvation (U87MG.wtEGFR + EGF). Either total RNA or total cell lysate was extracted from the tumor tissues and cells. A, upper gel: Northem blotting of bcl-XL expression in xenografted brain tumors and cells under serum starvation. Fifteen @gof total RNA were size fractionated in a 1% agarose-formaldehyde gel, transferred to a nylon membrane, and probed with a human bcl-XL eDNA. Numbers below the gel, relative expression levels of bcl-XL mRNA in each tumor normalized to that of U87MG tumor. Lower gel, methylene blue staining of 285 rRNA demonstrates similar loading of RNA in each lane. B, Western blotting of Bcl-XLexpression. Equal amounts of tumor or cell lysates were subjected to 12.5% SDS-polyacrylamide gel, transferred to a nitrocellulose membrane, and probed with anti-human Bcl-XL polyclonal antibody. C, Western blotting of Bcl-2 expression. D, Western blotting of Bax expression. Numbers below the gels, expression levels of proteins in each tumor normalized to that of the U87MG tumor. Similar experiments were done more than three times, and representative blots are shown. The experiments were reproduced in three independent tumor specimens. kD, kilodaltons. these characteristics became evident by culturing cells in completely serum-free conditions. Reduced cell death was inversely correlated with increased expression of Bcl-XL, but not of Bcl-2 or Bax. Finally, we showed that a kinase-deficient L@EGFRwas unable to provide glioblastoma cells with a similar growth advantage, suggesting that AEGFR signals in a way similar to other tyrosine kinases. Wild-type EGFR transduces mitogenic signals upon activation by binding to its specific ligands, EGF, or transforming growth factor a 5084 InVivo InVitro C, A (bcl-x) 1.0 4.1 0.7 1.1 â€¢Ã˜1 1.0 3.1 0.7 0.9 @ -@ â€” 1.0 1.1 0.9 0.9 (Bax) _@iT@= :::: 1.0 1.0 1.2 C (BcI-2) D - 26 kD - 21 kD Downloaded from cancerres.aacrjournals.org on May 29, 2017. © 1996 American Association for Cancer Research.  ENHANCED TUMORIGENICITYCONFERRED BY A MUTANT EGFR (33). Down-regulation of EGFR occurs rapidly after activation and its signals are attenuated in the process of receptor internalization and endocytosis (34, 35). In contrast to the tight regulation of the wild type, I@EGFR is ligand independent and constitutively tyrosine phos phorylated (18). Our present data showing that glioblastoma cells expressing L@EGFRhad higher proliferation indexes and lower rates of apoptosis when compared to cells expressing endogenous wild-type EGFR, or cells overexpressing exogenous wild-type EGFR, suggests that these quantitative differences in receptor signaling have important physiological impact. It would indicate that sustained ligand-indepen dent signaling of proliferation by U@EGFRin these cells is more potent than that through overexpressed wild-type EGFR, which must be acutely stimulated in tumors in an autocrine or paracrine fashion by ligand for maximum activation, as seen in Fig. 2C (16). In addition to stimulating proliferation, AEGFR also reduced ap optosis, another important determinant of net tumor growth. Interest ingly, several investigators have demonstrated that proliferative ac tivity and apoptosis rate are positively correlated in several types of malignant tumors including gliomas (23, 36) and that Al increases with higher grades of tumors (37), although the reason for this correlation remains unclear and is not invariable (38). The inhibition of cell death, in addition to augmenting tumor growth, also results in the survival of cells where abnormality would ordinarily have led to apoptosis, thereby contributing to the selective expansion of increas ingly malignant cells. For example, it has been shown that oncogeni cally transformed cells become resistant to hypoxia-induced apoptosis and thus survive and expand clonally under hypoxic condition when they lose the wild-type p.53 tumor suppressor gene or overexpress the apoptosis-inhibiting protein Bcl-2 (39). It is likely that tumor cells implanted in the brain are readily exposed to the selective pressures engendered by ischemia, malnutrition, and immune reactions which are also known to induce spontaneous apoptosis in tumors (22). Our observation showing significant suppression of apoptosis in U87MG.AEGFR tumors grown as xenografts in the brain as well as these cells cultured under serum starvation indicates that glioblastoma cells expressing L@EGFRhave a greater potential to survive in such environments and perhaps attain more growth factor independence. Whether this is also true in primary human brain tumors should be investigated by combining detection of I@EGFR with the TUNEL assay in serial sections of clinical tumor specimens. The reduction of apoptosis in U87MG.L@EGFR tumor cells was inversely correlated with the expression of Bcl-XL at both the mRNA and protein levels, whereas Bcl-2 and Bax expression were un changed. The extent of Bcl-XL up-regulation by U87MG.L@EGFR cells was slightly lower in vitro than in vivo, suggesting that there are differences in the regulatory signals present in these two environ ments. Nevertheless, in both cases there was a specific increase in Bcl-XL that correlated positively with a reduced apoptotic rate and depended on the presence of the U@EGFR.The increase in Bcl-XL mRNA implicates the constitutively active truncated EGFR in the transcriptional regulation of Bcl-XL and suggests a close relationship of Bcl-XL expression with the inhibition of apoptosis in this cell type. Further evidence for the link between AEGFR signaling and Bcl-XL expression comes from experiments on the tyrosine-mutated recep tors: the degree of Bcl-XL up-regulation in these tumors was in proportion to the extent of tyrosine autophosphorylation (data not shown) and completely absent in the tumor expressing the kinase deficient mutant. Moreover, this mechanism is postulated to be Bcl-XL specific since neither Bcl-2 nor Bax expression was affected. The involvement of Bcl-XL in the suppression of glioma cell apoptosis is consistent with what is known about the function and expression of the gene. Bcl-XL is the product of the long-form splice variant of the bcl-x gene, structurally and functionally closely resem bles Bcl-2 as a potent inhibitor of cell death (24), and has been shown to prevent hypoxia-induced apoptosis more efficiently than Bcl-2 (40) and to block X-ray irradiation-inducedapoptosis(41). Bcl-XLis expressed widely in development, especially in the brain and kidney, and its expression is retained in the adult CNS unlike Bcl-2 (24, 42, 43). Expression of Bcl-XL is also detected frequently in neuroblas toma cell lines (44) and some glioma cell lines including U87MG.4 These observations suggest a role of Bcl-XL in CNS homeostasis and therefore make it a candidate for involvement in gliomagenesis. The observation that the proportions of the cells changed strikingly in favor of U87MG.L@EGFR cells between the original mixture and the final brain tumor suggests it is unlikely that L@EGFRhad positive effects on surrounding cells. Indeed, the very low final number of U87MG cells in the resultant tumors may itself require explanation in terms of the effects that the faster growing @EGFR-expressing variant has on its parental counterpart. For example, U87MG cells might experience limiting nutrients in the course of being overtaken by U87MG.I@EGFR cells in vivo. Alternatively, it is possible that U87MG.L@EGFR cells have direct cytotoxic effects on cells without this mutant receptor in vivo. Conversely, the ability of only 10 or 50 U87MG.L@EGFR cells to survive and expand in the brain is probably dependent on the surrounding parental U87MG cells. The hypothesis that cell clonal variants growing in a tumor may be protected from the host immunosurveillance by surrounding parental cells is an intrigu ing one. The experiments described here could be used as a model for tumor progression by clonal evolution, where a few cells acquire an additional genetic alteration that is beneficial for their additional growth (19). By altering the properties of the parental U87MG cells and the new variant U87MG.I@EGFR cells and studying the effect on the growth rate and composition of the resultant tumors, the interplay between the two cellular components of an evolving glioma could be elucidated. Among the genetic alterations involved in the progression of human glial tumors, amplification and rearrangement of EGFR is considered a key event that occurs at the transition to glioblastoma multiforme (3). This notion has been supported by intensive genetic analyses of human gliomas that revealed a highly restricted occurrence of EGFR alterations in those gliomas with extreme malignancy (14â€”16).Our investigations of the biological effect of this receptor on glioblastoma cells show that its dual effect on proliferation and apoptosis contrib utes to an explanation of its pivotal role in glioma progression. It will be of importance to elucidate the nature and functions of such mutant EGFR in glioblastoma in more detail as well as to investigate their potential role in new therapeutic strategies against this intractable brain tumor.   Abstract  The epidermal growth factor receptor (EGFR) and its ligands figure prominently in the biology of gliomas, the most common tumors of the central nervous system (CNS). Although their histologic classification seems to be straightforward, these tumors constitute a heterogeneous class of related neoplasms. They are associated with a variety of molecular abnormalities affecting signal transduction, transcription factors, apoptosis, angiogensesis, and the extracellular matrix. Under normal conditions, these same interacting factors drive CNS growth and development. We are now recognizing the diverse molecular genetic heterogeneity that underlies tumors classified histologically into three distinct grades. This recognition is leading to new therapeutic strategies targeted directly at specific molecular subtypes. In this article, we will review the role of EGFR and related molecular pathways in the genesis of the normal CNS and their relationship to glial tumorigenesis. We will discuss barriers to effective treatment as they relate to anatomic specialization of the CNS. We will also consider the ways in which specific EGFR alterations common to glioma reflect outcomes following treatment with targeted therapies, all with an eye towards applying this understanding to improved patient outcomes.  Epidermal growth factor (EGF) was identified by embryologist Stanley Cohen in the early 1960s and its receptor (EGFR) was identified a decade later (1, 2). During that time and in the decades that followed, a family of related factors and their receptors were identified. The complex intracellular effects that follow activation of these receptors were subsequently elaborated. Together, this would prove to be of critical importance in our understanding of both normal growth and development and neoplastic transformation. This review will focus on the role of these receptor-ligand interactions in primary brain tumor biology. We will briefly discuss their importance in the development of normal brain and the implication of those roles for tumorigenesis. Following a review of specific abnormalities in brain tumors, we will conclude with a summary of brain tumor therapies that include EGFR-mediated signaling in their rationale.  Receptor-Ligand Interactions  The EGFR was the first of the human EGF receptor (HER) family members to be discovered. The HER family proteins are members of the receptor tyrosine kinase (RTK) superfamily. The platelet-derived growth factor receptor (PDGFR), another RTK family member, is also important in brain tumor biology. The ligands for the HER family receptors are also diverse. They include the EGF family and the structurally related neuregulin family. Both are produced by many cell types, including neurons and glia. All of the ligands function via autocrine and/or paracrine signaling. The EGF family includes EGF, transforming growth factor-α (TGF-α), heparin-binding EGF, amphiregulin, betacellulin, and epiregulin (3).  The neuregulin (NRG) family consists of the protein products derived from the alternative splicing of four distinct genes: NRG-1, NRG-2, NRG-3, and NRG-4 (4). Several of the neuregulins were discovered by groups working in different fields and are known variously as glial growth factor, neu differentiation factor, acetylcholine receptor inducing activity, divergent of neuregulin-1, and heregulin. In this review, we will refer to them as NRG-1 to NRG-4. The receptors, their synonyms, dimerization partners, and ligands are outlined in Table 1 .  View inlineView popup Table 1. Relationships between EGFR family members and their ligands Once activated by ligand binding and subsequent dimerization, these receptors activate a variety of intracelluar phosphorylation-mediated signal transduction events. Those of special importance in glioma biology include the RAS/mitogen-activated protein kinase and phosphatidylinositol 3-kinase/AKT pathways.  EGFR-Mediated Signaling in Normal Brain Development  The RTK superfamily and their ligands play a critical role in the development and subsequent maintenance of the central nervous system (CNS). Although there is considerable redundancy among receptors and ligands, each plays a unique role in CNS development and/or survival. The EGFR family and its ligands are variably expressed from embryogenesis, throughout brain development, and into adulthood. These factors are involved in the proliferation, migration, differentiation, and survival of all CNS cell types and their precursors (5). Evidence from both transgenic and knockout mice has shown unique roles for each receptor and ligand in brain development (5, 6).  Mice genetically engineered to overexpress EGFR constitutively have been studied. These animals are invariably growth retarded but lack a distinct phenotype (6). Knockout mice, lacking all EGFR expression, have also been generated. These mice display a range of phenotypes depending on the strain in which they are created, some dying during gestation, others at birth, and some surviving for the first month of life. Those mice surviving beyond birth die of a neurodegenerative process (7). This suggests a critical role for EGFR-mediated processes in neurotropism. Mice with targeted deletions of EGFR display defects in neuronal and glial precursor proliferation, differentiation, and migration (6). Similar findings are observed in mice lacking TGF-α during development (8). EGF exogenously applied to the developing mouse brain increases the number of glial cells relative to neurons (9). These observations can be extended to many species. In fact, EGFR and several of its ligands influence glial differentiation and survival in nervous systems as simple as that of Drosophila (10).  EGFR has been shown to influence the migration of neural stem cells during development (5). Furthermore, overexpression of EGFR in postnatal immobile neural progenitor cells results in their migration both in vitro and in vivo (11). These influences on cell motility are, in part, responsible for a variety of syndromes characterized by neural migration defects and hamartoma formation such as neurofibromatosis type 1 (12).  During development, cellular constituents may be differentially distributed to daughter cells during cell division. Asymmetrical distribution of EGFR during cell division influences the fate of neural progenitor cells (13). In mouse model systems, brain cortical progenitors inheriting high levels of EGFR differentiate along astrocytic lines whereas those inheriting low levels are fated to become oligodrendrocytes. It has been hypothesized that this early differential EGFR expression has a corollary in oncogenesis as astrocytic glial tumors are more likely to express high levels of EGFR and its mutants than are oligodendroglial cells (14).  Neuregulins play fundamental roles in brain development. They act on the developing brain and influence synapse formation, oligodendrocyte development, the cell fate of sensory and sympathetic neurons, differentiation of Schwann cells in the peripheral nervous system, and the formation of acetylcholine receptors at the neuromuscular junction (4). Abnormalities in neuregulin function at given levels of the neuraxis are associated with diseases including schizophrenia, multiple sclerosis, amyotropic lateral sclerosis, autonomic and peripheral neuropathies, and myasthenia gravis (4).  Excepting oligodendrogliomas, evidence supporting a link between neuregulins, their role in CNS development, and subsequent gliomagenesis is less compelling than it is for the EGF family. There is evidence that NRG-1 mediates migration of neural progenitor cells in an ErbB4-dependent manner (15). Heregulin, the alternatively spliced NRG-1 variant, has been shown in both neurons and glia during embryogenesis. Furthermore, exposure of primary rat brain cultures to heregulin resulted in a growth and survival advantage for astrocytes (16). These are but a few of the examples from developmental neurobiology about the potential effect of EGFR-mediated signaling operative in normal brain development that might influence glial tumorigenesis.  Classification of Glial Tumors  Generally, glial tumors are those derived from macroglia, astrocytes and oligodendroglia, as well as their precursors. They are typically classified according to the WHO criteria based on their features observed with H&E staining (17). Although this scheme allows for reliable phenotypic comparisons between pathologists, it is clear that many of these tumors, although histologically similar, differ in important ways at the molecular level. There are several subtypes, including astrocytomas, oligodendroglioma, and tumors of mixed histologies. These tumors, their grades, and the histologic criteria used to distinguish between them are outlined in Table 2 . The histologic criteria used to grade these tumors have changed with time and may soon include molecular criteria. At present, however, the number of simple histologic criteria present in a tumor sample determines tumor grade.  View inlineView popup Table 2. Histologic criteria commonly used in grading infiltrating glial neoplasms The natural history, prognosis, and treatment responses observed in gliomas are related to both tumor category and grade. We will focus primarily on the so-called high-grade tumors (WHO grades 3 and 4) with particular reference to glioblastoma multiforme, the most common and aggressive glioma. Although EGFR family members, their ligands, and downstream signaling targets are dysregulated in many gliomas, glioblastoma multiforme remains the best characterized in this regard. Furthermore, our experience with targeted therapies is greatest in glioblastoma multiforme.  Glioblastoma multiforme arises either (a) de novo from normal glial cells or their precursors (primary glioblastoma multiforme) or (b) via progression from tumors of lower grade (secondary glioblastoma multiforme; ref. 18). Although primary glioblastoma multiforme is said to account for 95% of cases, this may be an overestimation, including tumors that progress rapidly to a grade 4 neoplasm without earlier histologic confirmation of a lower-grade lesion (19). Importantly, these tumors differ from one another in several ways at the molecular level. EGFR gene amplification figures most prominently in primary as opposed to secondary glioblastoma multiforme (40% versus 8%) as does EGFR receptor overexpression (>60% versus <10%; ref. 20). Inactivation of the tumor suppressor gene p53 is more often encountered in secondary glioblastoma multiforme. These two features, amplification of an oncogene (with overexpression of its product) and inactivation of a tumor suppressor gene, constitute the primary features that distinguish one group from the other.  Primary and secondary glioblastoma multiformes differ in many ways beyond their relative mutual exclusion of EGFR gene amplification and p53 gene mutation. For example, although primary glioblastoma multiformes seldom have p53 gene mutations, they are more likely than secondary tumors to amplify the murine double minute 2 (mdm2) gene and to overexpress its product, MDM2 (21). As MDM2 binds to and inactivates p53, many primary glioblastoma multiformes also have a mechanism that provides escape from p53-mediated growth controls. Furthermore, loss of heterozygosity on chromosome 10 and mutations in the tumor suppressor gene, phosphatase and tenascin homologue on chromosome 10 (PTEN), are more common in primary than in secondary glioblastoma multiforme (22). Again, this distinction has important clinical implications as the presence or absence of functional PTEN seems to be critical in determining the effectiveness of certain EGFR-targeted molecular therapies (see below).  Abnormalities in molecular pathways that do not distinguish between primary and secondary glioblastoma multiformes also exist (23). These include abnormalities in cyclin-dependent kinase pathways. For example, either mutation at the INK4a/ARF locus or overexpression of cyclin-dependent kinase 4 may lead to inactivation of the p16 tumor suppressor gene, encoded by the INK4a/ARF locus. Furthermore, RTK signaling can occur through more than one pathway (e.g., both EGFR and PDGFR in glioma; ref. 18). It seems clear from these examples that glial tumors commonly arise as a consequence of the dysregulation of several critical molecular pathways, with more than one route to that dysregulation. With time, the classification of primary and secondary glioblastoma multiformes may expand to include subcategories of tumors within these distinct groups to reflect the many potential combinations of molecular abnormalities.  Gliomas and Their Cell(s) of Origin  Controversy exists about the cell of origin in glial tumor formation. It is clear from a review of normal brain development that glial cells derive from a common precursor, or neural stem cell, and that many factors influence their fate. Pluripotent neural stem cells capable of generating both neurons and glia persist in distinct regions of the adult brain throughout life. Several distinct neural stem cell–derived glial restricted precursor cells have been identified. These include the oligodendrocyte-type 2 astrocyte progenitor, first derived from the rat optic nerve (24), and the A2B5 precursor, first derived from the developing rat spinal cord (25). Others are also described, including neural stem cells that persist in the mature (adult) nervous system. Whereas EGFR-mediated influences allow for continued division and differentiation of various neural stem cells during life, in vitro modeling suggests that this capacity diminishes with age (26).  The concept of a self-renewing cancer stem cell developed from the observation of the shared properties of stem cells and cancer cells (27). Substantial cellular heterogeneity exists within a given glial tumor and a small proportion of those cells have characteristics of neural stem cells. This has given rise to the idea that glial tumors arise from brain tumor stem cells, which, in turn, arise from neural stem cells (reviewed in ref. 28). CD133-expressing brain tumor stem cells are characterized by their ability to recapitulate the phenotype of a morphologically heterogeneous glial tumor in cell culture, thus supporting their role in glial tumor formation (28).  The cancer stem cell hypothesis has challenged the view that tumors arise from the dedifferentiation of more mature cell types. Modeling in vitro has shown the capacity for both differentiated astrocytes and neural stem cells to acquire malignant features following loss of key cell cycle regulators (Ink4a/Arf) in the presence of constitutive EGFR expression (29). Thus, it seems that dysregulation of specific molecular pathways can influence tumor phenotype independent of the cell of origin. Furthermore, transgenic modeling of gliomagenesis directed towards the formation of oligodendrogliomas has resulted in astrocytomas (30). Conversely, transgenic strategies directed towards the formation of astrocytoma have generated oligodendrogliomas (31), supporting the view that astrocytes and oligodendrocytes have a common origin, their bidirectional movement along differentiating pathways subject to influence by a host of factors.  EGFR-Mediated Signaling in Glial Tumorigenesis  As noted, activation of EGFR-mediated signal transduction pathways is common in many malignancies, including glial neoplasms, and may be caused by perturbations at several levels. The four common pathways include (a) increased expression of EGFR ligands; (b) amplification of EGFR genes; (c) overexpression of wild-type and mutant EGFR family members; and (d) activation of downstream pathways via EGFR-independent mechanisms. Many of these factors operate simultaneously in glial tumors and often act in concert to confer the malignant phenotype.  Ligand-Mediated Mechanisms of Activation in Gliomagenesis  EGFR and other RTK family ligands are often overexpressed in gliomas. For example, neoplastic glial cells produce increased levels of both TGF-α and heparin-binding EGF, leading to overactivity of EGFR via an autocrine loop. A similar mechanism is also observed with PDGF (32). The effects of TGF-α and heparin-binding EGF become more apparent as tumors progress from lower to higher grades. PDGF, on the other hand, seems to play a role early in tumorigenesis (33). Gliomagenesis is now recognized as a multistep process, and although increased ligand expression may influence the transition of a glial cell from a quiescent to an activated state, this alone is probably insufficient to confer a malignant phenotype.  Gene Amplification, Overexpression, and Mutation of EGFR Family Members in Glioma  EGFR is best characterized among the HER family members that are frequently overexpressed in glioma. In glioblastoma multiforme, EGFR overexpression is almost always associated with gene amplification. Most series note an incidence of ∼40%. Mutation is also common. This is in contrast to other EGFR family members that may be present but are not commonly amplified, overexpressed, or mutated. Mutations in gliomas typically lead to truncated proteins, with a very low prevalence of point mutations. Furthermore, different EGFR mutations may be present in the same glioblastoma multiforme (34). In other categories of gliomas (e.g., oligodendroglioma), increased EGFR expression often occurs independent of amplification, with rare evidence for mutation (14). EGFR aberrancy is not, however, typically seen in lower-grade glial tumors of astrocytic histology. EGFR gene amplification is found in one third to one half of glioblastoma multiforme and 90% of these show protein overexpression (19, 34). Many EGFR mutations result in constitutive activation (35). EGFR mutations seem to be restricted to those tumors with gene amplification (34). This differs from other tumors (e.g., non–small-cell lung cancer) where amplification of mutant EGFR genes is uncommon (36).  HER2/neu overexpression is less common in glioma than it is in other solid tumors, such as breast cancer. However, one retrospective study involving 149 subjects with glioblastoma multiforme revealed HER2/neu overexpression in 15% (37). A subsequent study involving 347 subjects found an incidence of HER2/neu overexpression of 10% (38). In both studies, HER2/neu overexpression was associated with poor survival.  Specific EGFR Mutations in Glioblastoma Multiforme  EGFR gene rearrangements and expression of their aberrant protein products are common in glioblastoma multiforme (39). The EGFR deletion mutations are outlined in Fig. 1 . Of the five common deletion mutations, EGFRvII, EGFRvIII, and EGFRvV are commonly found in glioblastoma multiforme. One study reviewing tissues from 44 glioblastoma multiformes identified mutations in 17 (38%) of the tumors, all of which had amplified EGFR (34). The EGFR mutations most commonly found in glioblastoma multiforme, EGFRvII and EGFRvIII, are in-frame deletion of regions of the extracellular domain. The most common deletion, accounting for 60% to 70% of EGFR mutations in glioblastoma multiforme, involves exons 2 to 7 of the extracellular domain, forming a receptor known as EGFRvIII (ref. 40; see below). EGFRvII and EGFRvV each account for ∼15% of the EGFR mutations in glioblastoma multiforme. The first of these is EGFRvII, a deletion of exons 14 and 15 (amino acids 521-603) in the cysteine-rich portion of the extracellular region. This mutation may lend a growth advantage to the tumor cells (34). The other relatively common mutation seen in glioblastoma multiforme is a truncation of the intracellular region at amino acid 958, EGFRvV. Cells expressing this mutation exhibit increased ligand-dependent kinase activity (34). Mutations of the intracellular portion of EGFR are more common in other neoplasms (e.g., non-small cell lung cancer). In-frame tandem duplications of selected EGFR exons occur at a low frequency, as do missense mutations of the extracellular juxtamembrane region. Their significance is unclear. Of note, multiple mutations are sometimes seen in the same amplified EGFR gene, a finding unique to glioblastoma multiforme (34).  Fig. 1. Download figureOpen in new tabDownload powerpoint Fig. 1. A schematic representation of the EGFR protein and its common deletion mutations, EGFRvI to EGFRvV. The gene contains 26 exons transcribed into a protein containing 1,186 amino acids, with three distinct regions (extracellular, transmembrane, and intracellular) as shown. The extracellular region is divided into four functional domains. The intracellular region is divided into two domains. The amino acids separating these segments and domains are indicated by the boxed numbers and their relative position within the protein indicated by arrows. The percentages below each variant represent the proportion of each among glioblastoma multiforme harboring mutations. Mutants EGFRvII to EGFRvIV represent deletions of entire exons. The EGFRvI and EGFRvV mutants are the result of NH2-terminal and COOH-terminal truncations, respectively. The remainder (10% of all EGFR mutations found in glioblastoma multiforme; not shown) consist of a variety of tandem duplications of exons as well as missense and insertion mutations. As noted above, the most common in-frame deletion involves exons 2 to 7 (amino acids 6-273) creating the 145-kDa EGFRvIII mutant (41). Regions of multiple Alu repeats in introns 1 and 7 may play a role in mediating susceptibility to this specific gene rearrangement (42). The same mutation is also found in cancers of the breast, ovary, lung, and prostate (43–47). The joining of exons 1 to 8 creates a novel tertiary conformation of the extracellular domain, leaving the transmembrane and intracellular portions of the receptor intact. Although this mutant protein is unable to bind either EGF or TGF-α (and presumably its other ligands), it is constitutively active, phosphorylating downstream targets in the absence of ligand binding. Targets include those mediated by phosphatidylinositol 3-kinase and Ras-GTP, both of which are thought to confer a growth advantage on EGFRvIII-expressing cells (48, 49). In vitro, EGFRvIII-expressing glioblastoma multiforme cells have elevated expression of matrix metalloproteinases and extracellular matrix components when compared with those that do not express the variant receptor. These findings support a role for EGFR mutations in the increased invasiveness of tumor cells (50).  Alternate Mechanisms for Activating EGFR-Mediated Signal Transduction Pathways in Glioma  The activation of signal transduction pathways in glioblastoma multiforme by EGFR family receptor phosphorylation is also influenced by other factors. For example, phosphatidylinositol 3-kinase is an enzyme critical to the activation of Akt. Activation occurs as a consequence of phosphatidylinositol 3-kinase–mediated phosphorylation of phosphatidylinositol 4,5-bisphosphate to its triphosphate form, phosphatidylinositol 3,4,5-trisphosphate. Activated Akt inhibits apoptosis while promoting cell division (51). Phosphatidylinositol 3-kinase can be activated by PDGFR, thus bypassing EGFR-mediated signaling. Furthermore, PTEN converts phosphatidylinositol 3,4,5-trisphosphate to phosphatidylinositol 4,5-bisphosphate, inhibiting Akt activity (52). The PTEN gene is frequently deleted in glioblastoma multiforme, thus favoring activation of Akt signaling independent of and in concert with RTK-mediated signal transduction (52). This finding has important implications in the prediction of glioblastoma multiforme responsiveness to several tyrosine kinase inhibitors (see below).  RAS mutations lead to its constitutive activation in many tumors, resulting in cell proliferation. This is not the case for glioblastoma multiforme, where activation of the RAS/mitogen-activated protein kinase pathway is dependent on RTK overactivity (53). However, like phosphatidylinositol 3-kinase, RAS activation can occur through PDGFR phosphorylation, thus bypassing EGFR-mediated signaling. Furthermore, although targeted therapies directed at EGFR family members have not been devised to directly affect RAS/mitogen-activated protein kinase pathways, inhibitors of key enzymes important in the posttranslational modification of RAS are being investigated in glioblastoma multiforme (54). These agents, farnesyl transferase inhibitors, may prove synergistic when combined with tyrosine kinase inhibitor directed at EGFR family members.  The EGFR Family and Targeted Therapies in Malignant Glioma  The information outlined above suggests that glioblastoma multiforme is a reasonable tumor for the application of targeted molecular therapies. A number of EGFR family–associated targets have been studied for possible therapeutic interventions. Strategies have been aimed at (a) interference with ligand binding; (b) the reduction of tyrosine kinase activity; (c) interference with key regulatory elements of downstream signal transduction pathways; (d) using antibodies and ligands as vectors to target toxins to EGFR family proteins; and (e) the use of selectively expressed EGFR mutants as antigens to generate vaccines.  Many of these strategies have only been tested in preclinical settings. For example, antibodies directed against the EGFR ligand, TGF-α, have been evaluated in human glioma cell lines and have been shown to inhibit their growth (55). However, because EGFRvIII, the most common mutant EGFR family member in glioblastoma multiforme, is ligand independent, such strategies, when applied alone, are likely to show only modest benefit.  Antibodies. A number of monoclonal antibodies directed against HER family proteins have been developed and tested in preclinical models. Proof of principle for this application is seen with Herceptin, an anti-HER2 antibody, effective in the treatment of some breast cancer patients with systemic disease (i.e., not involving CNS) whose tumors overexpress this receptor. Herceptin has been shown to mediate cell death of HER2/neu–expressing glioblastoma cell lines in vitro, but in vivo activity against intra-CNS tumors is lacking, perhaps due to impaired uptake in the CNS (56). These include unmodified and modified antibodies, targeted against EGFRvIII. Significant anti–glioblastoma multiforme responses have also been shown with anti-EGFRvIII antibodies in s.c. animal models of glioblastoma multiforme. Potential mechanisms of action include antibody-dependent cellular cytotoxicity and receptor internalization (57).  The blood-brain barrier represents a significant obstacle to the successful treatment of intra-CNS tumors with systemically administered antibodies, despite the apparent increased region permeability within brain tumors (58–60). Most of the clinical trials to date have attempted to circumvent the blood-brain barrier via intracavitary antibody administration. None of these studies has yet employed an anti-EGFR product. However, responses have been seen in animal studies following the intracranial administration of boronated EGFRvIII-specific antibodies (61). Objective responses have been seen in humans following the intracavitary administration of monoclonal antibodies conjugated to radioisotopes (62). Antibodies can also be used as vectors to deliver other conjugated toxins to tumor cells. Potential toxins (e.g., tyrosine kinase inhibitors or bacterial toxins) can be either linked to the antibodies themselves or packaged in liposomes linked to antibody fragments (63). Although these antibody-based constructs will likely reach and affect tumor cells adjacent to the resection cavity, the most common site of early tumor recurrence, they may not reach tumor cells that lie at a distance from the site of their administration.  Results of a single phase I/II clinical trial with i.v. EGFR antibodies in malignant glioma have been reported from the Center of Molecular Immunology, Havana, Cuba (64). Twenty-nine patients with newly diagnosed malignant glioma received weekly infusions of a humanized monoclonal antibody (h-R3) directed against the EGFR extracellular domain. Sixteen of these patients had glioblastoma multiforme and, at the time of publication, median survival was reported to be 17.47 months. Immunoscintigraphy studies did show uptake of a technitium 99–labeled anti-EGFR antibody in some patients in the study, suggesting that the antibodies may be crossing from the systemic circulation into the CNS.  Another trial in development will combine temozolomide, radiation therapy, and cetuximab (IMC-C225), a humanized monoclonal antibody approved by the Food and Drug Administration in 2004 for the treatment of colorectal cancer. Cetuximab binds to both EGFRwt and EGFRvIII (65). Thus, it may be useful in tumors that express only the wild-type receptor.  Vaccines. EGFRvIII is currently being used as an antigen in a vaccine-based clinical trial for glioblastoma multiforme (66). Subjects with newly diagnosed glioblastoma multiforme receive s.c. injections of an EGFRvIII vaccine every 2 weeks for 8 weeks and then monthly until evidence for tumor progression is shown. Results are pending.  Tyrosine kinase inhibitors. Several small molecules that inhibit the tyrosine kinase activity (tyrosine kinase inhibitors) of HER family receptors have been developed and tested for activity against tumors, including malignant glioma (reviewed in refs. 67, 68). The greatest experience to date is with the reversible tyrosine kinase inhibitors erlotinib (OSI-774, Tarceva) and gefitinib (ZD1839, Iressa). Both of these agents were approved by the Food and Drug Administration for patients with advanced non–small-cell lung cancer who had failed prior cytotoxic chemotherapies. Sensitive tumors harbor specific in-frame deletions or amino acid substitutions that allow these drugs to act within EGFR ATP binding domains. Of note, gliomas do not carry the EGFR kinase domain mutations associated with tyrosine kinase inhibitor responses.  Given the high incidence of EGFRvIII mutations and associated amplification in gliomas, the activity of tyrosine kinase inhibitor in tumors with these mutations warranted investigation. The relative effects of gefitinib have been tested on cell lines that harbor wild-type EGFR and EGFRvIII (69). Higher doses of gefitinib were needed to inhibit downstream EGFR phosphorylation and signaling events from EGFRvIII. Furthermore, low levels of gefitinib actually seemed to confer a growth advantage to cells with EGFRvIII overexpression, measured both by rates of proliferation and anchorage-independent growth in this study. Erlotinib showed only modest efficacy in a mouse model of EGFRvIII-driven lung tumors (70). Erlotinib treatment resulted in decreased total EGFRvIII and phosphorylated EGFR levels. More robust responses were seen with the irreversible EGFR/HER2 inhibitor HKI-272.  As noted above, EGFRvIII is constitutively active, allowing phosphorylation and downstream activation of signal transduction cascades independent of ligand binding. Thus, one might predict resistance to agents designed to work through receptor-mediated blockade. Several lines of investigation support this view (71, 72). Furthermore, EGFRvIII is relatively resistant to degradation, allowing for prolonged activity at the cancer cell surface (73). Resistance to the current generation of tyrosine kinase inhibitors may be mediated, in part, by this effect.  Despite these findings, gefitinib and erlotinib have been evaluated for their safety and efficacy in the treatment of both newly diagnosed and recurrent malignant glioma, either alone or in combination with radiotherapy or chemotherapy. Some of these trials are ongoing, others closed to accrual with either no or only preliminary data available (67). Results of several completed trials have been published and summarized (74–77). The conclusions drawn from these studies to date are (a) these agents can be safely administered to glioma patients; (b) anticonvulsants that induce hepatic cytochrome P450 enzymes, commonly used in brain tumor patients, alter the metabolism of these agents; (c) the proportion of clinical responses are disappointing; (d) occasional long-lasting clinical responses are seen; and (e) complex mechanisms involving communication between various levels of EGFR-mediated signal transduction cascades account for responses to treatment.  Evaluation of tumor samples from two malignant glioma trials done by the North American Brain Tumor Consortium, one using gefitinib and the other using erlotinib, failed to show consistent and effective inhibition of phosphorylation both of EGFR and its downstream effectors, AKT and extracellular signal–regulated kinase (78). It is important to recognize that more than 100 subjects were enrolled in these studies and only a fraction had tumor available for molecular analysis (21 samples). Only one of the patients had a major response based on radiographic review. Of note, none of the EGFRvIII mutations were detected in this study and there was a lower incidence of EGFR gene amplification than expected. Rather, several new missense mutations were detected. Finally, tissue analysis of drug accumulation suggested that erlotinib penetration into tumors was disappointingly low in most cases and, at its highest, failed to correlate with inhibition of EGFR phosphorylation. Gefitinib, on the other hand, seemed to have higher penetration. Whereas the small numbers made the results from this preliminary study complex, systematic molecular analyses of this type will be important in future trials attempting to interpret the activity of these agents in glioma.  Occasional sustained responses to either gefitinib or erlotinib in select cases have led some to hypothesize that a multifactorial genetic predisposition involving both EGFR signaling and AKT activation was responsible for the beneficial responses to tyrosine kinase inhibitor observed in some patients (71, 72). EGFR-mediated phosphorylation of phosphatidylinositol 4,5-bisphosphate to its active state, phosphatidylinositol 3,4,5-trisphosphate, leads to activation of AKT and cell proliferation while simultaneously inhibiting apoptosis. PTEN ends the phosphatidylinositol 3,4,5-trisphosphate signal, converting the second messenger back to phosphatidylinositol 4,5-bisphosphate. Some patients both have constitutively active EGFR-mediated signaling and have also lost the PTEN tumor suppressor gene. This combination of molecular events is associated with high AKT activity and a poor prognosis. In contrast, preservation of PTEN, according to a study by Mellinghoff et al., provides a setting in which response to a tyrosine kinase inhibitor could be observed.  Mellinghoff et al. (79) retrospectively analyzed pretreatment tissue from 26 of 49 subjects with recurrent malignant glioma treated with either erlotinib or gefitinib. Response was defined as >25% reduction in the bidirectional area of contrast enhancing tumor on brain magnetic resonance imaging scans. Three of nine patients with a clinical response were noted to have the EGFRvIII mutation. They found and verified, in a second cohort of 33 patients, a significant association between coexpression of EGFRvIII and PTEN and clinical response (P < 0.001). This study provides a strong rationale for prospective testing of the molecular phenotype in malignant glioma, with a future goal of guiding treatment decisions.  Conclusions and Future Directions  The EGFR family plays a central role in both development and tumorigenesis within many organ systems, including the CNS. In this article, we have reviewed some of the relationships between normal brain development and glial tumorigenesis. These are summarized in Fig. 2 . The mechanisms whereby EGFR-mediated signal transduction influences normal brain and the glioma phenotype differ significantly from those of other organs and tumor types. This is not surprising, given the complexity of EGFR-mediated signaling and its diverse roles in normal growth and development as well as neoplasia. Because the EGFR family mediates so many interactions between cells and their environment, there are many ways in which these processes might be influenced. Regarding the CNS, normal development is highly dependent on temporal and spatial restrictions of progenitor cells of which the ultimate differentiation results in tremendous variability in function within a compact anatomic space. EGFR-mediated events are critical in this process.  Fig. 2. Download figureOpen in new tabDownload powerpoint Fig. 2. Summary diagram depicting the involvement of EGFR and EGFR-mediated signaling pathways in CNS development, carcinogenesis, and targeted therapeutics. (See Table 1 for an outline of the EGFRs and their involved ligands.) In normal CNS development, neurotransmitters are known to interact with EGFR-mediated processes to control growth and differentiation not only of the brain but also of the organism as a whole. This is well characterized in the events surrounding the initiation and subsequent inhibition of puberty (80). To date, the effects of neuron-to-glia signaling in gliomagenesis remain unstudied. It is possible that such inputs might contribute to the regional and temporal incidence of particular glial tumors characterized, to a large extent, by their location within the neuraxis and their usual age of onset.  Most EGFR-directed glioma therapies used to date have focused on single agents, with modest results at best. However, when the molecular mechanisms underlying the observed responses are understood, the possibility will exist that the overall effectiveness of a given therapy can be predicted a priori. A number of agents that target the EGFR family and other RTK family members important in glial tumorigenesis are under investigation and summarized in Table 3 . These agents, in addition to the other anti-EGFR strategies outlined above, should provide further insights into the complex physiology that underlies malignant glioma formation while offering new hope to brain tumor patients. Our current level of understanding of these pathways suggests the need to combine agents with different mechanisms of action, targeting signal transduction at several levels simultaneously, with close attention to pharmacodynamic constraints present in the unique environment of the CNS. Glioblastoma multiforme (GBM) is the most common brain tumour and has the worst prognosis. Epidermal growth factor receptor (EGFR) gene amplification, mutation and re-arrangement (all of which enhance tumour growth, survival, progression and resistance to therapy) are frequently observed in primary GBM. The most common EGFR variant in GBM, the EGFRvIII, is characterised by a deletion of 267 amino acids in the extracellular domain, leading to a receptor which is unable to bind ligand yet is constitutively active. Together with its impaired internalisation and degradation, the EGFRvIII enhances the tumourigenic potential of GBM by activating and sustaining mitogenic, anti-apoptotic and pro-invasive signalling pathways. This EGFRvIII-mediated enhanced tumourigenicity combined with the lack of EGFRvIII expression in normal tissue makes it an ideal candidate for targeted therapy. This review summarizes the current knowledge about the role of EGFRvIII in GBM and discusses therapeutic agents targeting EGFRvIII that are being evaluated as treatments for GBM.  ABSTRACT The development and neoplastic progression of human astrocytic tumors appears to result through an accumulation of genetic alterations occurring in a relatively defined order. One such alteration is amplification of the epidermal growth factor receptor (EGFR) gene. This episomal amplification occurs in 40-50% of glioblastomas, which also normally express endogenous receptors. Moreover, a significant fraction of amplified genes are rearranged to speifically eliminate a DNA fragment containing exons 2-7 of the gene, resulting in an in-frame deletion of 801 bp of the coding sequence of the extracellular domain. Here we used retroviral transfer of such a mutant receptor (de 2-7 EGFR) into glioblastoma cells expressing normal endogenous receptors to test whether the mutant receptor was able to augment their growth and malignancy. Western blotting analysis showed that these cells expressed endogenous EGFR of 170 kDa as well as the exogenous de 2-7 EGFR of 140-155 kDa. Although holoEGFRs were phosphorylated on tyrosine residues only after exposure of the cells to liand, de 2-7 EGFRs were constitutively phosphorylated. In tissue culture neither addition ofEGF nor expression of the mutant EGFR affected the rate of cell growth. However, when cells expressing mutant EGFR were implanted into nude mice subcutaneously or intracerebrally, tumorigenic capacity was greatly enhanced. These results suggest that a tumor-specific alteration of the EGFR plays a significant role in tumor progression perhaps by influencing interactions oftumor cells with their microenvironment in ways not easily assayed in vitro. The clonal evolution of neoplasms toward increasing malignancy has been ascribed to the accumulation of genetic alterations that are advantageous to growth (1). This hypothesis has gained experimental support in colorectal cancer (2), squamous carcinoma of the cervix (3), lymphoid tumors (4), and gliomas (5). In the latter case, a series ofevents occurring in a relatively ordered fashion have been identified and include loss of heterozygosity for chromosome 17p (6, 7), mutation of the p53 gene (5), deletions of the interferon gene cluster on chromosome 9p (8), amplification and alteration of the epidermal growth factor receptor (EGFR) gene (9, 10), and monosomy of chromosome 10 (11, 12). The involvement of EGFR in human cancer has been inferred from its ability to confer ligand-dependent transformation to NIH 3T3 cell transfectants (13) and by its frequent gene amplification and overexpression in carcinomas of the breast (14), ovary (15), cervix (15), kidney (15), and squamous cells (16). In glial tumors ofthe central nervous system, such alterations are restricted to grade III (anaplastic astrocytoma) and, especially, grade IV (glioblastoma multiforme) disease (10, 17-19). Several clinical and histopathological studies have shown that EGFR gene amplification is related to a shorter interval to relapse and poorer survival (17, 20, 21). Nearly half of tumors examined had significantly amplified EGFR genes (10, 18, 19) and, of these, the majority also show rearrangements ofthe gene resulting in alterations of its transcript so that such gliomas express both wild-type endogenous EGFR as well as the episomal mutant form. The most common of the rearrangements are genomic alterations leading to deletion of exons 2-7 in the EGFR mRNA, which causes an in-frame truncation of 801 bp in the extracellular domain of the molecule (9, 10, 22-24). Introduction of such a cDNA isolated from a glioma into murine NIH 3T3 cells resulted in expression of a constitutively phosphorylated membrane-associated 150-kDa receptor species, which caused weak ligand-independent cell transformation (25). Despite the common occurrence of these EGFR aberrations in gliomas, virtually all cell lines derived from them lose these features in tissue culture (26), suggesting either a growth disadvantage in vitro or a selection for EGFR overexpression in vivo. To study the biological effects of a mutant EGFR that lacks exons 2-7 (de 2-7 EGFR), we used retroviral transfer to achieve high-level expression of this protein in glioblastoma cells expressing wild-type endogenous receptor. In glioblastoma cells de 2-7 EGFR exhibited constitutively activated self-phosphorylation and exerted a pronounced enhancement of tumorigenicity in vivo. MATERIALS AND METHODS Construction and Production of Mutant EGFR Retroviral Vectors. Mutant EGFR (de 2-7 EGFR) was constructed from the full-length EGFR cDNA clone pRC/EGFR (a gift of M. G. Rosenfeld, Howard Hughes Medical Institute) by eliminating the sequences of exons 2-7. Two DNA fragments corresponding to nucleotides 83-274 and 1076-1357 (27) were synthesized by PCR using two pairs of primers, TCAGATCTCTAGAGAGC (sense)/AGACCTCCTTTTCTTTCCGGAGG (antisense) and AAGAAGTGTCCCATTAATTATGTGGT (sense)/AGGAGACCTAGGTGTCC (antisense), respectively. The first PCR product was digested with Xba I and Stu I and the second fragment was digested with Ase I, blunt-ended with the Klenow fragment of DNA polymerase, and digested with BamHI. The two resulting fragments were then ligated and subcloned into pBluescript SK- (Stratagene). The Xba I/Dra I EGFR fiagment was isolated from the plasmid and ligated with the Dra I/HindIII fragment from pRC/EGFR to create de 2-7 EGFR, which was cloned into pBluescript SK- and verified by DNA sequencing. De 2-7 EGFR was transferred to a retroviral vector, pLRNL (28), and introduced into 0-2 cells with Lipofectin (GIBCO/BRL). Abbreviations: EGF, epidermal growth factor; EGFR, epidermal growth factor receptor; FACS, fluorescence-activated cell sorter. ITo whom reprint requests should be addressed. 7727 The publication costs of this article were defrayed in part by page charge payment. This article must therefore be hereby marked \"advertisement\" in accordance with 18 U.S.C. §1734 solely to indicate this fact. Proc. Natl. Acad. Sci. USA 91 (1994) Virus was collected 48 hr later and was used to infect PA317. G418-resistant PA317 cells were assayed for mutant EGFR expression by fluorescence-activated cell sorter (FACS) analysis; cells staining positively were sorted, analyzed for de 2-7 EGFR expression by Western blotting analysis, and used as virus-producing stocks. FACS. The human glioblastoma cell line U87MG (29) was infected with virus and selected in a medium containing 400 ,ug of G418 per ml according to described procedures (28); 5 x 10W cells in 200 A4 of staining buffer [phosphate-buffered saline (PBS)/1% fetal bovine serum (FBS)/0.1% sodium azide] were exposed to anti-EGFR monoclonal antibody Ab-1 (clone 528) (30) or Ab-5 (clone EGFR.1) (31) (both obtained from Oncogene Science) for 30-60 min and then to fluorescein isothiocyanate-conjugated goat anti-mouse immunoglobulin antibody (PharMingen) for 30-60 min. Stained cells were analyzed with a FACSort (Becton Dickinson); for cell sorting, sodium azide was omitted from the staining buffer, and cells were sorted with a FACStar (Becton Dickinson). Western Blotting. Cells (4 x 104 cells per well) in a 12-well plate were cultured overnight in medium containing 10%1 FBS (PA317 and PA317 virus stock cells) or 2% dialyzed FBS (U87MG and U87MG.AEGFR). For antibody and ligand treatments, cells were washed with PBS and treated with anti-EGFR antibody Ab-1 (5 pg/ml) for 3 hr at 370C, and then 100 ng of EGF (Collaborative Biomedical Products, Bedford, MA) per ml for 5 min at room temperature. Cells were washed with PBS and lysed in 200 A4 of lysis buffer [0.064 M Tris HCI, pH 6.8/1.28% SDS/12.8% (vol/vol) glycerol/1.28% 2-mercaptoethanol/0.25% bromphenol blue] and then boiled. Frozen tumor samples were homogenized in a Dounce homogenizer containing lysis buffer, boiled, and centrifuged to remove cell debris. The lysates were electrophoresed through SDS/7% polyacrylamide gels and the separated proteins were transferred to nitrocellulose membranes before being probed with anti-EGFR rabbit polyclonal antibody 1964 (32), or anti-phosphotyrosine [anti-Tyr(P)] mouse monoclonal antibody PY20 (Transduction Laboratories, Lexington, KY), as described (33). Tumorigenicity. For subcutaneous inoculation, U87MG (1 x 106 cells) or U87MG.AEGFR (1 x 106 or 2 x 105 cells) was suspended in 0.2 ml of PBS and injected simultaneously into the left or right flanks, respectively, of 4- to 5-wk-old female nude mice ofBALB/c background. The growing tumors were measured twice a week with a caliper, and tumor volumes were calculated using width (a) and length (b) measurements (a2b/2, where a < b) (34). For intracerebral stereotactic implantation, 5 x 104-5 x 105 cells in 5 /4 of PBS were inoculated into the corpus striatum in theright hemisphere (2.5-3.0 mm deep; 1 mm anterior and 1.8 mm lateral to the bregma) of the nude mouse brain (35). Brains were removed at various time points, embedded in OCT compound (Miles), frozen in liquid nitrogen, and stored at -80°C. Thin cryostat sections (5-7 pm) were stained with hematoxylin, and tumor size was microscopically determined. Imnuinobhstochemlstry. Cryostat sections (5,uim) were airdried overnight, washed with PBS, and fixed in cold acetone for 10 min. Endogenous peroxidase activity was quenched with 2% hydrogen peroxide in PBS, and nonspecific binding sites were blocked with 1% bovine serum albumin for 15 min at room temperature. The sections were incubated with the anti-EGFR antibody Ab-1 (5 tpg/ml) for 1 hr at room temperature. Biotinylated horse anti-mouse IgG(H+L) antibody and avidin/biotin/horseradish peroxidase were then applied to the sections as recommended (Vector Laboratories). Immunoreactivity was revealed by adding the peroxidase substrate 3,3\\'-diaminobenzidine tetrahydrochloride (Vector Laboratories), and sections were counterstained with hematoxylin. RESULTS Expression of Mutant EGFR in Glioblastoma Ceils. U87MG cells were infected with the amphotropic virus, which carries a neomycin-resistance gene and an EGFR gene deleted for exons 2-7. Cells were selected for G418 resistance and assessed for the presence of EGFR on their cell surfaces by FACS analysis using the Ab-1 monoclonal antibody, which recognizes an epitope shared by wild-type and de 2-7 EGFRs; 85% of the cells were more heavily stained than the parental U87MG cells, indicating expression of the exogenous gene; the 7% of cells that were most brightly stained were sorted and the population (U87MG.AEGFR) was characterized further by FACS analysis. U87MG.AEGFR reacted equally as well as parental cells with the monoclonal antibody Ab-5, which reacted with wild-type but not de 2-7 EGFR, suggesting that this population expressed levels of the endogenous EGFR similar to that of parental cells (Fig. 1A). However, when cells were stained with Ab-1, the mean of fluorescence intensity for U87MG.AEGFR was 4-5 times higher than that of parental cells (Fig. 1B). By careful concordant comparisons, this level of expression was equivalent to that of the A431 squamous cell carcinoma line, which has been shown to express 1-3 x 106 EGFR molecules on its surface (36-38). Moreover, the U87MG.AEGFR population was relatively homogeneous in its expression as demonstrated by the narrow fluorescence intensity profiles. These characteristics remained stable in the presence of G418 selection over many months of culture. Self-phosphorylation of EGFR occurs rapidly after ligand binding and functions in assembly of signaling components that contain SH2 domains (39). The transforming protein 50 25. S. 2 z i0 S .5\\' 0 11 50 25- u bo A a B 101 102 1 Fluorescence Intensity 104 FiG. 1. Surface expression of endogenous and de 2-7 EGFR in U87MG and U87MG.AEGFR cells as determined by FACS analysis. U87MG (dotted lines) and U87MG.AEGFR (solid lines) were stained with anti-EGFR monoclonal antibody Ab-5 (A), which recognizes endogenous EGFR exclusively, or Ab-1 (B), which reacts with both endogenous and de 2-7 EGFR. Bars on the abscissa in A and B represent maximum staining of cells in the absence of primary antibody. ~~~~-4 7728 Genetics: Nishikawa et al. °1 -------9 16, .02 1,03 164 Proc. NatL. Acad. Sci. USA 91 (1994) 7729 v-erbB is constitutively active and self-phosphorylated (40, 41). To determine the activation state of de 2-7 EGFR, U87MG and U87MG.AEGFR cells were treated without or with EGF, and receptor self-phosphorylation was measured with monoclonal anti-Tyr(P) antibody. Probing Western blots with anti-EGFR antibody showed that parental U87MG and U87MG.AEGFR expressed the same level of 170-kDa endogenous EGFR (Fig. 2A, lanes 1 and 5). Endogenous EGFR was tyrosine phosphorylated only when ligand was added to the medium (Fig. 2B, lanes 2 and 6). Anti-EGFR antibody also detected molecules of 140 and 150-155 kDa, which were unique to U87MG.AEGFR cells, and corresponded to posttranslationally modified de 2-7 EGFR. The ratio of 170-kDa species to 140- to 155-kDa molecules was similar to that observed in FACS analysis, suggesting that the majority of de 2-7 EGFRs produced in U87MG.AEGFR are expressed on their surface. The 150- to 155-kDa molecules reacted with anti-Tyr(P) antibody but, in contrast to endogenous EGFR, their phosphorylation was constitutive and not dependent on exogenous ligand (Fig. 2B, lanes 5 and 6). To rule out the possibility that constitutive phosphorylation of the 150- to 155-kDa proteins was due to residual ligand in the culture medium, an anti-EGFR antibody that blocks ligand binding to the receptor and subsequent autophosphorylation of the receptor was used (42). The effectiveness of this antibody is demonstrated for parental U87MG cells in Fig. 2B (lanes 3 and 4). However, the antibody was not able to inhibit phosphorylation of the 150- to 155-kDa molecules in U87MG.AEGFR cells, even though phosphorylation of the endogenous 170-kDa species in these cells occurred in response to EGF and was blocked by exposure to the antibody (lanes 7 and 8). The inability of EGF to stimulate tyrosine self-phosphorylation of de 2-7 EGFR suggested that removal of exons 2-7 disrupted ligand binding. At least one point of contact between receptor and ligand, Tyr101 (43), located in exon 4 (44) is lost. To examine this directly, levels of 17-I-labeled EGF binding to U87MG and U87MG.AEGFR cells were compared. Both the Kd and B, values were similar (Kd = 3.8 and 3.2 nM; B. = 2.1 and 2.3 x 10\\' EGF binding sites per cell, for U87MG and U87MG.AEGFR cells, respectively). Be- 1 2 3 4 5 6 7 8 Ab EGF A 190 - A 125- 88 - B 190- A 125 - -++- -++ cause the two cells contain similar amounts of holo-EGFR (Fig. 2A), the observation that similar amounts ofligand were bound argues that EGF does not bind to the 4- to 5-fold excess amount of de 2-7 EGFR present on the surface of U87MG.AEGFR cells. Effect of Mutant EGFR on Cell Growth in Vitro. Because de 2-7 EGFR was expressed at a high level in glioblastoma cells and was constitutively active as assessed by tyrosine phosphorylation, we analyzed its influence on cellular growth rate. Fig. 3A shows that the growth of these two cell populations was identical when they were cultured in medium containing 10%1 FBS. Since 2% dialyzed serum was the minimum requirement to sustain growth, we cultured U87MG and U87MG.AEGFR cells in medium containing 2% dialyzed FBS and various concentrations of the ligands, EGF, or transforming growth factor a through a concentration range of 0.1-100 ng/ml. In no case did either ligand affect the growth rates of either cell population and, although U87MG.AEGFR grew slightly faster than U87MG in medium containing 2% dialyzed serum, their saturation densities were the same (Fig. 3B). To assume that the retroviral vector itself was not playing a role in these results, we also infected U87MG cells with a similar construct expressing luciferase. These cells and the parental cells had identical growth characteristics. Effect of Mutant EGFR on Tumorgenicity in Vivo. We.next tested whether the expression of a mutant EGFR had any effect on the ability of glioblastoma cells to form tumors in animals. U87MG.AEGFR cells grew remarkably faster than parental U87MG cells (or U87MG cells infected with the luciferase viral construct) when they were implanted subcutaneously into the flanks of nude mice (Fig. 4). U87MG.AEGFR caused palpable tumors in 1 wk when 1 x 106 cells were inoculated, whereas the same number of parental U87MG cells required 3 wk to grow to a similar extent; at 1 x 106 cell inoculation, U87MG.AEGFR and U87MG cells required 2 and 5 wk, respectively, to form tumor masses of 1 cm3. Even when the number of U87MG.AEGFR cells implanted was reduced 5-fold to 2 x 10\\', they still grew more vigorously than U87MG cells, as shown in Fig. 4. To determine whether this growth advantage had site specificity, the glioblastoma cells were stereotactically implanted into the brains of nude mice. Cells (2 x 10\\') were inoculated and after 2 wk, three ofthree animals that received U87MG.AEGFR cells had tumors of 5 mm diameter or greater, whereas only one of five mice implanted with U87MG cells developed a tumor that had a diameter of 2 mm. Examples are shown in Fig. 5. Four animals that received 5 x 10\\' U87MG.AEGFR cells died within 4 wk after implan- .8 E z - 88 - 105 FIG. 2. Western blotting analysis of EGFR expression and tyrosine phosphorylation in U87MG (lanes 1-4) and U87MG.AEGFR (lanes 5-8). Cells were treated (+) or not (-) with the monoclonal antibody Ab-1, which is a competitive antagonist for EGF binding, for 3 hr at 37°C and incubated in the presence (+) or absence (-) of EGF (100 ng/ml) for 5 min at room temperature before being lysed in SDS-containing buffer. (A) Blots were probed with anti-EGFR polyclonal antibody 1964. (B) Blots were probed with antiphosphotyrosine monoclonal antibody PY20. Endogenous EGFR of 170 kDa (.) and tyrosine-phosphorylated proteins of 150-155 kDa (A) are indicated on the left. Numbers are molecular mass markers (kDa). The experiment was reproduced four separate times. 104L 0 A r . . .___ 10 20 0 Days 10 20 FIG. 3. Lack of in vitro effects on growth of U87MG cells expressing de 2-7 EGFR. U87MG (o and *) and U87MG.AEGFR (o and o) cells were grown in 10%6 FBS (A) or 2% dialyzed FBS (B) in the presence ( and o) or absence ( and o) of EGF (100 ng/ml). Cell numbers were determined at 3- to 4-day intervals and are shown as means + SD of triplicate samples. Similar experiments were done four separate times. Genetics: Nishikawa et al. 22600 Proc. Natl. Acad. Sci. USA 91 (1994) 3000 E E 2000 E 1000 0 0 10 20 30 Days Fio. 4. Effect of de 2-7 EGFR expression on U87MG cell tumorigenicity in subcutaneous implantation. Either 1 x 106 (o) or 2 x 10W () U87MG.AEGFR cells were injected into the right flank of nude mice while 1 x 106 (A&) U87MG cells were inoculated into the left flank of the same animals. Tumor size was quantitated at the indicated times. U87MG, n = 10; U87MG.AEGFR, n = 5. Similar experiments were done three separate times. tation due to the overgrowth of tumors, which attained sizes of >10 mm in diameter, while four mice inoculated with the same number of parental U87MG cells or four animals implanted with U87MG cells infected with retroviral luciferase construct formed small tumors of <4 mm in diameter. Another group of animals that received 5 x 104-1 x 105 cells also showed a similar outcome after 3 wk: two of two mice that received U87MG.AEGFR cells developed tumors of >5 mm in diameter. None of the five animals implanted with parental U87MG cells produced tumors of >1 mm. Thus, the expression of de 2-7 EGFR in glioblastoma cells that also express endogenous EGFR conferred a substantial growth advantage in vivo. Previous studies showed that amplification of EGFR expression is an unstable situation in vitro (26). To confirm that the enhanced tumorigenicity ofU87MG.AEGFR cells was due to the effect of the mutant EGFR, we analyzed EGFR expression in tumors excised from the implanted animals. Western blotting analysis showed that tumors obtained from subcutaneous and intracerebral implantations ofU87MG.AEGFR expressed endogenous EGFR of 170 kDa as well as smaller species of 140 and 150-155 kDa, in which the 150- to 155-kDa receptors were tyrosine-phosphorylated and the ratio of endogenous to mutant receptor was comparable to that of U87MG.AEGFR grown in vitro (as in Fig. 2), indicating stability of expression of the mutant E1GFR (data not shown). Moreover, the expression of EGFRs In U87MG.AEGFR and U87MG tumor cells was homogeneous as shown by the immunohistochemical staining of cryostat sections (Fig. 5). DISCUSSION Expression of a mutant EGFR, which occurs commonly in human glial tumors, enhances the tumorigenic activity of U87MG glioblastoma cells in vivo in the absence of detectable in vitro growth advantages. Holo-EGFR is activated by ligand binding, which causes rapid dimerization (45), conformational change (46), activation of its intrinsic protein-tyrosine kinase activity, and autophosphorylation (47). Binding ofligand to the receptor also induces receptor interalizaton, downregulation, and attenuation of signaling (48). Thus, EGFR activities are normally under stringent positive as well as negative regulation. Disruption ofthe control process by mutations that either affect ligand-dependent tyrosine kinase activation mechanisms or impair ligand-induced receptor internalization may then lead to a constitutive elevation of receptor activity (48, 49). The mutant receptor we studied here may represent the former case, since it is constitutively tyrosine phosphorylated. Deletion of the portion encoded by exons 2-7 of the FIG. 5. Effect of de 2-7 EGFR on U87MG tumorigenicity in intracerebral implantation. Either 2 x 10W U87MG (A) or 2 x 10W U87MG.AEGFR (B) cells were stereotactically inoculated into the brains of nude mice. Two weeks postimplantation, brains were resected for analysis. Frozen sections were stained with the antiEGFR antibody Ab-1 and counterstained with hematoxylin. The small tumor shown for U87MG in A is the largest obtained in the series, whereas the tumor shown for U87MG.AEGFR in B was of average size in the series. Tumors are marked by arrows. Similar experiments were done with 5 x 104, 1 x 10\\', and 5 x 105 cells of both types in at least two animals each. (Bars = 1 mm.) extracellular domain appears to mimic the effect of ligand binding to the receptor, thereby inducing the active conformational change of the receptor and, as a consequence, activating the intrinsic tyrosine kinase activity and so enhancing their tumorigenic activity analogous to v-erbB (40, 41). Because this mutation removes only =\\'1/4th of the N-terminal portion ofthe extracellular domain, it is also possible that this form of EGFR is able to interact with unknown ligands in vivo or that it forms heterodimers with other members ofthe EGFR family, thereby eliciting the biological effects. This could explain why expression of the mutant EGFR in U87MG cells influences their phenotype in vivo but not in vitro. Nevertheless, previous findings have shown that a mutant EGFR derived by a large deletion of virtually the entire extracellular domain of the receptor was constitutively active and capable oftransforming established rodent fibroblasts to a tumorigenic phenotype (50), suggesting by analogy that the enhancing effect on tumorigenicity observed results, at least partially, 7730 Genetics: Nishikawa et al. Proc. NatL Acad. Sci. USA 91 (1994) 7731 from ligand-independent activation of the receptor. This is further supported by the observation that mutant EGFR in tumors isolated from implanted animals displayed constitutive tyrosine self-phosphorylation patterns similar to those of the cells cultured in vitro. In spite of its remarkable effect in vivo, expression of the mutant EGFR did not influence the growth rate in vitro. Perhaps this is because U87MG is a fully transformed cell line so that its in vitro responses are no longer affected by the added stimulation provided by the constitutive phosphorylation of the mutant receptor. Although EGF activates holo receptors in these cells, EGF did not influence the rate of cell growth. We have expressed high amounts ofwild-type EGFR in U87MG and found this to have little effect on in vitro growth. Yamazaki et al. (25) have also demonstrated that overexpression of a similarly mutated EGFR in NIH 3T3 cells only slightly affects the cellular morphology regardless of the presence of EGF, while expression ofa high level ofwild-type EGFR in the same cells resulted in alterations in cellular morphology and the ability of the cell to form foci in a ligand-dependent manner (25). Effects on growth clearly depend on cell type and the presence ofother mutational events. Amplification and rearrangement of the EGFR gene occurs almost exclusively at the latest stage ofthe glioma malignancy process, and our results argue for a strong selection for cells carrying such receptors. Since this selection takes place in vivo but not in vitro, perhaps the mutant receptor plays a role in the interactions of tumor cells with their environment rather than directly on cell growth. The late stage of glioma at which this occurs suggests several potential processes that may be influenced in this way. Tumors at this point are dependent on neovascularization and perhaps the signals transduced by the mutant EGFR enhance the ability of the tumor cells to induce this process. At this stage, the tumors also become increasingly necrotic and perhaps the tumor cells that carry these mutant receptors are better able to survive in these ischemic conditions. Whatever the case, the present studies provide an experimentally malleable system within which to delineate this particular type of in vivo growth advantage.  The epidermal growth factor receptor (EGFR) is overexpressed in a variety of human epithelial tumors, often as a consequence of gene amplification. Tumors with EGFR gene amplification frequently contain EGFR gene rearrangements, with the most common extracellular domain mutation being EGFRvIII. This mutation leads to a deletion of exons 2–7 of the EGFR gene and renders the mutant receptor incapable of binding any known ligand. Despite this, EGFRvIII displays low-level constitutive signaling that is augmented by reduced internalization and downregulation. Aberrant EGFRvIII signaling has been shown to be important in driving tumor progression and often correlates with poor prognosis. It is clear that EGFRvIII is expressed in a considerable proportion of patients with glioblastoma multiforme (GBM). The presence of EGFRvIII in other tumor types, however, remains controversial. In this review, we critically analyze the evidence for the expression of EGFRvIII in a range of tumor types and discuss recent findings pertinent to its function and biology in GBM. Abbreviations AMPK AMP-activated protein kinase CDK cyclin-dependent kinase DOCK180 dedicator of cytokinesis 1 EGFR epidermal growth factor receptor FDG fluorodeoxyglucose FOX forkhead GBM glioblastoma multiforme HGF hepatocyte growth factor IHC immunohistochemistry IL-8 interleukin-8 LIF leukemia inhibitory factor MAPK mitogen-activated protein kinase mTOR mammalian target of rapamycin mTORC2 CREB-regulated transcription co-activator 2 NF-κB nuclear factor κB NSCLC non-small-cell lung cancer PDGFRβ platelet-derived growth factor receptor β PI3K phosphatidylinositol 3-kinase RTK receptor tyrosine kinase SFKs Src family kinases Stat3 signal transducer and activator of transcription 3 VEGF vascular endothelial growth factor VEGFR2 VEGF receptor 2 WB western blotting wt wild-type Introduction  The epidermal growth factor receptor (EGFR) is a pivotal regulator of normal cellular growth in tissues of epithelial origin. Dysregulated EGFR signaling (resulting from mechanisms such as cell-surface overexpression, autocrine activation and EGFR gene mutation) contributes to the formation of many epithelial malignancies in humans [1, 2]. Where dysregulation is the result of cell-surface EGFR overexpression, there is often associated gene amplification [1] and/or mutation [3, 4]. Although several EGFR mutations have been described, the most common extracellular mutation is EGFRvIII (also known as de2-7EGFR and ΔEGFR). EGFRvIII is a tumor-specific mutation that results from in-frame deletion of 801 base pairs spanning exons 2–7 of the coding sequence [5-7]. This deletion removes 267 amino acids from the extracellular domain, creating a junction site between exons 1 and 8 and a new glycine residue (Fig. 1) [8, 9]. EGFRvIII has a molecular mass of approximately 145 kDa [10] and has similarities to the v-ErbB transforming protein of avian erythroblastosis virus, which also is an EGFR-related auto-activating oncogene created by a large extracellular deletion [11].  Figure 1.  Figure 1.  Open in figure viewer Schematic structure of EGFRvIII. Compared with the epidermal growth factor receptor (EGFR), EGFRvIII has an in-frame deletion of exons 2–7, resulting in a shorter extracellular domain and the insertion of a novel glycine residue at the junction site. The resultant truncated receptor is constitutively active and cannot bind any known ligand. The juxtamembrane and intracellular regions of the EGFR are conserved in EGFRvIII. The domains of EGFR are shown, labeled domains L1 (also known as domain 1), CR1 (domain 2), L2 (domain 3) and CR2 (domain 4) respectively. COOH, carboxy terminus; NH2, amino terminus; SFK, Src family kinase. Detection and analysis of EGFRvIII  The EGFRvIII literature must be approached with an appreciation of the technical limitations. Most reports on EGFRvIII are based on transfected cell lines [4, 10, 12-112]. This has been necessary because, when traditional cell lines are grown in vitro, they lose EGFR gene amplification and mutation [113-117]. This consistent negative in vitro selection for EGFRvIII [9, 116-118] is poorly understood and raises concerns about data based solely on transfected cell lines. More recently, researchers have been able to naturally maintain EGFRvIII expression if the lines are generated directly from EGFRvIII-positive tumors and propagated in vivo [9, 116-118], as neurospheres in three-dimensional cultures [119, 120] in specialized media [121].  There are also challenges in the detection of EGFRvIII in human tissues. Several monoclonal antibodies specific for EGFRvIII have been described [7, 12, 14, 19, 27, 30, 40, 47, 63, 73, 122-125], but their availability to researchers is restricted by patenting and licensing issues [126]. Furthermore, there are concerns about whether these antibodies are indeed specific for EGFRvIII. Limited data suggest that the false positive rate of some of these antibodies ranges from 4% to 100% [7, 66, 115, 126-129].  The lack of readily available antibodies against EGFRvIII has encouraged some research groups to use PCR to search for the presence of EGFRvIII in human tumors. This approach is also problematic as the presence of EGFRvIII transcripts does not necessarily correlate with protein levels of EGFRvIII that are functionally relevant. Between 0% and 63% of tumors that are positive for EGFRvIII by PCR have been reported to have no associated EGFRvIII protein [7, 115, 130-132]. To obtain the most reliable and valid data, future studies should use two complementary techniques to detect EGFRvIII: a PCR-based technique for detecting the presence of EGFRvIII transcripts and at least one protein-based method. Immunohistochemistry (IHC) or fluorescence activated cell sorting are preferred because western blotting (WB) can detect trivial amounts of protein; very limited data suggest that 69% of positive WB results are not confirmed by another protein detection technique [131]. Using the double-assay approach, EGFRvIII would be considered to be present only if there were concordant positive results using both methods [126].  Finally, the overwhelming evidence indicates that EGFRvIII is a tumor-specific receptor, with almost all published studies showing that normal tissues are devoid of EGFRvIII [5, 7, 9, 103, 115, 129-131, 133-145]. Only one study found EGFRvIII expression in normal lung tissue, using IHC [144]. We consider that this finding is an artifact of the IHC because the result has not been reproduced. EGFRvIII expression in glioblastoma multiforme (GBM)  Most EGFRvIII research has been performed on high-grade brain tumors such as GBM, the most common and lethal malignant neoplasm of the brain in adults. GBM is increasing in incidence, is difficult to treat and has a very poor prognosis – the 1-year survival rate for patients who are newly diagnosed with GBM is 18% [9, 17, 146-149]. Amplification of the EGFR gene and the subsequent overexpression of the EGFR protein is the most common genetic alteration in GBM, occurring at a frequency of approximately 34–63% [137, 140, 142, 150-153]. Of these cases, 63–75% also carry rearrangements of the EGFR gene, resulting in tumors expressing both wild-type (wt) and mutated EGFR [135, 141, 143, 150, 154, 155]. The most common mutation in GBM is EGFRvIII, which occurs at an overall frequency of 25–64% when assessed by multiple techniques in the same tumor (Table 1) [5, 7, 9, 103, 129-131, 133-135, 137, 138, 140-143]. EGFRvIII expression in GBM is almost always associated with amplification and co-expression of the wt EGFR gene [129, 134, 145, 150]. Some studies have suggested that EGFRvIII expression may be associated with a poor prognosis in patients with GBM [131, 140]. One of the larger reported studies analyzed the prognostic value of EGFR gene amplification and mutation in 87 adult patients newly diagnosed with GBM. This study demonstrated that EGFRvIII expression, in combination with EGFR gene amplification, was an independent and significant negative prognostic factor for survival, even after adjusting for established prognostic factors [140]. Not all studies, however, have shown an association between EGFRvIII and patient outcomes [128, 129, 145].  Table 1. EGFRvIII expression in human malignancies. IHC, immunohistochemistry; SB, Southern blotting; WB, western blotting; N/D, not determined; HNSCC, head and neck squamous cell carcinoma; AD, adenocarcinoma; LA, large cell carcinoma; SCC, squamous cell carcinoma; PNST, peripheral nerve sheath tumor; MFH, malignant fibrous histiocytoma Tumor type\\tSingle technique used\\tMultiple complementary techniques used EGFRvIII frequency (%)\\tTechnique\\tStudies\\tEGFRvIII frequency (%)\\tTechniques\\tStudies a Either pure oligodendrogliomas or mixed oligoastrocytomas. b Total sample size < 5 samples. c Includes adenosquamous carcinomas. Brain (glioma) Grade IV (glioblastoma)\\t0–81\\tIHC\\t[5, 7, 126, 129, 134, 137, 140, 180, 224-227]\\t64\\tRT-PCR, IHC\\t[130] 26–55\\tPCR\\t[5, 118, 126, 133, 228]\\t25\\tRT-PCR, IHC, WB\\t[131] 17\\tSB\\t[141]\\t \\t \\t  Grade III\\t14–25\\tIHC\\t[7, 129]\\t \\tN/D\\t  12–40\\tPCR\\t[133, 228]\\t \\t \\t  Grade II\\t0\\tPCR\\t[228]\\t \\tN/D\\t  86\\tWB\\t[138]\\t \\tN/D\\t  Brain (oligodendroglial)a Grade II\\t0\\tIHC\\t[229]\\t \\t \\t  Grade III\\t0–7\\tIHC\\t[145, 229]\\t \\t \\t  Bladder\\t0–50\\tIHC\\t[5, 66]\\t0\\tIHC, PCR, WB\\t[66] Breast\\t0–32\\tIHC\\t[5, 128, 162]\\t20–36\\tIHC, RT-PCR\\t[7, 115] 0–68\\tPCR\\t[114, 115, 163, 165, 230]\\t \\tN/D\\t  78\\tWB\\t[138]\\t \\t \\t  Colorectal\\t0–34\\tIHC\\t[5, 161]\\t \\tN/D\\t  0\\tPCR\\t[65, 160]\\t \\t \\t  0\\tWB\\t[160]\\t \\t \\t  Esophageal\\t0\\tIHC\\t[5]\\t \\tN/D\\t  HNSCC\\t0–42\\tIHC\\t[5, 72]\\t33b\\tIHC, PCR\\t[72] Lung AD\\t0–41\\tIHC\\t[5, 136, 144, 164]\\t \\tN/D\\t  0–1\\tPCR\\t[132, 166-168]\\t \\t \\t  LA\\t0–33b\\tIHC\\t[5, 144, 164]\\t \\tN/D\\t  14\\tPCR\\t[165]\\t \\t \\t  SCC\\t0–46\\tIHC\\t[5, 136, 144, 164]\\t5\\tIHC, PCR\\t[132] 0–8\\tPCR\\t[165, 167, 168]\\t \\tN/D\\t  Othersc\\t0–100b\\tPCR\\t[136, 165]\\t \\t \\t  Melanoma\\t0\\tIHC\\t[5]\\t \\tN/D\\t  Ovarian\\t0\\tIHC\\t[5, 157]\\t0\\tIHC, PCR\\t[157, 158] 0–75\\tPCR\\t[138, 157]\\t0\\tPCR, WB\\t[159] PNST\\t0\\tIHC\\t[231]\\t \\tN/D\\t  Prostate\\t100\\tIHC\\t[139, 172]\\t \\tN/D\\t  Sarcoma\\t \\t \\t \\t \\tN/D\\t  Leiomyosarcoma\\t0\\tIHC\\t[5]\\t \\t \\t  Synovial sarcoma\\t0\\tIHC\\t[5, 231]\\t \\t \\t  MFH\\t0\\tIHC\\t[5]\\t \\t \\t  Osteosarcoma\\t0\\tPCR\\t[232]\\t \\t \\t  Testicular seminoma\\t0\\tIHC\\t[5]\\t \\tN/D\\t  Thyroid\\t \\t \\t \\t \\tN/D\\t  Papillary\\t75\\tIHC\\t[156]\\t \\t \\t  Follicular\\t43\\tIHC\\t[156]\\t \\t \\t  Medullary\\t0\\tIHC\\t[156]\\t \\t \\t  EGFRvIII expression in other human malignancies  EGFRvIII expression has also been reported in malignancies outside the central nervous system (Table 1). Given some of the technical problems with EGFRvIII detection, especially those that occur when using polyclonal antibodies to detect EGFRvIII [127, 139, 156], several of these reports remain controversial. Our overview (Table 1) is restricted to studies that were conducted using primary tumor tissue from adults and distinguishes between frequencies defined using only one technique (left half of Table 1) and those defined using two or more complementary techniques (right half of Table 1), with the latter group likely to be more reliable. Many of the data are limited and conflicting. For example, Moscatello et al. [138] reported EGFRvIII expression in 24 of 32 (75%) ovarian carcinoma samples examined by WB. Subsequently, several large independent studies using more reliable methods for EGFRvIII detection were performed to verify EGFRvIII expression in ovarian cancer. Importantly, EGFRvIII was not detected at the mRNA or protein level in any of the ovarian carcinoma tissue samples or cell lines tested [5, 114, 157-159], indicating that EGFRvIII is not expressed in ovarian cancer (Table 1). Similarly, EGFRvIII is also not expressed in colorectal cancer, with three [5, 65, 160] of four studies [5, 65, 160, 161] confirming the lack of EGFRvIII expression at both the mRNA level and the protein level. For all other tumor types that have been tested, the data remain preliminary, especially given that most of the data are derived from solitary studies that used only one detection technique. We await corroborative data for these tumor studies, which hopefully will use more reliable methods and reagents for EGFRvIII detection at the mRNA and protein levels. EGFRvIII expression in breast carcinoma  The proportion of breast carcinomas that express EGFRvIII remains unresolved. Some studies have reported EGFRvIII expression in 20–78% of breast tumor tissues [7, 115, 138, 162], whereas others have found negligible expression in tumor tissues (0–4%) [114, 128, 163] (Table 1). In a study conducted by Ge et al. [115], using laser capture microdissection and reverse transcription PCR (RT-PCR) on tissue from 28 patients with breast cancer, approximately 68% (19 of 28) of primary invasive breast cancers expressed high levels of EGFRvIII mRNA. The tumors from 11 of these patients were also investigated by IHC using the EGFRvIII-specific monoclonal antibody Ab-18 (also known as DH8.3; no longer available commercially), and 10 were found to also have EGFRvIII expression. No detectable levels of EGFRvIII were observed in normal breast tissue. Furthermore, 57% of the infiltrating breast carcinoma samples (16 of 28) were shown to co-express EGFRvIII and wt EGFR [115]. Using the monoclonal antibodies L8A4 and Y10, Wikstrand et al. [7] found that breast cancers express EGFRvIII in 27% of cases. By contrast, Rae et al. [114] showed that 55 breast cancer cell lines and 170 primary breast cancers were devoid of EGFRvIII mRNA, as determined by RT-PCR, and Nieto et al. [128] found that only 4% of patients with advanced breast carcinoma (9 of 225), whose samples were screened by IHC using the EGFRvIII-specific antibody DH8.3, showed evidence of EGFRvIII expression.  More recently, Yu et al. [162] examined 182 normal breast tissue and breast carcinoma samples, including lymph node tissue taken from patients with metastatic breast carcinoma. EGFRvIII expression was assessed by IHC using the EGFRvIII-specific antibody DH8.3 and was found in 18% of ductal carcinomas in situ, 32% of primary invasive breast carcinomas and 54% of lymph node tissue samples taken from patients with metastatic disease [162]. No EGFRvIII was detected in normal breast tissues, consistent with previously reported findings [115]. Collectively, these reports demonstrate that EGFRvIII expression is correlated with disease progression in breast cancer and may be involved in tumor metastasis. EGFRvIII expression in lung carcinomas  The proportion of non-small-cell lung cancers (NSCLCs) that express EGFRvIII is also controversial (Table 1). EGFRvIII expression was found in 16–39% of NSCLC samples in two studies using either the polyclonal antibody ZMD.82 or the monoclonal antibody G100 [136, 164]. Both of these antibodies, however, non-specifically stained non-malignant cells and tissues such as type II pneumocytes, normal ciliated bronchial epithelium and submucosal bronchial glands. A consensus view has emerged that the proportion of NSCLCs that express EGFRvIII is much lower than is reported in these studies and that the expression of EGFRvIII seems to be strongly influenced by the histological subtype of NSCLC and inversely correlated with the presence of kinase domain mutations in exons 18–21 of wt EGFR. Several studies have shown that tumors with an adenocarcinoma subtype frequently express EGFR with kinase mutations (11–52% of tumors) [165-167], but they are universally devoid of the EGFRvIII mutation [132, 165-167] despite the significant correlation between receptor tyrosine kinase (RTK) mutations and wt EGFR overexpression. By contrast, kinase mutations are less frequent in tumors of a non-adenocarcinoma subtype (14%) [165], but a small proportion of these tumors (0–8%) reproducibly express EGFRvIII [132, 165, 167, 168]. EGFRvIII confers enhanced tumorigenicity through multiple mechanisms and pathways  EGFRvIII has been shown to increase the tumor-igenicity of cells of diverse origins, including those in GBM [4, 7, 17, 21, 34, 51, 82, 115, 146, 153, 169-184], NIH3T3 murine fibroblasts [10, 16, 123] and in breast [162], lung [132] and ovarian [74] cancers. Given the consistency of the results across cell types, we focus predominantly on the data for GBM. GBM is a highly proliferative, vascular and locally invasive tumor [185-187]. Consistent with these clinicopathological features, EGFRvIII-transfected GBM cell lines show increased rates of proliferation [146, 173], an increased ability to form tumor xenografts [4, 146, 183], reduced apoptosis [146], increased angiogenesis [184] and increased invasiveness [188-192] compared with matched parental cell lines.  These pro-tumorigenic effects of EGFRvIII seem to be mediated by several key signaling pathways downstream of EGFRvIII, including phosphatidylinositol 3-kinase (PI3K)/Akt [7, 51, 179-182], Ras/Raf/mitogen-activated protein kinase (MAPK) [169], signal transducer and activator of transcription 3 (Stat3) [34, 82, 115, 170-172] and nuclear factor κB (NF-κB) [184] (Fig. 2). Interestingly, the growth advantage conferred by EGFRvIII seems to be more prominent in vivo than in vitro. For example, transfection with EGFRvIII had only a modest effect on the in vitro growth characteristics of the GBM cell line U87MG, even under low serum conditions, but these cells had a marked growth advantage when grown as tumor xenografts in nude mice [4, 146, 183]. Furthermore, the growth advantage conferred by EGFRvIII in vivo resulted in the selective clonal expansion of EGFRvIII-expressing cells in vivo [146]. This was demonstrated by mixing U87MG.Δ2-7 cells (U87MG parental cells which have been transfected to overexpress EGFRvIII) with parental U87MG cells at a ratio of 1 : 10 000 and subsequently implanting them into the brains of nude mice. At 4–5 weeks post inoculation, this cell ratio was 20 : 1 [146].  Figure 2.  Figure 2.  Open in figure viewer Downstream signaling pathways of EGFRvIII. Multiple downstream signaling pathways are initiated by EGFRvIII. The phosphatidylinositol 3-kinase (PI3K)/Akt pathway has been identified as the signaling cascade that is preferentially activated by EGFRvIII, particularly in glioblastoma multiforme (GBM). The pro-tumorigenic effects of EGFRvIII seem to rely directly on its ability to signal [17, 146, 188], because transfection of a kinase-dead version of EGFRvIII into U87MG cells did not enhance xenograft growth. Furthermore, all five major auto-phosphorylation sites in the carboxy-terminal region of EGFRvIII are required for effective signaling, because even single point mutations at these sites, particularly of the tyrosine residues Tyr1068, Tyr1148 and Tyr1173, were sufficient to abolish most of the in vivo growth advantage conferred by EGFRvIII [17]. Lastly, although EGFR gene amplification is nearly always found together with EGFRvIII expression, wt EGFR expression is not required for the pro-tumorigenic effects of EGFRvIII as these effects have been shown even in the absence of wt EGFR expression [52].  The strength of the constitutive signaling mediated by EGFRvIII on the cell surface is a matter of debate, but many studies suggest that it is approximately 10% of the intensity of ligand-induced wt EGFR signaling [10, 16, 17, 123]. EGFRvIII signaling, however, is significantly enhanced by the prolonged retention of this receptor on the cell surface because of its impaired endocytosis [17, 193]. Paradoxically, it is the low level of EGFRvIII signaling, especially the hypo-phosphorylation at Tyr1045, that prevents the receptor from engaging with the endocytic proteins Cbl, CIN85 and endophilin and leads to its retention on the cell surface [68]. Pharmacological attenuation of wt EGFR signaling to approximately 20% of normal activity levels results in impaired interaction with the Cbl–CIN85–endophilin complex, as well as the absence of polyubiquitination and reduced internalization of the receptor; these observations are consistent with EGFRvIII downregulation. The importance of impaired downregulation has been further confirmed by Davies et al. [67], who showed that overexpression of the Cbl family of ubiquitin ligases made it possible to induce the polyubiquitination and degradation of EGFRvIII with a subsequent loss of tumorigenicity.  EGFRvIII activates several downstream pathways (Fig. 2), but a considerable amount of evidence indicates that it preferentially activates the PI3K/Akt signal transduction pathway [7, 51, 179-181]. EGFRvIII expression is tightly correlated with the activation of downstream targets of PI3K/Akt, including the mammalian target of rapamycin (mTOR), the forkhead (FOX) transcription factor family and S6 [182]. Recently, EGFRvIII has been shown to activate mTORC2 (CREB-regulated transcription co-activator 2) via the PI3K/Akt pathway, which in turn leads to stimulation of the NF-κB pathway and resistance to chemotherapy [194]. Another consequence of EGFRvIII activation of PI3K/Akt is increased proliferation and cell cycle progression mediated by a decrease in the level of p27KIP1, a cyclin-dependent kinase (CDK) inhibitor that binds and inactivates CDK2–cyclin E complexes and thereby inhibits the transition of cells from the G1 phase to the S phase [173]. The significance of PI3K/Akt activation by EGFRvIII has been confirmed by Klingler-Hoffmann et al. [178], who showed that treatment of U87MG.Δ2-7 cells with the PI3K inhibitor wortmannin, or by reconstitution of the physiological levels of PTEN (a negative regulator of PI3K), resulted in the ablation of the EGFRvIII-conferred growth advantage. Narita et al. [173] also demonstrated that overexpression of dominant-negative Akt in U87MG.Δ2-7 cells restored p27KIP1 to levels similar to those seen in parental U87MG cells, causing G1 cell cycle arrest and the loss of tumorigenicity. Selective activation of the PI3K/Akt pathway by EGFRvIII is thought to mediate the resistance to radiation that is observed in EGFRvIII-positive GBM [21, 50, 182, 195-197]. EGFRvIII signaling via the PI3K/Akt pathway may be facilitated by associated loss or mutation of the PTEN gene, which occurs in approximately 40% of GBM samples that express EGFRvIII [153, 174-176].  Several studies have shown that EGFRvIII and Src family kinases (SFKs) work cooperatively to enhance GBM tumorigenicity [103, 198]. The expression of EGFRvIII caused activating phosphorylation of the receptor and physical association of the SFK members Src and Fyn, which in turn promoted tumor growth and motility [103]. Genetic or pharmacological inhibition of SFKs inhibited cell motility in vitro and growth of EGFRvIII-expressing GBM xenografts in vivo [103, 199]. More recently, it was shown that the phosphorylation of Tyr772 on SFKs by EGFRvIII leads to the phosphorylation of dedicator of cytokinesis 1 (DOCK1; also known as DOCK180), a guanine nucleotide exchange factor with roles in cell motility, survival and proliferation [199]. Genetic ablation of DOCK180 blocked the EGFRvIII-mediated tumorigenicity of GBM cells [199].  The enhanced tumorigenicity of GBM cells expressing EGFRvIII is also associated with enhanced angiogenesis. Wu et al. [184] showed that the increased tumorigenicity conferred by transfection of GBM cell lines with EGFRvIII could be reversed by genetically inhibiting the NF-κB pathway. This reversal was associated with concurrent reductions in vascular endothelial growth factor (VEGF) and interleukin-8 (IL-8) expression and with reductions in tumor angiogenesis. These studies were confirmed and extended by Bonavia et al. [200], who showed that EGFRvIII transfection increased IL-8 expression in GBM cells by 60-fold through the NF-κB pathway. RNA-interference-mediated knockdown of the IL-8 pathway or the NF-κB pathway inhibited GBM xenograft growth and attenuated angiogenesis. EGFRvIII dimerization  Most publications indicate that EGFRvIII predominantly forms transient homodimers [17, 52, 201]. The findings of an early publication, showing constitutive dimerization of EGFRvIII, have not been reproduced [16], and another study reporting significant dimerization used non-physiological conditions that included an extensive and lengthy purification [29]. As auto-phosphorylation of the EGFR only occurs in trans after the formation of an asymmetric dimer [202], transient EGFRvIII homodimers must exist. By using differentially tagged EGFRvIII molecules, we were able to demonstrate this transient interaction [52]. Furthermore, we have also shown that EGFRvIII can interact transiently with wt EGFR [52]. Recently, we demonstrated that the transient EGFRvIII homodimer is stabilized by disulfide bonds through the free amino-terminal cysteines that are created by the deletion event [203]. Finally, forced dimerization of EGFRvIII by using genetic strategies enhances receptor phosphorylation and downstream signaling, leading to increased GBM cell proliferation [204]. Taken together, these data confirm the critical role of EGFRvIII dimerization for signaling. EGFRvIII activates multiple receptors through differing mechanisms  The direct interplay between EGFRvIII and wt EGFR has not been examined extensively. Co-expression of the two receptors in IL-3 dependent BaF/3 caused increased survival and proliferation in the absence of IL-3 compared with expression of EGFRvIII alone; wt EGFR expressed alone has no activity in these cells in the absence of ligand [52]. When EGFRvIII is expressed in U251-MG glioma cells it induces expression of the EGFR ligands TGF-α and HB-EGF [70]. Furthermore, neutralization of HB-EGF with antibodies reduced U251-MG cell proliferation suggesting that this EGFRvIII induced autocrine loop has a role in EGFRvIII-mediated tumorigenicity.  White and colleagues [205] conducted a mass spectral analysis of the phospho-proteome activated by EGFRvIII in U87MG cells. Along with numerous intracellular targets phosphorylated by EGFRvIII, the RTK c-Met also appeared to be phosphorylated in response to EGFRvIII expression. Additional studies have confirmed that c-Met is activated by EGFRvIII and have shown that this occurs independently of hepatocyte growth factor (HGF), the c-Met ligand [205]. Recently, it was shown that the activation of c-Met by EGFRvIII is a major mechanism underlying the previously reported activation of the transcription factor Stat3 [206].  Using phosphorylated RTK arrays, we established that EGFRvIII activates not only c-Met in the U87MG cell line but many other RTKs, including platelet-derived growth factor receptor β (PDGFRβ) and VEGF receptor 2 (VEGFR2) [106]. Using panitumumab, an EGFR-specific antibody that inhibits EGFRvIII activation and phosphorylation, we were able to block the phosphorylation of c-Met and PDGFRβ [106]. This verified that the activation of RTKs was mediated by EGFRvIII and was not an artifact of clonal selection following transfection. The ligand-independent activation of c-Met was confirmed by using neutralizing antibodies specific for HGF. Although the evidence suggests that EGFRvIII mediates the activation of c-Met through a direct interaction, the possibility of a downstream mechanism cannot be completely dismissed. Notably, these observations help explain the potent tumorigenicity of EGFRvIII despite its low level of activation.  EGFRvIII has been shown to stimulate the production of cytokines, including IL-6 and leukemia inhibitory factor (LIF), both of which signal through the gp130 complex [207]. Importantly, these cytokines have been shown to activate overexpressed wt EGFR in neighboring GBM cells through a mechanism involving crosstalk between gp130 and wt EGFR, and activation of wt EGFR by this mechanism led to enhanced GBM cell proliferation [208]. Thus, EGFRvIII contributes to the growth of surrounding GBM cells through this field effect. More generally, this indicates that EGFRvIII actively contributes to the heterogeneity of GBM by acting indirectly on neighboring cells that are EGFRvIII negative. This hypothesis is entirely consistent with the observation that wt EGFR amplification and EGFRvIII expression are usually observed together. It may also explain why, in patients, the pronounced growth advantage mediated by EGFRvIII does not lead to a homogeneous population of cells all of which express the EGFRvIII receptor. Role of EGFRvIII in metabolism  The PI3K/Akt/mTOR pathway, which is preferentially activated by EGFRvIII [7, 51, 179-181], is associated with changes in cell metabolism, including increased anaerobic glycolysis and lipogenesis [97, 209, 210]; these processes are important for continuous tumor cell proliferation [209, 210]. AMP-activated protein kinase (AMPK), an enzyme that plays a key role in regulating cellular metabolic processes including lipogenesis, inhibits mTOR signaling when activated. Using the AMPK agonist AICAR, which reportedly mediates its anti-tumor effects through forced AMPK activation and subsequent mTOR inhibition [211, 212], Guo et al. [97] examined the link between EGFRvIII, the PI3K/Akt/mTOR pathway (Fig. 2) and lipogenesis. Expression of EGFRvIII in U87MG cells caused an increase in intracellular fatty acid production in vitro and enhanced 18F-fluorodeoxyglucose (FDG) uptake in vivo [97], suggesting a role for EGFRvIII in lipogenesis and cell metabolism. Treatment with AICAR was found to significantly decrease 18F-FDG uptake in vivo and intracellular fatty acid production in vitro [97]. Collectively, these results suggest that activation of the PI3K/Akt/mTOR pathway by EGFRvIII promotes tumor growth through coordinated regulation of cell metabolism and lipogenesis. Finally, it has been shown that SFK activation of EGFRvIII stimulates the translocation of EGFRvIII to the mitochondria, where it enhances survival in low glucose conditions by an unknown mechanism [198]. Targeting EGFRvIII as a therapeutic modality  Whilst it is beyond the scope of this review to comprehensively discuss the therapeutic value of inhibiting EGFRvIII, a brief summary is useful to emphasize the importance of EGFRvIII as a clinical target. Pre-clinical in vivo studies consistently show that targeting of EGFRvIII can be therapeutically useful in brain tumors [32, 62, 73, 95, 118, 213, 214]. However, compared with wt EGFR, EGFRvIII appears to be relatively resistant to treatment with conventional anti-EGFR agents such as ligand blocking monoclonal antibodies [72, 76, 85, 215-218] or EGFR tyrosine kinase inhibitors [37, 50, 132, 219]. A number of alternative approaches have shown clinical promise and are in clinical trials, notably mAb806 and rindopepimut (CDX-110). mAb806 is a novel anti-EGFR antibody which was generated by vaccinating mice with EGFR-negative NR6 fibroblast cells transfected with EGFRvIII [5]. mAb806 targets a conformationally exposed epitope on wt EGFR when it is overexpressed in tumor cells or in the presence of oncogenic mutations such as EGFRvIII [220]. Not only is mAb806 less toxic than other anti-EGFR antibodies [221], it shows excellent blood–brain barrier penetration and localization to in situ GBM in patients [222]. mAb806, currently named ABT-806, has just completed phase 1 testing (trial identifiers NCT01255657, NCT01472003 and NCT01406119). A trial of ABT-414, an antibody–drug conjugate based on mAb806, is currently recruiting patients with newly diagnosed or recurrent GBM for treatment with concurrent radiotherapy and/or temozolomide (trial identifier NCT01800695). Rindopepimut is a 14-amino-acid peptide bridging the unique glycine at the site of the EGFRvIII mutation that is conjugated with keyhole limpet hemocyanin [223]. Phase 2 data combining rindopepimut vaccination with first-line chemo-radiation for newly diagnosed GBM were encouraging compared with historical data and treatment was well tolerated [223]. A phase 3 study of rindopepimut in the same setting is currently under way (trial identifier NCT01480479) as is a study of rindopepimut in patients with relapsed GBM in combination with bevacizumab, a monoclonal antibody against VEGF, or as a single agent in those who have failed bevacizumab (trial identifier NCT01498328). The role of anti-EGFRvIII therapy in other tumor types is still to be addressed. Conclusion  EGFRvIII is expressed at a relatively high frequency in GBM tumors but determining its frequency and significance in other tumor types will require additional studies. Recent work has begun to unravel the signaling networks that are used by EGFRvIII. Clearly, there is overlap with the pathways that are activated by wt EGFR; however, significant differences are starting to emerge, especially with respect to the co-activation of other cell-surface receptors by EGFRvIII. So far, these co-activation experiments have been performed in a limited range of transfected GBM cell lines and need to be reproducible in EGFRvIII-expressing xenografts, GBM spheroid lines or other cell lines that naturally express EGFRvIII. The assortment of RTKs that can be activated by EGFRvIII remains unknown; it is unclear whether EGFRvIII can activate any RTK that is expressed on the cell surface at a significant level or whether it only activates a specific subset. Although further insight into the tumor-promoting activities of EGFRvIII will improve our understanding of GBM development and progression, it is the broader implications – such as the maintenance of tumor heterogeneity, the rewiring of signaling pathways, the co-opting of other receptors and the metabolism of cancer cells – that make understanding this receptor so important. Lastly, the therapeutic potential of targeting EGFRvIII in brain and other tumors is becoming increasingly apparent, with agents against EGFRvIII now in clinical trials.  Alterations of the epidermal growth factor receptor (EGFR) gene are common in some forms of cancer and the most frequent is a deletion of exons 2–7. We have previously shown that this mutant receptor, called ΔEGFR, confers enhanced tumorigenicity to glioblastoma cells through elevated proliferation and reduced apoptotic rates of the tumor cells in vivo. To understand the molecular mechanisms that underlie ΔEGFR-enhanced proliferation, we examined the gene products that control cell cycle progression. We found that levels of the cyclin-dependent kinase (CDK) inhibitor, p27, were lower in U87MG.ΔEGFR tumors than in parental U87MG or control U87MG.DK tumors. Consequently, CDK2-cyclin A activity was also elevated, concomitant with the RB protein hyperphosphorylation. In addition, activated phosphatidylinositol 3-kinase (PI3-K) and phosphorylated Akt levels were also elevated in the U87MG.ΔEGFR tumors. U87MG.ΔEGFR cells failed to arrest in G1 in response to serum starvation in vitro and while maintaining high levels of PI3-K activity and hyperphosphorylated RB. Treatment of U87MG.ΔEGFR cells with LY294002, a PI3-K inhibitor, caused reduced levels of phosphorylated Akt and concomitantly up-regulated levels of p27. Expression of a kinase dead dominant-negative Akt mutant in the U87MG.ΔEGFR cells similarly resulted in up-regulation of p27 and down-regulation of tumorigenicity in vivo. These results suggest that the constitutively active ΔEGFR can enhance cell proliferation in part by down-regulation of p27 through activation of the PI3-K/Akt pathway. This pathway may represent another therapeutic target for treatment of those aggressive glioblastomas expressing ΔEGFR.  INTRODUCTION  Patients harboring glioblastoma multiforme have 5-year survival rates of less than 10% (1) , despite treatment with combinations of surgery, radiotherapy, and chemotherapy. About 30–40% of glioblastomas have alterations of the EGFR 3 gene (2 , 3) , the most frequent of which is an in-frame deletion of exons 2–7 that is named ΔEGFR or EGFRvIII (4) . This truncated receptor lacks part of the extracellular ligand binding domain and is constitutively activated despite being unable to bind to ligands such as EGF (5 , 6) . We have previously shown that ΔEGFR enhances the growth of glioblastoma cells as xenografts in immunodeficient mice and that these tumors showed an elevated proliferation rate as indicated by Ki-67 staining (3 , 7 , 8) , suggesting accelerated progression from G1 into S phase. This transition requires phosphorylation and activation of CDKs by cyclins. As a counterbalance to this, CDKIs can associate with CDK-cyclin complexes and inhibit their activities (9 , 10) . Glioblastomas have been shown to express increased levels of cyclin D (11 , 12) , contain amplification of CDK4/6 genes (13 , 14) , or have reduced levels of the CDKI, p27 (15, 16, 17) , suggesting that CDK-cyclin kinase activities are elevated in glioblastomas. However, whether ΔEGFR signaling influences the levels or activities of these molecules involved in G1-S phase transition is not known.  ΔEGFR has been shown to activate PI3-K, one of the signal transducers that plays an important role in tumorigenesis and cell cycle progression (6 , 18 , 19) . PI3-K transduces cellular signals by phosphorylating PIP2 to generate the second messenger PIP3, which leads to the activation of Akt kinase activity (20) . Recent studies show that activated and phosphorylated Akt, in turn, phosphorylates and promotes mobilization of the Forkhead transcription factor from the nucleus to the cytoplasm, which reduces the transcription levels of the cell cycle inhibitor p27 (21, 22, 23, 24) . p27 induces G1 arrest in tumor cells by binding to CDK-cyclin complexes and inhibiting their activities. Therefore PI3-K/Akt signaling can lead to cell cycle progression in tumor cells through the down-regulation of p27.  Here, we present evidence that active ΔEGFR can enhance cell proliferation in vitro and in vivo in various glioblastoma cell lines, by down-regulation of p27, through the activation of the PI3-K/Akt pathway. A direct demonstration of the importance of this cascade arises from data showing that chemical PI3-K inhibitors or dominant-negative KD-Akt mutants reverse ΔEGFR-mediated cell or tumor growth enhancement. These results provide additional insight into the basic mechanisms of tumor cell growth regulation by ΔEGFR and indicate that the PI3-K/Akt pathway may be a target for intervention.  MATERIALS AND METHODS  Cell Culture and Biological Reagents.  The human glioblastoma cell lines U87MG, A1207, and LNZ308 were cultured in DMEM with 10% fetal bovine serum, 2 mm glutamine, 100 units/ml penicillin, and 100 mg/ml streptomycin in a 95% air/5% CO2 atmosphere at 37°C. U87MG cells, LNZ308 cells, and A1207 cells, which express ΔEGFR, and U87MG.DK cells that express kinase-defective ΔEGFR, used in this study, were described previously (3) .  PIP2 and LY294002 (PI3-K inhibitor), were purchased from Sigma Chemical Co. (St. Louis, MO) and Histone H1 was purchased from Amersham Pharmacia Biotech (Piscataway, NJ). The pcDNA KD-Akt vector expressing KD-Akt was regenerated from pCMV5-HA-Akt(K179M), which was a gift of Dr. Philip Tsichlis (Thomas Jefferson Medical College, Philadelphia, PA; Ref. 25 ). To generate KD-AKT-U87MG.ΔEGFR cells, U87MG.ΔEGFR cells were transfected with pcDNA KD-Akt by the calcium phosphate method (26) and were selected under 100 μg/ml Zeocin and 10% serum. Several stable transfectants were obtained.  Tumors in Nude Mice.  Cells (5 × 105) of each cell line in 5 μl PBS were implanted into the right corpus striatum of 4–6-week-old female nude BALB/c mice (Simonsen Laboratories, Gilroy, CA) as described previously (3) . Mice were killed on the development of neurological symptoms. For s.c. tumors, 1 × 106 cells of the various cell lines in 0.1 ml of PBS were injected into the right flanks of nude mice. Tumor volumes were defined as (longest diameter) × (shortest diameter)2 × 0.5. All of the procedures were approved by the animal care and use committee of the University of California, San Diego.  Flow Cytometric Analysis.  U87MG.ΔEGFR cells were transiently transfected with control pcDNA, pcDNA KD-Akt, or pcDNA p27, cultured for 48 h under 2% serum, and harvested by trypsinization. Nuclei of transfected or nontransfected cells under various conditions were stained with propidium iodide as described previously (26) . A total of 20,000 nuclei were analyzed in a FACS Flow Cytometer (Becton Dickinson, Franklin Lakes, NJ) and DNA histograms were obtained and analyzed using Modfit-software (Verity, Topsham, ME).  Immunohistochemistry.  An assessment of tumor cell proliferation was performed by Ki-67 immunohistochemistry on formalin-fixed paraffin-embedded tumor tissues as described previously (7) . The Ki-67(MIB-1) staining index was determined as the percentage of Ki-67-labeled nuclei in the total number of nuclei in a high-power field (×400). Approximately 2000 nuclei were counted in each case by a process of systematic random sampling.  Western Blotting.  Tissue specimens of intracranial xenografts were snap-frozen in liquid nitrogen immediately after surgical resection and stored at −80°C until use. Western blotting was performed as described previously (7) . Each protein sample (20 μg) was separated with appropriate polyacrylamide/SDS gels and electroblotted onto nitrocellulose membranes (ECL membrane; Amersham Pharmacia Biotech). Primary antibodies used in this study were CDK2 (Santa Cruz Biotechnology, Santa Cruz, CA); CDK4 (Santa Cruz Biotechnology); CDK6 (Santa Cruz Biotechnology); p21 (PharMingen, San Diego, CA); p27 (BD Biosciences, San Jose, CA); cyclin A (Santa Cruz Biotechnology); cyclin D1 (Neo Markers, Fremont, CA); cyclin D2 (Neo Markers); cyclin D3 (Neo Markers); cyclin E (Santa Cruz Biotechnology); unphosphorylated RB (PharMingen); phosphorylated RB (PharMingen); Akt (Cell Signaling, Beverly, MA); phosphorylated Akt (Cell Signaling); phosphotyrosine 4G10 (UBI, Waltham, MA); PI3-K p85 (UBI); β-actin (Sigma). Western blotting for each antibody was performed more than three times, and representative blots are shown. β-actin was used as loading control for each result (data not shown). Quantitation was performed using laser scanning densitometry of the autoradiogram.  CDK2-, CDK4- or CDK6-associated Kinase Assays.  CDK2-, CDK4- or CDK6-associated kinase activity assays were performed as described previously (27) . Briefly, CDK2-cyclin A/E, CDK2-cyclin A and CDK2-cyclin E kinase activities were assayed by incubating Histone H1 and [γ-32P]ATP with 200 μg immunocomplexes of CDK2, cyclin A and cyclin E, respectively. CDK4-cyclin D, CDK6-cyclin D, and cyclin D1-CDK4/6 kinase activities were also examined by incubating GST-RB protein and [γ-32P]ATP with immunocomplexes of CDK4, CDK6, and cyclin D1, respectively. This RB fusion protein was generated in Escherichia coli (27) . Relative kinase activity signals were determined using a PhosphorImager (Molecular Dynamics, Sunnyvale, CA).  PI3-K Assay.  PI3-K assays were performed as described previously (6) . Cells or frozen tissue specimens were lysed in PI3-K lysis buffer [137 mm NaCl, 20 mm Tris-HCl (pH 7.4), 1 mm CaCl2, 1 mm MgCl2, 1% (v/v) NP40, 0.1 mm Na3VO4, 1 μg/ml aprotinin, 1 μg/ml leupeptin, and 100 μg/ml phenylmethylsulfonyl fluoride]. Lysate (500 μg) was incubated with anti-p85 antibody and protein G/protein A-Sepharose for 4 h at 4°C with rotation. Immunocomplexes were washed three times with PI3-K lysis buffer, twice with PI3-K washing buffer [0.5 m LiCl in 100 mm, Tris-HCl (pH 7.5), and 100 μm Na3VO4], and twice with reaction buffer [25 mm Tris-HCl (pH 7.5), 100 mm NaCl, 6.25 mm MgCl2, and 0.625 mm EDTA]. The beads were resuspended in 40 μl of reaction buffer containing 2 μg PIP2 and 30 μCi [γ-32P]ATP and then were incubated at room temperature for 10 min. The reaction was terminated by the addition of 80 μl of 1 m HCl. Phospholipids were extracted with 160 μl of CHCl3:CH3OH (1:1), and the CHCl3 layer was saved. The lipids were chromatographed on thin-layer chromatography plates (precoated with potassium oxalate and baked at 100°C for 1 h) in CHCl3:CH3OH:2.5 m NH4OH (9:7:2, v/v/v). PI3-K activity was determined by PhosphorImager quantitation of the PIP3 spot and was compared with that in U87MG cells cultured under 10% serum.  RESULTS  ΔEGFR Increases the Proliferation of Glioblastoma Cells.  We initially examined the effect of the introduction of ΔEGFR in various human glioma cell lines in vitro and in vivo. Mice bearing U87MG.ΔEGFR, A1207.ΔEGFR, or LNZ308.ΔEGFR cells showed symptoms such as hemiparesis or body weight loss 3–5 weeks after the inoculation. To compare tumor sizes between parental cells and their ΔEGFR transfectants, mice bearing U87MG and U87MG.ΔEGFR cells, A1207, and A1207.ΔEGFR cells or LNZ308 and LNZ308.ΔEGFR cells were killed 3, 4, or 5 weeks after the injection of tumor cells, respectively. ΔEGFR enhanced tumor formation by all of the cell lines (Table 1) ⇓ . The tumor volumes of ΔEGFR-expressing U87MG, A1207, and LNZ308 were about 80, 5, and 40 times larger, respectively, than those of the parental cell lines (Table 1) ⇓ .  View inlineView popup Table 1 Tumor volume and Ki-67 staining index in intracranial brain tumors of nude mice from parental and ΔEGFR engineered cell lines To assess tumor cell proliferation, the Ki-67(MIB) staining index was determined for each tumor sample. The Ki-67 indices of U87MG.ΔEGFR cells, A1207.ΔEGFR cells, and LNZ308.ΔEGFR were 46.6, 46.5, and 56.3%, respectively, values that were approximately 2-fold greater than those of parental cells (Table 1) ⇓ . Because growth enhancement and proliferation were similarly affected in each of the three cell types by ΔEGFR, we concentrated further on the U87MG series.  To determine whether cell cycle profiles differed between parental cells and ΔEGFR transfectants, the DNA content in U87MG and U87MG.ΔEGFR cells cultured with and without serum was compared. No obvious differences between cells cultured with 10% serum were detected, whereas, compared with parental U87MG cells, U87MG.ΔEGFR cells arrested to a much lesser extent after serum starvation (Fig. 1) ⇓ .  Fig. 1. Download figureOpen in new tabDownload powerpoint Fig. 1. ΔEGFR enhances cell proliferation in vitro. U87MG and U87MG.ΔEGFR cells were cultured in 10% serum or serum starvation conditions for 48 h. Their nuclei were stained with propidium iodide, and cell cycle analysis was performed by flow cytometry. U87MG.ΔEGFR Cells Have Lower Expression Levels of p27 Than Parental U87MG Cells in Vitro and in Vivo.  To understand the molecular mechanisms that underlie ΔEGFR-enhanced proliferation, we examined the gene products that control cell cycle progression in vitro and in vivo (Fig. 2) ⇓ . There was no obvious difference in the expression levels of CDK2, CDK4, CDK6, cyclin D1, cyclin D3, cyclin E, and p21 when examined in cell lines cultured with or without serum (cyclin D2 was not detected in U87MG cells). Lower levels of p27 (0.5-fold) and slightly elevated cyclin A levels were detected in U87MG.ΔEGFR cells compared with parental U87MG cells only under serum starvation conditions. Similarly, the expression levels of p27were 0.5-fold lower and the levels of cyclin A, D1, and D3 and of CDK2 were more than 2-fold higher in U87MG.ΔEGFR- than in parental U87MG- or U87MG.DK- generated tumors. There was no obvious difference in the expression levels of the other cell cycle-related proteins among these tumors.  Fig. 2. Download figureOpen in new tabDownload powerpoint Fig. 2. Expression of cell cycle-related proteins in U87MG and U87MG.ΔEGFR cells in vitro and in vivo. U87MG and U87MG.ΔEGFR cells were cultured in 10% serum or serum starvation conditions for 48 h, and then cell cycle-related proteins were analyzed by immunoblotting. In a similar fashion, tumors derived from U87MG, U87MG.DK, or U87MG.ΔEGFR cells were examined. Down-regulation of p27 and up-regulation of cyclin A levels in cultured U87MG.ΔEGFR cells under serum starvation conditions. Levels of p27 (arrow) were lower, and the levels of cyclin A and D1 were higher, in U87MG.ΔEGFR than in parental U87MG- or U87MG.DK-generated brain tumors. Triangle indicates immunoglobulin light chain, which originates from mouse tissue. There was no obvious difference in the expression levels of other proteins among these tumors. Below each panel, relative ratios to U87MG cells under 10% serum or parental U87MG-generated brain tumors measured by densitometry. U87MG.ΔEGFR Cells Have Higher CDK2-Cyclin A Activity and Maintenance of Hyperphosphorylated RB Proteins Than Parental U87MG Cells under Serum Starvation Conditions.  Because p27 binds to CDK2-cyclin A/E complexes and suppresses their activities, we examined these kinase activities by incubating Histone H1 with immunocomplexes of CDK2. CDK2-cyclin A/E kinase activities in U87MG.ΔEGFR were selectively 2-fold higher than in parental cells under serum starvation conditions (Fig. 3A) ⇓ . Consistent with this higher kinase activity, the RB protein in U87MG.ΔEGFR cells was hyperphosphorylated even under serum starvation conditions. When CDK2-cyclin A and CDK2-cyclin E kinase activities were measured separately, only CDK2-cyclin A kinase activity was up-regulated in vitro (Fig. 3A) ⇓ . There was no obvious difference in CDK4-cyclin D, CDK6-cyclin D, and cyclin D1-CDK4/6 kinase activities between cell lines.  Fig. 3. Download figureOpen in new tabDownload powerpoint Fig. 3. CDK-cyclin Kinase activities and phosphorylation of the RB protein. In A, CDK2 and CDK4/CDK6 kinase activities were measured using Histone H1 or GST-RB protein as substrates, respectively. CDK2-cyclin A/E kinase activities were higher in U87MG.ΔEGFR cells than in parental U87MG cells under serum starvation conditions. CDK2-cyclin A and CDK2-cyclin E kinase activities, obtained from immunocomplexes of cyclin A or cyclin E, were also measured, and only CDK2-cyclin A kinase activity was maintained in U87MG.ΔEGFR cells. Consistent with high CDK2-cyclin A kinase activity, the RB protein was hyperphosphorylated even under serum starvation conditions in U87MG.ΔEGFR cells. There was no obvious difference in CDK4-cyclin D, CDK6-cyclin D, and cyclin D1-CDK4/6 kinase activities among these cells with or without serum. In B, similarly, U87MG.ΔEGFR tumors showed higher activity of CDK2-cyclin A and hyperphosphorylated RB proteins, compared with parental U87MG and U87MG.DK tumors. CDK4-cyclin D and CDK6-cyclin D kinase activities were not altered among these tumors. Similar results were observed with in vivo tumors because U87MG.ΔEGFR tumors showed a 3-fold higher activity of CDK2-cyclin A and increased hyperphosphorylated RB protein (Fig. 3B) ⇓ when compared with parental U87MG and U87MG.DK tumors. CDK4-cyclin D, CDK6-cyclin D kinase activities were not altered among these tumors.  U87MG.ΔEGFR Cells Have More PI3-K Activity and Maintain Phosphorylation of Akt in Contrast with Parental U87MG Cells in Vitro and in Vivo.  Because wild-type EGFR signaling leads to PI3-K activation and subsequent Akt phosphorylation, we examined these parameters in U87MG and U87MG.ΔEGFR cells. PI3-K activity was decreased along with phosphorylated Akt levels in parental U87MG cells, grown under serum starvation conditions (Fig. 4) ⇓ . In contrast, these levels were unaltered under identical growth conditions for U87MG.ΔEGFR cells in vitro. U87MG.ΔEGFR tumors expressed similarly greater levels of activated PI3-K and phosphorylated Akt in vivo (Fig. 4) ⇓ .  Fig. 4. Download figureOpen in new tabDownload powerpoint Fig. 4. PI3-K activity and phosphorylation of Akt. PI3-K activity was obtained from anti-p85 immunoprecipitates. 32P-phosphorylated PIP3 product was analyzed using thin-layer chromatography method as described in “Materials and Methods.” Relative PI3-K activity, compared with that in U87MG cells cultured under 10% serum, was determined by PhosphorImager quantitation of PIP3 spots. Whereas PI3-K activity in parental U87MG was decreased to 75% when cells were cultured without serum, it was highly maintained in U87MG.ΔEGFR cells. Levels of total Akt and phosphorylated Akt were compared in each cell line by immunoblotting. Phosphorylation of Akt was not altered in U87MG.ΔEGFR cells even under serum starvation conditions in vitro. Activated PI3-K and phosphorylated Akt levels were also determined for U87MG-, U87MG DK-, and U87MG.ΔEGFR-generated tumors. Inhibition of PI3-K or Akt in U87MG.ΔEGFR Cells Induces G1 Arrest and Up-Regulation of p27.  To test whether activation of PI3-K and phosphorylation of Akt influences the expression of p27, U87MG.ΔEGFR cells were cultured for 24 h without serum and then treated with 20 μm LY294002, a PI3-K inhibitor. This treatment caused reduced levels of phosphorylated Akt and also 3-fold up-regulated levels of p27 (Fig. 5A) ⇓ . In a similar fashion, overexpression of the dominant-negative KD-Akt mutant in U87MG.ΔEGFR cells also produced 2-fold increased levels of p27 (Fig. 5B) ⇓ and a G1 arrest (Fig. 5C) ⇓ , when compared with vector control (pcDNA) transfected cells. Exogenous expression of p27 in U87MG.ΔEGFR cells led to a cell cycle block similar to that of KD-Akt (Fig. 5C) ⇓ , which indicated that p27 levels increased by least 2-fold are sufficient for the G1 arrest.  Fig. 5. Download figureOpen in new tabDownload powerpoint Fig. 5. Effects of inhibiting PI3-K/Akt activity on in vitro growth of the U87MG.ΔEGFR cells. In A, U87MG.ΔEGFR cells were cultured for 24 h without serum and then treated with 20 μm LY294002. Treatment with LY294002 up-regulated levels of p27. In B, U87MG.ΔEGFR cells were transiently transfected with control pcDNA, pcDNA KD-Akt, or pcDNA p27 and cultured for 48 h under 2% serum. Overexpression of p27 in U87MG.ΔEGFR cells induced G1 arrest and, similarly, overexpression of KD-Akt resulted in up-regulation of p27 and induced G1 arrest (C). Inhibition of Akt Causes Decreased Tumorigenicity of U87MG.ΔEGFR Cells in Vivo.  Because U87MG.ΔEGFR tumors enhance tumorigenicity in vivo and maintain high PI3-K activity and phosphorylation of Akt, thus leading to low expression levels of p27, we examined whether Akt inhibition, by exogenous expression of KD-Akt, inhibits their tumorigenicity. U87MG.ΔEGFR cells were transfected with pcDNA KD-Akt, and several stable transfectants were obtained. There was no obvious difference in cell growth among U87MG.ΔEGFR, U87MG.ΔEGFR-KD-Akt 1, and U87MG.ΔEGFR-KD-Akt 2 cells under 10% serum conditions in vitro. When these transfectants were injected s.c. into nude mice, the KD-Akt-transfected clones produced substantially smaller tumors when compared with U87MG.ΔEGFR cells and, like our in vitro results, expressed up-regulated p27 (Fig. 6) ⇓ . These results indicate that ΔEGFR-mediated down-regulation of p27 expression, through the PI3-K/Akt pathway, plays a direct role in the enhanced tumor growth of U87MG.ΔEGFR cells.  Fig. 6. Download figureOpen in new tabDownload powerpoint Fig. 6. The effect of inhibiting Akt activity on U87MG.ΔEGFR tumorigenicity. U87MG.ΔEGFR cells were transfected with control pcDNA or pcDNA KD-Akt, and the stable transfectants, KD-Akt 1 and KD-Akt 2, were obtained. Cells (1 × 106) of parental U87MG, U87MG.ΔEGFR with pcDNA, KD-Akt 1, and KD-Akt 2 cells in 0.1 ml of PBS were injected s.c. into the right flanks of nude mice. Tumor volumes were compared 3 weeks after the injection. Expression of KD-Akt in the U87MG.ΔEGFR cells resulted in up-regulation of p27 and substantial down-regulation of tumorigenicity in vivo. DISCUSSION  We have demonstrated that expression of ΔEGFR can increase the proliferation of various glioma cell lines in vivo. The volume of brain tumors obtained from each ΔEGFR transfectant was more than five times greater than that of parental cells, mirrored by a higher Ki-67 proliferation index. There was no obvious difference in cell growth in vitro between parental U87MG cells and U87MG.ΔEGFR cells under normal serum conditions. However, parental U87MG cells underwent a G1 arrest in serum starvation conditions, whereas U87MG.ΔEGFR cells were not completely arrested in identical conditions. To elucidate the mechanism by which ΔEGFR signaling increased proliferation in vivo and in vitro under serum starvation conditions, we initially compared the expression levels of cell cycle-related proteins and found that ΔEGFR-expressing cells maintained low expression levels of p27 and high levels of cyclin A. Consistent with these results, CDK2-cyclin A kinase activity was higher in U87MG.ΔEGFR cells in vitro without serum than the activity in parental U87MG cells. Tumors derived from U87MG.ΔEGFR cells also showed higher CDK2-cyclin A kinase activity than did parental tumors, which led to hyperphosphorylation of the RB protein and to cell cycle progression. Our study showed that high levels of PI3-K activity and phosphorylated Akt are found in U87MG.ΔEGFR cells in vivo and in vitro even under serum starvation conditions. All of the results illustrating the expression levels and kinase activities of cell cycle-related proteins in cultured cells in serum starvation conditions are very similar to the levels and activities in intracranial brain tumors. This suggests that some of the ΔEGFR-signaling phenotype mediated in vivo can be mimicked by serum starvation conditions in vitro.  p27 is a member of the CDKIs, which also includes p21WAF1 and p57KIP1. These molecules interact with cyclin A, E, and D, which bind to CDK 2, 4 and 6. CDK-cyclin complexes mediate the G1-S phase transition by phosphorylating the RB protein, which in turn leads to the activation of the E2F family of transcription factors. We demonstrated that parental U87MG cells underwent a G1 cell cycle arrest under serum starvation conditions mediated by the up-regulation of p27, which resulted in a decrease in CDK2-cyclin A kinase activity, and hypophosphorylation of the RB protein. CDK4-cyclin D and CDK6-cyclin D kinase activities were not altered under serum starvation conditions, presumably because of U87MG being mutated for p16 (28) , which mainly binds CDK4-cyclin D and CDK6-cyclin D complexes and inhibits their activities. Because the higher activities of CDK4-cyclin D and CDK6-cyclin D were maintained even under serum starvation conditions, the expression levels of p27, followed by the alteration of CDK2-cyclin A kinase activity, appear to play an important role in regulating cell cycle progression from G1 into S phase in these glioblastoma cells. p27 has been shown to be a bona fide tumor suppressor gene because p27−/− mice develop pituitary tumors spontaneously (29) , similar to RB mutant mice (30) . Mutations of p27 are rare in glioblastomas (15) ; however, the expression of p27 is reduced in high-grade gliomas and is inversely related to the prognosis of the patients (15, 16, 17) .  We also demonstrated that treatment with LY294002 caused reduced levels of phosphorylated Akt and concomitantly up-regulated levels of p27, and that overexpression of KD-Akt in the U87MG.ΔEGFR cells restored p27 levels similar to parental U87MG cells, which resulted in a G1 arrest. In contrast to U87MG.ΔEGFR cells, p27 is up-regulated in U87MG.wt EGFR cells under serum starvation conditions similar to U87MG parental cells, but it is also decreased after stimulation with EGF (data not shown). However, pretreatment with LY294002 can completely block phosphorylation of Akt and decrease p27 even after stimulation with EGF in U87MG.wt EGFR cells (data not shown). LY294002 also completely inhibits the phosphorylation of Akt and increases p27 levels in U87MG, DK, and ΔEGFR cell lines(data not shown). These results suggest that the activation of PI3-K, followed by phosphorylation of Akt, regulates the expression levels of p27 in glioblastoma cells. It has been reported that Akt phosphorylates and inactivates the Forkhead transcription factors, which mediate the transcription of p27 (21, 22, 23, 24) . Therefore the activation of the ΔEGFR/PI3-K/Akt pathway inhibits the expression levels of p27, which contributes to enhanced proliferation of the tumor cells.  Thirty to 40% of glioblastomas have a mutated PTEN gene (31 , 32) , which can inhibit PI3-K/Akt signaling by dephosphorylating PIP3. Introduction of PTEN in glioblastoma cells mediates a G1 cell cycle arrest by up-regulating p27 through the inhibition of the PI3-K/Akt pathway (33 , 34) . About 60% of glioblastomas have amplification of EGFR or mutated PTEN genes (32) , and, thus, the PI3-K/Akt pathway is activated in many, if not most, glioblastomas. Because Akt is phosphorylated in most glioblastomas (35) and the PI3-K/Akt pathway plays a pivotal role in tumor proliferation in glioblastomas (35 , 36) , other mechanisms for activating the PI3-K/Akt pathway, in addition to amplification or rearrangement of the EGFR gene or mutation of the PTEN gene, are likely and may provide additional insight into tumor growth regulation by this signal transduction cascade. In any case, its central and substantial role in affecting tumorigenic behavior suggests that this pathway may be a suitable target for therapeutic intervention in tumors expressing this mutated receptor.'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "txt3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "textome1 = txt1 + ' ' + txt3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "tokens1 = process_text1(txt1)\n",
    "tokens2 = process_text1(txt3)\n",
    "tokens_agg = process_text1(textome1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>Create Mutation Table</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mut_table1 = create_mutation_words_table(tokens1)\n",
    "mut_table2 = create_mutation_words_table(tokens2)\n",
    "mut_table_agg = create_mutation_words_table(tokens_agg)\n",
    "mut_table = pd.DataFrame([mut_table1, mut_table2, mut_table_agg])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Total</th>\n",
       "      <th>amplification</th>\n",
       "      <th>deletion</th>\n",
       "      <th>duplication</th>\n",
       "      <th>epigenetic</th>\n",
       "      <th>frame</th>\n",
       "      <th>fusion</th>\n",
       "      <th>insertion</th>\n",
       "      <th>overexpression</th>\n",
       "      <th>promoter</th>\n",
       "      <th>splice</th>\n",
       "      <th>subtype</th>\n",
       "      <th>truncation</th>\n",
       "      <th>wildtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>124</td>\n",
       "      <td>37</td>\n",
       "      <td>26</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>32</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>142</td>\n",
       "      <td>40</td>\n",
       "      <td>32</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>36</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Total  amplification  deletion  duplication  epigenetic  frame  fusion  \\\n",
       "0     18              3         6            1           0      0       3   \n",
       "1    124             37        26            2           0     10       3   \n",
       "2    142             40        32            3           0     10       6   \n",
       "\n",
       "   insertion  overexpression  promoter  splice  subtype  truncation  wildtype  \n",
       "0          0               4         0       0        1           0         0  \n",
       "1          2              32         1       3        5           3         0  \n",
       "2          2              36         1       3        6           3         0  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mut_table1 = create_mutation_words_table(tokens1, normed=True)\n",
    "mut_table2 = create_mutation_words_table(tokens2, normed=True)\n",
    "mut_table = pd.DataFrame([mut_table1, mut_table2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>amplification</th>\n",
       "      <th>deletion</th>\n",
       "      <th>duplication</th>\n",
       "      <th>epigenetic</th>\n",
       "      <th>frame</th>\n",
       "      <th>fusion</th>\n",
       "      <th>insertion</th>\n",
       "      <th>overexpression</th>\n",
       "      <th>promoter</th>\n",
       "      <th>splice</th>\n",
       "      <th>subtype</th>\n",
       "      <th>truncation</th>\n",
       "      <th>wildtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.298387</td>\n",
       "      <td>0.209677</td>\n",
       "      <td>0.016129</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.080645</td>\n",
       "      <td>0.024194</td>\n",
       "      <td>0.016129</td>\n",
       "      <td>0.258065</td>\n",
       "      <td>0.008065</td>\n",
       "      <td>0.024194</td>\n",
       "      <td>0.040323</td>\n",
       "      <td>0.024194</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   amplification  deletion  duplication  epigenetic     frame    fusion  \\\n",
       "0       0.166667  0.333333     0.055556         0.0  0.000000  0.166667   \n",
       "1       0.298387  0.209677     0.016129         0.0  0.080645  0.024194   \n",
       "\n",
       "   insertion  overexpression  promoter    splice   subtype  truncation  \\\n",
       "0   0.000000        0.222222  0.000000  0.000000  0.055556    0.000000   \n",
       "1   0.016129        0.258065  0.008065  0.024194  0.040323    0.024194   \n",
       "\n",
       "   wildtype  \n",
       "0       0.0  \n",
       "1       0.0  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>Create a sparse matrix for gene-ish words space</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "genes = list(class_train['Gene'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "glike_words1 = get_gene_like_words(tokens1, gene_list=genes)\n",
    "glike_words2 = get_gene_like_words(tokens2, gene_list=genes)\n",
    "glike_words_agg = get_gene_like_words(tokens_agg, gene_list=genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "c1 = dict(Counter(glike_words1))\n",
    "c2 = dict(Counter(glike_words2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "gene_table = pd.DataFrame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "gene_table = gene_table.append(c1, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "gene_table = gene_table.append(c2, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ABI</th>\n",
       "      <th>AML</th>\n",
       "      <th>AML,38</th>\n",
       "      <th>AML.15</th>\n",
       "      <th>AML.15,23,27</th>\n",
       "      <th>AND</th>\n",
       "      <th>ATG</th>\n",
       "      <th>BCR-ABL</th>\n",
       "      <th>BD</th>\n",
       "      <th>BRAF</th>\n",
       "      <th>...</th>\n",
       "      <th>TUNEL-positive</th>\n",
       "      <th>UBI</th>\n",
       "      <th>US7MG</th>\n",
       "      <th>USA</th>\n",
       "      <th>UT</th>\n",
       "      <th>VEGF</th>\n",
       "      <th>WB</th>\n",
       "      <th>WHO</th>\n",
       "      <th>ZD1839</th>\n",
       "      <th>ZMD.82</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 267 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   ABI  AML  AML,38  AML.15  AML.15,23,27  AND  ATG  BCR-ABL   BD  BRAF  \\\n",
       "0  1.0  2.0     1.0     1.0           1.0  1.0  1.0      3.0  1.0   1.0   \n",
       "1  NaN  NaN     NaN     NaN           NaN  3.0  NaN      NaN  1.0   NaN   \n",
       "\n",
       "    ...    TUNEL-positive  UBI  US7MG  USA   UT  VEGF    WB  WHO  ZD1839  \\\n",
       "0   ...               NaN  NaN    NaN  NaN  NaN   NaN   NaN  NaN     NaN   \n",
       "1   ...               1.0  2.0    4.0  4.0  1.0   5.0  11.0  2.0     1.0   \n",
       "\n",
       "   ZMD.82  \n",
       "0     NaN  \n",
       "1     1.0  \n",
       "\n",
       "[2 rows x 267 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='sec4'></a>\n",
    "# Compiling the entire text-ome - full mutation table (<a href='#sec0'>Back To Top</a>)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Cyclin-dependent kinases (CDKs) regulate a var...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>Recent evidence has demonstrated that acquired...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>Oncogenic mutations in the monomeric Casitas B...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID                                               Text\n",
       "0   0  Cyclin-dependent kinases (CDKs) regulate a var...\n",
       "1   1   Abstract Background  Non-small cell lung canc...\n",
       "2   2   Abstract Background  Non-small cell lung canc...\n",
       "3   3  Recent evidence has demonstrated that acquired...\n",
       "4   4  Oncogenic mutations in the monomeric Casitas B..."
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text_train.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a whole list of dictionaries first and then convert to DF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3min 44s, sys: 64.3 ms, total: 3min 44s\n",
      "Wall time: 3min 45s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "mut_words_list = []\n",
    "for i in range(len(text_train)):\n",
    "    text = text_train.loc[i, 'Text']\n",
    "    tokens = process_text1(text)\n",
    "    mut_words = create_mutation_words_table(tokens, normed='mutatio_types')\n",
    "    mut_words_list.append(mut_words)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "full_mutation_table = pd.DataFrame(mut_words_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>amplification</th>\n",
       "      <th>deletion</th>\n",
       "      <th>duplication</th>\n",
       "      <th>epigenetic</th>\n",
       "      <th>frame</th>\n",
       "      <th>fusion</th>\n",
       "      <th>insertion</th>\n",
       "      <th>overexpression</th>\n",
       "      <th>promoter</th>\n",
       "      <th>splice</th>\n",
       "      <th>subtype</th>\n",
       "      <th>truncation</th>\n",
       "      <th>wildtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.527778</td>\n",
       "      <td>0.027778</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027778</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.027778</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.138889</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027778</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.020833</td>\n",
       "      <td>0.458333</td>\n",
       "      <td>0.010417</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.104167</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>0.020833</td>\n",
       "      <td>0.041667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.135417</td>\n",
       "      <td>0.145833</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.175000</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.025000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.300000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.325000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.025000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>0.068966</td>\n",
       "      <td>0.275862</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.103448</td>\n",
       "      <td>0.137931</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.310345</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>0.025773</td>\n",
       "      <td>0.005155</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.020619</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005155</td>\n",
       "      <td>0.005155</td>\n",
       "      <td>0.005155</td>\n",
       "      <td>0.840206</td>\n",
       "      <td>0.005155</td>\n",
       "      <td>0.087629</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3291</th>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3292</th>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3293</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3294</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3295</th>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3296</th>\n",
       "      <td>0.057143</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>0.114286</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3297</th>\n",
       "      <td>0.027027</td>\n",
       "      <td>0.171171</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.198198</td>\n",
       "      <td>0.108108</td>\n",
       "      <td>0.036036</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.171171</td>\n",
       "      <td>0.081081</td>\n",
       "      <td>0.207207</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3298</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3299</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.178571</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.357143</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3300</th>\n",
       "      <td>0.024691</td>\n",
       "      <td>0.209877</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.197531</td>\n",
       "      <td>0.049383</td>\n",
       "      <td>0.024691</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.271605</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3301</th>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.206897</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.293103</td>\n",
       "      <td>0.103448</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.051724</td>\n",
       "      <td>0.068966</td>\n",
       "      <td>0.172414</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3302</th>\n",
       "      <td>0.065217</td>\n",
       "      <td>0.434783</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>0.108696</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.217391</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3303</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3304</th>\n",
       "      <td>0.065217</td>\n",
       "      <td>0.434783</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>0.108696</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.217391</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3305</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3306</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.066667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.133333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3307</th>\n",
       "      <td>0.106667</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>0.006667</td>\n",
       "      <td>0.006667</td>\n",
       "      <td>0.040000</td>\n",
       "      <td>0.393333</td>\n",
       "      <td>0.093333</td>\n",
       "      <td>0.080000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.073333</td>\n",
       "      <td>0.040000</td>\n",
       "      <td>0.026667</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3308</th>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.307692</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3309</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.178571</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.357143</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3310</th>\n",
       "      <td>0.062500</td>\n",
       "      <td>0.046875</td>\n",
       "      <td>0.015625</td>\n",
       "      <td>0.078125</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.203125</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>0.156250</td>\n",
       "      <td>0.281250</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.093750</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3311</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.078947</td>\n",
       "      <td>0.447368</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.078947</td>\n",
       "      <td>0.263158</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3312</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.122807</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017544</td>\n",
       "      <td>0.754386</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.070175</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035088</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3313</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.178571</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.357143</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3314</th>\n",
       "      <td>0.032787</td>\n",
       "      <td>0.213115</td>\n",
       "      <td>0.016393</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.131148</td>\n",
       "      <td>0.081967</td>\n",
       "      <td>0.081967</td>\n",
       "      <td>0.081967</td>\n",
       "      <td>0.147541</td>\n",
       "      <td>0.131148</td>\n",
       "      <td>0.049180</td>\n",
       "      <td>0.032787</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3315</th>\n",
       "      <td>0.030769</td>\n",
       "      <td>0.097436</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.117949</td>\n",
       "      <td>0.502564</td>\n",
       "      <td>0.056410</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.128205</td>\n",
       "      <td>0.035897</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3316</th>\n",
       "      <td>0.027027</td>\n",
       "      <td>0.162162</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.162162</td>\n",
       "      <td>0.108108</td>\n",
       "      <td>0.081081</td>\n",
       "      <td>0.081081</td>\n",
       "      <td>0.243243</td>\n",
       "      <td>0.027027</td>\n",
       "      <td>0.054054</td>\n",
       "      <td>0.054054</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3317</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3318</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.165049</td>\n",
       "      <td>0.611650</td>\n",
       "      <td>0.106796</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.038835</td>\n",
       "      <td>0.048544</td>\n",
       "      <td>0.019417</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.009709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3319</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027778</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.138889</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.611111</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3320</th>\n",
       "      <td>0.024691</td>\n",
       "      <td>0.209877</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.197531</td>\n",
       "      <td>0.049383</td>\n",
       "      <td>0.024691</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.271605</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3321 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      amplification  deletion  duplication  epigenetic     frame    fusion  \\\n",
       "0          0.000000  0.527778     0.027778    0.000000  0.027778  0.166667   \n",
       "1          0.000000  0.666667     0.000000    0.000000  0.000000  0.000000   \n",
       "2          0.000000  0.666667     0.000000    0.000000  0.000000  0.000000   \n",
       "3          0.166667  0.333333     0.055556    0.000000  0.000000  0.166667   \n",
       "4          0.000000  0.000000     0.000000    0.000000  1.000000  0.000000   \n",
       "5          0.000000  0.000000     0.000000    0.000000  1.000000  0.000000   \n",
       "6          0.000000  0.000000     0.000000    0.000000  1.000000  0.000000   \n",
       "7          0.020833  0.458333     0.010417    0.000000  0.104167  0.062500   \n",
       "8          0.000000  0.000000     0.000000    0.000000  1.000000  0.000000   \n",
       "9          0.000000  0.000000     0.000000    0.000000  0.000000  0.000000   \n",
       "10         0.000000  0.000000     0.000000    0.000000  1.000000  0.000000   \n",
       "11         0.000000  0.000000     0.000000    0.000000  0.333333  0.000000   \n",
       "12         0.100000  0.175000     0.050000    0.025000  0.000000  0.300000   \n",
       "13         0.000000  0.000000     0.000000    0.000000  0.333333  0.000000   \n",
       "14         0.000000  0.000000     0.000000    0.000000  1.000000  0.000000   \n",
       "15         0.000000  0.000000     0.000000    0.000000  0.000000  0.000000   \n",
       "16         0.068966  0.275862     0.000000    0.000000  0.103448  0.137931   \n",
       "17         0.000000  0.333333     0.333333    0.000000  0.000000  0.000000   \n",
       "18         0.000000  0.000000     0.000000    0.000000  1.000000  0.000000   \n",
       "19         0.000000  0.333333     0.333333    0.000000  0.000000  0.000000   \n",
       "20         0.000000  0.571429     0.000000    0.000000  0.142857  0.000000   \n",
       "21         0.000000  0.000000     0.000000    0.000000  1.000000  0.000000   \n",
       "22         0.000000  0.000000     0.000000    0.000000  1.000000  0.000000   \n",
       "23         0.166667  0.333333     0.055556    0.000000  0.000000  0.166667   \n",
       "24         0.166667  0.333333     0.055556    0.000000  0.000000  0.166667   \n",
       "25         0.166667  0.333333     0.055556    0.000000  0.000000  0.166667   \n",
       "26         0.000000  0.000000     0.000000    0.000000  0.250000  0.500000   \n",
       "27         0.000000  0.000000     0.000000    0.000000  0.000000  0.000000   \n",
       "28         0.025773  0.005155     0.000000    0.020619  0.000000  0.005155   \n",
       "29         0.000000  0.000000     0.000000    0.000000  0.000000  0.000000   \n",
       "...             ...       ...          ...         ...       ...       ...   \n",
       "3291       0.500000  0.000000     0.000000    0.000000  0.500000  0.000000   \n",
       "3292       0.500000  0.000000     0.000000    0.000000  0.500000  0.000000   \n",
       "3293       0.000000  0.000000     0.000000    0.000000  0.000000  0.000000   \n",
       "3294       1.000000  0.000000     0.000000    0.000000  0.000000  0.000000   \n",
       "3295       0.250000  0.250000     0.000000    0.000000  0.250000  0.000000   \n",
       "3296       0.057143  0.000000     0.000000    0.000000  0.028571  0.000000   \n",
       "3297       0.027027  0.171171     0.000000    0.000000  0.198198  0.108108   \n",
       "3298       0.000000  0.076923     0.000000    0.000000  0.230769  0.153846   \n",
       "3299       0.000000  0.178571     0.000000    0.035714  0.107143  0.250000   \n",
       "3300       0.024691  0.209877     0.000000    0.000000  0.197531  0.049383   \n",
       "3301       0.034483  0.206897     0.000000    0.000000  0.293103  0.103448   \n",
       "3302       0.065217  0.434783     0.000000    0.021739  0.108696  0.086957   \n",
       "3303       0.000000  0.076923     0.000000    0.000000  0.230769  0.153846   \n",
       "3304       0.065217  0.434783     0.000000    0.021739  0.108696  0.086957   \n",
       "3305       0.000000  0.076923     0.000000    0.000000  0.230769  0.153846   \n",
       "3306       0.000000  0.000000     0.000000    0.000000  0.200000  0.200000   \n",
       "3307       0.106667  0.033333     0.006667    0.006667  0.040000  0.393333   \n",
       "3308       0.153846  0.230769     0.076923    0.000000  0.000000  0.076923   \n",
       "3309       0.000000  0.178571     0.000000    0.035714  0.107143  0.250000   \n",
       "3310       0.062500  0.046875     0.015625    0.078125  0.000000  0.203125   \n",
       "3311       0.000000  0.052632     0.000000    0.026316  0.078947  0.447368   \n",
       "3312       0.000000  0.122807     0.000000    0.000000  0.017544  0.754386   \n",
       "3313       0.000000  0.178571     0.000000    0.035714  0.107143  0.250000   \n",
       "3314       0.032787  0.213115     0.016393    0.000000  0.131148  0.081967   \n",
       "3315       0.030769  0.097436     0.000000    0.000000  0.117949  0.502564   \n",
       "3316       0.027027  0.162162     0.000000    0.000000  0.162162  0.108108   \n",
       "3317       0.000000  0.076923     0.000000    0.000000  0.230769  0.153846   \n",
       "3318       0.000000  0.000000     0.000000    0.000000  0.165049  0.611650   \n",
       "3319       0.000000  0.166667     0.000000    0.000000  0.055556  0.000000   \n",
       "3320       0.024691  0.209877     0.000000    0.000000  0.197531  0.049383   \n",
       "\n",
       "      insertion  overexpression  promoter    splice   subtype  truncation  \\\n",
       "0      0.027778        0.055556  0.000000  0.138889  0.000000    0.027778   \n",
       "1      0.000000        0.000000  0.000000  0.166667  0.166667    0.000000   \n",
       "2      0.000000        0.000000  0.000000  0.166667  0.166667    0.000000   \n",
       "3      0.000000        0.222222  0.000000  0.000000  0.055556    0.000000   \n",
       "4      0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "5      0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "6      0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "7      0.020833        0.041667  0.000000  0.135417  0.145833    0.000000   \n",
       "8      0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "9      0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "10     0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "11     0.000000        0.666667  0.000000  0.000000  0.000000    0.000000   \n",
       "12     0.000000        0.325000  0.000000  0.000000  0.025000    0.000000   \n",
       "13     0.000000        0.666667  0.000000  0.000000  0.000000    0.000000   \n",
       "14     0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "15     0.000000        1.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "16     0.034483        0.000000  0.000000  0.310345  0.034483    0.034483   \n",
       "17     0.000000        0.333333  0.000000  0.000000  0.000000    0.000000   \n",
       "18     0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "19     0.000000        0.333333  0.000000  0.000000  0.000000    0.000000   \n",
       "20     0.000000        0.000000  0.000000  0.142857  0.142857    0.000000   \n",
       "21     0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "22     0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "23     0.000000        0.222222  0.000000  0.000000  0.055556    0.000000   \n",
       "24     0.000000        0.222222  0.000000  0.000000  0.055556    0.000000   \n",
       "25     0.000000        0.222222  0.000000  0.000000  0.055556    0.000000   \n",
       "26     0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "27     0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "28     0.005155        0.005155  0.840206  0.005155  0.087629    0.000000   \n",
       "29     0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "...         ...             ...       ...       ...       ...         ...   \n",
       "3291   0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "3292   0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "3293   0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "3294   0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "3295   0.000000        0.000000  0.250000  0.000000  0.000000    0.000000   \n",
       "3296   0.000000        0.200000  0.571429  0.028571  0.114286    0.000000   \n",
       "3297   0.036036        0.000000  0.171171  0.081081  0.207207    0.000000   \n",
       "3298   0.076923        0.230769  0.000000  0.076923  0.000000    0.153846   \n",
       "3299   0.000000        0.000000  0.357143  0.000000  0.035714    0.035714   \n",
       "3300   0.024691        0.000000  0.111111  0.111111  0.271605    0.000000   \n",
       "3301   0.034483        0.051724  0.068966  0.172414  0.000000    0.034483   \n",
       "3302   0.043478        0.000000  0.000000  0.217391  0.021739    0.000000   \n",
       "3303   0.076923        0.230769  0.000000  0.076923  0.000000    0.153846   \n",
       "3304   0.043478        0.000000  0.000000  0.217391  0.021739    0.000000   \n",
       "3305   0.076923        0.230769  0.000000  0.076923  0.000000    0.153846   \n",
       "3306   0.400000        0.000000  0.066667  0.000000  0.000000    0.000000   \n",
       "3307   0.093333        0.080000  0.100000  0.073333  0.040000    0.026667   \n",
       "3308   0.307692        0.076923  0.076923  0.000000  0.000000    0.000000   \n",
       "3309   0.000000        0.000000  0.357143  0.000000  0.035714    0.035714   \n",
       "3310   0.062500        0.156250  0.281250  0.000000  0.093750    0.000000   \n",
       "3311   0.000000        0.078947  0.263158  0.026316  0.026316    0.000000   \n",
       "3312   0.000000        0.000000  0.070175  0.000000  0.035088    0.000000   \n",
       "3313   0.000000        0.000000  0.357143  0.000000  0.035714    0.035714   \n",
       "3314   0.081967        0.081967  0.147541  0.131148  0.049180    0.032787   \n",
       "3315   0.056410        0.000000  0.128205  0.035897  0.025641    0.000000   \n",
       "3316   0.081081        0.081081  0.243243  0.027027  0.054054    0.054054   \n",
       "3317   0.076923        0.230769  0.000000  0.076923  0.000000    0.153846   \n",
       "3318   0.106796        0.000000  0.038835  0.048544  0.019417    0.000000   \n",
       "3319   0.027778        0.000000  0.138889  0.000000  0.611111    0.000000   \n",
       "3320   0.024691        0.000000  0.111111  0.111111  0.271605    0.000000   \n",
       "\n",
       "      wildtype  \n",
       "0     0.000000  \n",
       "1     0.000000  \n",
       "2     0.000000  \n",
       "3     0.000000  \n",
       "4     0.000000  \n",
       "5     0.000000  \n",
       "6     0.000000  \n",
       "7     0.000000  \n",
       "8     0.000000  \n",
       "9     0.000000  \n",
       "10    0.000000  \n",
       "11    0.000000  \n",
       "12    0.000000  \n",
       "13    0.000000  \n",
       "14    0.000000  \n",
       "15    0.000000  \n",
       "16    0.000000  \n",
       "17    0.000000  \n",
       "18    0.000000  \n",
       "19    0.000000  \n",
       "20    0.000000  \n",
       "21    0.000000  \n",
       "22    0.000000  \n",
       "23    0.000000  \n",
       "24    0.000000  \n",
       "25    0.000000  \n",
       "26    0.250000  \n",
       "27    0.000000  \n",
       "28    0.000000  \n",
       "29    0.000000  \n",
       "...        ...  \n",
       "3291  0.000000  \n",
       "3292  0.000000  \n",
       "3293  0.000000  \n",
       "3294  0.000000  \n",
       "3295  0.000000  \n",
       "3296  0.000000  \n",
       "3297  0.000000  \n",
       "3298  0.000000  \n",
       "3299  0.000000  \n",
       "3300  0.000000  \n",
       "3301  0.000000  \n",
       "3302  0.000000  \n",
       "3303  0.000000  \n",
       "3304  0.000000  \n",
       "3305  0.000000  \n",
       "3306  0.133333  \n",
       "3307  0.000000  \n",
       "3308  0.000000  \n",
       "3309  0.000000  \n",
       "3310  0.000000  \n",
       "3311  0.000000  \n",
       "3312  0.000000  \n",
       "3313  0.000000  \n",
       "3314  0.000000  \n",
       "3315  0.005128  \n",
       "3316  0.000000  \n",
       "3317  0.000000  \n",
       "3318  0.009709  \n",
       "3319  0.000000  \n",
       "3320  0.000000  \n",
       "\n",
       "[3321 rows x 13 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_mutation_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a DF and it's updated as new row appended"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "full_mutation_table2 = pd.DataFrame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3min 42s, sys: 61.9 ms, total: 3min 42s\n",
      "Wall time: 3min 42s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "for i in range(len(text_train)):\n",
    "    text = text_train.loc[i, 'Text']\n",
    "    tokens = process_text1(text)\n",
    "    mut_words = create_mutation_words_table(tokens, normed='mutation_types')\n",
    "    full_mutation_table2 = full_mutation_table2.append(mut_words, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>amplification</th>\n",
       "      <th>deletion</th>\n",
       "      <th>duplication</th>\n",
       "      <th>epigenetic</th>\n",
       "      <th>frame</th>\n",
       "      <th>fusion</th>\n",
       "      <th>insertion</th>\n",
       "      <th>overexpression</th>\n",
       "      <th>promoter</th>\n",
       "      <th>splice</th>\n",
       "      <th>subtype</th>\n",
       "      <th>truncation</th>\n",
       "      <th>wildtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.527778</td>\n",
       "      <td>0.027778</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027778</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.027778</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.138889</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027778</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.020833</td>\n",
       "      <td>0.458333</td>\n",
       "      <td>0.010417</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.104167</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>0.020833</td>\n",
       "      <td>0.041667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.135417</td>\n",
       "      <td>0.145833</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.175000</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.025000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.300000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.325000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.025000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>0.068966</td>\n",
       "      <td>0.275862</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.103448</td>\n",
       "      <td>0.137931</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.310345</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>0.025773</td>\n",
       "      <td>0.005155</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.020619</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005155</td>\n",
       "      <td>0.005155</td>\n",
       "      <td>0.005155</td>\n",
       "      <td>0.840206</td>\n",
       "      <td>0.005155</td>\n",
       "      <td>0.087629</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3291</th>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3292</th>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3293</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3294</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3295</th>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3296</th>\n",
       "      <td>0.057143</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>0.114286</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3297</th>\n",
       "      <td>0.027027</td>\n",
       "      <td>0.171171</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.198198</td>\n",
       "      <td>0.108108</td>\n",
       "      <td>0.036036</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.171171</td>\n",
       "      <td>0.081081</td>\n",
       "      <td>0.207207</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3298</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3299</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.178571</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.357143</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3300</th>\n",
       "      <td>0.024691</td>\n",
       "      <td>0.209877</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.197531</td>\n",
       "      <td>0.049383</td>\n",
       "      <td>0.024691</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.271605</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3301</th>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.206897</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.293103</td>\n",
       "      <td>0.103448</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.051724</td>\n",
       "      <td>0.068966</td>\n",
       "      <td>0.172414</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.034483</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3302</th>\n",
       "      <td>0.065217</td>\n",
       "      <td>0.434783</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>0.108696</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.217391</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3303</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3304</th>\n",
       "      <td>0.065217</td>\n",
       "      <td>0.434783</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>0.108696</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.217391</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3305</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3306</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.066667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.133333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3307</th>\n",
       "      <td>0.106667</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>0.006667</td>\n",
       "      <td>0.006667</td>\n",
       "      <td>0.040000</td>\n",
       "      <td>0.393333</td>\n",
       "      <td>0.093333</td>\n",
       "      <td>0.080000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.073333</td>\n",
       "      <td>0.040000</td>\n",
       "      <td>0.026667</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3308</th>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.307692</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3309</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.178571</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.357143</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3310</th>\n",
       "      <td>0.062500</td>\n",
       "      <td>0.046875</td>\n",
       "      <td>0.015625</td>\n",
       "      <td>0.078125</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.203125</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>0.156250</td>\n",
       "      <td>0.281250</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.093750</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3311</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.078947</td>\n",
       "      <td>0.447368</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.078947</td>\n",
       "      <td>0.263158</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3312</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.122807</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017544</td>\n",
       "      <td>0.754386</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.070175</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035088</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3313</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.178571</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.357143</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3314</th>\n",
       "      <td>0.032787</td>\n",
       "      <td>0.213115</td>\n",
       "      <td>0.016393</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.131148</td>\n",
       "      <td>0.081967</td>\n",
       "      <td>0.081967</td>\n",
       "      <td>0.081967</td>\n",
       "      <td>0.147541</td>\n",
       "      <td>0.131148</td>\n",
       "      <td>0.049180</td>\n",
       "      <td>0.032787</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3315</th>\n",
       "      <td>0.030769</td>\n",
       "      <td>0.097436</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.117949</td>\n",
       "      <td>0.502564</td>\n",
       "      <td>0.056410</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.128205</td>\n",
       "      <td>0.035897</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3316</th>\n",
       "      <td>0.027027</td>\n",
       "      <td>0.162162</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.162162</td>\n",
       "      <td>0.108108</td>\n",
       "      <td>0.081081</td>\n",
       "      <td>0.081081</td>\n",
       "      <td>0.243243</td>\n",
       "      <td>0.027027</td>\n",
       "      <td>0.054054</td>\n",
       "      <td>0.054054</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3317</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3318</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.165049</td>\n",
       "      <td>0.611650</td>\n",
       "      <td>0.106796</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.038835</td>\n",
       "      <td>0.048544</td>\n",
       "      <td>0.019417</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.009709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3319</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027778</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.138889</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.611111</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3320</th>\n",
       "      <td>0.024691</td>\n",
       "      <td>0.209877</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.197531</td>\n",
       "      <td>0.049383</td>\n",
       "      <td>0.024691</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.271605</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3321 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      amplification  deletion  duplication  epigenetic     frame    fusion  \\\n",
       "0          0.000000  0.527778     0.027778    0.000000  0.027778  0.166667   \n",
       "1          0.000000  0.666667     0.000000    0.000000  0.000000  0.000000   \n",
       "2          0.000000  0.666667     0.000000    0.000000  0.000000  0.000000   \n",
       "3          0.166667  0.333333     0.055556    0.000000  0.000000  0.166667   \n",
       "4          0.000000  0.000000     0.000000    0.000000  1.000000  0.000000   \n",
       "5          0.000000  0.000000     0.000000    0.000000  1.000000  0.000000   \n",
       "6          0.000000  0.000000     0.000000    0.000000  1.000000  0.000000   \n",
       "7          0.020833  0.458333     0.010417    0.000000  0.104167  0.062500   \n",
       "8          0.000000  0.000000     0.000000    0.000000  1.000000  0.000000   \n",
       "9          0.000000  0.000000     0.000000    0.000000  0.000000  0.000000   \n",
       "10         0.000000  0.000000     0.000000    0.000000  1.000000  0.000000   \n",
       "11         0.000000  0.000000     0.000000    0.000000  0.333333  0.000000   \n",
       "12         0.100000  0.175000     0.050000    0.025000  0.000000  0.300000   \n",
       "13         0.000000  0.000000     0.000000    0.000000  0.333333  0.000000   \n",
       "14         0.000000  0.000000     0.000000    0.000000  1.000000  0.000000   \n",
       "15         0.000000  0.000000     0.000000    0.000000  0.000000  0.000000   \n",
       "16         0.068966  0.275862     0.000000    0.000000  0.103448  0.137931   \n",
       "17         0.000000  0.333333     0.333333    0.000000  0.000000  0.000000   \n",
       "18         0.000000  0.000000     0.000000    0.000000  1.000000  0.000000   \n",
       "19         0.000000  0.333333     0.333333    0.000000  0.000000  0.000000   \n",
       "20         0.000000  0.571429     0.000000    0.000000  0.142857  0.000000   \n",
       "21         0.000000  0.000000     0.000000    0.000000  1.000000  0.000000   \n",
       "22         0.000000  0.000000     0.000000    0.000000  1.000000  0.000000   \n",
       "23         0.166667  0.333333     0.055556    0.000000  0.000000  0.166667   \n",
       "24         0.166667  0.333333     0.055556    0.000000  0.000000  0.166667   \n",
       "25         0.166667  0.333333     0.055556    0.000000  0.000000  0.166667   \n",
       "26         0.000000  0.000000     0.000000    0.000000  0.250000  0.500000   \n",
       "27         0.000000  0.000000     0.000000    0.000000  0.000000  0.000000   \n",
       "28         0.025773  0.005155     0.000000    0.020619  0.000000  0.005155   \n",
       "29         0.000000  0.000000     0.000000    0.000000  0.000000  0.000000   \n",
       "...             ...       ...          ...         ...       ...       ...   \n",
       "3291       0.500000  0.000000     0.000000    0.000000  0.500000  0.000000   \n",
       "3292       0.500000  0.000000     0.000000    0.000000  0.500000  0.000000   \n",
       "3293       0.000000  0.000000     0.000000    0.000000  0.000000  0.000000   \n",
       "3294       1.000000  0.000000     0.000000    0.000000  0.000000  0.000000   \n",
       "3295       0.250000  0.250000     0.000000    0.000000  0.250000  0.000000   \n",
       "3296       0.057143  0.000000     0.000000    0.000000  0.028571  0.000000   \n",
       "3297       0.027027  0.171171     0.000000    0.000000  0.198198  0.108108   \n",
       "3298       0.000000  0.076923     0.000000    0.000000  0.230769  0.153846   \n",
       "3299       0.000000  0.178571     0.000000    0.035714  0.107143  0.250000   \n",
       "3300       0.024691  0.209877     0.000000    0.000000  0.197531  0.049383   \n",
       "3301       0.034483  0.206897     0.000000    0.000000  0.293103  0.103448   \n",
       "3302       0.065217  0.434783     0.000000    0.021739  0.108696  0.086957   \n",
       "3303       0.000000  0.076923     0.000000    0.000000  0.230769  0.153846   \n",
       "3304       0.065217  0.434783     0.000000    0.021739  0.108696  0.086957   \n",
       "3305       0.000000  0.076923     0.000000    0.000000  0.230769  0.153846   \n",
       "3306       0.000000  0.000000     0.000000    0.000000  0.200000  0.200000   \n",
       "3307       0.106667  0.033333     0.006667    0.006667  0.040000  0.393333   \n",
       "3308       0.153846  0.230769     0.076923    0.000000  0.000000  0.076923   \n",
       "3309       0.000000  0.178571     0.000000    0.035714  0.107143  0.250000   \n",
       "3310       0.062500  0.046875     0.015625    0.078125  0.000000  0.203125   \n",
       "3311       0.000000  0.052632     0.000000    0.026316  0.078947  0.447368   \n",
       "3312       0.000000  0.122807     0.000000    0.000000  0.017544  0.754386   \n",
       "3313       0.000000  0.178571     0.000000    0.035714  0.107143  0.250000   \n",
       "3314       0.032787  0.213115     0.016393    0.000000  0.131148  0.081967   \n",
       "3315       0.030769  0.097436     0.000000    0.000000  0.117949  0.502564   \n",
       "3316       0.027027  0.162162     0.000000    0.000000  0.162162  0.108108   \n",
       "3317       0.000000  0.076923     0.000000    0.000000  0.230769  0.153846   \n",
       "3318       0.000000  0.000000     0.000000    0.000000  0.165049  0.611650   \n",
       "3319       0.000000  0.166667     0.000000    0.000000  0.055556  0.000000   \n",
       "3320       0.024691  0.209877     0.000000    0.000000  0.197531  0.049383   \n",
       "\n",
       "      insertion  overexpression  promoter    splice   subtype  truncation  \\\n",
       "0      0.027778        0.055556  0.000000  0.138889  0.000000    0.027778   \n",
       "1      0.000000        0.000000  0.000000  0.166667  0.166667    0.000000   \n",
       "2      0.000000        0.000000  0.000000  0.166667  0.166667    0.000000   \n",
       "3      0.000000        0.222222  0.000000  0.000000  0.055556    0.000000   \n",
       "4      0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "5      0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "6      0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "7      0.020833        0.041667  0.000000  0.135417  0.145833    0.000000   \n",
       "8      0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "9      0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "10     0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "11     0.000000        0.666667  0.000000  0.000000  0.000000    0.000000   \n",
       "12     0.000000        0.325000  0.000000  0.000000  0.025000    0.000000   \n",
       "13     0.000000        0.666667  0.000000  0.000000  0.000000    0.000000   \n",
       "14     0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "15     0.000000        1.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "16     0.034483        0.000000  0.000000  0.310345  0.034483    0.034483   \n",
       "17     0.000000        0.333333  0.000000  0.000000  0.000000    0.000000   \n",
       "18     0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "19     0.000000        0.333333  0.000000  0.000000  0.000000    0.000000   \n",
       "20     0.000000        0.000000  0.000000  0.142857  0.142857    0.000000   \n",
       "21     0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "22     0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "23     0.000000        0.222222  0.000000  0.000000  0.055556    0.000000   \n",
       "24     0.000000        0.222222  0.000000  0.000000  0.055556    0.000000   \n",
       "25     0.000000        0.222222  0.000000  0.000000  0.055556    0.000000   \n",
       "26     0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "27     0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "28     0.005155        0.005155  0.840206  0.005155  0.087629    0.000000   \n",
       "29     0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "...         ...             ...       ...       ...       ...         ...   \n",
       "3291   0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "3292   0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "3293   0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "3294   0.000000        0.000000  0.000000  0.000000  0.000000    0.000000   \n",
       "3295   0.000000        0.000000  0.250000  0.000000  0.000000    0.000000   \n",
       "3296   0.000000        0.200000  0.571429  0.028571  0.114286    0.000000   \n",
       "3297   0.036036        0.000000  0.171171  0.081081  0.207207    0.000000   \n",
       "3298   0.076923        0.230769  0.000000  0.076923  0.000000    0.153846   \n",
       "3299   0.000000        0.000000  0.357143  0.000000  0.035714    0.035714   \n",
       "3300   0.024691        0.000000  0.111111  0.111111  0.271605    0.000000   \n",
       "3301   0.034483        0.051724  0.068966  0.172414  0.000000    0.034483   \n",
       "3302   0.043478        0.000000  0.000000  0.217391  0.021739    0.000000   \n",
       "3303   0.076923        0.230769  0.000000  0.076923  0.000000    0.153846   \n",
       "3304   0.043478        0.000000  0.000000  0.217391  0.021739    0.000000   \n",
       "3305   0.076923        0.230769  0.000000  0.076923  0.000000    0.153846   \n",
       "3306   0.400000        0.000000  0.066667  0.000000  0.000000    0.000000   \n",
       "3307   0.093333        0.080000  0.100000  0.073333  0.040000    0.026667   \n",
       "3308   0.307692        0.076923  0.076923  0.000000  0.000000    0.000000   \n",
       "3309   0.000000        0.000000  0.357143  0.000000  0.035714    0.035714   \n",
       "3310   0.062500        0.156250  0.281250  0.000000  0.093750    0.000000   \n",
       "3311   0.000000        0.078947  0.263158  0.026316  0.026316    0.000000   \n",
       "3312   0.000000        0.000000  0.070175  0.000000  0.035088    0.000000   \n",
       "3313   0.000000        0.000000  0.357143  0.000000  0.035714    0.035714   \n",
       "3314   0.081967        0.081967  0.147541  0.131148  0.049180    0.032787   \n",
       "3315   0.056410        0.000000  0.128205  0.035897  0.025641    0.000000   \n",
       "3316   0.081081        0.081081  0.243243  0.027027  0.054054    0.054054   \n",
       "3317   0.076923        0.230769  0.000000  0.076923  0.000000    0.153846   \n",
       "3318   0.106796        0.000000  0.038835  0.048544  0.019417    0.000000   \n",
       "3319   0.027778        0.000000  0.138889  0.000000  0.611111    0.000000   \n",
       "3320   0.024691        0.000000  0.111111  0.111111  0.271605    0.000000   \n",
       "\n",
       "      wildtype  \n",
       "0     0.000000  \n",
       "1     0.000000  \n",
       "2     0.000000  \n",
       "3     0.000000  \n",
       "4     0.000000  \n",
       "5     0.000000  \n",
       "6     0.000000  \n",
       "7     0.000000  \n",
       "8     0.000000  \n",
       "9     0.000000  \n",
       "10    0.000000  \n",
       "11    0.000000  \n",
       "12    0.000000  \n",
       "13    0.000000  \n",
       "14    0.000000  \n",
       "15    0.000000  \n",
       "16    0.000000  \n",
       "17    0.000000  \n",
       "18    0.000000  \n",
       "19    0.000000  \n",
       "20    0.000000  \n",
       "21    0.000000  \n",
       "22    0.000000  \n",
       "23    0.000000  \n",
       "24    0.000000  \n",
       "25    0.000000  \n",
       "26    0.250000  \n",
       "27    0.000000  \n",
       "28    0.000000  \n",
       "29    0.000000  \n",
       "...        ...  \n",
       "3291  0.000000  \n",
       "3292  0.000000  \n",
       "3293  0.000000  \n",
       "3294  0.000000  \n",
       "3295  0.000000  \n",
       "3296  0.000000  \n",
       "3297  0.000000  \n",
       "3298  0.000000  \n",
       "3299  0.000000  \n",
       "3300  0.000000  \n",
       "3301  0.000000  \n",
       "3302  0.000000  \n",
       "3303  0.000000  \n",
       "3304  0.000000  \n",
       "3305  0.000000  \n",
       "3306  0.133333  \n",
       "3307  0.000000  \n",
       "3308  0.000000  \n",
       "3309  0.000000  \n",
       "3310  0.000000  \n",
       "3311  0.000000  \n",
       "3312  0.000000  \n",
       "3313  0.000000  \n",
       "3314  0.000000  \n",
       "3315  0.005128  \n",
       "3316  0.000000  \n",
       "3317  0.000000  \n",
       "3318  0.009709  \n",
       "3319  0.000000  \n",
       "3320  0.000000  \n",
       "\n",
       "[3321 rows x 13 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_mutation_table2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_mutation_table.equals(full_mutation_table2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Two methods gave identical result and were equally fast. It seemed like CPU was heating up more with the latter case, I'll use the whole list method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='sec5'></a>\n",
    "# Compiling the entire text-ome - full gene-like words table (<a href='#sec0'>Back To Top</a>)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Cyclin-dependent kinases (CDKs) regulate a var...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>Recent evidence has demonstrated that acquired...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>Oncogenic mutations in the monomeric Casitas B...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID                                               Text\n",
       "0   0  Cyclin-dependent kinases (CDKs) regulate a var...\n",
       "1   1   Abstract Background  Non-small cell lung canc...\n",
       "2   2   Abstract Background  Non-small cell lung canc...\n",
       "3   3  Recent evidence has demonstrated that acquired...\n",
       "4   4  Oncogenic mutations in the monomeric Casitas B..."
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text_train.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "genes = list(class_train['Gene'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3min 53s, sys: 74.9 ms, total: 3min 53s\n",
      "Wall time: 3min 53s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "glike_words_list = []\n",
    "for i in range(len(text_train)):\n",
    "    text = text_train.loc[i, 'Text']\n",
    "    tokens = process_text1(text)\n",
    "    glike_words = get_gene_like_words(tokens, gene_list=genes)\n",
    "    c = dict(Counter(glike_words))\n",
    "    glike_words_list.append(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "glike_words_table = pd.DataFrame(glike_words_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA</th>\n",
       "      <th>AA-3</th>\n",
       "      <th>AA-3'5</th>\n",
       "      <th>AA-30</th>\n",
       "      <th>AA-3555</th>\n",
       "      <th>AA-3V</th>\n",
       "      <th>AA-3′</th>\n",
       "      <th>AA-A00N</th>\n",
       "      <th>AA-N131Y</th>\n",
       "      <th>AA-V600EB-Raf</th>\n",
       "      <th>...</th>\n",
       "      <th>ZW3</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZ</th>\n",
       "      <th>ZZ-TAZ2</th>\n",
       "      <th>ZZ-type</th>\n",
       "      <th>ZZO</th>\n",
       "      <th>ZZQ</th>\n",
       "      <th>ZZZQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3291</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3292</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3293</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3294</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3295</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3296</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3297</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3298</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3299</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3300</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3301</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3302</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3303</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3304</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3305</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3306</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3307</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3308</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3309</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3310</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3311</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3312</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3313</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3314</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3315</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3316</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3317</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3318</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3319</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3320</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3321 rows × 47683 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       AA  AA-3  AA-3'5  AA-30  AA-3555  AA-3V  AA-3′  AA-A00N  AA-N131Y  \\\n",
       "0     NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "1     NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "2     NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3     NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "4     NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "5     NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "6     NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "7     NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "8     NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "9     NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "10    NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "11    NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "12    NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "13    NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "14    NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "15    NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "16    NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "17    NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "18    NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "19    NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "20    NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "21    NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "22    NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "23    NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "24    NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "25    NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "26    NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "27    NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "28    5.0   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "29    NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "...   ...   ...     ...    ...      ...    ...    ...      ...       ...   \n",
       "3291  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3292  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3293  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3294  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3295  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3296  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3297  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3298  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3299  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3300  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3301  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3302  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3303  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3304  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3305  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3306  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3307  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3308  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3309  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3310  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3311  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3312  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3313  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3314  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3315  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3316  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3317  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3318  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3319  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "3320  NaN   NaN     NaN    NaN      NaN    NaN    NaN      NaN       NaN   \n",
       "\n",
       "      AA-V600EB-Raf  ...   ZW3  ZWILCH  ZWINT  ZYX  ZZ  ZZ-TAZ2  ZZ-type  ZZO  \\\n",
       "0               NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "1               NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "2               NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3               NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "4               NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "5               NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "6               NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "7               NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "8               NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "9               NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "10              NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "11              NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "12              NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "13              NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "14              NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "15              NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "16              NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "17              NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "18              NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "19              NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "20              NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "21              NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "22              NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "23              NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "24              NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "25              NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "26              NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "27              NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "28              NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "29              NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "...             ...  ...   ...     ...    ...  ...  ..      ...      ...  ...   \n",
       "3291            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3292            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3293            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3294            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3295            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3296            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3297            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3298            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3299            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3300            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3301            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3302            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3303            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3304            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3305            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3306            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3307            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3308            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3309            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3310            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3311            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3312            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3313            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3314            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3315            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3316            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3317            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3318            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3319            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "3320            NaN  ...   NaN     NaN    NaN  NaN NaN      NaN      NaN  NaN   \n",
       "\n",
       "      ZZQ  ZZZQ  \n",
       "0     NaN   NaN  \n",
       "1     NaN   NaN  \n",
       "2     NaN   NaN  \n",
       "3     NaN   NaN  \n",
       "4     NaN   NaN  \n",
       "5     NaN   NaN  \n",
       "6     NaN   NaN  \n",
       "7     NaN   NaN  \n",
       "8     NaN   NaN  \n",
       "9     NaN   NaN  \n",
       "10    NaN   NaN  \n",
       "11    NaN   NaN  \n",
       "12    NaN   NaN  \n",
       "13    NaN   NaN  \n",
       "14    NaN   NaN  \n",
       "15    NaN   NaN  \n",
       "16    NaN   NaN  \n",
       "17    NaN   NaN  \n",
       "18    NaN   NaN  \n",
       "19    NaN   NaN  \n",
       "20    NaN   NaN  \n",
       "21    NaN   NaN  \n",
       "22    NaN   NaN  \n",
       "23    NaN   NaN  \n",
       "24    NaN   NaN  \n",
       "25    NaN   NaN  \n",
       "26    NaN   NaN  \n",
       "27    NaN   NaN  \n",
       "28    NaN   NaN  \n",
       "29    NaN   NaN  \n",
       "...   ...   ...  \n",
       "3291  NaN   NaN  \n",
       "3292  NaN   NaN  \n",
       "3293  NaN   NaN  \n",
       "3294  NaN   NaN  \n",
       "3295  NaN   NaN  \n",
       "3296  NaN   NaN  \n",
       "3297  NaN   NaN  \n",
       "3298  NaN   NaN  \n",
       "3299  NaN   NaN  \n",
       "3300  NaN   NaN  \n",
       "3301  NaN   NaN  \n",
       "3302  NaN   NaN  \n",
       "3303  NaN   NaN  \n",
       "3304  NaN   NaN  \n",
       "3305  NaN   NaN  \n",
       "3306  NaN   NaN  \n",
       "3307  NaN   NaN  \n",
       "3308  NaN   NaN  \n",
       "3309  NaN   NaN  \n",
       "3310  NaN   NaN  \n",
       "3311  NaN   NaN  \n",
       "3312  NaN   NaN  \n",
       "3313  NaN   NaN  \n",
       "3314  NaN   NaN  \n",
       "3315  NaN   NaN  \n",
       "3316  NaN   NaN  \n",
       "3317  NaN   NaN  \n",
       "3318  NaN   NaN  \n",
       "3319  NaN   NaN  \n",
       "3320  NaN   NaN  \n",
       "\n",
       "[3321 rows x 47683 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glike_words_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='sec6'></a>\n",
    "# Compiling the entire gene-ome - full gene table (not genome)(<a href='#sec0'>Back To Top</a>)\n",
    "- This is NOT the gene-like words from the text\n",
    "- This shows which gene is annotated for each ID in the 'variants' file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from sklearn.preprocessing import OneHotEncoder, LabelEncoder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X_gene = np.array(class_train.Gene)\n",
    "X_gene_int = LabelEncoder().fit_transform(X_gene.ravel()).reshape(-1, 1)\n",
    "X_gene_bin = OneHotEncoder().fit_transform(X_gene_int).toarray()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ 85],\n",
       "       [ 39],\n",
       "       [ 39],\n",
       "       ..., \n",
       "       [221],\n",
       "       [221],\n",
       "       [221]])"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_gene_int"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "full_gene_table = pd.DataFrame(X_gene_bin)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>254</th>\n",
       "      <th>255</th>\n",
       "      <th>256</th>\n",
       "      <th>257</th>\n",
       "      <th>258</th>\n",
       "      <th>259</th>\n",
       "      <th>260</th>\n",
       "      <th>261</th>\n",
       "      <th>262</th>\n",
       "      <th>263</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3291</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3292</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3293</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3294</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3295</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3296</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3297</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3298</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3299</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3300</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3301</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3302</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3303</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3304</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3305</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3306</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3307</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3308</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3309</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3310</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3311</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3312</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3313</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3314</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3315</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3316</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3317</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3318</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3319</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3320</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3321 rows × 264 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      0    1    2    3    4    5    6    7    8    9   ...   254  255  256  \\\n",
       "0     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "1     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "2     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "4     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "5     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "6     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "7     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "8     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "9     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "10    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "11    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "12    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "13    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "14    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "15    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "16    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "17    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "18    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "19    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "20    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "21    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "22    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "23    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "24    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "25    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "26    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "27    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "28    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "29    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "...   ...  ...  ...  ...  ...  ...  ...  ...  ...  ... ...   ...  ...  ...   \n",
       "3291  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3292  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3293  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3294  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3295  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3296  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3297  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3298  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3299  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3300  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3301  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3302  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3303  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3304  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3305  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3306  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3307  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3308  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3309  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3310  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3311  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3312  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3313  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3314  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3315  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3316  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3317  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3318  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3319  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "3320  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0   \n",
       "\n",
       "      257  258  259  260  261  262  263  \n",
       "0     0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "1     0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "2     0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3     0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "4     0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "5     0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "6     0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "7     0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "8     0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "9     0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "10    0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "11    0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "12    0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "13    0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "14    0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "15    0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "16    0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "17    0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "18    0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "19    0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "20    0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "21    0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "22    0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "23    0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "24    0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "25    0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "26    0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "27    0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "28    0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "29    0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "...   ...  ...  ...  ...  ...  ...  ...  \n",
       "3291  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3292  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3293  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3294  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3295  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3296  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3297  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3298  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3299  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3300  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3301  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3302  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3303  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3304  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3305  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3306  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3307  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3308  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3309  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3310  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3311  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3312  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3313  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3314  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3315  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3316  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3317  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3318  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3319  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "3320  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "\n",
       "[3321 rows x 264 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_gene_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    0.0\n",
       "1    1.0\n",
       "2    1.0\n",
       "3    1.0\n",
       "4    1.0\n",
       "5    1.0\n",
       "6    1.0\n",
       "7    1.0\n",
       "8    1.0\n",
       "9    1.0\n",
       "Name: 39, dtype: float64"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_gene_table.loc[:, 39].head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='sec7'></a>\n",
    "# Convert Mutation_Types in Class file (<a href='#sec0'>Back To Top</a>)\n",
    "- Import convert_mutation_type\n",
    "- Use the label encoding to make it a sparse matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def convert_mutation_type(data):\n",
    "    '''\n",
    "    Convert the 'Variant' Data into mutation_type in a new column, returns the new data with a new column\n",
    "\n",
    "    Input\n",
    "    =====\n",
    "    data : DataFrame\n",
    "        The train or test data containing Variant information\n",
    "\n",
    "    Output\n",
    "    ======\n",
    "    data : DataFrame\n",
    "        'mutation_type' is added to the original data from the input\n",
    "    '''\n",
    "    # Copy the Variation into a new column (this could be just an empty copy with Nones)\n",
    "    data['mutation_type'] = data['Variation']\n",
    "\n",
    "    # Define regex pattern for point mutants\n",
    "    point_mutation_pattern = \\\n",
    "        r\"[ARNDCEQGHILKMFPSTWYV]{1}[0-9]{1,4}[ARNDCEQGHILKMFPSTWYV*]?$\"\n",
    "\n",
    "    # Define new mutation types\n",
    "    major_types = ['Truncation', 'Point Mutation', 'Deletion', 'Promoter Mutations',\n",
    "       'Amplification', 'Epigenetic', 'Frame Shift', 'Overexpression',\n",
    "       'Deletion-Insertion', 'Duplication', 'Insertion',\n",
    "       'Gene Subtype', 'Fusion', 'Splice', 'Copy Number Loss', 'Wildtype']\n",
    "\n",
    "    # Convert the Variant information to mutation types\n",
    "    data.loc[(data['Variation'].str.match(point_mutation_pattern)), 'mutation_type']= 'Point Mutation'\n",
    "    data.loc[(data['Variation'].str.contains('missense', case=False)), 'mutation_type']= 'Point Mutation'\n",
    "    data.loc[(data['Variation'].str.contains('fusion', case=False)), 'mutation_type']= 'Fusion'\n",
    "    data.loc[(data['Variation'].str.contains('deletion', case=False)), 'mutation_type']= 'Deletion'\n",
    "    data.loc[((data['Variation'].str.contains('del', case=False))\\\n",
    "            &(data['Variation'].str.contains('delins', case=False) == False)),\n",
    "            'mutation_type']= 'Deletion'\n",
    "    data.loc[((data['Variation'].str.contains('ins', case=False))\\\n",
    "            &(data['Variation'].str.contains('delins', case=False) == False)),\n",
    "            'mutation_type']= 'Insertion'\n",
    "    data.loc[((data['Variation'].str.contains('del', case=False))\\\n",
    "            &(data['Variation'].str.contains('delins', case=False))),\n",
    "            'mutation_type']= 'Deletion-Insertion'\n",
    "    data.loc[(data['Variation'].str.contains('dup', case=False)), 'mutation_type']= 'Duplication'\n",
    "    data.loc[(data['Variation'].str.contains('trunc', case=False)), 'mutation_type']= 'Truncation'\n",
    "    data.loc[(data['Variation'].str.contains('fs', case=False)), 'mutation_type']= 'Frame Shift'\n",
    "    data.loc[(data['Variation'].str.contains('splice', case=False)), 'mutation_type']= 'Splice'\n",
    "    data.loc[(data['Variation'].str.contains('exon', case=False)), 'mutation_type']= 'Point Mutation'\n",
    "    data.loc[((data['Variation'].str.contains('EGFR', case=False))\\\n",
    "            |(data['Variation'].str.contains('AR', case=True))\\\n",
    "            |(data['Variation'].str.contains('MYC-nick', case=True))\\\n",
    "            |(data['Variation'].str.contains('TGFBR1', case=True))\\\n",
    "            |(data['Variation'].str.contains('CASP8L', case=True))),\n",
    "            'mutation_type']= 'Gene Subtype'\n",
    "    data.loc[((data['Variation'].str.contains('Hypermethylation', case=False))\\\n",
    "            |(data['Variation'].str.contains('Epigenetic', case=False))),\n",
    "             'mutation_type']= 'Epigenetic'\n",
    "    data.loc[(data['mutation_type'].isin(major_types) == False),\n",
    "            'mutation_type']= 'Others'\n",
    "\n",
    "    # rearrange order of columns\n",
    "    if 'Class' in data.columns:\n",
    "        data = data[['ID', 'Gene', 'Variation', 'mutation_type', 'Class']]\n",
    "    else:\n",
    "        data = data[['ID', 'Gene', 'Variation', 'mutation_type']]\n",
    "\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "new_table = convert_mutation_type(class_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X_mtype = np.array(new_table['mutation_type'])\n",
    "X_mtype_int = LabelEncoder().fit_transform(X_mtype.ravel()).reshape(-1, 1)\n",
    "X_mtype_bin = OneHotEncoder().fit_transform(X_mtype_int).toarray()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[15],\n",
       "       [12],\n",
       "       [12],\n",
       "       ..., \n",
       "       [ 7],\n",
       "       [12],\n",
       "       [12]])"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_mtype_int"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "full_mtype_table = pd.DataFrame(X_mtype_bin)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3291</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3292</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3293</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3294</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3295</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3296</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3297</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3298</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3299</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3300</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3301</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3302</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3303</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3304</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3305</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3306</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3307</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3308</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3309</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3310</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3311</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3312</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3313</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3314</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3315</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3316</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3317</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3318</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3319</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3320</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3321 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       0    1    2    3    4    5    6    7    8    9    10   11   12   13  \\\n",
       "0     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "1     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "2     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "4     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "5     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "6     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "7     0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "8     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "9     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "10    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "11    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "12    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "13    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "14    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "15    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "16    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "17    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "18    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "19    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "20    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "21    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "22    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "23    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "24    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "25    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "26    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "27    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "28    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "29    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "...   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
       "3291  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3292  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3293  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3294  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3295  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3296  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3297  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3298  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3299  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3300  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3301  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3302  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3303  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3304  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3305  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "3306  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3307  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "3308  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3309  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3310  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "3311  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "3312  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "3313  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3314  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3315  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "3316  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3317  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3318  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "3319  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "3320  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   \n",
       "\n",
       "       14   15   16  \n",
       "0     0.0  1.0  0.0  \n",
       "1     0.0  0.0  0.0  \n",
       "2     0.0  0.0  0.0  \n",
       "3     0.0  0.0  0.0  \n",
       "4     0.0  0.0  0.0  \n",
       "5     0.0  0.0  0.0  \n",
       "6     0.0  0.0  0.0  \n",
       "7     0.0  0.0  0.0  \n",
       "8     0.0  0.0  0.0  \n",
       "9     0.0  0.0  0.0  \n",
       "10    0.0  0.0  0.0  \n",
       "11    0.0  0.0  0.0  \n",
       "12    0.0  0.0  0.0  \n",
       "13    0.0  0.0  0.0  \n",
       "14    0.0  0.0  0.0  \n",
       "15    0.0  0.0  0.0  \n",
       "16    0.0  1.0  0.0  \n",
       "17    0.0  0.0  0.0  \n",
       "18    0.0  0.0  0.0  \n",
       "19    0.0  0.0  0.0  \n",
       "20    0.0  0.0  0.0  \n",
       "21    0.0  0.0  0.0  \n",
       "22    0.0  0.0  0.0  \n",
       "23    0.0  0.0  0.0  \n",
       "24    0.0  0.0  0.0  \n",
       "25    0.0  0.0  0.0  \n",
       "26    0.0  0.0  0.0  \n",
       "27    0.0  0.0  0.0  \n",
       "28    0.0  0.0  0.0  \n",
       "29    0.0  0.0  0.0  \n",
       "...   ...  ...  ...  \n",
       "3291  0.0  0.0  0.0  \n",
       "3292  0.0  0.0  0.0  \n",
       "3293  0.0  0.0  0.0  \n",
       "3294  0.0  0.0  0.0  \n",
       "3295  0.0  0.0  0.0  \n",
       "3296  0.0  0.0  0.0  \n",
       "3297  0.0  0.0  0.0  \n",
       "3298  0.0  0.0  0.0  \n",
       "3299  0.0  0.0  0.0  \n",
       "3300  0.0  0.0  0.0  \n",
       "3301  0.0  0.0  0.0  \n",
       "3302  0.0  0.0  0.0  \n",
       "3303  0.0  0.0  0.0  \n",
       "3304  0.0  0.0  0.0  \n",
       "3305  0.0  0.0  0.0  \n",
       "3306  0.0  0.0  0.0  \n",
       "3307  0.0  1.0  0.0  \n",
       "3308  0.0  0.0  0.0  \n",
       "3309  0.0  0.0  0.0  \n",
       "3310  0.0  0.0  0.0  \n",
       "3311  0.0  0.0  0.0  \n",
       "3312  0.0  0.0  0.0  \n",
       "3313  0.0  0.0  0.0  \n",
       "3314  0.0  0.0  0.0  \n",
       "3315  0.0  0.0  0.0  \n",
       "3316  0.0  0.0  0.0  \n",
       "3317  0.0  0.0  0.0  \n",
       "3318  0.0  0.0  0.0  \n",
       "3319  0.0  0.0  0.0  \n",
       "3320  0.0  0.0  0.0  \n",
       "\n",
       "[3321 rows x 17 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_mtype_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='sec8'></a>\n",
    "# Combined All! (<a href='#sec0'>Back To Top</a>)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "full_mutation_table = full_mutation_table.fillna(value=0)\n",
    "glike_words_table = glike_words_table.fillna(value=0)\n",
    "full_gene_table = full_gene_table.fillna(value=0)\n",
    "full_mtype_table = full_mtype_table.fillna(value=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "features = pd.concat([full_mutation_table, \n",
    "                      glike_words_table,\n",
    "                      full_gene_table,\n",
    "                      full_mtype_table],\n",
    "                      axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3321, 47977)"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3321,)"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class_train.Class.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='sec9'></a>\n",
    "# Test with Random Forest (<a href='#sec0'>Back To Top</a>)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X = np.array(features).astype(float)\n",
    "y = np.array(class_train.Class).astype(int).ravel()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import accuracy_score, recall_score, precision_score, f1_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 10 s, sys: 83.8 ms, total: 10.1 s\n",
      "Wall time: 10.1 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "rfc = RandomForestClassifier(n_estimators=50, max_depth=30)\n",
    "rfc.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "y_pred = rfc.predict(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.670676691729\n"
     ]
    }
   ],
   "source": [
    "print(accuracy_score(y_test, y_pred))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='sec10'></a>\n",
    "# Test with Simple SVM (<a href='#sec0'>Back To Top</a>)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from sklearn.preprocessing import scale\n",
    "from sklearn.svm import LinearSVC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X_scale = scale(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X_train2, X_test2, y_train2, y_test2 = train_test_split(X_scale, y, test_size=0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 31min 37s, sys: 282 ms, total: 31min 38s\n",
      "Wall time: 31min 43s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "clf = LinearSVC()\n",
    "clf.fit(X_train2, y_train2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "y_pred2 = clf.predict(X_test2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.470676691729\n"
     ]
    }
   ],
   "source": [
    "print(accuracy_score(y_test2, y_pred2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
